<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.427">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Callum Arnold">
<meta name="author" content="Matthew Ferrari">
<title>SISMID Module 2 Materials (2023) - 7&nbsp; R Session 02</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./L06_parameter-estimation.html" rel="next">
<link href="./L05_age-structure.html" rel="prev">
<link href="././images/sismid2023-logo-green.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<link href="site_libs/quarto-contrib/line-highlight-1.0.0/line-highlight.css" rel="stylesheet"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script type="module" src="site_libs/quarto-ojs/quarto-ojs-runtime.js"></script><link href="site_libs/quarto-ojs/quarto-ojs.css" rel="stylesheet">
<script>
  window.MathJax = {
    loader: {
      load: ['[tex]/physics']
    },
    tex: {
      packages: {'[+]': ['physics']}
    }
  };
</script><script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./L04_heterogeneity.html">Day 2</a></li><li class="breadcrumb-item"><a href="./r-session-02.html"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">R Session 02</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="./index.html" class="sidebar-logo-link">
      <img src="././images/sismid2023-logo-green.png" alt="" class="sidebar-logo py-0 d-lg-inline d-none"></a>
    <div class="sidebar-title mb-0 py-0">
      <a href="./">SISMID Module 2 Materials (2023)</a> 
        <div class="sidebar-tools-main tools-wide">
    <a href="https://github.com/arnold-c/SISMID-Module-02_2023" rel="" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><strong>Welcome</strong></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Pre-Requisites</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./exercise-requirements.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercise Requirements</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./install-r.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Install and Setup R &amp; RStudio</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./just-enough-rstudio.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Just Enough RStudio</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./just-enough-r.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Just Enough R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./project-management.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Organizing A Project</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Day 1</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./L01_intro-to-modeling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Intro to Modeling</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./L02_sir-basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Basics of SIR Models</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./r-session-01.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">R Session 01</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./L03_hit-and-vaccinations.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Herd Immunity &amp; Vaccination</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Day 2</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./L04_heterogeneity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Heterogeneity in Models</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./L05_age-structure.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Age Structured Models</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./r-session-02.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">R Session 02</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./L06_parameter-estimation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Parameter Estimation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./r-session-03.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">R Session 03</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Day 3</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./L07_models-and-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Confronting Models with Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./L08_stochastic-models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Stochastic Models</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><strong>References</strong></span></a>
  </div>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Page Items</h2>
   
  <ul>
<li><a href="#load-packages" id="toc-load-packages" class="nav-link active" data-scroll-target="#load-packages"><span class="header-section-number">7.1</span> Load Packages</a></li>
  <li><a href="#a-model-with-2-classes" id="toc-a-model-with-2-classes" class="nav-link" data-scroll-target="#a-model-with-2-classes"><span class="header-section-number">7.2</span> A Model With 2 Classes</a></li>
  <li>
<a href="#a-model-with-2-age-classes" id="toc-a-model-with-2-age-classes" class="nav-link" data-scroll-target="#a-model-with-2-age-classes"><span class="header-section-number">7.3</span> A Model With 2 Age Classes</a>
  <ul class="collapse">
<li><a href="#exercise-1-use-this-code-to-plot-the-number-of-susceptible-infected-and-recovered-individuals-over-time" id="toc-exercise-1-use-this-code-to-plot-the-number-of-susceptible-infected-and-recovered-individuals-over-time" class="nav-link" data-scroll-target="#exercise-1-use-this-code-to-plot-the-number-of-susceptible-infected-and-recovered-individuals-over-time"><span class="header-section-number">7.3.1</span> Exercise 1: Use this code to plot the number of susceptible, infected, and recovered individuals over time</a></li>
  </ul>
</li>
  <li>
<a href="#getting-more-realistic-adding-more-age-classes" id="toc-getting-more-realistic-adding-more-age-classes" class="nav-link" data-scroll-target="#getting-more-realistic-adding-more-age-classes"><span class="header-section-number">7.4</span> Getting more realistic: adding more age classes</a>
  <ul class="collapse">
<li><a href="#exercise-2-what-can-you-say-about-its-structure-how-are-the-different-age-groups-in-contact-with-each-other" id="toc-exercise-2-what-can-you-say-about-its-structure-how-are-the-different-age-groups-in-contact-with-each-other" class="nav-link" data-scroll-target="#exercise-2-what-can-you-say-about-its-structure-how-are-the-different-age-groups-in-contact-with-each-other"><span class="header-section-number">7.4.1</span> Exercise 2: What can you say about its structure? How are the different age groups in contact with each other?</a></li>
  <li><a href="#sec-ex-3" id="toc-sec-ex-3" class="nav-link" data-scroll-target="#sec-ex-3"><span class="header-section-number">7.4.2</span> Exercise 3: What does the equilibrium age-specific seroprevalence look like in this example?</a></li>
  <li><a href="#exercise-4-updating-the-contact-matrix" id="toc-exercise-4-updating-the-contact-matrix" class="nav-link" data-scroll-target="#exercise-4-updating-the-contact-matrix"><span class="header-section-number">7.4.3</span> Exercise 4: Updating the contact matrix</a></li>
  </ul>
</li>
  <li>
<a href="#r0-and-the-mean-age-of-infection" id="toc-r0-and-the-mean-age-of-infection" class="nav-link" data-scroll-target="#r0-and-the-mean-age-of-infection"><span class="header-section-number">7.5</span> R0 and the Mean Age of Infection</a>
  <ul class="collapse">
<li><a href="#exercise-5-mean-age-of-infection-interactions" id="toc-exercise-5-mean-age-of-infection-interactions" class="nav-link" data-scroll-target="#exercise-5-mean-age-of-infection-interactions"><span class="header-section-number">7.5.1</span> Exercise 5: mean age of infection interactions</a></li>
  </ul>
</li>
  <li>
<a href="#bonus-materials" id="toc-bonus-materials" class="nav-link" data-scroll-target="#bonus-materials"><span class="header-section-number">7.6</span> Bonus Materials</a>
  <ul class="collapse">
<li><a href="#calculating-r_0-with-the-next-generation-matrix" id="toc-calculating-r_0-with-the-next-generation-matrix" class="nav-link" data-scroll-target="#calculating-r_0-with-the-next-generation-matrix"><span class="header-section-number">7.6.1</span> Calculating <span class="math inline">\(R_0\)</span> with the Next Generation Matrix</a></li>
  <li><a href="#sec-mean-age-r-code" id="toc-sec-mean-age-r-code" class="nav-link" data-scroll-target="#sec-mean-age-r-code"><span class="header-section-number">7.6.2</span> Mean age of infection <code>R</code> code</a></li>
  <li><a href="#what-do-real-contact-networks-look-like" id="toc-what-do-real-contact-networks-look-like" class="nav-link" data-scroll-target="#what-do-real-contact-networks-look-like"><span class="header-section-number">7.6.3</span> What do real contact networks look like?</a></li>
  </ul>
</li>
  </ul><div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/arnold-c/SISMID-Module-02_2023/edit/main/r-session-02.qmd" class="toc-action">Edit this page</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">
<span class="chapter-number">7</span>&nbsp; <span class="chapter-title">R Session 02</span>
</h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
<p class="subtitle lead">Heterogeneity and Age Structure in SIR Models</p>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Authors</div>
  <div class="quarto-title-meta-heading">Affiliation</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author"><a href="https://callumarnold.com">Callum Arnold</a> <a href="https://orcid.org/0000-0002-3245-6956" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Pennsylvania State University
          </p>
      </div>
      <div class="quarto-title-meta-contents">
    <p class="author"><a href="http://theferrarilab.com">Matthew Ferrari</a> <a href="https://orcid.org/0000-0001-5251-8168" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Pennsylvania State University
          </p>
      </div>
    </div>

<div class="quarto-title-meta">

      
  
    
  </div>
  
<div>
  <div class="abstract">
    <div class="abstract-title"></div>
    <p><em>Materials adapted from Helen Wearing and Aaron King <span class="citation" data-cites="kingAgeStructuredModels2011">(<a href="references.html#ref-kingAgeStructuredModels2011" role="doc-biblioref">King and Wearing 2011</a>)</span></em></p>
  </div>
</div>

</header><section id="load-packages" class="level2" data-number="7.1"><h2 data-number="7.1" class="anchored" data-anchor-id="load-packages">
<span class="header-section-number">7.1</span> Load Packages</h2>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-1_7ef239b405ebe349db20b30876fba9e3">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb1" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="fu">library</span>(diagram)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="fu">library</span>(deSolve)</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="fu">library</span>(gt)</span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="fu">library</span>(rio)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-2_49be22abac60e52d9b277249d1e6d052">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb2" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_minimal</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section><section id="a-model-with-2-classes" class="level2 page-columns page-full" data-number="7.2"><h2 data-number="7.2" class="anchored" data-anchor-id="a-model-with-2-classes">
<span class="header-section-number">7.2</span> A Model With 2 Classes</h2>
<p>We’ll start with the simplest mechanistic model of two classes we can think of, which has separate classes for two groups <span class="math inline">\(a\)</span> and <span class="math inline">\(b\)</span>. These groups could represent different socioeconomic classes, for example.</p>
<div class="cell page-columns page-full" data-hash="r-session-02_cache/html/unnamed-chunk-3_d3436093244a142304f95a6e3ab31283">
<div class="cell-output-display column-body">
<p><img src="r-session-02_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid" style="width:100.0%"></p>
</div>
</div>
<p>Which can be written in equations as, <span class="math display">\[
\begin{aligned}
    \frac{\dd{S_a}}{\dd{t}} &amp;= -\lambda_a\,S_a \phantom{-\gamma\,I_b}\\
    \frac{\dd{S_b}}{\dd{t}} &amp;= -\lambda_b\,S_b \phantom{-\gamma\,I_b}\\
    \frac{\dd{I_a}}{\dd{t}} &amp;= \phantom{-}\lambda_a\,S_a -\gamma\,I_a\\
    \frac{\dd{I_b}}{\dd{t}} &amp;= \phantom{-}\lambda_b\,S_b-\gamma\,I_b\\
    \frac{\dd{R_a}}{\dd{t}} &amp;= \phantom{-\lambda_a\,S_b}+\gamma\,I_a\\
    \frac{\dd{R_b}}{\dd{t}} &amp;= \phantom{-\lambda_a\,S_b}+\gamma\,I_b\\
  \end{aligned}
\]</span></p>
<p>The <span class="math inline">\(\lambda\)</span>s denote the group-specific force of infections:</p>
<p><span class="math display">\[
\begin{aligned}
        \lambda_a &amp;= \beta_{aa}\,I_a+\beta_{ab}\,I_b\\
        \lambda_b &amp;= \beta_{ba}\,I_a+\beta_{bb}\,I_b
\end{aligned}
\]</span></p>
<p>In this model, each population can infect each other but the infection moves through the populations separately. Let’s simulate such a model. To make things concrete, we’ll assume that the transmission rates <span class="math inline">\(\beta\)</span> are greater within groups than between them.</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-4_b677c162717b7b7161472d8790080c31">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb3" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="co"># Create a named parameter vector that we can index by name in the model</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>ab_params <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb3-3"><a href="#cb3-3"></a>    <span class="at">beta_within =</span> <span class="fl">0.025</span>,</span>
<span id="cb3-4"><a href="#cb3-4"></a>    <span class="at">beta_between =</span> <span class="fl">0.005</span>,</span>
<span id="cb3-5"><a href="#cb3-5"></a>    <span class="at">recovery =</span> <span class="dv">10</span></span>
<span id="cb3-6"><a href="#cb3-6"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-5_7c1ebb30a21bba272b5a85dd49baddc1">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb4" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="co"># Here we set up the ODE model that matches the equations above</span></span>
<span id="cb4-2"><a href="#cb4-2"></a>ab_model <span class="ot">&lt;-</span> <span class="cf">function</span> (t, x, p, ...) {</span>
<span id="cb4-3"><a href="#cb4-3"></a>    <span class="co"># Unpack the state variables</span></span>
<span id="cb4-4"><a href="#cb4-4"></a>    Sa <span class="ot">&lt;-</span> x[<span class="st">"Sa"</span>]</span>
<span id="cb4-5"><a href="#cb4-5"></a>    Sb <span class="ot">&lt;-</span> x[<span class="st">"Sb"</span>]</span>
<span id="cb4-6"><a href="#cb4-6"></a>    Ia <span class="ot">&lt;-</span> x[<span class="st">"Ia"</span>]</span>
<span id="cb4-7"><a href="#cb4-7"></a>    Ib <span class="ot">&lt;-</span> x[<span class="st">"Ib"</span>]</span>
<span id="cb4-8"><a href="#cb4-8"></a>    </span>
<span id="cb4-9"><a href="#cb4-9"></a>    <span class="co"># Unpack the parameters</span></span>
<span id="cb4-10"><a href="#cb4-10"></a>    beta_within <span class="ot">&lt;-</span> p[<span class="st">"beta_within"</span>]</span>
<span id="cb4-11"><a href="#cb4-11"></a>    beta_between <span class="ot">&lt;-</span> p[<span class="st">"beta_between"</span>]</span>
<span id="cb4-12"><a href="#cb4-12"></a>    recovery <span class="ot">&lt;-</span> p[<span class="st">"recovery"</span>]</span>
<span id="cb4-13"><a href="#cb4-13"></a></span>
<span id="cb4-14"><a href="#cb4-14"></a>    <span class="co"># group A force of infection</span></span>
<span id="cb4-15"><a href="#cb4-15"></a>    lambda_a <span class="ot">&lt;-</span> beta_within <span class="sc">*</span> Ia <span class="sc">+</span> beta_between <span class="sc">*</span> Ib</span>
<span id="cb4-16"><a href="#cb4-16"></a></span>
<span id="cb4-17"><a href="#cb4-17"></a>    <span class="co"># group B force of infection</span></span>
<span id="cb4-18"><a href="#cb4-18"></a>    lambda_b <span class="ot">&lt;-</span> beta_within <span class="sc">*</span> Ib <span class="sc">+</span> beta_between <span class="sc">*</span> Ia</span>
<span id="cb4-19"><a href="#cb4-19"></a>    </span>
<span id="cb4-20"><a href="#cb4-20"></a>    <span class="co"># The ODEs</span></span>
<span id="cb4-21"><a href="#cb4-21"></a>    dSadt <span class="ot">&lt;-</span> <span class="sc">-</span> lambda_a <span class="sc">*</span> Sa</span>
<span id="cb4-22"><a href="#cb4-22"></a>    dSbdt <span class="ot">&lt;-</span> <span class="sc">-</span> lambda_b <span class="sc">*</span> Sb</span>
<span id="cb4-23"><a href="#cb4-23"></a>    dIadt <span class="ot">&lt;-</span> lambda_a <span class="sc">*</span> Sa <span class="sc">-</span> recovery <span class="sc">*</span> Ia</span>
<span id="cb4-24"><a href="#cb4-24"></a>    dIbdt <span class="ot">&lt;-</span> lambda_b <span class="sc">*</span> Sb <span class="sc">-</span> recovery <span class="sc">*</span> Ib</span>
<span id="cb4-25"><a href="#cb4-25"></a>    dRadt <span class="ot">&lt;-</span> recovery <span class="sc">*</span> Ia</span>
<span id="cb4-26"><a href="#cb4-26"></a>    dRbdt <span class="ot">&lt;-</span> recovery <span class="sc">*</span> Ib</span>
<span id="cb4-27"><a href="#cb4-27"></a></span>
<span id="cb4-28"><a href="#cb4-28"></a>    <span class="co"># Return the derivatives</span></span>
<span id="cb4-29"><a href="#cb4-29"></a>    <span class="fu">list</span>(<span class="fu">c</span>(</span>
<span id="cb4-30"><a href="#cb4-30"></a>        dSadt,</span>
<span id="cb4-31"><a href="#cb4-31"></a>        dSbdt,</span>
<span id="cb4-32"><a href="#cb4-32"></a>        dIadt,</span>
<span id="cb4-33"><a href="#cb4-33"></a>        dIbdt,</span>
<span id="cb4-34"><a href="#cb4-34"></a>        dRadt,</span>
<span id="cb4-35"><a href="#cb4-35"></a>        dRbdt</span>
<span id="cb4-36"><a href="#cb4-36"></a>    ))</span>
<span id="cb4-37"><a href="#cb4-37"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-6_dd1c1c833decb3b73f836b96a33a6555">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb5" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="co"># initial conditions</span></span>
<span id="cb5-2"><a href="#cb5-2"></a>ab_yinit <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="at">Sa =</span> <span class="dv">1000</span>, <span class="at">Sb =</span> <span class="dv">2000</span>, <span class="at">Ia =</span> <span class="dv">1</span>, <span class="at">Ib =</span> <span class="dv">1</span>, <span class="at">Ra =</span> <span class="dv">0</span>, <span class="at">Rb =</span> <span class="dv">0</span>)</span>
<span id="cb5-3"><a href="#cb5-3"></a></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="co"># Run the ODE solver from the deSolve package</span></span>
<span id="cb5-5"><a href="#cb5-5"></a>ab_sol <span class="ot">&lt;-</span> deSolve<span class="sc">::</span><span class="fu">ode</span>(</span>
<span id="cb5-6"><a href="#cb5-6"></a>    <span class="at">y =</span> ab_yinit,</span>
<span id="cb5-7"><a href="#cb5-7"></a>    <span class="at">times =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">2</span>, <span class="at">by =</span> <span class="fl">0.001</span>),</span>
<span id="cb5-8"><a href="#cb5-8"></a>    <span class="at">func =</span> ab_model,</span>
<span id="cb5-9"><a href="#cb5-9"></a>    <span class="at">parms =</span> ab_params,</span>
<span id="cb5-10"><a href="#cb5-10"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-7_ec5885002c8190b1b8ca6cc70a887939">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb6" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>ab_df <span class="ot">&lt;-</span> ab_sol <span class="sc">%&gt;%</span></span>
<span id="cb6-2"><a href="#cb6-2"></a>    <span class="co"># Convert the solution to a tibble for manipulation</span></span>
<span id="cb6-3"><a href="#cb6-3"></a>    <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb6-4"><a href="#cb6-4"></a>    <span class="co"># Create and modify columns</span></span>
<span id="cb6-5"><a href="#cb6-5"></a>    <span class="fu">mutate</span>(</span>
<span id="cb6-6"><a href="#cb6-6"></a>        <span class="co"># Convert all columns into type numeric</span></span>
<span id="cb6-7"><a href="#cb6-7"></a>        <span class="fu">across</span>(<span class="fu">everything</span>(), as.numeric),</span>
<span id="cb6-8"><a href="#cb6-8"></a>        <span class="co"># Create new columns to track pop sizes in each group</span></span>
<span id="cb6-9"><a href="#cb6-9"></a>        <span class="at">Na =</span> Sa <span class="sc">+</span> Ia <span class="sc">+</span> Ra,</span>
<span id="cb6-10"><a href="#cb6-10"></a>        <span class="at">Nb =</span> Sb <span class="sc">+</span> Ib <span class="sc">+</span> Rb</span>
<span id="cb6-11"><a href="#cb6-11"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb6-12"><a href="#cb6-12"></a>    <span class="co"># Go from a wide to long dataframe for ggplot</span></span>
<span id="cb6-13"><a href="#cb6-13"></a>    <span class="fu">pivot_longer</span>(</span>
<span id="cb6-14"><a href="#cb6-14"></a>        <span class="at">cols =</span> <span class="sc">-</span>time,</span>
<span id="cb6-15"><a href="#cb6-15"></a>        <span class="at">names_to =</span> <span class="fu">c</span>(<span class="st">"state"</span>, <span class="st">"group"</span>),</span>
<span id="cb6-16"><a href="#cb6-16"></a>        <span class="at">names_sep =</span> <span class="dv">1</span>,</span>
<span id="cb6-17"><a href="#cb6-17"></a>        <span class="at">values_to =</span> <span class="st">"value"</span></span>
<span id="cb6-18"><a href="#cb6-18"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb6-19"><a href="#cb6-19"></a>    <span class="co"># Clean pivoted columns for ordered plots</span></span>
<span id="cb6-20"><a href="#cb6-20"></a>    <span class="fu">mutate</span>(</span>
<span id="cb6-21"><a href="#cb6-21"></a>        <span class="at">state =</span> <span class="fu">factor</span>(state, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"S"</span>, <span class="st">"I"</span>, <span class="st">"R"</span>, <span class="st">"N"</span>)),</span>
<span id="cb6-22"><a href="#cb6-22"></a>        <span class="at">group =</span> <span class="fu">paste</span>(<span class="st">"Group"</span>, <span class="fu">str_to_upper</span>(group))</span>
<span id="cb6-23"><a href="#cb6-23"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell page-columns page-full" data-hash="r-session-02_cache/html/unnamed-chunk-8_38a30e09d3bd8346a435451814b266ee">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb7" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="co"># Create a vector of colors to be used throughout the ggplots</span></span>
<span id="cb7-2"><a href="#cb7-2"></a>SIRcolors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"#1f77b4"</span>, <span class="st">"#ff7f0e"</span>, <span class="st">"#FF3851"</span>, <span class="st">"#591099"</span>)</span>
<span id="cb7-3"><a href="#cb7-3"></a></span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="fu">ggplot</span>(ab_df, <span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> value, <span class="at">color =</span> state)) <span class="sc">+</span></span>
<span id="cb7-5"><a href="#cb7-5"></a>    <span class="fu">geom_line</span>(<span class="at">linewidth =</span> <span class="fl">1.5</span>) <span class="sc">+</span></span>
<span id="cb7-6"><a href="#cb7-6"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span>group, <span class="at">scales =</span> <span class="st">"free_y"</span>) <span class="sc">+</span></span>
<span id="cb7-7"><a href="#cb7-7"></a>    <span class="fu">scale_color_manual</span>(</span>
<span id="cb7-8"><a href="#cb7-8"></a>        <span class="at">values =</span> SIRcolors,</span>
<span id="cb7-9"><a href="#cb7-9"></a>        <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"Susceptible"</span>, <span class="st">"Infected"</span>, <span class="st">"Recovered"</span>, <span class="st">"Total"</span>)</span>
<span id="cb7-10"><a href="#cb7-10"></a>    ) <span class="sc">+</span></span>
<span id="cb7-11"><a href="#cb7-11"></a>    <span class="fu">labs</span>(</span>
<span id="cb7-12"><a href="#cb7-12"></a>        <span class="at">x =</span> <span class="st">"Time"</span>,</span>
<span id="cb7-13"><a href="#cb7-13"></a>        <span class="at">y =</span> <span class="st">"Number of individuals"</span>,</span>
<span id="cb7-14"><a href="#cb7-14"></a>        <span class="at">color =</span> <span class="st">"State"</span></span>
<span id="cb7-15"><a href="#cb7-15"></a>    ) <span class="sc">+</span></span>
<span id="cb7-16"><a href="#cb7-16"></a>    <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display column-body">
<p><img src="r-session-02_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" style="width:100.0%"></p>
</div>
</div>
<div class="callout-question">
<p>Despite using the same transmission rates, the epidemic in group B is much larger than in group A. Why do you think this is?</p>
</div>
<p>Now let’s plot the proportion of individuals in each state for the two groups.</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-9_c4180640934e58326f17cf0ca06456b9">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb8" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>ab_df_props <span class="ot">&lt;-</span> ab_df <span class="sc">%&gt;%</span></span>
<span id="cb8-2"><a href="#cb8-2"></a>    <span class="co"># Remove total pop count as we only want the group-specific values</span></span>
<span id="cb8-3"><a href="#cb8-3"></a>    <span class="fu">filter</span>(state <span class="sc">!=</span> <span class="st">"N"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb8-4"><a href="#cb8-4"></a>    <span class="fu">mutate</span>(</span>
<span id="cb8-5"><a href="#cb8-5"></a>        <span class="co"># Concatenate the state variable and the group letter for each row</span></span>
<span id="cb8-6"><a href="#cb8-6"></a>        <span class="at">state_group =</span> <span class="fu">paste0</span>(state, <span class="fu">str_extract_all</span>(group, <span class="st">"[^Group ]"</span>)),</span>
<span id="cb8-7"><a href="#cb8-7"></a>        <span class="co"># Factor new variable for nicer plotting</span></span>
<span id="cb8-8"><a href="#cb8-8"></a>        <span class="at">state_group =</span> <span class="fu">factor</span>(state_group, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"RA"</span>, <span class="st">"RB"</span>, <span class="st">"IA"</span>, <span class="st">"IB"</span>, <span class="st">"SA"</span>, <span class="st">"SB"</span>))</span>
<span id="cb8-9"><a href="#cb8-9"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb8-10"><a href="#cb8-10"></a>    <span class="co"># Group by time and state_group so we can calculate the relevant proportions over time</span></span>
<span id="cb8-11"><a href="#cb8-11"></a>    <span class="fu">group_by</span>(time, state_group) <span class="sc">%&gt;%</span></span>
<span id="cb8-12"><a href="#cb8-12"></a>    <span class="fu">mutate</span>(</span>
<span id="cb8-13"><a href="#cb8-13"></a>        <span class="at">prop =</span> value <span class="sc">/</span> <span class="fu">sum</span>(ab_yinit)</span>
<span id="cb8-14"><a href="#cb8-14"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb8-15"><a href="#cb8-15"></a>    <span class="fu">ungroup</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell page-columns page-full" data-hash="r-session-02_cache/html/unnamed-chunk-10_f69fa79acdad0dfa50543f66f0dab914">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb9" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="co"># Create new vectors of colors as using 6: one of each for A and J groups</span></span>
<span id="cb9-2"><a href="#cb9-2"></a>Scolors <span class="ot">&lt;-</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">3</span>, <span class="st">"Blues"</span>)[<span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">3</span>)]</span>
<span id="cb9-3"><a href="#cb9-3"></a>Icolors <span class="ot">&lt;-</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">3</span>, <span class="st">"Oranges"</span>)[<span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">3</span>)]</span>
<span id="cb9-4"><a href="#cb9-4"></a>Rcolors <span class="ot">&lt;-</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">3</span>, <span class="st">"Greens"</span>)[<span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">3</span>)]</span>
<span id="cb9-5"><a href="#cb9-5"></a></span>
<span id="cb9-6"><a href="#cb9-6"></a><span class="fu">ggplot</span>(ab_df_props, <span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> prop, <span class="at">fill =</span> state_group)) <span class="sc">+</span></span>
<span id="cb9-7"><a href="#cb9-7"></a>    <span class="fu">geom_area</span>() <span class="sc">+</span></span>
<span id="cb9-8"><a href="#cb9-8"></a>    <span class="fu">scale_fill_manual</span>(</span>
<span id="cb9-9"><a href="#cb9-9"></a>        <span class="at">values =</span> <span class="fu">c</span>(Scolors, Icolors, Rcolors),</span>
<span id="cb9-10"><a href="#cb9-10"></a>        <span class="at">limits =</span> <span class="fu">c</span>(<span class="st">"SA"</span>, <span class="st">"SB"</span>, <span class="st">"IA"</span>, <span class="st">"IB"</span>, <span class="st">"RA"</span>, <span class="st">"RB"</span>),</span>
<span id="cb9-11"><a href="#cb9-11"></a>    ) <span class="sc">+</span></span>
<span id="cb9-12"><a href="#cb9-12"></a>    <span class="fu">labs</span>(</span>
<span id="cb9-13"><a href="#cb9-13"></a>        <span class="at">x =</span> <span class="st">"Time"</span>,</span>
<span id="cb9-14"><a href="#cb9-14"></a>        <span class="at">y =</span> <span class="st">"Proportion of individuals"</span>,</span>
<span id="cb9-15"><a href="#cb9-15"></a>        <span class="at">fill =</span> <span class="st">"State"</span></span>
<span id="cb9-16"><a href="#cb9-16"></a>    ) <span class="sc">+</span></span>
<span id="cb9-17"><a href="#cb9-17"></a>    <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display column-body">
<p><img src="r-session-02_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" style="width:100.0%"></p>
</div>
</div>
</section><section id="a-model-with-2-age-classes" class="level2 page-columns page-full" data-number="7.3"><h2 data-number="7.3" class="anchored" data-anchor-id="a-model-with-2-age-classes">
<span class="header-section-number">7.3</span> A Model With 2 Age Classes</h2>
<p>Note that age is a special kind of heterogeneity in an epidemic model because individuals necessarily move from one class (younger) to another class (older) in a directional fashion that is independent of the infection and recovery process.</p>
<!-- Today's first lecture showed how force of infection can vary with age. -->
<!-- What sort of mechanisms might give rise to these effects? -->
<!-- Here we'll see to what extent we can infer these mechanisms on the basis of age-specific incidence and seroprevalence data. -->
<p>We’ll start by introducing age into the model above. So now <span class="math inline">\(a\)</span> becomes juveniles and <span class="math inline">\(b\)</span> becomes adults. And, independent of the disease process, juveniles (of any category) age into adults. Additionally, new juveniles are added through births (always first susceptible) and old individuals are lost to death.</p>
<div class="cell page-columns page-full" data-hash="r-session-02_cache/html/unnamed-chunk-11_9b9d42994e2693d0bb3d314331c39ab6">
<div class="cell-output-display column-body">
<p><img src="r-session-02_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" style="width:100.0%"></p>
</div>
</div>
<p>We can do this very simply using the same ingredients that go into the basic SIR model. In that model, the waiting times in the S and I classes are exponential. Let’s assume the same thing about the aging process. We’ll also add in births into the juvenile susceptible class and deaths from the adult classes.</p>
<p><span class="math display">\[
  \begin{aligned}
    \frac{\dd{S_J}}{\dd{t}} &amp;= B -\lambda_J\,S_J \phantom{- \gamma\,I_A} -\alpha\,S_J \phantom{-\mu\,S_A}\\
    \frac{\dd{S_A}}{\dd{t}} &amp;= \phantom{B} - \lambda_A\,S_A \phantom{- \gamma\,I_A} +\alpha\,S_J -\mu\,S_A\\
    \frac{\dd{I_J}}{\dd{t}} &amp;= \phantom{B} +\lambda_J\,S_J - \gamma\,I_J -\alpha\,I_J \phantom{-\mu\,S_A}\\
    \frac{\dd{I_A}}{\dd{t}} &amp;= \phantom{B} +\lambda_A\,S_A - \gamma\,I_A + \alpha\,I_J - \mu\,I_A\\
    \frac{\dd{R_J}}{\dd{t}} &amp;= \phantom{B - \lambda_J\,S_A} + \gamma\,I_J - \alpha\,R_J \phantom{- \mu\,S_A}\\
    \frac{\dd{R_A}}{\dd{t}} &amp;= \phantom{B - \lambda_J\,S_A} + \gamma\,I_A + \alpha\,R_J -\mu\,R_A\\
  \end{aligned}
\]</span></p>
<p>Now, let’s simulate this model, under the same assumptions about transmission rates as above.</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-12_5fa09780cd2bddf13e2224e3f73088ef">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb10" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="co"># define the parameters for the demographic model</span></span>
<span id="cb10-2"><a href="#cb10-2"></a>demog_params <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb10-3"><a href="#cb10-3"></a>    <span class="at">beta_within =</span> <span class="fl">0.004</span>,</span>
<span id="cb10-4"><a href="#cb10-4"></a>    <span class="at">beta_between =</span> <span class="fl">0.002</span>,</span>
<span id="cb10-5"><a href="#cb10-5"></a>    <span class="at">recovery =</span> <span class="dv">10</span>,</span>
<span id="cb10-6"><a href="#cb10-6"></a>    <span class="at">births =</span> <span class="dv">100</span>,</span>
<span id="cb10-7"><a href="#cb10-7"></a>    <span class="co"># Width of age bands in years</span></span>
<span id="cb10-8"><a href="#cb10-8"></a>    <span class="at">age_band_j =</span> <span class="dv">20</span>,</span>
<span id="cb10-9"><a href="#cb10-9"></a>    <span class="at">age_band_a =</span> <span class="dv">60</span></span>
<span id="cb10-10"><a href="#cb10-10"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-13_6c3ac27cf828a890aae43603666d106e">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb11" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>demog_model <span class="ot">&lt;-</span> <span class="cf">function</span> (t, x, p, ...) {</span>
<span id="cb11-2"><a href="#cb11-2"></a>    <span class="co"># Unpack states</span></span>
<span id="cb11-3"><a href="#cb11-3"></a>    Sj <span class="ot">&lt;-</span> x[<span class="st">"Sj"</span>]</span>
<span id="cb11-4"><a href="#cb11-4"></a>    Sa <span class="ot">&lt;-</span> x[<span class="st">"Sa"</span>]</span>
<span id="cb11-5"><a href="#cb11-5"></a>    Ij <span class="ot">&lt;-</span> x[<span class="st">"Ij"</span>]</span>
<span id="cb11-6"><a href="#cb11-6"></a>    Ia <span class="ot">&lt;-</span> x[<span class="st">"Ia"</span>]</span>
<span id="cb11-7"><a href="#cb11-7"></a>    Rj <span class="ot">&lt;-</span> x[<span class="st">"Rj"</span>]</span>
<span id="cb11-8"><a href="#cb11-8"></a>    Ra <span class="ot">&lt;-</span> x[<span class="st">"Ra"</span>]</span>
<span id="cb11-9"><a href="#cb11-9"></a></span>
<span id="cb11-10"><a href="#cb11-10"></a>    <span class="co"># Unpack parameters from vector</span></span>
<span id="cb11-11"><a href="#cb11-11"></a>    beta_within <span class="ot">&lt;-</span> p[<span class="st">"beta_within"</span>]</span>
<span id="cb11-12"><a href="#cb11-12"></a>    beta_between <span class="ot">&lt;-</span> p[<span class="st">"beta_between"</span>]</span>
<span id="cb11-13"><a href="#cb11-13"></a>    recovery <span class="ot">&lt;-</span> p[<span class="st">"recovery"</span>]</span>
<span id="cb11-14"><a href="#cb11-14"></a>    births <span class="ot">&lt;-</span> p[<span class="st">"births"</span>]</span>
<span id="cb11-15"><a href="#cb11-15"></a>    <span class="co"># Calculate rate of aging from each age group</span></span>
<span id="cb11-16"><a href="#cb11-16"></a>    aging_j <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> p[<span class="st">"age_band_j"</span>]</span>
<span id="cb11-17"><a href="#cb11-17"></a>    aging_a <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> p[<span class="st">"age_band_a"</span>]</span>
<span id="cb11-18"><a href="#cb11-18"></a></span>
<span id="cb11-19"><a href="#cb11-19"></a>    <span class="co"># juv. force of infection</span></span>
<span id="cb11-20"><a href="#cb11-20"></a>    lambda_j <span class="ot">&lt;-</span> beta_within <span class="sc">*</span> Ij <span class="sc">+</span> beta_between <span class="sc">*</span> Ia</span>
<span id="cb11-21"><a href="#cb11-21"></a></span>
<span id="cb11-22"><a href="#cb11-22"></a>    <span class="co"># adult. force of infection</span></span>
<span id="cb11-23"><a href="#cb11-23"></a>    lambda_a <span class="ot">&lt;-</span> beta_within <span class="sc">*</span> Ia <span class="sc">+</span> beta_between <span class="sc">*</span> Ij</span>
<span id="cb11-24"><a href="#cb11-24"></a></span>
<span id="cb11-25"><a href="#cb11-25"></a>    <span class="co"># Calculate the ODEs</span></span>
<span id="cb11-26"><a href="#cb11-26"></a>    dSjdt <span class="ot">&lt;-</span> births <span class="sc">-</span> (lambda_j <span class="sc">*</span> Sj) <span class="sc">-</span> (aging_j <span class="sc">*</span> Sj)</span>
<span id="cb11-27"><a href="#cb11-27"></a>    dSadt <span class="ot">&lt;-</span> <span class="sc">-</span>(lambda_a <span class="sc">*</span> Sa) <span class="sc">+</span> (aging_j <span class="sc">*</span> Sj) <span class="sc">-</span> (aging_a <span class="sc">*</span> Sa)</span>
<span id="cb11-28"><a href="#cb11-28"></a>    dIjdt <span class="ot">&lt;-</span> (lambda_j <span class="sc">*</span> Sj) <span class="sc">-</span> (recovery <span class="sc">*</span> Ij) <span class="sc">-</span> (aging_j <span class="sc">*</span> Ij)</span>
<span id="cb11-29"><a href="#cb11-29"></a>    dIadt <span class="ot">&lt;-</span> (lambda_a <span class="sc">*</span> Sa) <span class="sc">-</span> (recovery <span class="sc">*</span> Ia) <span class="sc">+</span> (aging_j <span class="sc">*</span> Ij) <span class="sc">-</span> (aging_a <span class="sc">*</span> Ia)</span>
<span id="cb11-30"><a href="#cb11-30"></a>    dRjdt <span class="ot">&lt;-</span> (recovery <span class="sc">*</span> Ij) <span class="sc">-</span> (aging_j <span class="sc">*</span> Rj)</span>
<span id="cb11-31"><a href="#cb11-31"></a>    dRadt <span class="ot">&lt;-</span> (recovery <span class="sc">*</span> Ia) <span class="sc">+</span> (aging_j <span class="sc">*</span> Rj) <span class="sc">-</span> (aging_a <span class="sc">*</span> Ra)</span>
<span id="cb11-32"><a href="#cb11-32"></a></span>
<span id="cb11-33"><a href="#cb11-33"></a>    <span class="co"># Return the ODEs</span></span>
<span id="cb11-34"><a href="#cb11-34"></a>    <span class="fu">list</span>(<span class="fu">c</span>(</span>
<span id="cb11-35"><a href="#cb11-35"></a>        dSjdt,</span>
<span id="cb11-36"><a href="#cb11-36"></a>        dSadt,</span>
<span id="cb11-37"><a href="#cb11-37"></a>        dIjdt,</span>
<span id="cb11-38"><a href="#cb11-38"></a>        dIadt,</span>
<span id="cb11-39"><a href="#cb11-39"></a>        dRjdt,</span>
<span id="cb11-40"><a href="#cb11-40"></a>        dRadt</span>
<span id="cb11-41"><a href="#cb11-41"></a>    ))</span>
<span id="cb11-42"><a href="#cb11-42"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Note that in this function, <span class="math inline">\(\mu=\)</span> <code>aging_a</code> <span class="math inline">\(=\)</span> <code>1 / p["age_band_a"]</code>, i.e., death, is just like another age class.</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-14_22ff9929e73a38091aae54343c1a4f0f">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb12" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="co"># initial conditions</span></span>
<span id="cb12-2"><a href="#cb12-2"></a>demog_yinit <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="at">Sj =</span> <span class="dv">2000</span>, <span class="at">Sa =</span> <span class="dv">3000</span>, <span class="at">Ij =</span> <span class="dv">0</span>, <span class="at">Ia =</span> <span class="dv">1</span>, <span class="at">Rj =</span> <span class="dv">0</span>, <span class="at">Ra =</span> <span class="dv">1000</span>)</span>
<span id="cb12-3"><a href="#cb12-3"></a></span>
<span id="cb12-4"><a href="#cb12-4"></a><span class="co"># Solve the demographic model</span></span>
<span id="cb12-5"><a href="#cb12-5"></a>demog_sol <span class="ot">&lt;-</span> deSolve<span class="sc">::</span><span class="fu">ode</span>(</span>
<span id="cb12-6"><a href="#cb12-6"></a>    <span class="at">y =</span> demog_yinit,</span>
<span id="cb12-7"><a href="#cb12-7"></a>    <span class="at">times =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">200</span>, <span class="at">by =</span> <span class="fl">0.1</span>),</span>
<span id="cb12-8"><a href="#cb12-8"></a>    <span class="at">func =</span> demog_model,</span>
<span id="cb12-9"><a href="#cb12-9"></a>    <span class="at">parms =</span> demog_params</span>
<span id="cb12-10"><a href="#cb12-10"></a>)</span>
<span id="cb12-11"><a href="#cb12-11"></a></span>
<span id="cb12-12"><a href="#cb12-12"></a>demog_df <span class="ot">&lt;-</span> demog_sol <span class="sc">%&gt;%</span></span>
<span id="cb12-13"><a href="#cb12-13"></a>    <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb12-14"><a href="#cb12-14"></a>    <span class="fu">mutate</span>(</span>
<span id="cb12-15"><a href="#cb12-15"></a>        <span class="fu">across</span>(<span class="fu">everything</span>(), as.numeric),</span>
<span id="cb12-16"><a href="#cb12-16"></a>        <span class="at">Nj =</span> Sj <span class="sc">+</span> Ij <span class="sc">+</span> Rj,</span>
<span id="cb12-17"><a href="#cb12-17"></a>        <span class="at">Na =</span> Sa <span class="sc">+</span> Ia <span class="sc">+</span> Ra,</span>
<span id="cb12-18"><a href="#cb12-18"></a>        <span class="co"># Calculate total population as need for proportional area plots</span></span>
<span id="cb12-19"><a href="#cb12-19"></a>        <span class="at">N =</span> Nj <span class="sc">+</span> Na</span>
<span id="cb12-20"><a href="#cb12-20"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb12-21"><a href="#cb12-21"></a>    <span class="fu">pivot_longer</span>(</span>
<span id="cb12-22"><a href="#cb12-22"></a>        <span class="at">cols =</span> <span class="sc">-</span><span class="fu">c</span>(time, N),</span>
<span id="cb12-23"><a href="#cb12-23"></a>        <span class="at">names_to =</span> <span class="fu">c</span>(<span class="st">"state"</span>, <span class="st">"group"</span>),</span>
<span id="cb12-24"><a href="#cb12-24"></a>        <span class="at">names_sep =</span> <span class="dv">1</span>,</span>
<span id="cb12-25"><a href="#cb12-25"></a>        <span class="at">values_to =</span> <span class="st">"value"</span></span>
<span id="cb12-26"><a href="#cb12-26"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb12-27"><a href="#cb12-27"></a>    <span class="fu">mutate</span>(</span>
<span id="cb12-28"><a href="#cb12-28"></a>        <span class="at">state =</span> <span class="fu">factor</span>(state, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"S"</span>, <span class="st">"I"</span>, <span class="st">"R"</span>, <span class="st">"N"</span>)),</span>
<span id="cb12-29"><a href="#cb12-29"></a>        <span class="at">group =</span> <span class="fu">paste</span>(<span class="st">"Group"</span>, <span class="fu">str_to_upper</span>(group))</span>
<span id="cb12-30"><a href="#cb12-30"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="exercise-1-use-this-code-to-plot-the-number-of-susceptible-infected-and-recovered-individuals-over-time" class="level3 exercise" data-number="7.3.1"><h3 data-number="7.3.1" class="anchored" data-anchor-id="exercise-1-use-this-code-to-plot-the-number-of-susceptible-infected-and-recovered-individuals-over-time">
<span class="header-section-number">7.3.1</span> Exercise 1: Use this code to plot the number of susceptible, infected, and recovered individuals over time</h3>
</section><div class="cell page-columns page-full" data-hash="r-session-02_cache/html/unnamed-chunk-15_3b1c5cd3b3af797b5d36678d3f7cad04">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb13" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="fu">ggplot</span>(demog_df, <span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> value, <span class="at">color =</span> state)) <span class="sc">+</span></span>
<span id="cb13-2"><a href="#cb13-2"></a>    <span class="fu">geom_line</span>(<span class="at">linewidth =</span> <span class="fl">1.5</span>) <span class="sc">+</span></span>
<span id="cb13-3"><a href="#cb13-3"></a>    <span class="fu">facet_wrap</span>(</span>
<span id="cb13-4"><a href="#cb13-4"></a>        <span class="sc">~</span>group, <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">scales =</span> <span class="st">"free_y"</span>,</span>
<span id="cb13-5"><a href="#cb13-5"></a>        <span class="at">labeller =</span> <span class="fu">as_labeller</span>(<span class="fu">c</span>(</span>
<span id="cb13-6"><a href="#cb13-6"></a>            <span class="st">`</span><span class="at">Group A</span><span class="st">`</span> <span class="ot">=</span> <span class="st">"Adults"</span>,</span>
<span id="cb13-7"><a href="#cb13-7"></a>            <span class="st">`</span><span class="at">Group J</span><span class="st">`</span> <span class="ot">=</span> <span class="st">"Juveniles"</span></span>
<span id="cb13-8"><a href="#cb13-8"></a>        ))</span>
<span id="cb13-9"><a href="#cb13-9"></a>    ) <span class="sc">+</span></span>
<span id="cb13-10"><a href="#cb13-10"></a>    <span class="fu">scale_color_manual</span>(</span>
<span id="cb13-11"><a href="#cb13-11"></a>        <span class="at">values =</span> SIRcolors,</span>
<span id="cb13-12"><a href="#cb13-12"></a>        <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"Susceptible"</span>, <span class="st">"Infected"</span>, <span class="st">"Recovered"</span>, <span class="st">"Total"</span>)</span>
<span id="cb13-13"><a href="#cb13-13"></a>    ) <span class="sc">+</span></span>
<span id="cb13-14"><a href="#cb13-14"></a>    <span class="fu">labs</span>(</span>
<span id="cb13-15"><a href="#cb13-15"></a>        <span class="at">x =</span> <span class="st">"Time"</span>,</span>
<span id="cb13-16"><a href="#cb13-16"></a>        <span class="at">y =</span> <span class="st">"Number of individuals"</span>,</span>
<span id="cb13-17"><a href="#cb13-17"></a>        <span class="at">color =</span> <span class="st">"State"</span></span>
<span id="cb13-18"><a href="#cb13-18"></a>    ) <span class="sc">+</span></span>
<span id="cb13-19"><a href="#cb13-19"></a>    <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display column-body">
<p><img src="r-session-02_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" style="width:100.0%"></p>
</div>
</div>
<p>Note that now that births are replenishing susceptibles, infection persists. The results of the above are plotted here:</p>
<p>Now let’s plot the proportion of individuals in each state for the two groups.</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-16_1be7ba4a64bbb73ba96cd855fc48375f">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb14" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="co"># Calculate the proportions in each state and group at each time point</span></span>
<span id="cb14-2"><a href="#cb14-2"></a>demog_df_props <span class="ot">&lt;-</span> demog_df <span class="sc">%&gt;%</span></span>
<span id="cb14-3"><a href="#cb14-3"></a>    <span class="fu">filter</span>(state <span class="sc">!=</span> <span class="st">"N"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb14-4"><a href="#cb14-4"></a>    <span class="fu">mutate</span>(</span>
<span id="cb14-5"><a href="#cb14-5"></a>        <span class="at">state_group =</span> <span class="fu">paste0</span>(state, <span class="fu">str_extract_all</span>(group, <span class="st">"[^Group ]"</span>)),</span>
<span id="cb14-6"><a href="#cb14-6"></a>        <span class="at">state_group =</span> <span class="fu">factor</span>(</span>
<span id="cb14-7"><a href="#cb14-7"></a>          state_group,</span>
<span id="cb14-8"><a href="#cb14-8"></a>          <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"RJ"</span>, <span class="st">"RA"</span>, <span class="st">"IJ"</span>, <span class="st">"IA"</span>, <span class="st">"SJ"</span>, <span class="st">"SA"</span>)</span>
<span id="cb14-9"><a href="#cb14-9"></a>        )</span>
<span id="cb14-10"><a href="#cb14-10"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb14-11"><a href="#cb14-11"></a>    <span class="fu">group_by</span>(time, state_group) <span class="sc">%&gt;%</span></span>
<span id="cb14-12"><a href="#cb14-12"></a>    <span class="fu">mutate</span>(</span>
<span id="cb14-13"><a href="#cb14-13"></a>        <span class="co"># Calculate the proportion of the total population, not the group pop</span></span>
<span id="cb14-14"><a href="#cb14-14"></a>        <span class="at">prop =</span> value <span class="sc">/</span> N</span>
<span id="cb14-15"><a href="#cb14-15"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb14-16"><a href="#cb14-16"></a>    <span class="fu">ungroup</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell page-columns page-full" data-hash="r-session-02_cache/html/unnamed-chunk-17_f2e4e832a1feb17d023409b3475b4b0f">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb15" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a><span class="fu">ggplot</span>(demog_df_props, <span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> prop, <span class="at">fill =</span> state_group)) <span class="sc">+</span></span>
<span id="cb15-2"><a href="#cb15-2"></a>    <span class="fu">geom_area</span>() <span class="sc">+</span></span>
<span id="cb15-3"><a href="#cb15-3"></a>    <span class="fu">scale_fill_manual</span>(</span>
<span id="cb15-4"><a href="#cb15-4"></a>        <span class="at">values =</span> <span class="fu">c</span>(Scolors, Icolors, Rcolors),</span>
<span id="cb15-5"><a href="#cb15-5"></a>        <span class="at">limits =</span> <span class="fu">c</span>(<span class="st">"SJ"</span>, <span class="st">"SA"</span>, <span class="st">"IJ"</span>, <span class="st">"IA"</span>, <span class="st">"RJ"</span>, <span class="st">"RA"</span>)</span>
<span id="cb15-6"><a href="#cb15-6"></a>    ) <span class="sc">+</span></span>
<span id="cb15-7"><a href="#cb15-7"></a>    <span class="fu">labs</span>(</span>
<span id="cb15-8"><a href="#cb15-8"></a>        <span class="at">x =</span> <span class="st">"Time"</span>,</span>
<span id="cb15-9"><a href="#cb15-9"></a>        <span class="at">y =</span> <span class="st">"Proportion of individuals"</span>,</span>
<span id="cb15-10"><a href="#cb15-10"></a>        <span class="at">fill =</span> <span class="st">"State"</span></span>
<span id="cb15-11"><a href="#cb15-11"></a>    ) <span class="sc">+</span></span>
<span id="cb15-12"><a href="#cb15-12"></a>    <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display column-body">
<p><img src="r-session-02_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" style="width:100.0%"></p>
</div>
</div>
<p>Now let’s plot the equilibrium seroprevalence for each age group.</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-18_e95c8730515b7fc5ef4219ff2a85f6e6">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb16" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a><span class="co"># Select the last row (time point) of the data frame</span></span>
<span id="cb16-2"><a href="#cb16-2"></a>demog_equil_seroprev <span class="ot">&lt;-</span> <span class="fu">tail</span>(demog_df) <span class="sc">%&gt;%</span></span>
<span id="cb16-3"><a href="#cb16-3"></a>    <span class="fu">mutate</span>(</span>
<span id="cb16-4"><a href="#cb16-4"></a>        <span class="co"># Calculate the proportion of individuals in each state and age group</span></span>
<span id="cb16-5"><a href="#cb16-5"></a>        <span class="at">prop =</span> value <span class="sc">/</span> <span class="fu">sum</span>(value),</span>
<span id="cb16-6"><a href="#cb16-6"></a>        <span class="co"># Relabel groups for plots</span></span>
<span id="cb16-7"><a href="#cb16-7"></a>        <span class="at">group =</span> <span class="fu">case_when</span>(group <span class="sc">==</span> <span class="st">"Group J"</span> <span class="sc">~</span> <span class="st">"Juveniles"</span>, <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"Adults"</span>),</span>
<span id="cb16-8"><a href="#cb16-8"></a>        <span class="at">group =</span> <span class="fu">factor</span>(group, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Juveniles"</span>, <span class="st">"Adults"</span>)),</span>
<span id="cb16-9"><a href="#cb16-9"></a>        <span class="at">.by =</span> group</span>
<span id="cb16-10"><a href="#cb16-10"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb16-11"><a href="#cb16-11"></a>    <span class="fu">filter</span>(state <span class="sc">==</span> <span class="st">"R"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell page-columns page-full" data-hash="r-session-02_cache/html/unnamed-chunk-19_e461ebc225729dcd90f54c7d8ff58c5f">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb17" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a><span class="co"># Create vector of colors to distinguish between age groups</span></span>
<span id="cb17-2"><a href="#cb17-2"></a>age_group_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"#2980B9"</span>, <span class="st">"#154360"</span>)</span>
<span id="cb17-3"><a href="#cb17-3"></a></span>
<span id="cb17-4"><a href="#cb17-4"></a><span class="fu">ggplot</span>(demog_equil_seroprev, <span class="fu">aes</span>(<span class="at">x =</span> group, <span class="at">y =</span> prop, <span class="at">fill =</span> group)) <span class="sc">+</span></span>
<span id="cb17-5"><a href="#cb17-5"></a>    <span class="fu">geom_col</span>(<span class="at">position =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb17-6"><a href="#cb17-6"></a>    <span class="fu">scale_fill_manual</span>(</span>
<span id="cb17-7"><a href="#cb17-7"></a>        <span class="at">values =</span> age_group_colors</span>
<span id="cb17-8"><a href="#cb17-8"></a>    ) <span class="sc">+</span></span>
<span id="cb17-9"><a href="#cb17-9"></a>    <span class="fu">labs</span>(</span>
<span id="cb17-10"><a href="#cb17-10"></a>        <span class="at">x =</span> <span class="st">"Age group"</span>,</span>
<span id="cb17-11"><a href="#cb17-11"></a>        <span class="at">y =</span> <span class="st">"Equilibrium seroprevalence"</span>,</span>
<span id="cb17-12"><a href="#cb17-12"></a>        <span class="at">fill =</span> <span class="st">"Age group"</span></span>
<span id="cb17-13"><a href="#cb17-13"></a>    ) <span class="sc">+</span></span>
<span id="cb17-14"><a href="#cb17-14"></a>    <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display column-body">
<p><img src="r-session-02_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" style="width:100.0%"></p>
</div>
</div>
<p>One thing we are often interested in is the <span class="math inline">\(R_0\)</span> of a system. The details are beyond the scope of this workshop and are not required to complete the exercises in this worksheet, but we have outlined them in <a href="#sec-simple-ngm"><span>Section&nbsp;7.6.1.1</span></a>, particularly in <a href="#eq-simple-ngm">Equation&nbsp;<span>7.3</span></a>, at the end of this page.</p>
<p>In our system, <span class="math inline">\(R_0 =\)</span> 2.66.</p>
</section><section id="getting-more-realistic-adding-more-age-classes" class="level2 page-columns page-full" data-number="7.4"><h2 data-number="7.4" class="anchored" data-anchor-id="getting-more-realistic-adding-more-age-classes">
<span class="header-section-number">7.4</span> Getting more realistic: adding more age classes</h2>
<p>In the models above, the aging process follows an exponential distribution, which means that whether an individual is 1~year old or 10 years old, the chance of them becoming an adult is the same! To improve on this, we can assume that the time a juvenile must wait before becoming an adult follows a gamma distribution. This is equivalent to saying that the waiting time is a sum of some number of exponential distributions. This suggests that we can achieve such a distribution by adding age classes to the model, so that becoming an adult means passing through some number of stages. We’ll use 30 age classes, and since they don’t have to be of equal duration, we’ll assume that they’re not. Specifically, we’ll have 20 1-yr age classes to take us up to adulthood and break adults into 10 age classes of 5~yr duration each. The last age class covers age 66-80.</p>
<p>Now, when we had just two age classes, we could write out each of the equations easily enough, but now that we’re going to have 30, we’ll need to be more systematic. In particular, we’ll need to think of <span class="math inline">\(\beta\)</span> as a matrix of transmission rates. Let’s see how to define such a matrix in <code>R</code>. So that we don’t change too many things all at once, let’s keep the same contact structure as in the juvenile-adult model.</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-21_75335ed45d69925a39a8078fddd98b7b">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb18" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="co"># Set up the parameters for model that incorporates a more realistic age matrix</span></span>
<span id="cb18-2"><a href="#cb18-2"></a>ages_params <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb18-3"><a href="#cb18-3"></a>    <span class="at">beta_j =</span> <span class="fl">0.02</span>,</span>
<span id="cb18-4"><a href="#cb18-4"></a>    <span class="at">beta_a =</span> <span class="fl">0.01</span>,</span>
<span id="cb18-5"><a href="#cb18-5"></a>    <span class="at">beta_aj =</span> <span class="fl">0.01</span> <span class="sc">/</span> <span class="dv">2</span>,</span>
<span id="cb18-6"><a href="#cb18-6"></a>    <span class="at">recovery =</span> <span class="dv">10</span>,</span>
<span id="cb18-7"><a href="#cb18-7"></a>    <span class="at">births =</span> <span class="dv">100</span></span>
<span id="cb18-8"><a href="#cb18-8"></a>)</span>
<span id="cb18-9"><a href="#cb18-9"></a></span>
<span id="cb18-10"><a href="#cb18-10"></a><span class="co"># Create a vector of ages</span></span>
<span id="cb18-11"><a href="#cb18-11"></a>ages <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">seq</span>(<span class="dv">1</span>, <span class="dv">20</span>, <span class="at">by =</span> <span class="dv">1</span>), <span class="fu">seq</span>(<span class="dv">25</span>, <span class="dv">65</span>, <span class="at">by =</span> <span class="dv">5</span>), <span class="dv">80</span>)</span>
<span id="cb18-12"><a href="#cb18-12"></a></span>
<span id="cb18-13"><a href="#cb18-13"></a><span class="co"># Calculate the widths of the age bands</span></span>
<span id="cb18-14"><a href="#cb18-14"></a>da_ages <span class="ot">&lt;-</span> <span class="fu">diff</span>(<span class="fu">c</span>(<span class="dv">0</span>, ages))                  <span class="co"># widths of age classes</span></span>
<span id="cb18-15"><a href="#cb18-15"></a></span>
<span id="cb18-16"><a href="#cb18-16"></a><span class="co"># set up a matrix of contact rates between classes -- more contact within juveniles and adults than between</span></span>
<span id="cb18-17"><a href="#cb18-17"></a>ages_beta_mat <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">nrow =</span> <span class="dv">30</span>, <span class="at">ncol =</span> <span class="dv">30</span>)</span>
<span id="cb18-18"><a href="#cb18-18"></a></span>
<span id="cb18-19"><a href="#cb18-19"></a><span class="co"># transmission rate for juveniles</span></span>
<span id="cb18-20"><a href="#cb18-20"></a>ages_beta_mat[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>] <span class="ot">&lt;-</span> ages_params[<span class="st">"beta_j"</span>]</span>
<span id="cb18-21"><a href="#cb18-21"></a></span>
<span id="cb18-22"><a href="#cb18-22"></a><span class="co"># transmission rate for adults</span></span>
<span id="cb18-23"><a href="#cb18-23"></a>ages_beta_mat[<span class="dv">21</span><span class="sc">:</span><span class="dv">30</span>, <span class="dv">21</span><span class="sc">:</span><span class="dv">30</span>] <span class="ot">&lt;-</span> ages_params[<span class="st">"beta_a"</span>]</span>
<span id="cb18-24"><a href="#cb18-24"></a></span>
<span id="cb18-25"><a href="#cb18-25"></a><span class="co"># lower transmission rate between juveniles and adults</span></span>
<span id="cb18-26"><a href="#cb18-26"></a>ages_beta_mat[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="dv">21</span><span class="sc">:</span><span class="dv">30</span>] <span class="ot">&lt;-</span> ages_params[<span class="st">"beta_aj"</span>]</span>
<span id="cb18-27"><a href="#cb18-27"></a></span>
<span id="cb18-28"><a href="#cb18-28"></a><span class="co"># lower transmission rate between juveniles and adults</span></span>
<span id="cb18-29"><a href="#cb18-29"></a>ages_beta_mat[<span class="dv">21</span><span class="sc">:</span><span class="dv">30</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>] <span class="ot">&lt;-</span> ages_params[<span class="st">"beta_aj"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p><a id="fig-beta-mat"></a></p>
<div class="cell page-columns page-full" data-hash="r-session-02_cache/html/unnamed-chunk-22_95ed5e80d8ebecbcf8b2e46ff36f2c08">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb19" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>ages_beta_mat <span class="sc">%&gt;%</span></span>
<span id="cb19-2"><a href="#cb19-2"></a>    <span class="co"># Turn into a data.frame so we can use ggplot()</span></span>
<span id="cb19-3"><a href="#cb19-3"></a>    <span class="fu">as.data.frame.table</span>() <span class="sc">%&gt;%</span></span>
<span id="cb19-4"><a href="#cb19-4"></a>    <span class="fu">mutate</span>(</span>
<span id="cb19-5"><a href="#cb19-5"></a>        <span class="at">age_contactor =</span> <span class="fu">rep</span>(ages, <span class="dv">30</span>),</span>
<span id="cb19-6"><a href="#cb19-6"></a>        <span class="co"># Repeat each age in ages vector 30 times, and place in a list, then unlist </span></span>
<span id="cb19-7"><a href="#cb19-7"></a>        <span class="co"># to be stacked into a single dataframe column</span></span>
<span id="cb19-8"><a href="#cb19-8"></a>        <span class="at">age_contactee =</span> <span class="fu">unlist</span>(purrr<span class="sc">::</span><span class="fu">map</span>(<span class="at">.x =</span> ages, <span class="sc">~</span><span class="fu">rep</span>(.x, <span class="dv">30</span>))),</span>
<span id="cb19-9"><a href="#cb19-9"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb19-10"><a href="#cb19-10"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> age_contactor, <span class="at">y =</span> age_contactee, <span class="at">z =</span> Freq)) <span class="sc">+</span></span>
<span id="cb19-11"><a href="#cb19-11"></a>    <span class="fu">geom_contour_filled</span>(<span class="at">bins =</span> <span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb19-12"><a href="#cb19-12"></a>    <span class="fu">scale_fill_brewer</span>(</span>
<span id="cb19-13"><a href="#cb19-13"></a>        <span class="at">palette =</span> <span class="st">"Reds"</span>,</span>
<span id="cb19-14"><a href="#cb19-14"></a>        <span class="co"># Don't drop unused bins as useful for ensuring extremities properly binned</span></span>
<span id="cb19-15"><a href="#cb19-15"></a>        <span class="at">drop =</span> <span class="cn">FALSE</span></span>
<span id="cb19-16"><a href="#cb19-16"></a>    ) <span class="sc">+</span></span>
<span id="cb19-17"><a href="#cb19-17"></a>    <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Age of Contactor"</span>, <span class="at">y =</span> <span class="st">"Age of Contactee"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display column-body">
<p><img src="r-session-02_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid" style="width:100.0%"></p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<p>You could also use the <code><a href="https://rdrr.io/r/graphics/filled.contour.html">filled.contour()</a></code> base-R function to plot the beta matrix without needing to do any dataframe modifications, but it doesn’t look as nice …</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-23_f01ef42304be36f9efa7bcf0814ab531">
<details><summary>Code</summary><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a><span class="fu">filled.contour</span>(</span>
<span id="cb20-2"><a href="#cb20-2"></a>    ages, ages, ages_beta_mat,</span>
<span id="cb20-3"><a href="#cb20-3"></a>    <span class="at">plot.title =</span> <span class="fu">title</span>(</span>
<span id="cb20-4"><a href="#cb20-4"></a>        <span class="at">xlab =</span> <span class="st">"Age of Contactor"</span>,</span>
<span id="cb20-5"><a href="#cb20-5"></a>        <span class="at">ylab =</span> <span class="st">"Age of Contactee"</span></span>
<span id="cb20-6"><a href="#cb20-6"></a>    )</span>
<span id="cb20-7"><a href="#cb20-7"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display">
<p><img src="r-session-02_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
</div>
</div>
<p>We’ll assume that, at the time of introduction, all children are susceptible, as are adults over 45, but that individuals aged 20–45 have seen the pathogen before and are immune. The vector <code>yinit</code> expresses these initial conditions.</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-24_ddaf2fb8b14273b649ccfb4f31c3d61a">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb21" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="co"># Create a long vector of initial states with only one</span></span>
<span id="cb21-2"><a href="#cb21-2"></a><span class="co"># initial infection in age 50</span></span>
<span id="cb21-3"><a href="#cb21-3"></a>demog_yinit_ages <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb21-4"><a href="#cb21-4"></a>    <span class="at">S =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">100</span>, <span class="dv">20</span>), <span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">5</span>), <span class="fu">rep</span>(<span class="dv">200</span>, <span class="dv">5</span>)),</span>
<span id="cb21-5"><a href="#cb21-5"></a>    <span class="at">I =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">25</span>), <span class="dv">1</span>, <span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">4</span>)),</span>
<span id="cb21-6"><a href="#cb21-6"></a>    <span class="at">R =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">20</span>), <span class="fu">rep</span>(<span class="dv">1000</span>, <span class="dv">5</span>), <span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">5</span>))</span>
<span id="cb21-7"><a href="#cb21-7"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Note that we’re starting out with 1 infected individual in the 26th age class (age 50).</p>
<p>The codes that follow will be a bit easier to follow if we introduce some indexes that will allow us to pick out certain bits of the <code>yinit</code> vector.</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-25_c7577af972628b50f6d2917e2730d7f0">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb22" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="co"># Create vectors of indices relating to each state (ordered S1-80, I1-80, R1-80)</span></span>
<span id="cb22-2"><a href="#cb22-2"></a>sindex <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span></span>
<span id="cb22-3"><a href="#cb22-3"></a>iindex <span class="ot">&lt;-</span> <span class="dv">31</span><span class="sc">:</span><span class="dv">60</span></span>
<span id="cb22-4"><a href="#cb22-4"></a>rindex <span class="ot">&lt;-</span> <span class="dv">61</span><span class="sc">:</span><span class="dv">90</span></span>
<span id="cb22-5"><a href="#cb22-5"></a><span class="co"># Create vectors of indices relating to age group</span></span>
<span id="cb22-6"><a href="#cb22-6"></a>juvies <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span></span>
<span id="cb22-7"><a href="#cb22-7"></a>adults <span class="ot">&lt;-</span> <span class="dv">21</span><span class="sc">:</span><span class="dv">30</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Now, to capture the aging process, it’s convenient to define another matrix to hold the rates of movement between age classes. Generally, this matrix would look like this:</p>
<p><span id="eq-aging-mat"><span class="math display">\[
\begin{pmatrix}
    -\alpha_1 &amp; 0 &amp; 0 &amp; \cdots &amp; 0\\
    \alpha_1 &amp; -\alpha_2 &amp; 0 &amp; \cdots &amp; 0\\
    0 &amp; \alpha_2 &amp; -\alpha_3 &amp; \cdots &amp; 0\\
    \vdots &amp;  &amp; \ddots &amp; \ddots &amp; \vdots \\
    0 &amp; \cdots &amp; &amp; \alpha_{29} &amp; -\alpha_{30}\\
\end{pmatrix}
\tag{7.1}\]</span></span></p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-26_0aeac495c9d81e812953cd646315c358">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb23" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a><span class="co"># Create a diagonal matrix that holds the rates of aging out of each age class</span></span>
<span id="cb23-2"><a href="#cb23-2"></a><span class="co"># The rows represent the age class you're in, the columns represent the age class</span></span>
<span id="cb23-3"><a href="#cb23-3"></a><span class="co"># you're moving to</span></span>
<span id="cb23-4"><a href="#cb23-4"></a>aging_mat <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="sc">-</span><span class="dv">1</span> <span class="sc">/</span> da_ages)</span>
<span id="cb23-5"><a href="#cb23-5"></a></span>
<span id="cb23-6"><a href="#cb23-6"></a><span class="co"># Fill in the rates of aging into each age class</span></span>
<span id="cb23-7"><a href="#cb23-7"></a>aging_mat[<span class="fu">row</span>(aging_mat) <span class="sc">-</span> <span class="fu">col</span>(aging_mat) <span class="sc">==</span> <span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> <span class="fu">head</span>(da_ages, <span class="sc">-</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Have a look at the aging matrix, for example by doing:</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-27_11c5427d9c0aa9bc0d79413a9a4c582c">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb24" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="co"># Move fast through the 1-year age classes - negatives are moves out, positives are moves in</span></span>
<span id="cb24-2"><a href="#cb24-2"></a><span class="co"># Cannot move between non-adjacent age classes</span></span>
<span id="cb24-3"><a href="#cb24-3"></a>aging_mat[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre data-code-line-numbers=""><code>     [,1] [,2] [,3] [,4] [,5]
[1,]   -1    0    0    0    0
[2,]    1   -1    0    0    0
[3,]    0    1   -1    0    0
[4,]    0    0    1   -1    0
[5,]    0    0    0    1   -1</code></pre>
</div>
</div>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-28_2097e1afc61a72a3e7edeac137dda095">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb26" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a><span class="co"># Move slowly between the wider age classes</span></span>
<span id="cb26-2"><a href="#cb26-2"></a>aging_mat[<span class="dv">25</span><span class="sc">:</span><span class="dv">30</span>, <span class="dv">25</span><span class="sc">:</span><span class="dv">30</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre data-code-line-numbers=""><code>     [,1] [,2] [,3] [,4] [,5]        [,6]
[1,] -0.2  0.0  0.0  0.0  0.0  0.00000000
[2,]  0.2 -0.2  0.0  0.0  0.0  0.00000000
[3,]  0.0  0.2 -0.2  0.0  0.0  0.00000000
[4,]  0.0  0.0  0.2 -0.2  0.0  0.00000000
[5,]  0.0  0.0  0.0  0.2 -0.2  0.00000000
[6,]  0.0  0.0  0.0  0.0  0.2 -0.06666667</code></pre>
</div>
</div>
<div class="cell page-columns page-full" data-hash="r-session-02_cache/html/unnamed-chunk-29_b4bd10806ad1ca16c47d270596f4787e">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb28" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a>aging_mat <span class="sc">%&gt;%</span></span>
<span id="cb28-2"><a href="#cb28-2"></a>    <span class="fu">as.data.frame.table</span>() <span class="sc">%&gt;%</span></span>
<span id="cb28-3"><a href="#cb28-3"></a>    <span class="fu">mutate</span>(</span>
<span id="cb28-4"><a href="#cb28-4"></a>        <span class="at">age_recipient =</span> <span class="fu">rep</span>(ages, <span class="dv">30</span>),</span>
<span id="cb28-5"><a href="#cb28-5"></a>        <span class="at">age_source =</span> <span class="fu">unlist</span>(purrr<span class="sc">::</span><span class="fu">map</span>(<span class="at">.x =</span> ages, <span class="sc">~</span><span class="fu">rep</span>(.x, <span class="dv">30</span>)))</span>
<span id="cb28-6"><a href="#cb28-6"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb28-7"><a href="#cb28-7"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> age_source, <span class="at">y =</span> age_recipient, <span class="at">z =</span> Freq)) <span class="sc">+</span></span>
<span id="cb28-8"><a href="#cb28-8"></a>    <span class="fu">geom_contour_filled</span>() <span class="sc">+</span></span>
<span id="cb28-9"><a href="#cb28-9"></a>    <span class="fu">scale_fill_brewer</span>(</span>
<span id="cb28-10"><a href="#cb28-10"></a>        <span class="at">palette =</span> <span class="st">"RdBu"</span>,</span>
<span id="cb28-11"><a href="#cb28-11"></a>        <span class="at">drop =</span> <span class="cn">FALSE</span></span>
<span id="cb28-12"><a href="#cb28-12"></a>    ) <span class="sc">+</span></span>
<span id="cb28-13"><a href="#cb28-13"></a>    <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Source Age Group"</span>, <span class="at">y =</span> <span class="st">"Recipient Age Group"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display column-body">
<p><img src="r-session-02_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid" style="width:100.0%"></p>
</div>
</div>
<section id="exercise-2-what-can-you-say-about-its-structure-how-are-the-different-age-groups-in-contact-with-each-other" class="level3 exercise" data-number="7.4.1"><h3 data-number="7.4.1" class="anchored" data-anchor-id="exercise-2-what-can-you-say-about-its-structure-how-are-the-different-age-groups-in-contact-with-each-other">
<span class="header-section-number">7.4.1</span> Exercise 2: What can you say about its structure? How are the different age groups in contact with each other?</h3>
</section><p>Now we can put the pieces together to write a simulator for the age-structured SIR dynamics.</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-30_bea3f39852d90ee1ec7bf73241c93e6a">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb29" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a><span class="co"># Using a list instead of a vector to hold the parameters, as ages_beta_mat and aging </span></span>
<span id="cb29-2"><a href="#cb29-2"></a><span class="co"># are both matrices, so we want to keep them as matrices, rather than flattening</span></span>
<span id="cb29-3"><a href="#cb29-3"></a>multistage_params <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb29-4"><a href="#cb29-4"></a>    <span class="at">beta_mat =</span> ages_beta_mat,</span>
<span id="cb29-5"><a href="#cb29-5"></a>    <span class="at">recovery =</span> ages_params[<span class="st">"recovery"</span>],</span>
<span id="cb29-6"><a href="#cb29-6"></a>    <span class="at">births =</span> ages_params[<span class="st">"births"</span>],</span>
<span id="cb29-7"><a href="#cb29-7"></a>    <span class="at">aging_mat =</span> aging_mat</span>
<span id="cb29-8"><a href="#cb29-8"></a>)</span>
<span id="cb29-9"><a href="#cb29-9"></a></span>
<span id="cb29-10"><a href="#cb29-10"></a>multistage_model <span class="ot">&lt;-</span> <span class="cf">function</span> (t, x, p, ...) {</span>
<span id="cb29-11"><a href="#cb29-11"></a>    <span class="co"># Unpack all states from the vector using the relevant indices</span></span>
<span id="cb29-12"><a href="#cb29-12"></a>    s <span class="ot">&lt;-</span> x[sindex]</span>
<span id="cb29-13"><a href="#cb29-13"></a>    i <span class="ot">&lt;-</span> x[iindex]</span>
<span id="cb29-14"><a href="#cb29-14"></a>    r <span class="ot">&lt;-</span> x[rindex]</span>
<span id="cb29-15"><a href="#cb29-15"></a>    </span>
<span id="cb29-16"><a href="#cb29-16"></a>    <span class="co"># Unpack parameters</span></span>
<span id="cb29-17"><a href="#cb29-17"></a>    beta_mat <span class="ot">&lt;-</span> p[[<span class="st">"beta_mat"</span>]]</span>
<span id="cb29-18"><a href="#cb29-18"></a>    recovery <span class="ot">&lt;-</span> p[[<span class="st">"recovery"</span>]]</span>
<span id="cb29-19"><a href="#cb29-19"></a>    births <span class="ot">&lt;-</span> p[[<span class="st">"births"</span>]]</span>
<span id="cb29-20"><a href="#cb29-20"></a>    aging_mat <span class="ot">&lt;-</span> p[[<span class="st">"aging_mat"</span>]]</span>
<span id="cb29-21"><a href="#cb29-21"></a></span>
<span id="cb29-22"><a href="#cb29-22"></a>    <span class="co"># Calculate force of infection using matrix multiplication</span></span>
<span id="cb29-23"><a href="#cb29-23"></a>    lambda <span class="ot">&lt;-</span> beta_mat <span class="sc">%*%</span> i</span>
<span id="cb29-24"><a href="#cb29-24"></a>    </span>
<span id="cb29-25"><a href="#cb29-25"></a>    <span class="co"># Calculate the ODEs at every time step</span></span>
<span id="cb29-26"><a href="#cb29-26"></a>    <span class="co"># Note that R add element-wise for vectors i.e. lambda * s results</span></span>
<span id="cb29-27"><a href="#cb29-27"></a>    <span class="co"># in a vector length 30 (30 age groups), as does aging_mat %*% s,</span></span>
<span id="cb29-28"><a href="#cb29-28"></a>    <span class="co"># so v1[i] + v2[i] for i in 1:30</span></span>
<span id="cb29-29"><a href="#cb29-29"></a>    dsdt <span class="ot">&lt;-</span> <span class="sc">-</span>lambda <span class="sc">*</span> s <span class="sc">+</span> aging_mat <span class="sc">%*%</span> s</span>
<span id="cb29-30"><a href="#cb29-30"></a>    didt <span class="ot">&lt;-</span> lambda <span class="sc">*</span> s <span class="sc">+</span> aging_mat <span class="sc">%*%</span> i <span class="sc">-</span> recovery <span class="sc">*</span> i</span>
<span id="cb29-31"><a href="#cb29-31"></a>    drdt <span class="ot">&lt;-</span> aging_mat <span class="sc">%*%</span> r <span class="sc">+</span> recovery <span class="sc">*</span> i </span>
<span id="cb29-32"><a href="#cb29-32"></a>    <span class="co"># Add the birth rate to the first age group</span></span>
<span id="cb29-33"><a href="#cb29-33"></a>    dsdt[<span class="dv">1</span>] <span class="ot">&lt;-</span> dsdt[<span class="dv">1</span>] <span class="sc">+</span> births</span>
<span id="cb29-34"><a href="#cb29-34"></a>    </span>
<span id="cb29-35"><a href="#cb29-35"></a>    <span class="co"># Return the ODEs in a list</span></span>
<span id="cb29-36"><a href="#cb29-36"></a>    <span class="fu">list</span>(<span class="fu">c</span>(dsdt, didt, drdt))</span>
<span id="cb29-37"><a href="#cb29-37"></a></span>
<span id="cb29-38"><a href="#cb29-38"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We can plug this into <code>ode</code> just as we did the simpler models to simulate an epidemic. We’ll then plot the epidemic curve.</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-31_b33e43b04e02830bacb559b5d9658d9e">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb30" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a><span class="co"># Solve the model with a realistic age matrix</span></span>
<span id="cb30-2"><a href="#cb30-2"></a>multistage_sol <span class="ot">&lt;-</span> deSolve<span class="sc">::</span><span class="fu">ode</span>(</span>
<span id="cb30-3"><a href="#cb30-3"></a>    <span class="at">y =</span> demog_yinit_ages,</span>
<span id="cb30-4"><a href="#cb30-4"></a>    <span class="at">times =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">100</span>, <span class="at">by =</span> <span class="fl">0.1</span>),</span>
<span id="cb30-5"><a href="#cb30-5"></a>    <span class="at">func =</span> multistage_model,</span>
<span id="cb30-6"><a href="#cb30-6"></a>    <span class="at">parms =</span> multistage_params</span>
<span id="cb30-7"><a href="#cb30-7"></a>)</span>
<span id="cb30-8"><a href="#cb30-8"></a></span>
<span id="cb30-9"><a href="#cb30-9"></a><span class="co"># Extract all infected age groups at all time points into a new vector</span></span>
<span id="cb30-10"><a href="#cb30-10"></a>multistage_infecteds <span class="ot">&lt;-</span> multistage_sol[, <span class="dv">1</span> <span class="sc">+</span> iindex]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-32_6adac213708381984f7457234f7febf5">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb31" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a><span class="co"># Create a dataframe of the sum of infectious individuals in Juv/Adult age groups</span></span>
<span id="cb31-2"><a href="#cb31-2"></a><span class="co"># at each time point</span></span>
<span id="cb31-3"><a href="#cb31-3"></a>multistage_df <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb31-4"><a href="#cb31-4"></a>        <span class="co"># Get all times from model run</span></span>
<span id="cb31-5"><a href="#cb31-5"></a>        <span class="at">time =</span> multistage_sol[, <span class="dv">1</span>],</span>
<span id="cb31-6"><a href="#cb31-6"></a>        <span class="co"># At each timepoint, apply the sum function to all juvenile infected individuals</span></span>
<span id="cb31-7"><a href="#cb31-7"></a>        <span class="at">Juveniles =</span> <span class="fu">apply</span>(multistage_infecteds[, juvies], <span class="dv">1</span>, sum),</span>
<span id="cb31-8"><a href="#cb31-8"></a>        <span class="co"># At each timepoint, apply the sum function to all adult infected individuals</span></span>
<span id="cb31-9"><a href="#cb31-9"></a>        <span class="at">Adults =</span> <span class="fu">apply</span>(multistage_infecteds[, adults], <span class="dv">1</span>, sum)</span>
<span id="cb31-10"><a href="#cb31-10"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb31-11"><a href="#cb31-11"></a>    <span class="co"># Pivot to create a long dataframe that works with ggplot</span></span>
<span id="cb31-12"><a href="#cb31-12"></a>    <span class="fu">pivot_longer</span>(</span>
<span id="cb31-13"><a href="#cb31-13"></a>        <span class="at">cols =</span> <span class="fu">c</span>(Juveniles, Adults),</span>
<span id="cb31-14"><a href="#cb31-14"></a>        <span class="at">names_to =</span> <span class="st">"age_group"</span>,</span>
<span id="cb31-15"><a href="#cb31-15"></a>        <span class="at">values_to =</span> <span class="st">"infections"</span></span>
<span id="cb31-16"><a href="#cb31-16"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb31-17"><a href="#cb31-17"></a>    <span class="co"># Turn new pivoted variable into a factor to plot nicely</span></span>
<span id="cb31-18"><a href="#cb31-18"></a>    <span class="fu">mutate</span>(</span>
<span id="cb31-19"><a href="#cb31-19"></a>        <span class="at">age_group =</span> <span class="fu">factor</span>(age_group, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Juveniles"</span>, <span class="st">"Adults"</span>))</span>
<span id="cb31-20"><a href="#cb31-20"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell page-columns page-full" data-hash="r-session-02_cache/html/unnamed-chunk-33_d47fe7ac9b32412ff963099aa5db7583">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb32" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a><span class="fu">ggplot</span>(multistage_df, <span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> infections, <span class="at">color =</span> age_group)) <span class="sc">+</span></span>
<span id="cb32-2"><a href="#cb32-2"></a>    <span class="fu">geom_line</span>(<span class="at">linewidth =</span> <span class="fl">1.5</span>) <span class="sc">+</span></span>
<span id="cb32-3"><a href="#cb32-3"></a>    <span class="fu">scale_color_manual</span>(</span>
<span id="cb32-4"><a href="#cb32-4"></a>        <span class="at">values =</span> age_group_colors</span>
<span id="cb32-5"><a href="#cb32-5"></a>    ) <span class="sc">+</span></span>
<span id="cb32-6"><a href="#cb32-6"></a>    <span class="fu">labs</span>(</span>
<span id="cb32-7"><a href="#cb32-7"></a>        <span class="at">x =</span> <span class="st">"Time"</span>,</span>
<span id="cb32-8"><a href="#cb32-8"></a>        <span class="at">y =</span> <span class="st">"Number of infections"</span>,</span>
<span id="cb32-9"><a href="#cb32-9"></a>        <span class="at">color =</span> <span class="st">"Age group"</span></span>
<span id="cb32-10"><a href="#cb32-10"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display column-body">
<p><img src="r-session-02_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid" style="width:100.0%"></p>
</div>
</div>
<p>Let’s mimic a situation where we have cross-sectional seroprevalence data (e.g.&nbsp;measures of antibodies that tell you someone is in the R class). In using such data, we’d typically assume that the system was at equilibrium.</p>
<section id="sec-ex-3" class="level3 exercise" data-number="7.4.2"><h3 data-number="7.4.2" class="anchored" data-anchor-id="sec-ex-3">
<span class="header-section-number">7.4.2</span> Exercise 3: What does the equilibrium age-specific seroprevalence look like in this example?</h3>
</section><p>Use the code below to display the age-specific seroprevalence (i.e., the seroprevalence for each age group at equilibrium)</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-34_cec3b638c44b95f4c26711b3e65db4ee">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb33" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1"></a><span class="co"># Get the last values for all individuals. drop() removes the column name, [-1] removes the time value</span></span>
<span id="cb33-2"><a href="#cb33-2"></a>multistage_equil <span class="ot">&lt;-</span> <span class="fu">drop</span>(<span class="fu">tail</span>(multistage_sol, <span class="dv">1</span>))[<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb33-3"><a href="#cb33-3"></a><span class="co"># Calculate the equilibrium pop sizes of each age group</span></span>
<span id="cb33-4"><a href="#cb33-4"></a>multistage_equil_n <span class="ot">&lt;-</span> multistage_equil[sindex] <span class="sc">+</span></span>
<span id="cb33-5"><a href="#cb33-5"></a>    multistage_equil[iindex] <span class="sc">+</span></span>
<span id="cb33-6"><a href="#cb33-6"></a>    multistage_equil[rindex]</span>
<span id="cb33-7"><a href="#cb33-7"></a></span>
<span id="cb33-8"><a href="#cb33-8"></a><span class="co"># Calculate equilibrium seroprevalence for each age group</span></span>
<span id="cb33-9"><a href="#cb33-9"></a>multistage_equil_seroprev <span class="ot">&lt;-</span> multistage_equil[rindex] <span class="sc">/</span> multistage_equil_n</span>
<span id="cb33-10"><a href="#cb33-10"></a></span>
<span id="cb33-11"><a href="#cb33-11"></a><span class="co"># Create a dataframe to store equilibrium seroprev for plotting</span></span>
<span id="cb33-12"><a href="#cb33-12"></a>multistage_equil_seroprev_df <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb33-13"><a href="#cb33-13"></a>    <span class="at">age =</span> ages,</span>
<span id="cb33-14"><a href="#cb33-14"></a>    <span class="at">seroprev =</span> multistage_equil_seroprev,</span>
<span id="cb33-15"><a href="#cb33-15"></a>    <span class="at">width =</span> da_ages</span>
<span id="cb33-16"><a href="#cb33-16"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell page-columns page-full" data-hash="r-session-02_cache/html/unnamed-chunk-35_80295f11211d332649017bc001766a31">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb34" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1"></a><span class="fu">ggplot</span>(multistage_equil_seroprev_df, <span class="fu">aes</span>(<span class="at">x =</span> age, <span class="at">y =</span> seroprev, <span class="at">fill =</span> age)) <span class="sc">+</span></span>
<span id="cb34-2"><a href="#cb34-2"></a>    <span class="co"># Set column width to width of age bands, and justify to start at lower bound</span></span>
<span id="cb34-3"><a href="#cb34-3"></a>    <span class="fu">geom_col</span>(<span class="at">width =</span> multistage_equil_seroprev_df<span class="sc">$</span>width, <span class="at">just =</span> <span class="fl">1.0</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb34-4"><a href="#cb34-4"></a>    <span class="fu">labs</span>(</span>
<span id="cb34-5"><a href="#cb34-5"></a>        <span class="at">x =</span> <span class="st">"Age"</span>,</span>
<span id="cb34-6"><a href="#cb34-6"></a>        <span class="at">y =</span> <span class="st">"Seroprevalence"</span></span>
<span id="cb34-7"><a href="#cb34-7"></a>    ) <span class="sc">+</span></span>
<span id="cb34-8"><a href="#cb34-8"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">80</span>, <span class="dv">10</span>)) <span class="sc">+</span></span>
<span id="cb34-9"><a href="#cb34-9"></a>    <span class="fu">scale_fill_continuous</span>(<span class="at">low =</span> age_group_colors[<span class="dv">1</span>], <span class="at">high =</span> age_group_colors[<span class="dv">2</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display column-body">
<p><img src="r-session-02_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid" style="width:100.0%"></p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="QUESTION">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
QUESTION
</div>
</div>
<div class="callout-body-container callout-body">
<p>At what age does the seroprevalence reach 75%?</p>
</div>
</div>
<p>Let’s also compute <span class="math inline">\(R_0\)</span>. And because we’ve added a lot of age structure, with transitions between the age groups, we can’t just copy and paste the previous Next Generation Matrix code (from <a href="#sec-simple-ngm"><span>Section&nbsp;7.6.1.1</span></a>). As before, the details of this computation are out of the workshop’s scope, but they are outlined in <a href="#sec-age-structure-ngm"><span>Section&nbsp;7.6.1.2</span></a>. We have created a function to calculate R0 for an age-structured SIR and have added some comments, but read <a href="#sec-age-structure-ngm"><span>Section&nbsp;7.6.1.2</span></a> for the full details and reasoning.</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-36_16d2ce0454389eec865bbd449e1ea29c">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb35" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a><span class="co"># Calculate the stable disease-free age distribution. Could also simulate without any infections.</span></span>
<span id="cb35-2"><a href="#cb35-2"></a>multistage_stable_n <span class="ot">&lt;-</span> <span class="fu">solve</span>(</span>
<span id="cb35-3"><a href="#cb35-3"></a>    aging_mat,</span>
<span id="cb35-4"><a href="#cb35-4"></a>    <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span> <span class="sc">*</span> multistage_params[[<span class="st">"births"</span>]], <span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">29</span>))</span>
<span id="cb35-5"><a href="#cb35-5"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p><a id="ngm-function"></a></p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-37_9b8b674cfe8add2ae0fc4ce5152e3040">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb36" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a><span class="co">#' Calculate R0</span></span>
<span id="cb36-2"><a href="#cb36-2"></a><span class="co">#'</span></span>
<span id="cb36-3"><a href="#cb36-3"></a><span class="co">#' Calculate the R0 of an SIR model using the next generation matrix approach described in @heffernanPerspectivesBasicReproductive2005</span></span>
<span id="cb36-4"><a href="#cb36-4"></a><span class="co">#'</span></span>
<span id="cb36-5"><a href="#cb36-5"></a><span class="co">#' @param beta_mat A matrix of beta parameter values</span></span>
<span id="cb36-6"><a href="#cb36-6"></a><span class="co">#' @param stable_n_mat A matrix of the stable age distributions</span></span>
<span id="cb36-7"><a href="#cb36-7"></a><span class="co">#' @param aging_mat A matrix of the aging rates between age compartments</span></span>
<span id="cb36-8"><a href="#cb36-8"></a><span class="co">#' @param recovery_rate The recover rate parameter value (of type double)</span></span>
<span id="cb36-9"><a href="#cb36-9"></a><span class="co">#'</span></span>
<span id="cb36-10"><a href="#cb36-10"></a><span class="co">#' @return The R0 value as type double</span></span>
<span id="cb36-11"><a href="#cb36-11"></a><span class="co">#' @examples</span></span>
<span id="cb36-12"><a href="#cb36-12"></a><span class="co">#' calculate_R0(</span></span>
<span id="cb36-13"><a href="#cb36-13"></a><span class="co">#'    beta_mat = multistage_params[["beta_mat"]],</span></span>
<span id="cb36-14"><a href="#cb36-14"></a><span class="co">#'    stable_n_mat = multistage_stable_n,</span></span>
<span id="cb36-15"><a href="#cb36-15"></a><span class="co">#'    aging_mat = multistage_params[["aging_mat"]],</span></span>
<span id="cb36-16"><a href="#cb36-16"></a><span class="co">#'    recovery_rate = multistage_params[["recovery"]]</span></span>
<span id="cb36-17"><a href="#cb36-17"></a><span class="co">#')</span></span>
<span id="cb36-18"><a href="#cb36-18"></a>calculate_R0 <span class="ot">&lt;-</span> <span class="cf">function</span>(beta_mat, stable_n_mat, aging_mat, recovery_rate) {</span>
<span id="cb36-19"><a href="#cb36-19"></a>    <span class="co"># evaluate new inf jac pde at dfe</span></span>
<span id="cb36-20"><a href="#cb36-20"></a>    f_mat <span class="ot">&lt;-</span> beta_mat <span class="sc">*</span> stable_n_mat</span>
<span id="cb36-21"><a href="#cb36-21"></a></span>
<span id="cb36-22"><a href="#cb36-22"></a>    <span class="co"># set off-diag of non-inf transition jac pde to neg aging of prev age group</span></span>
<span id="cb36-23"><a href="#cb36-23"></a>    <span class="co"># (use aging matrix as already calculated in correct places)</span></span>
<span id="cb36-24"><a href="#cb36-24"></a>    v_mat <span class="ot">&lt;-</span> <span class="sc">-</span>aging_mat</span>
<span id="cb36-25"><a href="#cb36-25"></a>    <span class="co"># Update the diagonal of non-inf transition jac to add recovery rate</span></span>
<span id="cb36-26"><a href="#cb36-26"></a>    <span class="fu">diag</span>(v_mat) <span class="ot">&lt;-</span> <span class="fu">diag</span>(v_mat) <span class="sc">+</span> recovery_rate</span>
<span id="cb36-27"><a href="#cb36-27"></a></span>
<span id="cb36-28"><a href="#cb36-28"></a>    <span class="do">## Alternative method of calculating using age bands directly</span></span>
<span id="cb36-29"><a href="#cb36-29"></a>    <span class="co"># v_mat &lt;- diag(recovery_rate + 1 / da_ages)</span></span>
<span id="cb36-30"><a href="#cb36-30"></a>    <span class="co"># v_mat[row(v_mat) - col(v_mat) == 1] &lt;- - 1 / head(da_ages, -1)</span></span>
<span id="cb36-31"><a href="#cb36-31"></a>    </span>
<span id="cb36-32"><a href="#cb36-32"></a>    <span class="co"># spectral trace</span></span>
<span id="cb36-33"><a href="#cb36-33"></a>    R0 <span class="ot">&lt;-</span> <span class="fu">max</span>(<span class="fu">Re</span>(<span class="fu">eigen</span>(<span class="fu">solve</span>(v_mat, f_mat), <span class="at">only.values =</span> <span class="cn">TRUE</span>)<span class="sc">$</span>values))   </span>
<span id="cb36-34"><a href="#cb36-34"></a>    </span>
<span id="cb36-35"><a href="#cb36-35"></a>    <span class="fu">return</span>(R0)</span>
<span id="cb36-36"><a href="#cb36-36"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-38_f6cba45736fbb36c0a4fa60f8618e6de">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb37" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1"></a><span class="fu">calculate_R0</span>(</span>
<span id="cb37-2"><a href="#cb37-2"></a>    <span class="at">beta_mat =</span> multistage_params[[<span class="st">"beta_mat"</span>]],</span>
<span id="cb37-3"><a href="#cb37-3"></a>    <span class="at">stable_n_mat =</span> multistage_stable_n,</span>
<span id="cb37-4"><a href="#cb37-4"></a>    <span class="at">aging_mat =</span> multistage_params[[<span class="st">"aging_mat"</span>]],</span>
<span id="cb37-5"><a href="#cb37-5"></a>    <span class="at">recovery_rate =</span> multistage_params[[<span class="st">"recovery"</span>]]</span>
<span id="cb37-6"><a href="#cb37-6"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre data-code-line-numbers=""><code>[1] 6.991242</code></pre>
</div>
</div>
<section id="exercise-4-updating-the-contact-matrix" class="level3 exercise" data-number="7.4.3"><h3 data-number="7.4.3" class="anchored" data-anchor-id="exercise-4-updating-the-contact-matrix">
<span class="header-section-number">7.4.3</span> Exercise 4: Updating the contact matrix</h3>
</section><div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>You will need to read and edit the following code carefully so that it runs with your updated parameters. We have highlighted the relevant lines in the code chunks, so hopefully you won’t miss them, though make sure you do copy all the code!</p>
</div>
</div>
<section id="change-the-juvenile-juvenile-contact-rate-to-be-0.025." class="level4" data-number="7.4.3.1"><h4 data-number="7.4.3.1" class="anchored" data-anchor-id="change-the-juvenile-juvenile-contact-rate-to-be-0.025.">
<span class="header-section-number">7.4.3.1</span> Change the juvenile-juvenile contact rate to be 0.025.</h4>
<div class="cell" data-source-line-numbers="2" data-hash="r-session-02_cache/html/unnamed-chunk-40_16ecd60775bd9e19842451eb4eaff576">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb39" data-source-line-numbers="2" data-code-line-numbers="2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1"></a>update_age_beta_mat <span class="ot">&lt;-</span> ages_beta_mat</span>
<span id="cb39-2"><a href="#cb39-2"></a>update_age_beta_mat[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>] <span class="ot">&lt;-</span> ?</span>
<span id="cb39-3"><a href="#cb39-3"></a></span>
<span id="cb39-4"><a href="#cb39-4"></a>update_age_params <span class="ot">&lt;-</span> multistage_params</span>
<span id="cb39-5"><a href="#cb39-5"></a>update_age_params[[<span class="st">"beta_mat"</span>]] <span class="ot">&lt;-</span> update_age_beta_mat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section><section id="simulate-and-plot-the-age-structured-sir-dynamics-under-your-assumptions-and-record-how-the-age-specific-seroprevalence-has-changed." class="level4" data-number="7.4.3.2"><h4 data-number="7.4.3.2" class="anchored" data-anchor-id="simulate-and-plot-the-age-structured-sir-dynamics-under-your-assumptions-and-record-how-the-age-specific-seroprevalence-has-changed.">
<span class="header-section-number">7.4.3.2</span> Simulate and plot the age-structured SIR dynamics under your assumptions and record how the age-specific seroprevalence has changed.</h4>
<div class="cell" data-source-line-numbers="5,15-17,21" data-hash="r-session-02_cache/html/unnamed-chunk-42_b6eb36c5a0cbb69f89395fefdd93ed50">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb40" data-source-line-numbers="5,15-17,21" data-code-line-numbers="5,15-17,21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1"></a>update_age_sol <span class="ot">&lt;-</span> deSolve<span class="sc">::</span><span class="fu">ode</span>(</span>
<span id="cb40-2"><a href="#cb40-2"></a>    <span class="at">y =</span> demog_yinit_ages,</span>
<span id="cb40-3"><a href="#cb40-3"></a>    <span class="at">times =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">400</span>, <span class="at">by =</span> <span class="fl">0.1</span>),</span>
<span id="cb40-4"><a href="#cb40-4"></a>    <span class="at">func =</span> multistage_model,</span>
<span id="cb40-5"><a href="#cb40-5"></a>    <span class="at">parms =</span> ?</span>
<span id="cb40-6"><a href="#cb40-6"></a>)</span>
<span id="cb40-7"><a href="#cb40-7"></a></span>
<span id="cb40-8"><a href="#cb40-8"></a><span class="co"># Get the time series for each infectious age group</span></span>
<span id="cb40-9"><a href="#cb40-9"></a>update_age_infecteds <span class="ot">&lt;-</span> update_age_sol[, <span class="dv">1</span> <span class="sc">+</span> iindex]</span>
<span id="cb40-10"><a href="#cb40-10"></a></span>
<span id="cb40-11"><a href="#cb40-11"></a><span class="co"># Get the last values in the time series</span></span>
<span id="cb40-12"><a href="#cb40-12"></a>update_age_equil <span class="ot">&lt;-</span> <span class="fu">drop</span>(<span class="fu">tail</span>(update_age_sol, <span class="dv">1</span>))[<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb40-13"><a href="#cb40-13"></a></span>
<span id="cb40-14"><a href="#cb40-14"></a><span class="co"># Calculate the number of individuals in each age group at the final timepoint</span></span>
<span id="cb40-15"><a href="#cb40-15"></a>update_age_equil_n <span class="ot">&lt;-</span> update_age_equil[ ? ] <span class="sc">+</span></span>
<span id="cb40-16"><a href="#cb40-16"></a>    update_age_equil[ ? ] <span class="sc">+</span></span>
<span id="cb40-17"><a href="#cb40-17"></a>    update_age_equil[ ? ]</span>
<span id="cb40-18"><a href="#cb40-18"></a></span>
<span id="cb40-19"><a href="#cb40-19"></a><span class="co"># Calculate final seroprevalence</span></span>
<span id="cb40-20"><a href="#cb40-20"></a><span class="co"># Hint: You need PREVIOUSLY infected individuals</span></span>
<span id="cb40-21"><a href="#cb40-21"></a>update_age_equil_seroprev <span class="ot">&lt;-</span> update_age_equil[ ? ] <span class="sc">/</span> update_age_equil_n</span>
<span id="cb40-22"><a href="#cb40-22"></a></span>
<span id="cb40-23"><a href="#cb40-23"></a>update_age_equil_seroprev_df <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb40-24"><a href="#cb40-24"></a>    <span class="at">age =</span> ages,</span>
<span id="cb40-25"><a href="#cb40-25"></a>    <span class="at">seroprev =</span> update_age_equil_seroprev,</span>
<span id="cb40-26"><a href="#cb40-26"></a>    <span class="at">width =</span> da_ages</span>
<span id="cb40-27"><a href="#cb40-27"></a>)</span>
<span id="cb40-28"><a href="#cb40-28"></a></span>
<span id="cb40-29"><a href="#cb40-29"></a><span class="fu">ggplot</span>(update_age_equil_seroprev_df, <span class="fu">aes</span>(<span class="at">x =</span> age, <span class="at">y =</span> seroprev, <span class="at">fill =</span> age)) <span class="sc">+</span></span>
<span id="cb40-30"><a href="#cb40-30"></a>    <span class="co"># Set column width to width of age bands, and justify to start at lower bound</span></span>
<span id="cb40-31"><a href="#cb40-31"></a>    <span class="fu">geom_col</span>(<span class="at">width =</span> update_age_equil_seroprev_df<span class="sc">$</span>width, <span class="at">just =</span> <span class="fl">1.0</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb40-32"><a href="#cb40-32"></a>    <span class="fu">labs</span>(</span>
<span id="cb40-33"><a href="#cb40-33"></a>        <span class="at">x =</span> <span class="st">"Age"</span>,</span>
<span id="cb40-34"><a href="#cb40-34"></a>        <span class="at">y =</span> <span class="st">"Seroprevalence"</span></span>
<span id="cb40-35"><a href="#cb40-35"></a>    ) <span class="sc">+</span></span>
<span id="cb40-36"><a href="#cb40-36"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">80</span>, <span class="dv">10</span>)) <span class="sc">+</span></span>
<span id="cb40-37"><a href="#cb40-37"></a>    <span class="fu">scale_fill_continuous</span>(<span class="at">low =</span> age_group_colors[<span class="dv">1</span>], <span class="at">high =</span> age_group_colors[<span class="dv">2</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="QUESTION">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
QUESTION
</div>
</div>
<div class="callout-body-container callout-body">
<p>At what age does the seroprevalence reach 75%? How does this compare to the answer in <a href="#sec-ex-3"><span>Section&nbsp;7.4.2</span></a>?</p>
</div>
</div>
</section><section id="compute-r_0-for-your-assumptions." class="level4" data-number="7.4.3.3"><h4 data-number="7.4.3.3" class="anchored" data-anchor-id="compute-r_0-for-your-assumptions.">
<span class="header-section-number">7.4.3.3</span> Compute <span class="math inline">\(R_0\)</span> for your assumptions.</h4>
<p>As described previously, the calculation for <span class="math inline">\(R_0\)</span> is difficult due to all the age categories and transitions. Use the <code>calculate_R0()</code> function we <a href="#ngm-function">defined earlier</a> to calculate <span class="math inline">\(R_0\)</span> for our updated system.</p>
<div class="cell" data-source-line-numbers="2-5" data-hash="r-session-02_cache/html/unnamed-chunk-43_e7beeea014499e385ffcf98cbd0ef141">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb41" data-source-line-numbers="2-5" data-code-line-numbers="2-5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1"></a><span class="fu">calculate_R0</span>(</span>
<span id="cb41-2"><a href="#cb41-2"></a>    <span class="at">beta_mat =</span> ?,</span>
<span id="cb41-3"><a href="#cb41-3"></a>    <span class="at">stable_n_mat =</span> ?,</span>
<span id="cb41-4"><a href="#cb41-4"></a>    <span class="at">aging_mat =</span> ?,</span>
<span id="cb41-5"><a href="#cb41-5"></a>    <span class="at">recovery_rate =</span> ?</span>
<span id="cb41-6"><a href="#cb41-6"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>If you’ve done everything correctly, you should get <span class="math inline">\(R_0 =\)</span> 7.29. This is higher than previously. Does this match your intuition, given our changes to the beta matrix?</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>The following is Aaron’s original code which provides slightly different results.</p>
</div>
</div>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-45_bf69726422bf710d2dd02a4b202de0f6">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb42" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1"></a>F_mat <span class="ot">&lt;-</span> <span class="fu">diag</span>(update_age_equil_n) <span class="sc">%*%</span> update_age_beta_mat <span class="sc">+</span></span>
<span id="cb42-2"><a href="#cb42-2"></a>    update_age_params[[<span class="st">"aging_mat"</span>]] <span class="sc">-</span></span>
<span id="cb42-3"><a href="#cb42-3"></a>    <span class="fu">diag</span>(<span class="fu">diag</span>(update_age_params[[<span class="st">"aging_mat"</span>]]))</span>
<span id="cb42-4"><a href="#cb42-4"></a></span>
<span id="cb42-5"><a href="#cb42-5"></a>V_mat <span class="ot">&lt;-</span> <span class="fu">diag</span>(update_age_params[[<span class="st">"recovery"</span>]] <span class="sc">-</span> <span class="fu">diag</span>(update_age_params[[<span class="st">"aging_mat"</span>]]))</span>
<span id="cb42-6"><a href="#cb42-6"></a></span>
<span id="cb42-7"><a href="#cb42-7"></a><span class="co"># R0 approximately 6.53</span></span>
<span id="cb42-8"><a href="#cb42-8"></a><span class="fu">max</span>(<span class="fu">Re</span>(<span class="fu">eigen</span>(<span class="fu">solve</span>(V_mat, F_mat), <span class="at">only.values =</span> <span class="cn">TRUE</span>)<span class="sc">$</span>values))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre data-code-line-numbers=""><code>[1] 7.034769</code></pre>
</div>
</div>
</section></section><section id="r0-and-the-mean-age-of-infection" class="level2" data-number="7.5"><h2 data-number="7.5" class="anchored" data-anchor-id="r0-and-the-mean-age-of-infection">
<span class="header-section-number">7.5</span> R0 and the Mean Age of Infection</h2>
<p>To develop some intuition about the relationship between <span class="math inline">\(R_0\)</span> and the mean age of infection, let’s play with an interactive plot. We will assume that the population is completely susceptible and that the force of infection is constant. We’ll also assume that there is heterogenous mixing i.e.&nbsp;no age structure.</p>
<p>As we’ve seen in <a href="./L05_age-structure.html">Matt’s lecture on age structure</a>, we can calculate the mean age of infection using the equation below:</p>
<p><span id="eq-mean-age"><span class="math display">\[
A \approx \frac{L}{R_E - 1}
\tag{7.2}\]</span></span></p>
<p>where <span class="math inline">\(L\)</span> is the life expectancy <span class="math inline">\(\left(L = \frac{1}{\mu}\right)\)</span> and <span class="math inline">\(R_E\)</span> is the effective reproductive number (<span class="math inline">\(R_E = R_0 * (1 - p)\)</span> where <span class="math inline">\(p\)</span> is the fraction of individuals vaccinated).</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<p>See <a href="#sec-mean-age-r-code"><span>Section&nbsp;7.6.2</span></a> for code you can run in <code>R</code> to investigate the relationship between <span class="math inline">\(R_0\)</span>, vaccination coverage, life expectancy, and the mean age of infection.</p>
</div>
</div>
</div>
<section id="exercise-5-mean-age-of-infection-interactions" class="level3 exercise" data-number="7.5.1"><h3 data-number="7.5.1" class="anchored" data-anchor-id="exercise-5-mean-age-of-infection-interactions">
<span class="header-section-number">7.5.1</span> Exercise 5: mean age of infection interactions</h3>
</section><section id="when-you-increase-r_0-from-2.0-to-4.0-what-happens-to-the-mean-age-of-infection" class="level4" data-number="7.5.1.1"><h4 data-number="7.5.1.1" class="anchored" data-anchor-id="when-you-increase-r_0-from-2.0-to-4.0-what-happens-to-the-mean-age-of-infection">
<span class="header-section-number">7.5.1.1</span> When you increase <span class="math inline">\(R_0\)</span> from 2.0 to 4.0, what happens to the mean age of infection?</h4>
<section id="is-there-a-linear-change-if-not-why-not" class="level5" data-number="7.5.1.1.1"><h5 data-number="7.5.1.1.1" class="anchored" data-anchor-id="is-there-a-linear-change-if-not-why-not">
<span class="header-section-number">7.5.1.1.1</span> Is there a linear change? If not, why not?</h5>
</section></section><section id="with-r_0-to-4.0-approximately-what-level-of-vaccination-coverage-is-required-for-a-mean-age-of-infection-of-40-years" class="level4" data-number="7.5.1.2"><h4 data-number="7.5.1.2" class="anchored" data-anchor-id="with-r_0-to-4.0-approximately-what-level-of-vaccination-coverage-is-required-for-a-mean-age-of-infection-of-40-years">
<span class="header-section-number">7.5.1.2</span> With <span class="math inline">\(R_0\)</span> to 4.0, approximately what level of vaccination coverage is required for a mean age of infection of 40 years?</h4>
</section><section id="leaving-r_0-and-vaccination-coverage-the-same-decrease-the-life-expectancy-to-50-years.-what-happens-to-the-mean-age-of-infection" class="level4" data-number="7.5.1.3"><h4 data-number="7.5.1.3" class="anchored" data-anchor-id="leaving-r_0-and-vaccination-coverage-the-same-decrease-the-life-expectancy-to-50-years.-what-happens-to-the-mean-age-of-infection">
<span class="header-section-number">7.5.1.3</span> Leaving <span class="math inline">\(R_0\)</span> and vaccination coverage the same, decrease the life expectancy to 50 years. What happens to the mean age of infection?</h4>
<section id="if-it-changed-why-do-you-think-it-did" class="level5" data-number="7.5.1.3.1"><h5 data-number="7.5.1.3.1" class="anchored" data-anchor-id="if-it-changed-why-do-you-think-it-did">
<span class="header-section-number">7.5.1.3.1</span> If it changed, why do you think it did?</h5>
<p><br></p>
<div class="cell">
<details class="hidden"><summary>Code</summary><div class="sourceCode cell-code hidden" id="cb44" data-startfrom="1229" data-source-offset="-0" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource js number-lines code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 1228;"><span id="cb44-1229"><a href="#cb44-1229"></a>init_R0 <span class="op">=</span> <span class="fl">2.0</span></span>
<span id="cb44-1230"><a href="#cb44-1230"></a>init_vacc <span class="op">=</span> <span class="fl">0.0</span></span>
<span id="cb44-1231"><a href="#cb44-1231"></a>init_lifeexp <span class="op">=</span> <span class="dv">75</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-display">
<div>
<div id="ojs-cell-1-1" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<div id="ojs-cell-1-2" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<div id="ojs-cell-1-3" data-nodetype="declaration">

</div>
</div>
</div>
</div>
<div class="cell">
<details class="hidden"><summary>Code</summary><div class="sourceCode cell-code hidden" id="cb45" data-startfrom="1235" data-source-offset="0" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource js number-lines code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 1234;"><span id="cb45-1235"><a href="#cb45-1235"></a><span class="kw">function</span> <span class="fu">set</span>(input<span class="op">,</span> value) {</span>
<span id="cb45-1236"><a href="#cb45-1236"></a>  input<span class="op">.</span><span class="at">value</span> <span class="op">=</span> value<span class="op">;</span></span>
<span id="cb45-1237"><a href="#cb45-1237"></a>  input<span class="op">.</span><span class="fu">dispatchEvent</span>(<span class="kw">new</span> <span class="bu">Event</span>(<span class="st">"input"</span><span class="op">,</span> {<span class="dt">bubbles</span><span class="op">:</span> <span class="kw">true</span>}))<span class="op">;</span></span>
<span id="cb45-1238"><a href="#cb45-1238"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-display">
<div id="ojs-cell-2" data-nodetype="declaration">

</div>
</div>
</div>
<div class="panel-grid layout-sidebar ms-md-0 layout-sidebar-left">
<div class="cell panel-sidebar card bg-light p-2 g-col-24 g-col-lg-7">
<details class="hidden"><summary>Code</summary><div class="sourceCode cell-code hidden" id="cb46" data-startfrom="1245" data-source-offset="-0" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource js number-lines code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 1244;"><span id="cb46-1245"><a href="#cb46-1245"></a>viewof reset <span class="op">=</span> Inputs<span class="op">.</span><span class="fu">button</span>([</span>
<span id="cb46-1246"><a href="#cb46-1246"></a>  [<span class="st">"Reset all sliders"</span><span class="op">,</span> () <span class="kw">=&gt;</span> {</span>
<span id="cb46-1247"><a href="#cb46-1247"></a>    <span class="kw">set</span>(viewof R0<span class="op">,</span> init_R0)</span>
<span id="cb46-1248"><a href="#cb46-1248"></a>    <span class="kw">set</span>(viewof vacc<span class="op">,</span> init_vacc)</span>
<span id="cb46-1249"><a href="#cb46-1249"></a>    <span class="kw">set</span>(viewof lifeexp<span class="op">,</span> init_lifeexp)</span>
<span id="cb46-1250"><a href="#cb46-1250"></a>  }]</span>
<span id="cb46-1251"><a href="#cb46-1251"></a>])</span>
<span id="cb46-1252"><a href="#cb46-1252"></a>viewof R0 <span class="op">=</span> Inputs<span class="op">.</span><span class="fu">range</span>(</span>
<span id="cb46-1253"><a href="#cb46-1253"></a>    [<span class="fl">1.0</span><span class="op">,</span> <span class="fl">10.0</span>]<span class="op">,</span></span>
<span id="cb46-1254"><a href="#cb46-1254"></a>    {<span class="dt">value</span><span class="op">:</span> <span class="fl">2.0</span><span class="op">,</span> <span class="dt">step</span><span class="op">:</span> <span class="fl">0.01</span><span class="op">,</span> <span class="dt">label</span><span class="op">:</span> <span class="fu">md</span><span class="vs">`</span><span class="sc">${</span><span class="fu">tex</span><span class="vs">`R_0`</span><span class="sc">}</span><span class="vs">`</span>}</span>
<span id="cb46-1255"><a href="#cb46-1255"></a>)</span>
<span id="cb46-1256"><a href="#cb46-1256"></a></span>
<span id="cb46-1257"><a href="#cb46-1257"></a>viewof vacc <span class="op">=</span> Inputs<span class="op">.</span><span class="fu">range</span>(</span>
<span id="cb46-1258"><a href="#cb46-1258"></a>    [<span class="fl">0.0</span><span class="op">,</span> <span class="fl">1.0</span>]<span class="op">,</span></span>
<span id="cb46-1259"><a href="#cb46-1259"></a>    {<span class="dt">value</span><span class="op">:</span> <span class="fl">0.0</span><span class="op">,</span> <span class="dt">step</span><span class="op">:</span> <span class="fl">0.01</span><span class="op">,</span> <span class="dt">label</span><span class="op">:</span> <span class="st">"Vaccination coverage"</span>}</span>
<span id="cb46-1260"><a href="#cb46-1260"></a>)</span>
<span id="cb46-1261"><a href="#cb46-1261"></a></span>
<span id="cb46-1262"><a href="#cb46-1262"></a>viewof lifeexp <span class="op">=</span> Inputs<span class="op">.</span><span class="fu">range</span>(</span>
<span id="cb46-1263"><a href="#cb46-1263"></a>    [<span class="dv">50</span><span class="op">,</span> <span class="dv">100</span>]<span class="op">,</span></span>
<span id="cb46-1264"><a href="#cb46-1264"></a>    {<span class="dt">value</span><span class="op">:</span> <span class="dv">75</span><span class="op">,</span> <span class="dt">step</span><span class="op">:</span> <span class="dv">1</span><span class="op">,</span> <span class="dt">label</span><span class="op">:</span> <span class="st">"Life expectancy"</span>}</span>
<span id="cb46-1265"><a href="#cb46-1265"></a>)</span>
<span id="cb46-1266"><a href="#cb46-1266"></a></span>
<span id="cb46-1267"><a href="#cb46-1267"></a><span class="fu">md</span><span class="vs">`</span><span class="sc">${</span><span class="fu">tex</span><span class="vs">`R_E = </span><span class="sc">${</span>Re_str<span class="sc">}</span><span class="vs">`</span><span class="sc">}</span><span class="vs">`</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-display">
<div>
<div id="ojs-cell-3-1" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<div id="ojs-cell-3-2" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<div id="ojs-cell-3-3" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<div id="ojs-cell-3-4" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<div id="ojs-cell-3-5" data-nodetype="expression">

</div>
</div>
</div>
<details class="hidden"><summary>Code</summary><div class="sourceCode cell-code hidden" id="cb47" data-startfrom="1268" data-source-offset="-503" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource js number-lines code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 1267;"><span id="cb47-1268"><a href="#cb47-1268"></a><span class="fu">md</span><span class="vs">`</span><span class="sc">${</span><span class="fu">tex</span><span class="vs">`</span><span class="sc">\t</span><span class="vs">ext{Mean age of infection} = </span><span class="sc">${</span>Re_mean_age_str<span class="sc">}</span><span class="vs">`</span><span class="sc">}</span><span class="vs">`</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-display">
<div>
<div id="ojs-cell-3-6" data-nodetype="expression">

</div>
</div>
</div>
</div>
<div class="panel-fill g-col-24 g-col-lg-17 pt-3 pt-lg-0">
<div class="cell">
<details class="hidden"><summary>Code</summary><div class="sourceCode cell-code hidden" id="cb48" data-startfrom="1273" data-source-offset="-0" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource js number-lines code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 1272;"><span id="cb48-1273"><a href="#cb48-1273"></a>Re <span class="op">=</span> R0 <span class="op">*</span> (<span class="dv">1</span> <span class="op">-</span> vacc)</span>
<span id="cb48-1274"><a href="#cb48-1274"></a>Re_str <span class="op">=</span> Re<span class="op">.</span><span class="fu">toPrecision</span>(<span class="dv">4</span>)<span class="op">.</span><span class="fu">toLocaleString</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-display">
<div>
<div id="ojs-cell-4-1" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<div id="ojs-cell-4-2" data-nodetype="declaration">

</div>
</div>
</div>
</div>
<div class="cell">
<details class="hidden"><summary>Code</summary><div class="sourceCode cell-code hidden" id="cb49" data-startfrom="1278" data-source-offset="0" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource js number-lines code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 1277;"><span id="cb49-1278"><a href="#cb49-1278"></a><span class="kw">function</span> <span class="fu">calc_mean_age</span>(Re<span class="op">,</span> lifeexp) {</span>
<span id="cb49-1279"><a href="#cb49-1279"></a>    <span class="cf">if</span>(Re <span class="op">&gt;=</span> <span class="dv">1</span>) {</span>
<span id="cb49-1280"><a href="#cb49-1280"></a>        <span class="kw">var</span> mean_age <span class="op">=</span> (lifeexp <span class="op">/</span> (Re <span class="op">-</span> <span class="dv">1</span>))</span>
<span id="cb49-1281"><a href="#cb49-1281"></a>    } <span class="cf">else</span> {</span>
<span id="cb49-1282"><a href="#cb49-1282"></a>        <span class="kw">var</span> mean_age <span class="op">=</span> <span class="kw">Infinity</span></span>
<span id="cb49-1283"><a href="#cb49-1283"></a>    }</span>
<span id="cb49-1284"><a href="#cb49-1284"></a>    <span class="cf">return</span> mean_age</span>
<span id="cb49-1285"><a href="#cb49-1285"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-display">
<div id="ojs-cell-5" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell">
<details class="hidden"><summary>Code</summary><div class="sourceCode cell-code hidden" id="cb50" data-startfrom="1292" data-source-offset="-0" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource js number-lines code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 1291;"><span id="cb50-1292"><a href="#cb50-1292"></a>R0_mean_age <span class="op">=</span> <span class="fu">calc_mean_age</span>(R0<span class="op">,</span> lifeexp)</span>
<span id="cb50-1293"><a href="#cb50-1293"></a>Re_mean_age <span class="op">=</span> <span class="fu">calc_mean_age</span>(Re<span class="op">,</span> lifeexp)</span>
<span id="cb50-1294"><a href="#cb50-1294"></a>Re_mean_age_str <span class="op">=</span> Re_mean_age<span class="op">.</span><span class="fu">toPrecision</span>(<span class="dv">4</span>)<span class="op">.</span><span class="fu">toLocaleString</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-display">
<div>
<div id="ojs-cell-6-1" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<div id="ojs-cell-6-2" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<div id="ojs-cell-6-3" data-nodetype="declaration">

</div>
</div>
</div>
</div>
<div class="cell">
<details class="hidden"><summary>Code</summary><div class="sourceCode cell-code hidden" id="cb51" data-startfrom="1299" data-source-offset="0" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource js number-lines code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 1298;"><span id="cb51-1299"><a href="#cb51-1299"></a><span class="im">import</span> { aq<span class="op">,</span> op } <span class="im">from</span> <span class="st">'@uwdata/arquero'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-display">
<div id="ojs-cell-7" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell">
<details class="hidden"><summary>Code</summary><div class="sourceCode cell-code hidden" id="cb52" data-startfrom="1304" data-source-offset="0" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource js number-lines code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 1303;"><span id="cb52-1304"><a href="#cb52-1304"></a><span class="kw">function</span> <span class="fu">calc_mean_age_arr</span>(vacc<span class="op">,</span> lifeexp<span class="op">,</span> R0_min<span class="op">,</span> R0_max<span class="op">,</span> dR0) {</span>
<span id="cb52-1305"><a href="#cb52-1305"></a>    <span class="kw">var</span> R0_sim <span class="op">=</span> R0_min</span>
<span id="cb52-1306"><a href="#cb52-1306"></a></span>
<span id="cb52-1307"><a href="#cb52-1307"></a>    <span class="kw">var</span> R0 <span class="op">=</span> []</span>
<span id="cb52-1308"><a href="#cb52-1308"></a>    <span class="kw">var</span> Re <span class="op">=</span> []</span>
<span id="cb52-1309"><a href="#cb52-1309"></a>    <span class="kw">var</span> R0_mean_age <span class="op">=</span> []</span>
<span id="cb52-1310"><a href="#cb52-1310"></a>    <span class="kw">var</span> Re_mean_age <span class="op">=</span> []</span>
<span id="cb52-1311"><a href="#cb52-1311"></a></span>
<span id="cb52-1312"><a href="#cb52-1312"></a>    <span class="cf">for</span> (R0_sim <span class="op">=</span> R0_min<span class="op">;</span> R0_sim <span class="op">&lt;=</span> R0_max<span class="op">;</span> R0_sim <span class="op">+=</span> dR0) {</span>
<span id="cb52-1313"><a href="#cb52-1313"></a>        <span class="kw">var</span> Re_sim <span class="op">=</span> R0_sim <span class="op">*</span> (<span class="dv">1</span> <span class="op">-</span> vacc)</span>
<span id="cb52-1314"><a href="#cb52-1314"></a>        <span class="kw">var</span> R0_mean_age_sim <span class="op">=</span> <span class="fu">calc_mean_age</span>(R0_sim<span class="op">,</span> lifeexp)</span>
<span id="cb52-1315"><a href="#cb52-1315"></a>        <span class="kw">var</span> Re_mean_age_sim <span class="op">=</span> <span class="fu">calc_mean_age</span>(Re_sim<span class="op">,</span> lifeexp)</span>
<span id="cb52-1316"><a href="#cb52-1316"></a></span>
<span id="cb52-1317"><a href="#cb52-1317"></a>        R0<span class="op">.</span><span class="fu">push</span>(R0_sim)</span>
<span id="cb52-1318"><a href="#cb52-1318"></a>        Re<span class="op">.</span><span class="fu">push</span>(Re_sim)</span>
<span id="cb52-1319"><a href="#cb52-1319"></a>        R0_mean_age<span class="op">.</span><span class="fu">push</span>(R0_mean_age_sim)</span>
<span id="cb52-1320"><a href="#cb52-1320"></a>        Re_mean_age<span class="op">.</span><span class="fu">push</span>(Re_mean_age_sim)</span>
<span id="cb52-1321"><a href="#cb52-1321"></a>    }</span>
<span id="cb52-1322"><a href="#cb52-1322"></a></span>
<span id="cb52-1323"><a href="#cb52-1323"></a>    <span class="cf">return</span> {</span>
<span id="cb52-1324"><a href="#cb52-1324"></a>        <span class="dt">Re</span><span class="op">:</span> aq<span class="op">.</span><span class="fu">table</span>({</span>
<span id="cb52-1325"><a href="#cb52-1325"></a>                <span class="dt">R0</span><span class="op">:</span> R0<span class="op">,</span></span>
<span id="cb52-1326"><a href="#cb52-1326"></a>                <span class="dt">mean_age</span><span class="op">:</span> Re_mean_age</span>
<span id="cb52-1327"><a href="#cb52-1327"></a>            })<span class="op">.</span><span class="fu">filter</span>((d) <span class="kw">=&gt;</span> d<span class="op">.</span><span class="at">mean_age</span> <span class="op">&lt;=</span> <span class="dv">100</span>)<span class="op">,</span></span>
<span id="cb52-1328"><a href="#cb52-1328"></a>        <span class="dt">R0</span><span class="op">:</span> aq<span class="op">.</span><span class="fu">table</span>({</span>
<span id="cb52-1329"><a href="#cb52-1329"></a>                <span class="dt">R0</span><span class="op">:</span> R0<span class="op">,</span></span>
<span id="cb52-1330"><a href="#cb52-1330"></a>                <span class="dt">mean_age</span><span class="op">:</span> R0_mean_age</span>
<span id="cb52-1331"><a href="#cb52-1331"></a>            })<span class="op">.</span><span class="fu">filter</span>((d) <span class="kw">=&gt;</span> d<span class="op">.</span><span class="at">mean_age</span> <span class="op">&lt;=</span> <span class="dv">100</span>)</span>
<span id="cb52-1332"><a href="#cb52-1332"></a>    }</span>
<span id="cb52-1333"><a href="#cb52-1333"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-display">
<div id="ojs-cell-8" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell">
<details class="hidden"><summary>Code</summary><div class="sourceCode cell-code hidden" id="cb53" data-startfrom="1338" data-source-offset="0" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource js number-lines code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 1337;"><span id="cb53-1338"><a href="#cb53-1338"></a>mean_age_arrs <span class="op">=</span> <span class="fu">calc_mean_age_arr</span>(vacc<span class="op">,</span> lifeexp<span class="op">,</span> <span class="fl">1.0</span><span class="op">,</span> <span class="fl">10.0</span><span class="op">,</span> <span class="fl">0.01</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-display">
<div id="ojs-cell-9" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell">
<details class="hidden"><summary>Code</summary><div class="sourceCode cell-code hidden" id="cb54" data-startfrom="1343" data-source-offset="0" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource js number-lines code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 1342;"><span id="cb54-1343"><a href="#cb54-1343"></a>mean_age_dots <span class="op">=</span> [({</span>
<span id="cb54-1344"><a href="#cb54-1344"></a>    <span class="dt">arrow_start</span><span class="op">:</span> R0_mean_age <span class="op">&lt;=</span> <span class="dv">100</span> <span class="op">?</span> R0_mean_age <span class="op">:</span> <span class="dv">100</span><span class="op">,</span></span>
<span id="cb54-1345"><a href="#cb54-1345"></a>    <span class="dt">arrow_end</span><span class="op">:</span> Re_mean_age <span class="op">&lt;=</span> <span class="dv">100</span> <span class="op">?</span> Re_mean_age <span class="op">:</span> <span class="dv">100</span><span class="op">,</span></span>
<span id="cb54-1346"><a href="#cb54-1346"></a>    <span class="dt">R0</span><span class="op">:</span> R0<span class="op">.</span><span class="fu">toPrecision</span>(<span class="dv">3</span>)<span class="op">,</span></span>
<span id="cb54-1347"><a href="#cb54-1347"></a>    <span class="dt">Re</span><span class="op">:</span> Re<span class="op">.</span><span class="fu">toPrecision</span>(<span class="dv">3</span>)<span class="op">,</span></span>
<span id="cb54-1348"><a href="#cb54-1348"></a>    R0_mean_age<span class="op">,</span></span>
<span id="cb54-1349"><a href="#cb54-1349"></a>    Re_mean_age</span>
<span id="cb54-1350"><a href="#cb54-1350"></a>})]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-display">
<div id="ojs-cell-10" data-nodetype="declaration">

</div>
</div>
</div>
<div class="panel-fill panel-grid">
<div class="g-col-24">
<div class="cell panel-fill">
<details class="hidden"><summary>Code</summary><div class="sourceCode cell-code hidden" id="cb55" data-startfrom="1356" data-source-offset="0" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource js number-lines code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 1355;"><span id="cb55-1356"><a href="#cb55-1356"></a>{</span>
<span id="cb55-1357"><a href="#cb55-1357"></a>    <span class="kw">let</span> R0Color <span class="op">=</span> <span class="st">"#1f77b4"</span></span>
<span id="cb55-1358"><a href="#cb55-1358"></a>    <span class="kw">let</span> ReColor <span class="op">=</span> <span class="st">"#ff7f0e"</span></span>
<span id="cb55-1359"><a href="#cb55-1359"></a></span>
<span id="cb55-1360"><a href="#cb55-1360"></a>    <span class="kw">let</span> plot <span class="op">=</span> Plot<span class="op">.</span><span class="fu">plot</span>({</span>
<span id="cb55-1361"><a href="#cb55-1361"></a>        <span class="dt">color</span><span class="op">:</span> {</span>
<span id="cb55-1362"><a href="#cb55-1362"></a>            <span class="dt">legend</span><span class="op">:</span> <span class="kw">true</span><span class="op">,</span></span>
<span id="cb55-1363"><a href="#cb55-1363"></a>            <span class="dt">domain</span><span class="op">:</span> [<span class="st">"Unvaccinated"</span><span class="op">,</span> <span class="st">"Vaccinated"</span>]<span class="op">,</span></span>
<span id="cb55-1364"><a href="#cb55-1364"></a>            <span class="dt">range</span><span class="op">:</span> [<span class="st">"#1f77b4"</span><span class="op">,</span> <span class="st">"#ff7f0e"</span>]</span>
<span id="cb55-1365"><a href="#cb55-1365"></a>        }<span class="op">,</span></span>
<span id="cb55-1366"><a href="#cb55-1366"></a>        <span class="dt">style</span><span class="op">:</span> {<span class="dt">fontSize</span><span class="op">:</span> <span class="st">"20px"</span>}<span class="op">,</span></span>
<span id="cb55-1367"><a href="#cb55-1367"></a>        <span class="dt">marginLeft</span><span class="op">:</span> <span class="dv">65</span><span class="op">,</span></span>
<span id="cb55-1368"><a href="#cb55-1368"></a>        <span class="dt">marginTop</span><span class="op">:</span> <span class="dv">40</span><span class="op">,</span></span>
<span id="cb55-1369"><a href="#cb55-1369"></a>        <span class="dt">marginBottom</span><span class="op">:</span> <span class="dv">55</span><span class="op">,</span></span>
<span id="cb55-1370"><a href="#cb55-1370"></a>        <span class="dt">grid</span><span class="op">:</span> <span class="kw">true</span><span class="op">,</span></span>
<span id="cb55-1371"><a href="#cb55-1371"></a>        <span class="dt">width</span><span class="op">:</span> <span class="dv">800</span><span class="op">,</span></span>
<span id="cb55-1372"><a href="#cb55-1372"></a>        <span class="dt">height</span><span class="op">:</span> <span class="dv">670</span><span class="op">,</span></span>
<span id="cb55-1373"><a href="#cb55-1373"></a>        <span class="dt">x</span><span class="op">:</span> {<span class="dt">label</span><span class="op">:</span> <span class="st">"R0"</span><span class="op">,</span> <span class="dt">domain</span><span class="op">:</span> [<span class="dv">0</span><span class="op">,</span> <span class="dv">10</span>]}<span class="op">,</span></span>
<span id="cb55-1374"><a href="#cb55-1374"></a>        <span class="dt">y</span><span class="op">:</span> {<span class="dt">label</span><span class="op">:</span> <span class="st">"Mean Age of Infection"</span><span class="op">,</span> <span class="dt">domain</span><span class="op">:</span> [<span class="dv">0</span><span class="op">,</span> <span class="dv">100</span>]}<span class="op">,</span></span>
<span id="cb55-1375"><a href="#cb55-1375"></a>        <span class="dt">marks</span><span class="op">:</span> [</span>
<span id="cb55-1376"><a href="#cb55-1376"></a>            Plot<span class="op">.</span><span class="fu">line</span>(mean_age_arrs<span class="op">.</span><span class="at">Re</span><span class="op">,</span> {<span class="dt">x</span><span class="op">:</span> <span class="st">"R0"</span><span class="op">,</span> <span class="dt">y</span><span class="op">:</span> <span class="st">"mean_age"</span><span class="op">,</span> <span class="dt">stroke</span><span class="op">:</span> ReColor<span class="op">,</span> <span class="dt">strokeWidth</span><span class="op">:</span> <span class="dv">6</span>})<span class="op">,</span></span>
<span id="cb55-1377"><a href="#cb55-1377"></a>            Plot<span class="op">.</span><span class="fu">line</span>(mean_age_arrs<span class="op">.</span><span class="at">R0</span><span class="op">,</span> {<span class="dt">x</span><span class="op">:</span> <span class="st">"R0"</span><span class="op">,</span> <span class="dt">y</span><span class="op">:</span> <span class="st">"mean_age"</span><span class="op">,</span> <span class="dt">stroke</span><span class="op">:</span> R0Color<span class="op">,</span> <span class="dt">strokeWidth</span><span class="op">:</span> <span class="dv">6</span>})<span class="op">,</span></span>
<span id="cb55-1378"><a href="#cb55-1378"></a>            Re_mean_age <span class="op">&lt;=</span> <span class="dv">100</span> <span class="op">?</span></span>
<span id="cb55-1379"><a href="#cb55-1379"></a>                [</span>
<span id="cb55-1380"><a href="#cb55-1380"></a>                    vacc <span class="op">&gt;</span> <span class="fl">0.00</span> <span class="op">?</span> Plot<span class="op">.</span><span class="fu">dot</span>(mean_age_dots<span class="op">,</span> {<span class="dt">x</span><span class="op">:</span> <span class="st">"R0"</span><span class="op">,</span> <span class="dt">y</span><span class="op">:</span> <span class="st">"Re_mean_age"</span><span class="op">,</span> <span class="dt">r</span><span class="op">:</span> <span class="dv">12</span><span class="op">,</span> <span class="dt">stroke</span><span class="op">:</span> ReColor<span class="op">,</span> <span class="dt">fill</span><span class="op">:</span>     ReColor<span class="op">,</span>  <span class="dt">fillOpacity</span><span class="op">:</span> <span class="fl">0.6</span>}) <span class="op">:</span> <span class="kw">null</span><span class="op">,</span></span>
<span id="cb55-1381"><a href="#cb55-1381"></a>                    Plot<span class="op">.</span><span class="fu">text</span>(</span>
<span id="cb55-1382"><a href="#cb55-1382"></a>                        mean_age_dots<span class="op">,</span></span>
<span id="cb55-1383"><a href="#cb55-1383"></a>                        {<span class="dt">x</span><span class="op">:</span> <span class="st">"R0"</span><span class="op">,</span> <span class="dt">y</span><span class="op">:</span> <span class="st">"Re_mean_age"</span><span class="op">,</span> <span class="dt">text</span><span class="op">:</span> (d) <span class="kw">=&gt;</span> <span class="vs">`Re = </span><span class="sc">${</span>d<span class="op">.</span><span class="at">Re</span><span class="sc">}</span><span class="vs">`</span><span class="op">,</span> <span class="dt">dx</span><span class="op">:</span> <span class="dv">55</span><span class="op">,</span> <span class="dt">dy</span><span class="op">:</span> <span class="op">-</span><span class="dv">25</span><span class="op">,</span> <span class="dt">fontWeight</span><span class="op">:</span> <span class="st">"bold"</span><span class="op">,</span> <span class="dt">fill</span><span class="op">:</span> ReColor}</span>
<span id="cb55-1384"><a href="#cb55-1384"></a>                    )</span>
<span id="cb55-1385"><a href="#cb55-1385"></a>                ] <span class="op">:</span></span>
<span id="cb55-1386"><a href="#cb55-1386"></a>            <span class="kw">null</span><span class="op">,</span></span>
<span id="cb55-1387"><a href="#cb55-1387"></a>            R0_mean_age <span class="op">&lt;=</span> <span class="dv">100</span> <span class="op">?</span></span>
<span id="cb55-1388"><a href="#cb55-1388"></a>                [</span>
<span id="cb55-1389"><a href="#cb55-1389"></a>                Plot<span class="op">.</span><span class="fu">dot</span>(mean_age_dots<span class="op">,</span> {<span class="dt">x</span><span class="op">:</span> <span class="st">"R0"</span><span class="op">,</span> <span class="dt">y</span><span class="op">:</span> <span class="st">"R0_mean_age"</span><span class="op">,</span> <span class="dt">r</span><span class="op">:</span> <span class="dv">12</span><span class="op">,</span> <span class="dt">stroke</span><span class="op">:</span> R0Color<span class="op">,</span> <span class="dt">fill</span><span class="op">:</span> R0Color<span class="op">,</span> <span class="dt">fillOpacity</span><span class="op">:</span> <span class="fl">0.6</span>})<span class="op">,</span></span>
<span id="cb55-1390"><a href="#cb55-1390"></a>                Plot<span class="op">.</span><span class="fu">text</span>(</span>
<span id="cb55-1391"><a href="#cb55-1391"></a>                    mean_age_dots<span class="op">,</span></span>
<span id="cb55-1392"><a href="#cb55-1392"></a>                    {<span class="dt">x</span><span class="op">:</span> <span class="st">"R0"</span><span class="op">,</span> <span class="dt">y</span><span class="op">:</span> <span class="st">"R0_mean_age"</span><span class="op">,</span> <span class="dt">text</span><span class="op">:</span> (d) <span class="kw">=&gt;</span> <span class="vs">`R0 = </span><span class="sc">${</span>d<span class="op">.</span><span class="at">R0</span><span class="sc">}</span><span class="vs">`</span><span class="op">,</span> <span class="dt">dx</span><span class="op">:</span> <span class="op">-</span><span class="dv">60</span><span class="op">,</span> <span class="dt">dy</span><span class="op">:</span> <span class="dv">30</span><span class="op">,</span> <span class="dt">fontWeight</span><span class="op">:</span> <span class="st">"bold"</span><span class="op">,</span> <span class="dt">fill</span><span class="op">:</span> R0Color}</span>
<span id="cb55-1393"><a href="#cb55-1393"></a>                )</span>
<span id="cb55-1394"><a href="#cb55-1394"></a>                ] <span class="op">:</span></span>
<span id="cb55-1395"><a href="#cb55-1395"></a>            <span class="kw">null</span><span class="op">,</span></span>
<span id="cb55-1396"><a href="#cb55-1396"></a>            Plot<span class="op">.</span><span class="fu">arrow</span>(mean_age_dots<span class="op">,</span> {<span class="dt">x1</span><span class="op">:</span> <span class="st">"R0"</span><span class="op">,</span> <span class="dt">x2</span><span class="op">:</span> <span class="st">"R0"</span><span class="op">,</span> <span class="dt">y1</span><span class="op">:</span> <span class="st">"arrow_start"</span><span class="op">,</span> <span class="dt">y2</span><span class="op">:</span> <span class="st">"arrow_end"</span><span class="op">,</span> <span class="dt">strokeWidth</span><span class="op">:</span> <span class="dv">4</span><span class="op">,</span> <span class="dt">headLength</span><span class="op">:</span> <span class="dv">5</span><span class="op">,</span> <span class="dt">inset</span><span class="op">:</span> <span class="dv">15</span>})<span class="op">,</span></span>
<span id="cb55-1397"><a href="#cb55-1397"></a>        ]</span>
<span id="cb55-1398"><a href="#cb55-1398"></a>    })<span class="op">;</span></span>
<span id="cb55-1399"><a href="#cb55-1399"></a></span>
<span id="cb55-1400"><a href="#cb55-1400"></a>  <span class="cf">return</span> plot<span class="op">;</span></span>
<span id="cb55-1401"><a href="#cb55-1401"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-display">
<div id="ojs-cell-11" data-nodetype="expression">

</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section></section></section><section id="bonus-materials" class="level2 page-columns page-full" data-number="7.6"><h2 data-number="7.6" class="anchored" data-anchor-id="bonus-materials">
<span class="header-section-number">7.6</span> Bonus Materials</h2>
<section id="calculating-r_0-with-the-next-generation-matrix" class="level3 page-columns page-full" data-number="7.6.1"><h3 data-number="7.6.1" class="anchored" data-anchor-id="calculating-r_0-with-the-next-generation-matrix">
<span class="header-section-number">7.6.1</span> Calculating <span class="math inline">\(R_0\)</span> with the Next Generation Matrix</h3>
<section id="sec-simple-ngm" class="level4" data-number="7.6.1.1"><h4 data-number="7.6.1.1" class="anchored" data-anchor-id="sec-simple-ngm">
<span class="header-section-number">7.6.1.1</span> Simple model structure</h4>
<p>To compute <span class="math inline">\(R_0\)</span>, we need to know the stable age distribution (the relative proportion in the juvenile and adult age classes) of the population, which we can find by solving for the disease-free equilibrium: <span class="math inline">\(S_J^*=B/\alpha\)</span> and <span class="math inline">\(S_A^*=B/\mu\)</span>. With the stable age distribution, we can calculate <span class="math inline">\(R_0\)</span> by constructing the next generation matrix. The code below outlines how the next generation matrix is constructed using the <span class="math inline">\(\alpha\)</span> (aging from juvenile to adult), <span class="math inline">\(\mu\)</span> (death), <span class="math inline">\(n\)</span> (total births), <span class="math inline">\(\gamma\)</span> (recovery), <span class="math inline">\(da\)</span> (width of age groups in years), and <span class="math inline">\(\beta\)</span> (transmission) parameters.</p>
<p>The next generation matrix is a matrix that specifies how many new age-specific infections are generated by a typical infected individual of each age class (in a fully susceptible population). For example, let’s consider an infected adult and ask how many new juvenile infections it generates: this is the product of the number of susceptible juveniles (from the stable age distribution), the per capita transmission rate from adults to juveniles and the average duration of infection, i.e.&nbsp;<span class="math inline">\(S_J^* \times \beta_{JA} \times 1/ (\gamma+\mu)\)</span>. This forms one element of our next generation matrix. The other elements look very similar, except there are extra terms when we consider an infected juvenile because there is a (very small) chance they may age during the infectious period and therefore cause new infections as an adult:</p>
<p><span id="eq-simple-ngm"><span class="math display">\[
\mathrm{NGM} = \begin{pmatrix}
        \frac{S_J^* \beta_{JJ}}{(\gamma + \alpha)} +
        \frac{\alpha}{(\gamma+\mu)} \frac{S_J^* \beta_{JA}}{(\gamma + \mu)} &amp;
        \frac{S_J^* \beta_{JA}}{(\gamma + \mu)} \\
        \frac{S_A^* \beta_{AJ}}{(\gamma + \alpha)} +
            \frac{\alpha}{(\gamma + \mu)} \frac{S_A^*\beta_{AA}}{(\gamma+\mu)} &amp;
            \frac{S_A^* \beta_{AA}}{(\gamma + \mu)}
    \end{pmatrix}
\tag{7.3}\]</span></span></p>
<p><span class="math inline">\(R_0\)</span> can then be computed as the dominant eigenvalue (i.e., the one with the largest real part) of this matrix. Let’s take an example from a model with 2 age classes, from above. First, let’s define the components of the next generation matrix:</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-57_ee102dd65d25f69ab8ef614552720657">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb56" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1"></a>ngm_params <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb56-2"><a href="#cb56-2"></a>    <span class="at">beta_within =</span> <span class="fl">0.011</span>,</span>
<span id="cb56-3"><a href="#cb56-3"></a>    <span class="at">beta_between =</span> <span class="fl">0.005</span>,</span>
<span id="cb56-4"><a href="#cb56-4"></a>    <span class="at">age_band_j =</span> <span class="dv">20</span>,</span>
<span id="cb56-5"><a href="#cb56-5"></a>    <span class="at">age_band_a =</span> <span class="dv">60</span>,</span>
<span id="cb56-6"><a href="#cb56-6"></a>    <span class="at">recovery =</span> <span class="dv">10</span></span>
<span id="cb56-7"><a href="#cb56-7"></a>)</span>
<span id="cb56-8"><a href="#cb56-8"></a></span>
<span id="cb56-9"><a href="#cb56-9"></a>alpha_ngm <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> ngm_params[<span class="st">"age_band_j"</span>]</span>
<span id="cb56-10"><a href="#cb56-10"></a>mu_ngm <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> ngm_params[<span class="st">"age_band_a"</span>]</span>
<span id="cb56-11"><a href="#cb56-11"></a>n_ngm <span class="ot">&lt;-</span> demog_params[<span class="st">"births"</span>] <span class="sc">/</span> <span class="fu">c</span>(alpha_ngm, mu_ngm)</span>
<span id="cb56-12"><a href="#cb56-12"></a></span>
<span id="cb56-13"><a href="#cb56-13"></a>beta_ngm <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(</span>
<span id="cb56-14"><a href="#cb56-14"></a>    ngm_params[<span class="st">"beta_within"</span>],</span>
<span id="cb56-15"><a href="#cb56-15"></a>    ngm_params[<span class="st">"beta_between"</span>],</span>
<span id="cb56-16"><a href="#cb56-16"></a>    ngm_params[<span class="st">"beta_between"</span>],</span>
<span id="cb56-17"><a href="#cb56-17"></a>    ngm_params[<span class="st">"beta_within"</span>]</span>
<span id="cb56-18"><a href="#cb56-18"></a>    ),</span>
<span id="cb56-19"><a href="#cb56-19"></a>    <span class="at">nrow =</span> <span class="dv">2</span>,</span>
<span id="cb56-20"><a href="#cb56-20"></a>    <span class="at">ncol =</span> <span class="dv">2</span></span>
<span id="cb56-21"><a href="#cb56-21"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>The Next Generation Matrix can be calculated in <code>R</code> as:</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-58_b32533a48edcf60265ad649b38a85c0b">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb57" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1"></a>ngm <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb57-2"><a href="#cb57-2"></a>    <span class="fu">c</span>(</span>
<span id="cb57-3"><a href="#cb57-3"></a>        n_ngm[<span class="dv">1</span>] <span class="sc">*</span> (beta_ngm[<span class="dv">1</span>, <span class="dv">1</span>] <span class="sc">/</span> (ngm_params[<span class="st">"recovery"</span>] <span class="sc">+</span> alpha_ngm)) <span class="sc">+</span></span>
<span id="cb57-4"><a href="#cb57-4"></a>            alpha_ngm <span class="sc">/</span> (ngm_params[<span class="st">"recovery"</span>] <span class="sc">+</span> mu_ngm) <span class="sc">*</span></span>
<span id="cb57-5"><a href="#cb57-5"></a>            n_ngm[<span class="dv">1</span>] <span class="sc">*</span> beta_ngm[<span class="dv">1</span>, <span class="dv">2</span>] <span class="sc">/</span> (ngm_params[<span class="st">"recovery"</span>] <span class="sc">+</span> mu_ngm),</span>
<span id="cb57-6"><a href="#cb57-6"></a>        </span>
<span id="cb57-7"><a href="#cb57-7"></a>        n_ngm[<span class="dv">2</span>] <span class="sc">*</span> beta_ngm[<span class="dv">2</span>, <span class="dv">1</span>] <span class="sc">/</span> (ngm_params[<span class="st">"recovery"</span>] <span class="sc">+</span> alpha_ngm) <span class="sc">+</span></span>
<span id="cb57-8"><a href="#cb57-8"></a>            alpha_ngm <span class="sc">/</span> (ngm_params[<span class="st">"recovery"</span>] <span class="sc">+</span> mu_ngm) <span class="sc">*</span></span>
<span id="cb57-9"><a href="#cb57-9"></a>            n_ngm[<span class="dv">2</span>] <span class="sc">*</span> (beta_ngm[<span class="dv">2</span>, <span class="dv">2</span>] <span class="sc">/</span> (ngm_params[<span class="st">"recovery"</span>] <span class="sc">+</span> mu_ngm)),</span>
<span id="cb57-10"><a href="#cb57-10"></a></span>
<span id="cb57-11"><a href="#cb57-11"></a>        n_ngm[<span class="dv">1</span>] <span class="sc">*</span> beta_ngm[<span class="dv">1</span>, <span class="dv">2</span>] <span class="sc">/</span> (ngm_params[<span class="st">"recovery"</span>] <span class="sc">+</span> mu_ngm),</span>
<span id="cb57-12"><a href="#cb57-12"></a></span>
<span id="cb57-13"><a href="#cb57-13"></a>        n_ngm[<span class="dv">2</span>] <span class="sc">*</span> beta_ngm[<span class="dv">2</span>, <span class="dv">2</span>] <span class="sc">/</span> (ngm_params[<span class="st">"recovery"</span>] <span class="sc">+</span> mu_ngm)</span>
<span id="cb57-14"><a href="#cb57-14"></a>    ),</span>
<span id="cb57-15"><a href="#cb57-15"></a>    <span class="at">nrow =</span> <span class="dv">2</span>,</span>
<span id="cb57-16"><a href="#cb57-16"></a>    <span class="at">ncol =</span> <span class="dv">2</span></span>
<span id="cb57-17"><a href="#cb57-17"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We can then calculate the eigenvalues and eigenvectors of this matrix:</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-59_7458e0b54ca680459b7bb073d6def9dc">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb58" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1"></a><span class="fu">eigen</span>(ngm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre data-code-line-numbers=""><code>eigen() decomposition
$values
[1] 7.191869 1.591188

$vectors
           [,1]       [,2]
[1,] -0.1958841 -0.8560336
[2,] -0.9806271  0.5169202</code></pre>
</div>
</div>
<p>We can also choose to just output the eigenvalues:</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-60_d88f8804b80086f52144454307b96c8a">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb60" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1"></a><span class="fu">eigen</span>(ngm, <span class="at">only.values =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre data-code-line-numbers=""><code>$values
[1] 7.191869 1.591188

$vectors
NULL</code></pre>
</div>
</div>
<p>Finally, let’s print <span class="math inline">\(R_0\)</span>:</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-61_41c42a2218abd901696a6f5dbb8be912">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb62" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1"></a><span class="fu">max</span>(</span>
<span id="cb62-2"><a href="#cb62-2"></a>    <span class="fu">Re</span>(</span>
<span id="cb62-3"><a href="#cb62-3"></a>        <span class="fu">eigen</span>(ngm, <span class="at">only.values =</span> <span class="cn">TRUE</span>)<span class="sc">$</span>values</span>
<span id="cb62-4"><a href="#cb62-4"></a>    )</span>
<span id="cb62-5"><a href="#cb62-5"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre data-code-line-numbers=""><code>[1] 7.191869</code></pre>
</div>
</div>
</section><section id="sec-age-structure-ngm" class="level4 page-columns page-full" data-number="7.6.1.2"><h4 data-number="7.6.1.2" class="anchored" data-anchor-id="sec-age-structure-ngm">
<span class="header-section-number">7.6.1.2</span> Age-structured NGM</h4>
<p>Many times it would be impractical to write out the NGM: there are often too many compartments in an age-structured model. In this instance, we want to use a slightly different approach, but the underlying principles are the same: each element of the NGM balances the number of new infections expected to be produced with the rates of individuals coming in and out of that compartment.</p>
<section id="stable-age-distribution" class="level5 page-columns page-full" data-number="7.6.1.2.1"><h5 data-number="7.6.1.2.1" class="anchored" data-anchor-id="stable-age-distribution">
<span class="header-section-number">7.6.1.2.1</span> Stable Age Distribution</h5>
<p>The first thing we need, as before, is the stable age distribution i.e., the disease-free equilibrium. There are two ways we can do this:</p>
<ol type="1">
<li>Simulate the model without any infections for a sufficiently long time (simple, but less accurate)</li>
<li>Do the math.</li>
</ol>
<section id="disease-free-simulation" class="level6 page-columns page-full" data-number="7.6.1.2.1.1"><h6 data-number="7.6.1.2.1.1" class="anchored" data-anchor-id="disease-free-simulation">
<span class="header-section-number">7.6.1.2.1.1</span> Disease-Free Simulation</h6>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-62_43dfa25a09484e9978556fe9a80e52e3">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb64" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1"></a><span class="co"># Set up initial conditions without any infections</span></span>
<span id="cb64-2"><a href="#cb64-2"></a>multistage_sonly_yinit <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb64-3"><a href="#cb64-3"></a>    <span class="at">S =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">250</span>, <span class="dv">30</span>)),</span>
<span id="cb64-4"><a href="#cb64-4"></a>    <span class="at">I =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">30</span>)),</span>
<span id="cb64-5"><a href="#cb64-5"></a>    <span class="at">R =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">30</span>))</span>
<span id="cb64-6"><a href="#cb64-6"></a>)</span>
<span id="cb64-7"><a href="#cb64-7"></a></span>
<span id="cb64-8"><a href="#cb64-8"></a><span class="co"># Solve disease free sim to get dfe</span></span>
<span id="cb64-9"><a href="#cb64-9"></a>multistage_sonly_sol <span class="ot">&lt;-</span> deSolve<span class="sc">::</span><span class="fu">ode</span>(</span>
<span id="cb64-10"><a href="#cb64-10"></a>    <span class="at">y =</span> multistage_sonly_yinit,</span>
<span id="cb64-11"><a href="#cb64-11"></a>    <span class="at">times =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">300</span>, <span class="at">by =</span> <span class="dv">1</span>),</span>
<span id="cb64-12"><a href="#cb64-12"></a>    <span class="at">func =</span> multistage_model,</span>
<span id="cb64-13"><a href="#cb64-13"></a>    <span class="at">parms =</span> multistage_params</span>
<span id="cb64-14"><a href="#cb64-14"></a>)</span>
<span id="cb64-15"><a href="#cb64-15"></a></span>
<span id="cb64-16"><a href="#cb64-16"></a><span class="co"># Calculate population size at each time point and save to dataframe</span></span>
<span id="cb64-17"><a href="#cb64-17"></a>multistage_sonly_pop <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb64-18"><a href="#cb64-18"></a>    <span class="at">time =</span> multistage_sonly_sol[, <span class="dv">1</span>],</span>
<span id="cb64-19"><a href="#cb64-19"></a>    <span class="at">pop =</span> <span class="fu">apply</span>(multistage_sonly_sol[, <span class="sc">-</span><span class="dv">1</span>], <span class="dv">1</span> , sum)</span>
<span id="cb64-20"><a href="#cb64-20"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell page-columns page-full" data-hash="r-session-02_cache/html/unnamed-chunk-63_9db2cc67b9718812b3280213403a7336">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb65" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1"></a><span class="fu">ggplot</span>(multistage_sonly_pop, <span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> pop)) <span class="sc">+</span></span>
<span id="cb65-2"><a href="#cb65-2"></a>    <span class="fu">geom_area</span>(<span class="at">fill =</span> SIRcolors[<span class="dv">4</span>], <span class="at">alpha =</span> <span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb65-3"><a href="#cb65-3"></a>    <span class="fu">labs</span>(</span>
<span id="cb65-4"><a href="#cb65-4"></a>        <span class="at">x =</span> <span class="st">"Time"</span>,</span>
<span id="cb65-5"><a href="#cb65-5"></a>        <span class="at">y =</span> <span class="st">"Population size"</span></span>
<span id="cb65-6"><a href="#cb65-6"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display column-body">
<p><img src="r-session-02_files/figure-html/unnamed-chunk-63-1.png" class="img-fluid" style="width:100.0%"></p>
</div>
</div>
</section><section id="doing-the-math" class="level6" data-number="7.6.1.2.1.2"><h6 data-number="7.6.1.2.1.2" class="anchored" data-anchor-id="doing-the-math">
<span class="header-section-number">7.6.1.2.1.2</span> Doing the Math</h6>
<p>Alternatively, we can get the stable age distribution by finding the population structure that balances the birth, aging, and death processes. We have already seen the aging matrix in <a href="#eq-aging-mat">Equation&nbsp;<span>7.1</span></a>, and at equilibrium, we have the matrix equation</p>
<p><span class="math display">\[
\begin{pmatrix}
    -\alpha_1 &amp; 0 &amp; 0 &amp; \cdots &amp; 0\\
    \alpha_1 &amp; -\alpha_2 &amp; 0 &amp; \cdots &amp; 0\\
    0 &amp; \alpha_2 &amp; -\alpha_3 &amp; \cdots &amp; 0\\
    \vdots &amp;  &amp; \ddots &amp; \ddots &amp; \vdots \\
    0 &amp; \cdots &amp; &amp; \alpha_{29} &amp; -\alpha_{30}\\
\end{pmatrix} .
\begin{pmatrix}
    n_1 \\ n_2 \\ n_3 \\ \vdots \\ n_{30}
\end{pmatrix} +
\begin{pmatrix}
    B \\ 0 \\ 0 \\ \vdots \\ 0
\end{pmatrix} =
\begin{pmatrix}
    0 \\ 0 \\ 0 \\ \vdots \\ 0
\end{pmatrix}
\]</span></p>
<p>To solve this equation in <code>R</code>, we can do</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-64_ea4abb60ee20b446c8c3a72f954db85c">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb66" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1"></a><span class="co"># solve(a, b) solves the equation a %*% x = b for x, so rearrange equation above so b is on the RHS of the equation</span></span>
<span id="cb66-2"><a href="#cb66-2"></a>multistage_stable_n <span class="ot">&lt;-</span> <span class="fu">solve</span>(</span>
<span id="cb66-3"><a href="#cb66-3"></a>    aging_mat,</span>
<span id="cb66-4"><a href="#cb66-4"></a>    <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span> <span class="sc">*</span> multistage_params[[<span class="st">"births"</span>]], <span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">29</span>))</span>
<span id="cb66-5"><a href="#cb66-5"></a>)</span>
<span id="cb66-6"><a href="#cb66-6"></a></span>
<span id="cb66-7"><a href="#cb66-7"></a>multistage_stable_n</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre data-code-line-numbers=""><code> [1]  100  100  100  100  100  100  100  100  100  100  100  100  100  100  100
[16]  100  100  100  100  100  500  500  500  500  500  500  500  500  500 1500</code></pre>
</div>
</div>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-65_72f861ef3c1c0da4d7c74e78b898cce4">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb68" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1"></a><span class="co"># Check the final pop value of the S-only sim is equal to the sum of the stable age distribution calculated above</span></span>
<span id="cb68-2"><a href="#cb68-2"></a><span class="fu">round</span>(<span class="fu">tail</span>(multistage_sonly_pop<span class="sc">$</span>pop, <span class="dv">1</span>)) <span class="sc">==</span> <span class="fu">sum</span>(multistage_stable_n)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre data-code-line-numbers=""><code>[1] TRUE</code></pre>
</div>
</div>
<p>The following lines then compute <span class="math inline">\(R_0\)</span> using the next generation matrix method. This calculation comes from a recipe described in detail previously <span class="citation" data-cites="diekmann2000mathematical heesterbeekBriefHistoryR02002 bjornstadAdvancedNextGenerationMatrix2018 heffernanPerspectivesBasicReproductive2005 hurfordNextgenerationToolsEvolutionary2009">(<a href="references.html#ref-diekmann2000mathematical" role="doc-biblioref">Diekmann and Heesterbeek 2000</a>; <a href="references.html#ref-heesterbeekBriefHistoryR02002" role="doc-biblioref">Heesterbeek 2002</a>; <a href="references.html#ref-bjornstadAdvancedNextGenerationMatrix2018" role="doc-biblioref">Bjørnstad 2018</a>; <a href="references.html#ref-heffernanPerspectivesBasicReproductive2005" role="doc-biblioref">Heffernan, Smith, and Wahl 2005</a>; <a href="references.html#ref-hurfordNextgenerationToolsEvolutionary2009" role="doc-biblioref">Hurford, Cownden, and Day 2009</a>)</span> (we would recommend starting with <span class="citation" data-cites="bjornstadAdvancedNextGenerationMatrix2018 heffernanPerspectivesBasicReproductive2005">(<a href="references.html#ref-bjornstadAdvancedNextGenerationMatrix2018" role="doc-biblioref">Bjørnstad 2018</a>; and <a href="references.html#ref-heffernanPerspectivesBasicReproductive2005" role="doc-biblioref">Heffernan, Smith, and Wahl 2005</a>)</span>).</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>This is Aaron’s original code and gives slightly different results than the calculations below.</p>
</div>
</div>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-66_a50e77c5dedd683abd886a6d7854dadb">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb70" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1"></a>F_mat <span class="ot">&lt;-</span> <span class="fu">diag</span>(multistage_stable_n) <span class="sc">%*%</span> multistage_params[[<span class="st">"beta_mat"</span>]] <span class="sc">+</span></span>
<span id="cb70-2"><a href="#cb70-2"></a>    multistage_params[[<span class="st">"aging_mat"</span>]] <span class="sc">-</span></span>
<span id="cb70-3"><a href="#cb70-3"></a>    <span class="fu">diag</span>(<span class="fu">diag</span>(multistage_params[[<span class="st">"aging_mat"</span>]]))</span>
<span id="cb70-4"><a href="#cb70-4"></a></span>
<span id="cb70-5"><a href="#cb70-5"></a>V_mat <span class="ot">&lt;-</span> <span class="fu">diag</span>(multistage_params[[<span class="st">"recovery"</span>]] <span class="sc">-</span> <span class="fu">diag</span>(multistage_params[[<span class="st">"aging_mat"</span>]]))</span>
<span id="cb70-6"><a href="#cb70-6"></a></span>
<span id="cb70-7"><a href="#cb70-7"></a><span class="fu">max</span>(</span>
<span id="cb70-8"><a href="#cb70-8"></a>    <span class="fu">Re</span>(</span>
<span id="cb70-9"><a href="#cb70-9"></a>        <span class="fu">eigen</span>(<span class="fu">solve</span>(V_mat, F_mat), <span class="at">only.values =</span> <span class="cn">TRUE</span>)<span class="sc">$</span>values</span>
<span id="cb70-10"><a href="#cb70-10"></a>    )</span>
<span id="cb70-11"><a href="#cb70-11"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre data-code-line-numbers=""><code>[1] 6.792003</code></pre>
</div>
</div>
<p>The steps below are copied from <span class="citation" data-cites="bjornstadAdvancedNextGenerationMatrix2018">(<a href="references.html#ref-bjornstadAdvancedNextGenerationMatrix2018" role="doc-biblioref">Bjørnstad 2018</a>)</span></p>
<ol type="1">
<li>Identify all n infected compartments</li>
<li>Construct a n × 1 matrix, <span class="math inline">\(\mathbf{F}\)</span>, that contains expressions for all completely new infections entering each infected compartment</li>
<li>Construct a n × 1 matrix, <span class="math inline">\(\mathbf{V^−}\)</span>, that contains expressions for all losses out of each infected compartment</li>
<li>Construct a n × 1 matrix, <span class="math inline">\(\mathbf{V^+}\)</span>, that contains expressions for all gains into each infected compartment that does not represent new infections but transfers among infectious classes</li>
<li>Construct a n × 1 matrix, <span class="math inline">\(\mathbf{V} = \mathbf{V^−} − \mathbf{V^+}\)</span>
</li>
<li>Generate two n × n Jacobian matrices <span class="math inline">\(f\)</span> and <span class="math inline">\(v\)</span> that are the partial derivatives of <span class="math inline">\(\mathbf{F}\)</span> and <span class="math inline">\(\mathbf{V}\)</span> with respect to the <span class="math inline">\(n\)</span> infectious state variables</li>
<li>Evaluate the matrices at the disease free equilibrium (dfe), and finally</li>
<li>
<span class="math inline">\(R_0\)</span> is the spectral trace (greatest non-negative real eigenvalue) of <span class="math inline">\(\mathbf{fv}^{−1}|_{\text{dfe}}\)</span>.</li>
</ol>
<p>Working through these steps looks like this:</p>
<ol type="1">
<li>Our only infected compartments are the <span class="math inline">\(I_i\)</span> states, for each age group (<span class="math inline">\(i \in [1, 30]\)</span>) To start, let’s write out our differential equation:</li>
</ol>
<p><span class="math display">\[\begin{equation}
    \frac{\dd{I_i}}{\dd{t}} = \lambda_i S_i - \gamma I_i + \alpha_{i-1} I_{i-1} - \alpha_i I_i
\end{equation}\]</span></p>
<ol start="2" type="1">
<li>We’ll now calculate <span class="math inline">\(\mathbf{F}\)</span> and <span class="math inline">\(\mathbf{f}\)</span>
</li>
</ol>
<p><span class="math display">\[\begin{align*}
    \mathbf{F} &amp;= \begin{pmatrix}
        \lambda_1 S_1 + \cancelto{0}{\alpha_0 I_0} \\
        \vdots \\
        \lambda_{30} S_{30} + \alpha_{29} I_{29}
    \end{pmatrix} \\ \\
    \mathbf{F} &amp;= \begin{pmatrix}
        \left(\beta_{1, 1} I_1 + \cdots + \beta_{1, 30} I_{30} \right)S_1 \\
        \vdots \\
        \left(\beta_{30, 1} I_1 + \cdots + \beta_{30, 30} I_{30} \right) S_{30} + \alpha_{29} I_{29}
    \end{pmatrix} \\
    \mathbf{f} &amp;= \begin{pmatrix}
        \frac{\partial F_1}{\partial I_1} &amp; \cdots &amp; \frac{\partial F_1}{\partial I_{30}} \\
        \vdots &amp; \ddots &amp; \vdots \\
        \frac{\partial F_{30}}{\partial I_1} &amp; \cdots &amp; \frac{\partial F_{30}}{\partial I_{30}}
    \end{pmatrix} &amp; \frac{\partial F_1}{\partial I_1} &amp;= \frac{\partial}{\partial I_1} \left( \beta_{1, 1} I_1 + \cancelto{0}{\beta_{1, 2} I_2 + \cdots + \beta_{1, 30} I_{30}}\right) S_1 \\
    &amp; &amp; \frac{\partial F_1}{\partial I_1} &amp;= \beta_{1, 1} S_1 \\
    \mathbf{f} &amp;= \begin{pmatrix}
        \beta_{1, 1} S_1 &amp; \cdots &amp; \beta_{1, 30} S_1 \\
        \vdots &amp; \ddots &amp; \vdots \\
        \beta_{30, 1} S_{30} &amp; \cdots &amp; \beta_{30, 30} S_{30}
    \end{pmatrix}
\end{align*}\]</span></p>
<ol start="3" type="1">
<li>Now let’s calculate <span class="math inline">\(\mathbf{V^-}\)</span>, <span class="math inline">\(\mathbf{V^+}\)</span>, <span class="math inline">\(\mathbf{V}\)</span>, and <span class="math inline">\(\mathbf{v}\)</span>
</li>
</ol>
<p><span class="math display">\[\begin{align*}
    \mathbf{V^-} &amp;= \begin{pmatrix}
        \gamma I_1 + \alpha_1 I_1 \\
        \gamma I_2 + \alpha_2 I_2 \\
        \vdots \\
        \gamma I_{30} + \alpha_{30} I_{30}
    \end{pmatrix} &amp; \mathbf{V^+} &amp;= \begin{pmatrix}
        \cancelto{0}{\alpha_0 I_0} \\
        \alpha_1 I_1 \\
        \vdots \\
        \alpha_{29} I_{29}
    \end{pmatrix} \\ \\
    \mathbf{V} &amp;= \mathbf{V^-} - \mathbf{V^+} = \begin{pmatrix}
        \gamma I_1 + \alpha_1 I_1 \\
        \gamma I_2 + \alpha_2 I_2 - \alpha_1 I_1\\
        \vdots \\
        \gamma I_{30} + \alpha_{30} I_{30} - \alpha_{29} I_{29}
    \end{pmatrix} \\ \\
    \mathbf{v} &amp;= \begin{pmatrix}
        \pdv{V_1}{I_1} &amp; \cdots &amp; \pdv{V_1}{I_{30}} \\
        \vdots &amp; \ddots &amp; \vdots \\
        \pdv{V_{30}}{I_1} &amp; \cdots &amp; \pdv{V_{30}}{I_{30}}
    \end{pmatrix} &amp; \pdv{V_1}{I_1} &amp;= \pdv{I_1} I_1 \left( \gamma + \alpha_1  \right)\\
    &amp; &amp; \pdv{V_1}{I_1} &amp;= \gamma + \alpha_1 \\ \\
    &amp; &amp; \pdv{V_2}{I_1} &amp;= \pdv{I_1} \left( \cancelto{0}{I_2 \left( \gamma + \alpha_2  \right)} - \alpha_1 I_1 \right)\\
    &amp; &amp; \pdv{V_2}{I_1} &amp;= - \alpha_1 \\ \\
    &amp; &amp; \pdv{V_1}{I_2} &amp;= \pdv{I_2} \cancelto{0}{I_1 \left( \gamma + \alpha_1  \right)} \\
    &amp; &amp; \pdv{V_1}{I_2} &amp;= 0 \\ \\
    \mathbf{v} &amp;= \begin{pmatrix}
        \gamma + \alpha_1 &amp; 0 &amp;  \cdots  &amp; 0\\
        - \alpha_1 &amp; \gamma + \alpha_2 &amp; \cdots &amp; 0 \\
        \vdots &amp; \ddots &amp; \ddots &amp; \vdots \\
        0 &amp; \cdots &amp; - \alpha_{29} &amp; \gamma + \alpha_{30}
    \end{pmatrix}
\end{align*}\]</span></p>
<ol start="4" type="1">
<li>To evaluate <span class="math inline">\(\mathbf{f}\)</span> and <span class="math inline">\(\mathbf{v}\)</span> at the disease-free equilibrium, we can use the results from our previous calculations. <span class="math inline">\(\mathbf{v}\)</span> doesn’t have any state terms in the equation, so it is already evaluated at <span class="math inline">\(\text{dfe}\)</span>. <span class="math inline">\(\mathbf{f}|_{\text{dfe}}\)</span> involves subsituting <span class="math inline">\(S_i\)</span> for the equilibrium population distribution that balances the births and aging processes.</li>
</ol>
<p>This translates to the function we <a href="#ngm-function">defined earlier</a></p>
</section></section></section></section><section id="sec-mean-age-r-code" class="level3 page-columns page-full" data-number="7.6.2"><h3 data-number="7.6.2" class="anchored" data-anchor-id="sec-mean-age-r-code">
<span class="header-section-number">7.6.2</span> Mean age of infection <code>R</code> code</h3>
<p>Now let’s look at how we can investigate our the relationships between the mean age of infection and <span class="math inline">\(R_0\)</span> and the vaccination coverage using <code>R</code>. Unlike the interactive plot that simply uses <a href="#eq-mean-age">Equation&nbsp;<span>7.2</span></a> to calculate the mean age of infection, we will use a more realistic age-structured model.</p>
<p>Let’s return to the earlier models with an age-class mixing matrix. But this time, we’ll calculate <span class="math inline">\(R_0\)</span>, the mean age of infection, and the number of cases that occur in individuals between 15-35 years as we increase the contact rate.</p>
<p>Recall from the rubella and congenital rubella syndrome (CRS) example that the risk of severe disease outcomes depends on the risk of infection in reproductive age women (here we’ll use individuals between 15 and 35 years as a proxy; in reality we would want to account for the differential rate of reproduction at different ages, including those above 35 years). Recall also that increasing vaccination reduces <span class="math inline">\(R_E\)</span> – for simplicity here, so we don’t have to add vaccination into the code, we’ll simply change <span class="math inline">\(R_0\)</span> because we already know that will give us outcomes that are dynamically equivalent to increasing the proportion of children born who are vaccinated. We’ll then calculate how the mean age of infection changes, and specifically how the absolute number of cases among individuals between the ages of 15-35 (as a proxy for reproductive age women) changes. To do so, we’ll make a loop and evaluate the code for each of 10 decreaing levels of mixing (which will reduce <span class="math inline">\(R_0\)</span> and we can interpret as analogous to the reduction in <span class="math inline">\(R_E\)</span> that would result from increasing vaccination).</p>
<div class="callout callout-style-default callout-note callout-titled" title="Note about `map()`">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-8-contents" aria-controls="callout-8" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note about <code>map()</code>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-8" class="callout-8-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<p>As you may have noticed previously, we often use the <code>map_*()</code> series of functions. We’ll use that again here (<code>map_dfr()</code>). The full reasons are too complicated to get into here, but broadly speaking, the <code>map_*()</code> functions provide us guarantees over the output of our loops. If it runs, we know that something didn’t get silently skipped, and that out output vector/list/dataframe is the same length as the inputs. The same can not be said for <code>for()</code> loops, and the base <code>apply</code> functions are more awkward to work with as they don’t have a consistent syntax across the family of functions.</p>
<p>To learn more, read <a href="">this section</a> of our <code>R</code> primer.</p>
</div>
</div>
</div>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-67_5e6280265b924250770cfc0879538c7c">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb72" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1"></a><span class="co"># Create vector of scalings to reduce R0</span></span>
<span id="cb72-2"><a href="#cb72-2"></a>scale_contact <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">1</span>, .<span class="dv">2</span>, <span class="at">length =</span> <span class="dv">10</span>)</span>
<span id="cb72-3"><a href="#cb72-3"></a></span>
<span id="cb72-4"><a href="#cb72-4"></a><span class="co"># Create a new transmission matrix</span></span>
<span id="cb72-5"><a href="#cb72-5"></a>beta_low <span class="ot">&lt;-</span> <span class="fl">0.007</span></span>
<span id="cb72-6"><a href="#cb72-6"></a>beta_medium <span class="ot">&lt;-</span> <span class="fl">0.02</span></span>
<span id="cb72-7"><a href="#cb72-7"></a>beta_high <span class="ot">&lt;-</span> <span class="fl">0.03</span></span>
<span id="cb72-8"><a href="#cb72-8"></a></span>
<span id="cb72-9"><a href="#cb72-9"></a>beta_mat <span class="ot">&lt;-</span> <span class="fu">matrix</span>(beta_low, <span class="at">nrow =</span> <span class="dv">30</span>, <span class="at">ncol =</span> <span class="dv">30</span>)</span>
<span id="cb72-10"><a href="#cb72-10"></a>beta_mat[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>] <span class="ot">&lt;-</span> beta_medium</span>
<span id="cb72-11"><a href="#cb72-11"></a>beta_mat[<span class="dv">6</span><span class="sc">:</span><span class="dv">16</span>, <span class="dv">6</span><span class="sc">:</span><span class="dv">16</span>] <span class="ot">&lt;-</span> beta_high</span>
<span id="cb72-12"><a href="#cb72-12"></a></span>
<span id="cb72-13"><a href="#cb72-13"></a>scaled_params <span class="ot">&lt;-</span> multistage_params</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-68_c6a6c7a8a8608b288bf88a2cd30afbd6">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb73" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1"></a><span class="co"># Create a dataframe where each row relates to a different R0 value</span></span>
<span id="cb73-2"><a href="#cb73-2"></a>R0_mean_age_contacts_df <span class="ot">&lt;-</span> <span class="fu">map_dfr</span>(</span>
<span id="cb73-3"><a href="#cb73-3"></a>    <span class="co"># Apply the function to each item in the vector of R0 scaling factors</span></span>
<span id="cb73-4"><a href="#cb73-4"></a>    <span class="at">.x =</span> scale_contact,</span>
<span id="cb73-5"><a href="#cb73-5"></a>    <span class="at">.f =</span> <span class="cf">function</span>(.x) {</span>
<span id="cb73-6"><a href="#cb73-6"></a>        <span class="co"># Scale contacts</span></span>
<span id="cb73-7"><a href="#cb73-7"></a>        scaled_beta_mat <span class="ot">&lt;-</span> beta_mat <span class="sc">*</span> .x</span>
<span id="cb73-8"><a href="#cb73-8"></a></span>
<span id="cb73-9"><a href="#cb73-9"></a>        <span class="co"># Set up parameters</span></span>
<span id="cb73-10"><a href="#cb73-10"></a>        scaled_params[[<span class="st">"beta_mat"</span>]] <span class="ot">&lt;-</span> scaled_beta_mat</span>
<span id="cb73-11"><a href="#cb73-11"></a></span>
<span id="cb73-12"><a href="#cb73-12"></a>        <span class="co"># Solve the model</span></span>
<span id="cb73-13"><a href="#cb73-13"></a>        sol <span class="ot">&lt;-</span> deSolve<span class="sc">::</span><span class="fu">ode</span>(</span>
<span id="cb73-14"><a href="#cb73-14"></a>            <span class="at">y =</span> demog_yinit_ages,</span>
<span id="cb73-15"><a href="#cb73-15"></a>            <span class="at">times =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">400</span>, <span class="at">by =</span> <span class="fl">0.1</span>),</span>
<span id="cb73-16"><a href="#cb73-16"></a>            <span class="at">func =</span> multistage_model,</span>
<span id="cb73-17"><a href="#cb73-17"></a>            <span class="at">parms =</span> scaled_params</span>
<span id="cb73-18"><a href="#cb73-18"></a>        )</span>
<span id="cb73-19"><a href="#cb73-19"></a></span>
<span id="cb73-20"><a href="#cb73-20"></a>        <span class="co"># Get stable age distribution</span></span>
<span id="cb73-21"><a href="#cb73-21"></a>        stable_n <span class="ot">&lt;-</span> <span class="fu">solve</span>(</span>
<span id="cb73-22"><a href="#cb73-22"></a>            scaled_params[[<span class="st">"aging_mat"</span>]],</span>
<span id="cb73-23"><a href="#cb73-23"></a>            <span class="sc">-</span><span class="fu">c</span>(scaled_params[[<span class="st">"births"</span>]], <span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">29</span>))</span>
<span id="cb73-24"><a href="#cb73-24"></a>        )</span>
<span id="cb73-25"><a href="#cb73-25"></a></span>
<span id="cb73-26"><a href="#cb73-26"></a>        R0 <span class="ot">&lt;-</span> <span class="fu">calculate_R0</span>(</span>
<span id="cb73-27"><a href="#cb73-27"></a>            <span class="at">beta_mat =</span> scaled_params[[<span class="st">"beta_mat"</span>]],</span>
<span id="cb73-28"><a href="#cb73-28"></a>            <span class="at">stable_n_mat =</span> stable_n,</span>
<span id="cb73-29"><a href="#cb73-29"></a>            <span class="at">aging_mat =</span> scaled_params[[<span class="st">"aging_mat"</span>]],</span>
<span id="cb73-30"><a href="#cb73-30"></a>            <span class="at">recovery =</span> scaled_params[[<span class="st">"recovery"</span>]]</span>
<span id="cb73-31"><a href="#cb73-31"></a>        )</span>
<span id="cb73-32"><a href="#cb73-32"></a></span>
<span id="cb73-33"><a href="#cb73-33"></a>        <span class="co"># Get final number of infected individuals for each I class</span></span>
<span id="cb73-34"><a href="#cb73-34"></a>        infecteds <span class="ot">&lt;-</span> sol[<span class="fu">dim</span>(sol)[<span class="dv">1</span>], <span class="dv">1</span> <span class="sc">+</span> iindex]</span>
<span id="cb73-35"><a href="#cb73-35"></a></span>
<span id="cb73-36"><a href="#cb73-36"></a>        <span class="co"># Calculate mean age of infection</span></span>
<span id="cb73-37"><a href="#cb73-37"></a>        mean_age <span class="ot">&lt;-</span> <span class="fu">sum</span>(ages <span class="sc">*</span> infecteds <span class="sc">/</span> <span class="fu">sum</span>(infecteds))</span>
<span id="cb73-38"><a href="#cb73-38"></a></span>
<span id="cb73-39"><a href="#cb73-39"></a>        <span class="co"># Calculate sum of cases between 15-35 years </span></span>
<span id="cb73-40"><a href="#cb73-40"></a>        <span class="co"># recall, from the figures above, that that this is the equilibrium prevalence</span></span>
<span id="cb73-41"><a href="#cb73-41"></a>        <span class="co"># of infection in these age classes, or the average numbef of individuals that are infected </span></span>
<span id="cb73-42"><a href="#cb73-42"></a>        <span class="co"># at any given time in these age classes. Note that we're not differentiating between </span></span>
<span id="cb73-43"><a href="#cb73-43"></a>        <span class="co"># individuals who can and cannot get pregnant here. So we're making an implicit assumption that</span></span>
<span id="cb73-44"><a href="#cb73-44"></a>        <span class="co"># there no difference in the risk of rubella infection in these groups so that if prevalence goes</span></span>
<span id="cb73-45"><a href="#cb73-45"></a>        <span class="co"># up in one group, it goes up in the other.</span></span>
<span id="cb73-46"><a href="#cb73-46"></a>        sum_cases <span class="ot">&lt;-</span> <span class="fu">sum</span>(infecteds[<span class="dv">15</span><span class="sc">:</span><span class="dv">23</span>])</span>
<span id="cb73-47"><a href="#cb73-47"></a></span>
<span id="cb73-48"><a href="#cb73-48"></a>        <span class="co"># Return a dataframe with the values</span></span>
<span id="cb73-49"><a href="#cb73-49"></a>        <span class="fu">return</span>(<span class="fu">tibble</span>(R0, mean_age, sum_cases))</span>
<span id="cb73-50"><a href="#cb73-50"></a>    }</span>
<span id="cb73-51"><a href="#cb73-51"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Now we can make a table of the results and plot mean age and the sum of cases between 15-35 years of age as a function of <span class="math inline">\(R_0\)</span>.</p>
<div class="cell page-columns page-full" data-hash="r-session-02_cache/html/unnamed-chunk-69_d2d773db58f1206985848694edcff80f">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb74" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1"></a><span class="co"># Create a table from the dataframe</span></span>
<span id="cb74-2"><a href="#cb74-2"></a><span class="fu">gt</span>(R0_mean_age_contacts_df) <span class="sc">%&gt;%</span></span>
<span id="cb74-3"><a href="#cb74-3"></a>    <span class="fu">fmt_number</span>(</span>
<span id="cb74-4"><a href="#cb74-4"></a>        <span class="at">columns =</span> <span class="fu">everything</span>(),</span>
<span id="cb74-5"><a href="#cb74-5"></a>        <span class="at">decimals =</span> <span class="dv">2</span></span>
<span id="cb74-6"><a href="#cb74-6"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb74-7"><a href="#cb74-7"></a>    <span class="co"># Relabel the column headers</span></span>
<span id="cb74-8"><a href="#cb74-8"></a>    <span class="fu">cols_label</span>(</span>
<span id="cb74-9"><a href="#cb74-9"></a>        <span class="at">R0 =</span> <span class="fu">md</span>(<span class="st">"**R0**"</span>),</span>
<span id="cb74-10"><a href="#cb74-10"></a>        <span class="at">mean_age =</span> <span class="fu">md</span>(<span class="st">"**Mean age of&lt;br&gt;infection**"</span>),</span>
<span id="cb74-11"><a href="#cb74-11"></a>        <span class="at">sum_cases =</span> <span class="fu">md</span>(<span class="st">"**Total cases between&lt;br&gt;15-35 years**"</span>)</span>
<span id="cb74-12"><a href="#cb74-12"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb74-13"><a href="#cb74-13"></a>    <span class="co"># Apply style to the table with gray alternating rows</span></span>
<span id="cb74-14"><a href="#cb74-14"></a>    <span class="fu">opt_stylize</span>(<span class="at">style =</span> <span class="dv">1</span>, <span class="at">color =</span> <span class="st">'gray'</span>) <span class="sc">%&gt;%</span></span>
<span id="cb74-15"><a href="#cb74-15"></a>    <span class="co"># Increate space between columns</span></span>
<span id="cb74-16"><a href="#cb74-16"></a>    <span class="fu">opt_horizontal_padding</span>(<span class="at">scale =</span> <span class="dv">3</span>) <span class="sc">%&gt;%</span></span>
<span id="cb74-17"><a href="#cb74-17"></a>    <span class="fu">cols_align</span>(<span class="st">"center"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display column-body">

<div id="cuwwzrecxg" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#cuwwzrecxg table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#cuwwzrecxg thead, #cuwwzrecxg tbody, #cuwwzrecxg tfoot, #cuwwzrecxg tr, #cuwwzrecxg td, #cuwwzrecxg th {
  border-style: none;
}

#cuwwzrecxg p {
  margin: 0;
  padding: 0;
}

#cuwwzrecxg .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #000000;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #000000;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#cuwwzrecxg .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#cuwwzrecxg .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 15px;
  padding-right: 15px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#cuwwzrecxg .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 15px;
  padding-right: 15px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#cuwwzrecxg .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#cuwwzrecxg .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #5F5F5F;
}

#cuwwzrecxg .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #5F5F5F;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #5F5F5F;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#cuwwzrecxg .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 15px;
  padding-right: 15px;
  overflow-x: hidden;
}

#cuwwzrecxg .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#cuwwzrecxg .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#cuwwzrecxg .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#cuwwzrecxg .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #5F5F5F;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#cuwwzrecxg .gt_spanner_row {
  border-bottom-style: hidden;
}

#cuwwzrecxg .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 15px;
  padding-right: 15px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #5F5F5F;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #5F5F5F;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#cuwwzrecxg .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #5F5F5F;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #5F5F5F;
  vertical-align: middle;
}

#cuwwzrecxg .gt_from_md > :first-child {
  margin-top: 0;
}

#cuwwzrecxg .gt_from_md > :last-child {
  margin-bottom: 0;
}

#cuwwzrecxg .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 15px;
  padding-right: 15px;
  margin: 10px;
  border-top-style: none;
  border-top-width: 1px;
  border-top-color: #D5D5D5;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D5D5D5;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D5D5D5;
  vertical-align: middle;
  overflow-x: hidden;
}

#cuwwzrecxg .gt_stub {
  color: #FFFFFF;
  background-color: #5F5F5F;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #5F5F5F;
  padding-left: 15px;
  padding-right: 15px;
}

#cuwwzrecxg .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 15px;
  padding-right: 15px;
  vertical-align: top;
}

#cuwwzrecxg .gt_row_group_first td {
  border-top-width: 2px;
}

#cuwwzrecxg .gt_row_group_first th {
  border-top-width: 2px;
}

#cuwwzrecxg .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 15px;
  padding-right: 15px;
}

#cuwwzrecxg .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #5F5F5F;
}

#cuwwzrecxg .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#cuwwzrecxg .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 15px;
  padding-right: 15px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #5F5F5F;
}

#cuwwzrecxg .gt_grand_summary_row {
  color: #333333;
  background-color: #D5D5D5;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 15px;
  padding-right: 15px;
}

#cuwwzrecxg .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 15px;
  padding-right: 15px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #5F5F5F;
}

#cuwwzrecxg .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 15px;
  padding-right: 15px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #5F5F5F;
}

#cuwwzrecxg .gt_striped {
  background-color: #F4F4F4;
}

#cuwwzrecxg .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #5F5F5F;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #5F5F5F;
}

#cuwwzrecxg .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#cuwwzrecxg .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 15px;
  padding-right: 15px;
}

#cuwwzrecxg .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#cuwwzrecxg .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 15px;
  padding-right: 15px;
}

#cuwwzrecxg .gt_left {
  text-align: left;
}

#cuwwzrecxg .gt_center {
  text-align: center;
}

#cuwwzrecxg .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#cuwwzrecxg .gt_font_normal {
  font-weight: normal;
}

#cuwwzrecxg .gt_font_bold {
  font-weight: bold;
}

#cuwwzrecxg .gt_font_italic {
  font-style: italic;
}

#cuwwzrecxg .gt_super {
  font-size: 65%;
}

#cuwwzrecxg .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#cuwwzrecxg .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#cuwwzrecxg .gt_indent_1 {
  text-indent: 5px;
}

#cuwwzrecxg .gt_indent_2 {
  text-indent: 10px;
}

#cuwwzrecxg .gt_indent_3 {
  text-indent: 15px;
}

#cuwwzrecxg .gt_indent_4 {
  text-indent: 20px;
}

#cuwwzrecxg .gt_indent_5 {
  text-indent: 25px;
}
</style>
<table class="gt_table" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead><tr class="gt_col_headings">
<th class="gt_col_heading gt_columns_bottom_border gt_center" rowspan="1" colspan="1" scope="col" id="<strong>R0</strong>"><strong>R0</strong></th>
      <th class="gt_col_heading gt_columns_bottom_border gt_center" rowspan="1" colspan="1" scope="col" id="<strong>Mean age of<br>infection</strong>"><strong>Mean age of<br>infection</strong></th>
      <th class="gt_col_heading gt_columns_bottom_border gt_center" rowspan="1" colspan="1" scope="col" id="<strong>Total cases between<br>15-35 years</strong>"><strong>Total cases between<br>15-35 years</strong></th>
    </tr></thead>
<tbody class="gt_table_body">
<tr>
<td headers="R0" class="gt_row gt_center">6.85</td>
<td headers="mean_age" class="gt_row gt_center">6.08</td>
<td headers="sum_cases" class="gt_row gt_center">0.54</td>
</tr>
<tr>
<td headers="R0" class="gt_row gt_center gt_striped">6.24</td>
<td headers="mean_age" class="gt_row gt_center gt_striped">6.61</td>
<td headers="sum_cases" class="gt_row gt_center gt_striped">0.66</td>
</tr>
<tr>
<td headers="R0" class="gt_row gt_center">5.63</td>
<td headers="mean_age" class="gt_row gt_center">7.28</td>
<td headers="sum_cases" class="gt_row gt_center">0.81</td>
</tr>
<tr>
<td headers="R0" class="gt_row gt_center gt_striped">5.02</td>
<td headers="mean_age" class="gt_row gt_center gt_striped">8.15</td>
<td headers="sum_cases" class="gt_row gt_center gt_striped">0.98</td>
</tr>
<tr>
<td headers="R0" class="gt_row gt_center">4.41</td>
<td headers="mean_age" class="gt_row gt_center">9.30</td>
<td headers="sum_cases" class="gt_row gt_center">1.20</td>
</tr>
<tr>
<td headers="R0" class="gt_row gt_center gt_striped">3.81</td>
<td headers="mean_age" class="gt_row gt_center gt_striped">10.87</td>
<td headers="sum_cases" class="gt_row gt_center gt_striped">1.45</td>
</tr>
<tr>
<td headers="R0" class="gt_row gt_center">3.20</td>
<td headers="mean_age" class="gt_row gt_center">13.11</td>
<td headers="sum_cases" class="gt_row gt_center">1.74</td>
</tr>
<tr>
<td headers="R0" class="gt_row gt_center gt_striped">2.59</td>
<td headers="mean_age" class="gt_row gt_center gt_striped">16.40</td>
<td headers="sum_cases" class="gt_row gt_center gt_striped">2.01</td>
</tr>
<tr>
<td headers="R0" class="gt_row gt_center">1.98</td>
<td headers="mean_age" class="gt_row gt_center">21.50</td>
<td headers="sum_cases" class="gt_row gt_center">2.11</td>
</tr>
<tr>
<td headers="R0" class="gt_row gt_center gt_striped">1.37</td>
<td headers="mean_age" class="gt_row gt_center gt_striped">29.72</td>
<td headers="sum_cases" class="gt_row gt_center gt_striped">1.48</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="cell page-columns page-full" data-hash="r-session-02_cache/html/unnamed-chunk-70_22648e13bba2319f42d75280115b249c">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb75" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1"></a>R0_mean_age_contacts_df <span class="sc">%&gt;%</span></span>
<span id="cb75-2"><a href="#cb75-2"></a>    <span class="co"># Convert to long data frame for facet plotting</span></span>
<span id="cb75-3"><a href="#cb75-3"></a>    <span class="fu">pivot_longer</span>(<span class="sc">-</span>R0, <span class="at">names_to =</span> <span class="st">"metric"</span>, <span class="at">values_to =</span> <span class="st">"value"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb75-4"><a href="#cb75-4"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> R0, <span class="at">y =</span> value)) <span class="sc">+</span></span>
<span id="cb75-5"><a href="#cb75-5"></a>    <span class="fu">geom_line</span>(<span class="at">color =</span> <span class="st">"slategray4"</span>) <span class="sc">+</span></span>
<span id="cb75-6"><a href="#cb75-6"></a>    <span class="fu">geom_point</span>(<span class="at">shape =</span> <span class="dv">21</span>, <span class="at">size =</span> <span class="dv">5</span>, <span class="at">fill =</span> <span class="st">"slategray4"</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>) <span class="sc">+</span></span>
<span id="cb75-7"><a href="#cb75-7"></a>    <span class="fu">facet_wrap</span>(</span>
<span id="cb75-8"><a href="#cb75-8"></a>        <span class="sc">~</span>metric,</span>
<span id="cb75-9"><a href="#cb75-9"></a>        <span class="at">scales =</span> <span class="st">"free_y"</span>,</span>
<span id="cb75-10"><a href="#cb75-10"></a>        <span class="at">labeller =</span> <span class="fu">as_labeller</span>(<span class="fu">c</span>(</span>
<span id="cb75-11"><a href="#cb75-11"></a>            <span class="at">mean_age =</span> <span class="st">"Mean Age of Infection"</span>,</span>
<span id="cb75-12"><a href="#cb75-12"></a>            <span class="at">sum_cases =</span> <span class="st">"Total cases between 15-35 years"</span></span>
<span id="cb75-13"><a href="#cb75-13"></a>        ))</span>
<span id="cb75-14"><a href="#cb75-14"></a>    ) <span class="sc">+</span></span>
<span id="cb75-15"><a href="#cb75-15"></a>    <span class="fu">labs</span>(</span>
<span id="cb75-16"><a href="#cb75-16"></a>        <span class="at">x =</span> <span class="st">"R0"</span>,</span>
<span id="cb75-17"><a href="#cb75-17"></a>        <span class="at">y =</span> <span class="st">"Value"</span></span>
<span id="cb75-18"><a href="#cb75-18"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display column-body">
<p><img src="r-session-02_files/figure-html/unnamed-chunk-70-1.png" class="img-fluid" style="width:100.0%"></p>
</div>
</div>
</section><section id="what-do-real-contact-networks-look-like" class="level3 page-columns page-full" data-number="7.6.3"><h3 data-number="7.6.3" class="anchored" data-anchor-id="what-do-real-contact-networks-look-like">
<span class="header-section-number">7.6.3</span> What do real contact networks look like?</h3>
<p>The POLYMOD study <span class="citation" data-cites="mossongSocialContactsMixing2008a">(<a href="references.html#ref-mossongSocialContactsMixing2008a" role="doc-biblioref">Mossong et al. 2008</a>)</span> was a journal-based look into the contact network in contemporary European society. Let’s have a look what these data tell us about the contact structure.</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-71_50842d88b0044549d9af9e45f4ea277f">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb76" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1"></a>mossong_cont_net <span class="ot">&lt;-</span> rio<span class="sc">::</span><span class="fu">import</span>(<span class="st">"https://raw.githubusercontent.com/arnold-c/SISMID-Module-02_2023/main/data/mossong-matrix.csv"</span>)</span>
<span id="cb76-2"><a href="#cb76-2"></a><span class="co"># mossong_cont_net &lt;- rio::import(here::here("data", "mossong-matrix.csv"))</span></span>
<span id="cb76-3"><a href="#cb76-3"></a></span>
<span id="cb76-4"><a href="#cb76-4"></a>mossong_ages <span class="ot">&lt;-</span> <span class="fu">unique</span>(mossong_cont_net<span class="sc">$</span>contactor)</span>
<span id="cb76-5"><a href="#cb76-5"></a>mossong_cont_net<span class="sc">$</span>contactor <span class="ot">&lt;-</span> <span class="fu">ordered</span>(mossong_cont_net<span class="sc">$</span>contactor, <span class="at">levels =</span> mossong_ages)</span>
<span id="cb76-6"><a href="#cb76-6"></a>mossong_cont_net<span class="sc">$</span>contactee <span class="ot">&lt;-</span> <span class="fu">ordered</span>(mossong_cont_net<span class="sc">$</span>contactee, <span class="at">levels =</span> mossong_ages)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Since contacts are symmetric, we’ll need to estimate the symmetric contact matrix.</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-72_984ec604bfeff9ec06391e59dde2ef74">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb77" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1"></a>mossong_mat <span class="ot">&lt;-</span> mossong_cont_net <span class="sc">%&gt;%</span></span>
<span id="cb77-2"><a href="#cb77-2"></a>    <span class="fu">pivot_wider</span>(</span>
<span id="cb77-3"><a href="#cb77-3"></a>        <span class="at">names_from =</span> contactor,</span>
<span id="cb77-4"><a href="#cb77-4"></a>        <span class="at">values_from =</span> contact.rate</span>
<span id="cb77-5"><a href="#cb77-5"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb77-6"><a href="#cb77-6"></a>    <span class="fu">select</span>(<span class="sc">-</span>contactee) <span class="sc">%&gt;%</span></span>
<span id="cb77-7"><a href="#cb77-7"></a>    <span class="fu">as.matrix</span>()</span>
<span id="cb77-8"><a href="#cb77-8"></a></span>
<span id="cb77-9"><a href="#cb77-9"></a><span class="fu">rownames</span>(mossong_mat) <span class="ot">&lt;-</span> mossong_ages</span>
<span id="cb77-10"><a href="#cb77-10"></a></span>
<span id="cb77-11"><a href="#cb77-11"></a><span class="co"># Create a symmetrical contact matrix</span></span>
<span id="cb77-12"><a href="#cb77-12"></a>mossong_mat_sym <span class="ot">&lt;-</span> (mossong_mat <span class="sc">+</span> <span class="fu">t</span>(mossong_mat)) <span class="sc">/</span> <span class="dv">2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Here we’ll use the <code>filled.contour</code> function to visualize the contact matrix, to show you an alternative way of visualizing contact matrices. Notices that we are using the raw matrix object, not a long dataframe, as previously.</p>
<div class="cell page-columns page-full" data-hash="r-session-02_cache/html/unnamed-chunk-73_1240ac7a6f8b38da266d0bce0f91cdde">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb78" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1"></a><span class="fu">filled.contour</span>(</span>
<span id="cb78-2"><a href="#cb78-2"></a>    ages, ages, <span class="fu">log10</span>(mossong_mat),</span>
<span id="cb78-3"><a href="#cb78-3"></a>    <span class="at">plot.title =</span> <span class="fu">title</span>(</span>
<span id="cb78-4"><a href="#cb78-4"></a>        <span class="at">main =</span> <span class="st">"Log10 of Raw Contact Rate"</span>,</span>
<span id="cb78-5"><a href="#cb78-5"></a>        <span class="at">xlab =</span> <span class="st">"Age of Contactor"</span>,</span>
<span id="cb78-6"><a href="#cb78-6"></a>        <span class="at">ylab =</span> <span class="st">"Age of Contactee"</span></span>
<span id="cb78-7"><a href="#cb78-7"></a>    )</span>
<span id="cb78-8"><a href="#cb78-8"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display column-body">
<p><img src="r-session-02_files/figure-html/unnamed-chunk-73-1.png" class="img-fluid" style="width:100.0%"></p>
</div>
</div>
<div class="cell page-columns page-full" data-hash="r-session-02_cache/html/unnamed-chunk-74_ea28b7a92f7c1690ba976e20a9fbe3bd">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb79" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1"></a><span class="fu">filled.contour</span>(</span>
<span id="cb79-2"><a href="#cb79-2"></a>    ages, ages, <span class="fu">log10</span>(mossong_mat_sym),</span>
<span id="cb79-3"><a href="#cb79-3"></a>    <span class="at">plot.title =</span> <span class="fu">title</span>(</span>
<span id="cb79-4"><a href="#cb79-4"></a>        <span class="at">main =</span> <span class="st">"Log10 of Symmetrical Contact Rate"</span>,</span>
<span id="cb79-5"><a href="#cb79-5"></a>        <span class="at">xlab =</span> <span class="st">"Age of Contactor"</span>,</span>
<span id="cb79-6"><a href="#cb79-6"></a>        <span class="at">ylab =</span> <span class="st">"Age of Contactee"</span></span>
<span id="cb79-7"><a href="#cb79-7"></a>    )</span>
<span id="cb79-8"><a href="#cb79-8"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display column-body">
<p><img src="r-session-02_files/figure-html/unnamed-chunk-74-1.png" class="img-fluid" style="width:100.0%"></p>
</div>
</div>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-75_ba8df601e1c0dbc92c4f6f2e768796d2">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb80" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1"></a>mossong_cont_sums <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb80-2"><a href="#cb80-2"></a>        <span class="at">age =</span> <span class="fu">factor</span>(mossong_ages, <span class="at">levels =</span> mossong_ages),</span>
<span id="cb80-3"><a href="#cb80-3"></a>        <span class="at">contactees =</span> <span class="fu">rowSums</span>(mossong_mat),</span>
<span id="cb80-4"><a href="#cb80-4"></a>        <span class="at">contactors =</span> <span class="fu">colSums</span>(mossong_mat)</span>
<span id="cb80-5"><a href="#cb80-5"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb80-6"><a href="#cb80-6"></a>    <span class="fu">pivot_longer</span>(<span class="sc">-</span>age, <span class="at">names_to =</span> <span class="st">"type"</span>, <span class="at">values_to =</span> <span class="st">"total_contacts"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell page-columns page-full" data-hash="r-session-02_cache/html/unnamed-chunk-76_99997c1528e9b2e9640e2ba65713cd45">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb81" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1"></a><span class="fu">ggplot</span>(</span>
<span id="cb81-2"><a href="#cb81-2"></a>        mossong_cont_sums,</span>
<span id="cb81-3"><a href="#cb81-3"></a>        <span class="fu">aes</span>(<span class="at">x =</span> age, <span class="at">y =</span> total_contacts, <span class="at">color =</span> type, <span class="at">fill =</span> type, <span class="at">group =</span> type)</span>
<span id="cb81-4"><a href="#cb81-4"></a>    ) <span class="sc">+</span></span>
<span id="cb81-5"><a href="#cb81-5"></a>    <span class="fu">geom_path</span>(<span class="at">linewidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb81-6"><a href="#cb81-6"></a>    <span class="fu">geom_point</span>(</span>
<span id="cb81-7"><a href="#cb81-7"></a>        <span class="at">position =</span> <span class="st">"identity"</span>,</span>
<span id="cb81-8"><a href="#cb81-8"></a>        <span class="at">alpha =</span> <span class="fl">0.8</span>,</span>
<span id="cb81-9"><a href="#cb81-9"></a>        <span class="at">shape =</span> <span class="dv">21</span>,</span>
<span id="cb81-10"><a href="#cb81-10"></a>        <span class="at">size =</span> <span class="dv">4</span></span>
<span id="cb81-11"><a href="#cb81-11"></a>    ) <span class="sc">+</span></span>
<span id="cb81-12"><a href="#cb81-12"></a>    <span class="fu">scale_color_manual</span>(</span>
<span id="cb81-13"><a href="#cb81-13"></a>        <span class="at">values =</span> <span class="fu">c</span>(<span class="st">"slategray4"</span>, <span class="st">"navy"</span>),</span>
<span id="cb81-14"><a href="#cb81-14"></a>        <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"Contactees"</span>, <span class="st">"Contactors"</span>),</span>
<span id="cb81-15"><a href="#cb81-15"></a>        <span class="at">aesthetics =</span> <span class="fu">c</span>(<span class="st">"color"</span>, <span class="st">"fill"</span>)</span>
<span id="cb81-16"><a href="#cb81-16"></a>    ) <span class="sc">+</span></span>
<span id="cb81-17"><a href="#cb81-17"></a>    <span class="fu">guides</span>(<span class="at">color =</span> <span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb81-18"><a href="#cb81-18"></a>    <span class="fu">labs</span>(</span>
<span id="cb81-19"><a href="#cb81-19"></a>        <span class="at">x =</span> <span class="st">"Age"</span>,</span>
<span id="cb81-20"><a href="#cb81-20"></a>        <span class="at">y =</span> <span class="st">"Total contacts"</span>,</span>
<span id="cb81-21"><a href="#cb81-21"></a>        <span class="at">fill =</span> <span class="st">"Type of contact"</span></span>
<span id="cb81-22"><a href="#cb81-22"></a>    ) <span class="sc">+</span></span>
<span id="cb81-23"><a href="#cb81-23"></a>    <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display column-body">
<p><img src="r-session-02_files/figure-html/unnamed-chunk-76-1.png" class="img-fluid" style="width:100.0%"></p>
</div>
</div>
<p>While this matrix tells us how many contacts are made per year by an individual of each age, it doesn’t tell us anything about the probability that a contact results in communication of infection. Let’s assume that each contact has a constant probability <span class="math inline">\(q\)</span> of resulting in a transmission event.</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-77_9d1f8237cfcd8b7c32da2da256a58c2f">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb82" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1"></a>q <span class="ot">&lt;-</span> <span class="fl">3e-5</span></span>
<span id="cb82-2"><a href="#cb82-2"></a>mossong_beta_mat <span class="ot">&lt;-</span> q <span class="sc">*</span> mossong_mat_sym</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell page-columns page-full" data-hash="r-session-02_cache/html/unnamed-chunk-78_b1127c715877d64c2d0f6c4c4df65849">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb83" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1"></a><span class="fu">filled.contour</span>(</span>
<span id="cb83-2"><a href="#cb83-2"></a>    ages, ages, <span class="fu">log10</span>(mossong_beta_mat),</span>
<span id="cb83-3"><a href="#cb83-3"></a>    <span class="at">plot.title =</span> <span class="fu">title</span>(</span>
<span id="cb83-4"><a href="#cb83-4"></a>        <span class="at">main =</span> <span class="st">"WAIFW matrix based on POLYMOD data"</span>,</span>
<span id="cb83-5"><a href="#cb83-5"></a>        <span class="at">xlab =</span> <span class="st">"Age"</span>,</span>
<span id="cb83-6"><a href="#cb83-6"></a>        <span class="at">ylab =</span> <span class="st">"Age"</span></span>
<span id="cb83-7"><a href="#cb83-7"></a>    )</span>
<span id="cb83-8"><a href="#cb83-8"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display column-body">
<p><img src="r-session-02_files/figure-html/unnamed-chunk-78-1.png" class="img-fluid" style="width:100.0%"></p>
</div>
</div>
<p>Now let’s simulate the introduction of such a pathogen into a population characterized by this contact structure.</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-79_3c0289c53d1f30190ce3cbf4aa228e45">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb84" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1"></a><span class="co"># Update the parameters with the POLYMOD-based beta matrix</span></span>
<span id="cb84-2"><a href="#cb84-2"></a>mossong_params <span class="ot">&lt;-</span> multistage_params</span>
<span id="cb84-3"><a href="#cb84-3"></a>mossong_params[[<span class="st">"beta_mat"</span>]] <span class="ot">&lt;-</span> mossong_beta_mat</span>
<span id="cb84-4"><a href="#cb84-4"></a></span>
<span id="cb84-5"><a href="#cb84-5"></a><span class="co"># Solve the model with the updated parameters</span></span>
<span id="cb84-6"><a href="#cb84-6"></a>mossong_sol <span class="ot">&lt;-</span> deSolve<span class="sc">::</span><span class="fu">ode</span>(</span>
<span id="cb84-7"><a href="#cb84-7"></a>    <span class="at">y =</span> demog_yinit_ages,</span>
<span id="cb84-8"><a href="#cb84-8"></a>    <span class="at">times =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">200</span>, <span class="at">by =</span> <span class="fl">0.5</span>),</span>
<span id="cb84-9"><a href="#cb84-9"></a>    <span class="at">func =</span> multistage_model,</span>
<span id="cb84-10"><a href="#cb84-10"></a>    <span class="at">parms =</span> mossong_params</span>
<span id="cb84-11"><a href="#cb84-11"></a>)</span>
<span id="cb84-12"><a href="#cb84-12"></a></span>
<span id="cb84-13"><a href="#cb84-13"></a><span class="co"># Extract the timeseries of infectious individuals</span></span>
<span id="cb84-14"><a href="#cb84-14"></a>mossong_infecteds <span class="ot">&lt;-</span> mossong_sol[, <span class="dv">1</span> <span class="sc">+</span> iindex]</span>
<span id="cb84-15"><a href="#cb84-15"></a></span>
<span id="cb84-16"><a href="#cb84-16"></a><span class="co"># Convert infectious individual timeseries to dataframe for plotting</span></span>
<span id="cb84-17"><a href="#cb84-17"></a>mossong_infecteds_df <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb84-18"><a href="#cb84-18"></a>        <span class="at">time =</span> mossong_sol[, <span class="dv">1</span>],</span>
<span id="cb84-19"><a href="#cb84-19"></a>        <span class="at">Juveniles =</span> <span class="fu">apply</span>(mossong_infecteds[, juvies], <span class="dv">1</span>, sum),</span>
<span id="cb84-20"><a href="#cb84-20"></a>        <span class="at">Adults =</span> <span class="fu">apply</span>(mossong_infecteds[, adults], <span class="dv">1</span>, sum)</span>
<span id="cb84-21"><a href="#cb84-21"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb84-22"><a href="#cb84-22"></a>    <span class="fu">pivot_longer</span>(</span>
<span id="cb84-23"><a href="#cb84-23"></a>        <span class="at">cols =</span> <span class="fu">c</span>(Juveniles, Adults),</span>
<span id="cb84-24"><a href="#cb84-24"></a>        <span class="at">names_to =</span> <span class="st">"age_group"</span>,</span>
<span id="cb84-25"><a href="#cb84-25"></a>        <span class="at">values_to =</span> <span class="st">"infections"</span></span>
<span id="cb84-26"><a href="#cb84-26"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb84-27"><a href="#cb84-27"></a>    <span class="fu">mutate</span>(</span>
<span id="cb84-28"><a href="#cb84-28"></a>        <span class="at">age_group =</span> <span class="fu">factor</span>(age_group, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Juveniles"</span>, <span class="st">"Adults"</span>))</span>
<span id="cb84-29"><a href="#cb84-29"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell page-columns page-full" data-hash="r-session-02_cache/html/unnamed-chunk-80_fae5470f75592b3b24f1e098d5a76d94">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb85" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1"></a><span class="fu">ggplot</span>(mossong_infecteds_df, <span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> infections, <span class="at">color =</span> age_group)) <span class="sc">+</span></span>
<span id="cb85-2"><a href="#cb85-2"></a>    <span class="fu">geom_line</span>(<span class="at">linewidth =</span> <span class="fl">1.5</span>) <span class="sc">+</span></span>
<span id="cb85-3"><a href="#cb85-3"></a>    <span class="fu">scale_color_manual</span>(</span>
<span id="cb85-4"><a href="#cb85-4"></a>        <span class="at">values =</span> age_group_colors</span>
<span id="cb85-5"><a href="#cb85-5"></a>    ) <span class="sc">+</span></span>
<span id="cb85-6"><a href="#cb85-6"></a>    <span class="fu">labs</span>(</span>
<span id="cb85-7"><a href="#cb85-7"></a>        <span class="at">x =</span> <span class="st">"Time"</span>,</span>
<span id="cb85-8"><a href="#cb85-8"></a>        <span class="at">y =</span> <span class="st">"Number of infections"</span>,</span>
<span id="cb85-9"><a href="#cb85-9"></a>        <span class="at">color =</span> <span class="st">"Age group"</span></span>
<span id="cb85-10"><a href="#cb85-10"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display column-body">
<p><img src="r-session-02_files/figure-html/unnamed-chunk-80-1.png" class="img-fluid" style="width:100.0%"></p>
</div>
</div>
<p>As before, we can also look at the equilibrium seroprevalence</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-81_43d0927541f6bdb1175ef3f053469a8a">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb86" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1"></a><span class="co"># Get last time point</span></span>
<span id="cb86-2"><a href="#cb86-2"></a>mossong_equil <span class="ot">&lt;-</span> <span class="fu">drop</span>(<span class="fu">tail</span>(mossong_sol, <span class="dv">1</span>))[<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb86-3"><a href="#cb86-3"></a></span>
<span id="cb86-4"><a href="#cb86-4"></a><span class="co"># Calculate number of individuals in each age group at end of simulation</span></span>
<span id="cb86-5"><a href="#cb86-5"></a>mossong_equil_n <span class="ot">&lt;-</span> mossong_equil[sindex] <span class="sc">+</span></span>
<span id="cb86-6"><a href="#cb86-6"></a>    mossong_equil[iindex] <span class="sc">+</span></span>
<span id="cb86-7"><a href="#cb86-7"></a>    mossong_equil[rindex]</span>
<span id="cb86-8"><a href="#cb86-8"></a></span>
<span id="cb86-9"><a href="#cb86-9"></a><span class="co"># Calculate equilibrium seroprevalence</span></span>
<span id="cb86-10"><a href="#cb86-10"></a>mossong_equil_seroprev <span class="ot">&lt;-</span> mossong_equil[rindex] <span class="sc">/</span> mossong_equil_n</span>
<span id="cb86-11"><a href="#cb86-11"></a></span>
<span id="cb86-12"><a href="#cb86-12"></a><span class="co"># Convert to dataframe for plotting</span></span>
<span id="cb86-13"><a href="#cb86-13"></a>mossong_equil_seroprev_df <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb86-14"><a href="#cb86-14"></a>    <span class="co"># We can reuse the ages vectors from before as they are the same</span></span>
<span id="cb86-15"><a href="#cb86-15"></a>    <span class="co"># as the POLYMOD data</span></span>
<span id="cb86-16"><a href="#cb86-16"></a>    <span class="at">age =</span> ages,</span>
<span id="cb86-17"><a href="#cb86-17"></a>    <span class="at">seroprev =</span> mossong_equil_seroprev,</span>
<span id="cb86-18"><a href="#cb86-18"></a>    <span class="at">width =</span> da_ages</span>
<span id="cb86-19"><a href="#cb86-19"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell page-columns page-full" data-hash="r-session-02_cache/html/unnamed-chunk-82_e30c1f9ae0667f694854d5b4e03c937b">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb87" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1"></a><span class="fu">ggplot</span>(mossong_equil_seroprev_df, <span class="fu">aes</span>(<span class="at">x =</span> age, <span class="at">y =</span> seroprev, <span class="at">fill =</span> age)) <span class="sc">+</span></span>
<span id="cb87-2"><a href="#cb87-2"></a>    <span class="fu">geom_col</span>(<span class="at">width =</span> mossong_equil_seroprev_df<span class="sc">$</span>width, <span class="at">just =</span> <span class="fl">1.0</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb87-3"><a href="#cb87-3"></a>    <span class="fu">labs</span>(</span>
<span id="cb87-4"><a href="#cb87-4"></a>        <span class="at">x =</span> <span class="st">"Age"</span>,</span>
<span id="cb87-5"><a href="#cb87-5"></a>        <span class="at">y =</span> <span class="st">"Seroprevalence"</span></span>
<span id="cb87-6"><a href="#cb87-6"></a>    ) <span class="sc">+</span></span>
<span id="cb87-7"><a href="#cb87-7"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">80</span>, <span class="dv">10</span>)) <span class="sc">+</span></span>
<span id="cb87-8"><a href="#cb87-8"></a>    <span class="fu">scale_fill_continuous</span>(<span class="at">low =</span> age_group_colors[<span class="dv">1</span>], <span class="at">high =</span> age_group_colors[<span class="dv">2</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display column-body">
<p><img src="r-session-02_files/figure-html/unnamed-chunk-82-1.png" class="img-fluid" style="width:100.0%"></p>
</div>
</div>
<p>and compute the <span class="math inline">\(R_0\)</span> for this infection.</p>
<div class="cell" data-hash="r-session-02_cache/html/unnamed-chunk-83_1ad195441d5c14c798aa677c678c50b5">
<details open=""><summary>Code</summary><div class="sourceCode cell-code" id="cb88" data-source-line-numbers="nil" data-code-line-numbers="nil"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1"></a>mossong_stable_n <span class="ot">&lt;-</span> <span class="fu">solve</span>(</span>
<span id="cb88-2"><a href="#cb88-2"></a>    mossong_params[[<span class="st">"aging_mat"</span>]],</span>
<span id="cb88-3"><a href="#cb88-3"></a>    <span class="sc">-</span><span class="fu">c</span>(mossong_params[[<span class="st">"births"</span>]], <span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">29</span>))</span>
<span id="cb88-4"><a href="#cb88-4"></a>)</span>
<span id="cb88-5"><a href="#cb88-5"></a></span>
<span id="cb88-6"><a href="#cb88-6"></a><span class="fu">calculate_R0</span>(</span>
<span id="cb88-7"><a href="#cb88-7"></a>    <span class="at">beta_mat =</span> mossong_params[[<span class="st">"beta_mat"</span>]],</span>
<span id="cb88-8"><a href="#cb88-8"></a>    <span class="at">stable_n_mat =</span> mossong_stable_n,</span>
<span id="cb88-9"><a href="#cb88-9"></a>    <span class="at">aging_mat =</span> mossong_params[[<span class="st">"aging_mat"</span>]],</span>
<span id="cb88-10"><a href="#cb88-10"></a>    <span class="at">recovery =</span> mossong_params[[<span class="st">"recovery"</span>]]</span>
<span id="cb88-11"><a href="#cb88-11"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre data-code-line-numbers=""><code>[1] 7.058675</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="QUESTION">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
QUESTION
</div>
</div>
<div class="callout-body-container callout-body">
<p>How does this R0 value compare to the R0 value obtained from <a href="#sec-ex-3"><span>Section&nbsp;7.4.2</span></a>?</p>
</div>
</div>
<!-- -->



<div id="refs" class="references csl-bib-body hanging-indent" role="list" style="display: none">
<div id="ref-bjornstadAdvancedNextGenerationMatrix2018" class="csl-entry" role="listitem">
Bjørnstad, Ottar N. 2018. <span>“Advanced: <span>The Next-Generation Matrix</span>.”</span> In <em>Epidemics: <span>Models</span> and <span>Data</span> Using <span>R</span></em>, 51. Use <span>R</span>! <span>Cham</span>: <span>Springer International Publishing</span>. <a href="https://doi.org/10.1007/978-3-319-97487-3">https://doi.org/10.1007/978-3-319-97487-3</a>.
</div>
<div id="ref-diekmann2000mathematical" class="csl-entry" role="listitem">
Diekmann, O., and J. A. P. Heesterbeek. 2000. <em>Mathematical Epidemiology of Infectious Diseases: <span>Model</span> Building, Analysis and Interpretation</em>. Wiley Series in Mathematical &amp; Computational Biology. <span>Wiley</span>. <a href="https://books.google.ca/books?id=5VjSaAf35pMC">https://books.google.ca/books?id=5VjSaAf35pMC</a>.
</div>
<div id="ref-heesterbeekBriefHistoryR02002" class="csl-entry" role="listitem">
Heesterbeek, J. A. P. 2002. <span>“A <span>Brief History</span> of <span>R0</span> and a <span>Recipe</span> for Its <span>Calculation</span>.”</span> <em>Acta Biotheoretica</em> 50 (3): 189–204. <a href="https://doi.org/10.1023/A:1016599411804">https://doi.org/10.1023/A:1016599411804</a>.
</div>
<div id="ref-heffernanPerspectivesBasicReproductive2005" class="csl-entry" role="listitem">
Heffernan, J. M, R. J Smith, and L. M Wahl. 2005. <span>“Perspectives on the Basic Reproductive Ratio.”</span> <em>J R Soc Interface</em> 2 (4): 281–93. <a href="https://doi.org/10.1098/rsif.2005.0042">https://doi.org/10.1098/rsif.2005.0042</a>.
</div>
<div id="ref-hurfordNextgenerationToolsEvolutionary2009" class="csl-entry" role="listitem">
Hurford, Amy, Daniel Cownden, and Troy Day. 2009. <span>“Next-Generation Tools for Evolutionary Invasion Analyses.”</span> <em>Journal of The Royal Society Interface</em> 7 (45): 561–71. <a href="https://doi.org/10.1098/rsif.2009.0448">https://doi.org/10.1098/rsif.2009.0448</a>.
</div>
<div id="ref-kingAgeStructuredModels2011" class="csl-entry" role="listitem">
King, Aaron A, and Helen J Wearing. 2011. <span>“Age <span>Structured Models</span>.”</span> In. <a href="https://ms.mcmaster.ca/~bolker/eeid/2011_eco/waifw.pdf">https://ms.mcmaster.ca/~bolker/eeid/2011_eco/waifw.pdf</a>.
</div>
<div id="ref-mossongSocialContactsMixing2008a" class="csl-entry" role="listitem">
Mossong, Joël, Niel Hens, Mark Jit, Philippe Beutels, Kari Auranen, Rafael Mikolajczyk, Marco Massari, et al. 2008. <span>“Social <span>Contacts</span> and <span>Mixing Patterns Relevant</span> to the <span>Spread</span> of <span>Infectious Diseases</span>.”</span> <em>PLOS Medicine</em> 5 (3): e74. <a href="https://doi.org/10.1371/journal.pmed.0050074">https://doi.org/10.1371/journal.pmed.0050074</a>.
</div>
</div>
</section></section></main><!-- /main --><script type="ojs-module-contents">
{"contents":[{"methodName":"interpret","cellName":"ojs-cell-1","inline":false,"source":"init_R0 = 2.0\ninit_vacc = 0.0\ninit_lifeexp = 75\n"},{"methodName":"interpret","cellName":"ojs-cell-2","inline":false,"source":"function set(input, value) {\n  input.value = value;\n  input.dispatchEvent(new Event(\"input\", {bubbles: true}));\n}\n"},{"methodName":"interpret","cellName":"ojs-cell-3","inline":false,"source":"viewof reset = Inputs.button([\n  [\"Reset all sliders\", () => {\n    set(viewof R0, init_R0)\n    set(viewof vacc, init_vacc)\n    set(viewof lifeexp, init_lifeexp)\n  }]\n])\nviewof R0 = Inputs.range(\n    [1.0, 10.0],\n    {value: 2.0, step: 0.01, label: md`${tex`R_0`}`}\n)\n\nviewof vacc = Inputs.range(\n    [0.0, 1.0],\n    {value: 0.0, step: 0.01, label: \"Vaccination coverage\"}\n)\n\nviewof lifeexp = Inputs.range(\n    [50, 100],\n    {value: 75, step: 1, label: \"Life expectancy\"}\n)\n\nmd`${tex`R_E = ${Re_str}`}`\nmd`${tex`\\text{Mean age of infection} = ${Re_mean_age_str}`}`\n"},{"methodName":"interpret","cellName":"ojs-cell-4","inline":false,"source":"Re = R0 * (1 - vacc)\nRe_str = Re.toPrecision(4).toLocaleString()\n"},{"methodName":"interpret","cellName":"ojs-cell-5","inline":false,"source":"function calc_mean_age(Re, lifeexp) {\n    if(Re >= 1) {\n        var mean_age = (lifeexp / (Re - 1))\n    } else {\n        var mean_age = Infinity\n    }\n    return mean_age\n}\n"},{"methodName":"interpret","cellName":"ojs-cell-6","inline":false,"source":"R0_mean_age = calc_mean_age(R0, lifeexp)\nRe_mean_age = calc_mean_age(Re, lifeexp)\nRe_mean_age_str = Re_mean_age.toPrecision(4).toLocaleString()\n"},{"methodName":"interpret","cellName":"ojs-cell-7","inline":false,"source":"import { aq, op } from '@uwdata/arquero'\n"},{"methodName":"interpret","cellName":"ojs-cell-8","inline":false,"source":"function calc_mean_age_arr(vacc, lifeexp, R0_min, R0_max, dR0) {\n    var R0_sim = R0_min\n\n    var R0 = []\n    var Re = []\n    var R0_mean_age = []\n    var Re_mean_age = []\n\n    for (R0_sim = R0_min; R0_sim <= R0_max; R0_sim += dR0) {\n        var Re_sim = R0_sim * (1 - vacc)\n        var R0_mean_age_sim = calc_mean_age(R0_sim, lifeexp)\n        var Re_mean_age_sim = calc_mean_age(Re_sim, lifeexp)\n\n        R0.push(R0_sim)\n        Re.push(Re_sim)\n        R0_mean_age.push(R0_mean_age_sim)\n        Re_mean_age.push(Re_mean_age_sim)\n    }\n\n    return {\n        Re: aq.table({\n                R0: R0,\n                mean_age: Re_mean_age\n            }).filter((d) => d.mean_age <= 100),\n        R0: aq.table({\n                R0: R0,\n                mean_age: R0_mean_age\n            }).filter((d) => d.mean_age <= 100)\n    }\n}\n"},{"methodName":"interpret","cellName":"ojs-cell-9","inline":false,"source":"mean_age_arrs = calc_mean_age_arr(vacc, lifeexp, 1.0, 10.0, 0.01)\n"},{"methodName":"interpret","cellName":"ojs-cell-10","inline":false,"source":"mean_age_dots = [({\n    arrow_start: R0_mean_age <= 100 ? R0_mean_age : 100,\n    arrow_end: Re_mean_age <= 100 ? Re_mean_age : 100,\n    R0: R0.toPrecision(3),\n    Re: Re.toPrecision(3),\n    R0_mean_age,\n    Re_mean_age\n})]\n"},{"methodName":"interpret","cellName":"ojs-cell-11","inline":false,"source":"{\n    let R0Color = \"#1f77b4\"\n    let ReColor = \"#ff7f0e\"\n\n    let plot = Plot.plot({\n        color: {\n            legend: true,\n            domain: [\"Unvaccinated\", \"Vaccinated\"],\n            range: [\"#1f77b4\", \"#ff7f0e\"]\n        },\n        style: {fontSize: \"20px\"},\n        marginLeft: 65,\n        marginTop: 40,\n        marginBottom: 55,\n        grid: true,\n        width: 800,\n        height: 670,\n        x: {label: \"R0\", domain: [0, 10]},\n        y: {label: \"Mean Age of Infection\", domain: [0, 100]},\n        marks: [\n            Plot.line(mean_age_arrs.Re, {x: \"R0\", y: \"mean_age\", stroke: ReColor, strokeWidth: 6}),\n            Plot.line(mean_age_arrs.R0, {x: \"R0\", y: \"mean_age\", stroke: R0Color, strokeWidth: 6}),\n            Re_mean_age <= 100 ?\n                [\n                    vacc > 0.00 ? Plot.dot(mean_age_dots, {x: \"R0\", y: \"Re_mean_age\", r: 12, stroke: ReColor, fill:     ReColor,  fillOpacity: 0.6}) : null,\n                    Plot.text(\n                        mean_age_dots,\n                        {x: \"R0\", y: \"Re_mean_age\", text: (d) => `Re = ${d.Re}`, dx: 55, dy: -25, fontWeight: \"bold\", fill: ReColor}\n                    )\n                ] :\n            null,\n            R0_mean_age <= 100 ?\n                [\n                Plot.dot(mean_age_dots, {x: \"R0\", y: \"R0_mean_age\", r: 12, stroke: R0Color, fill: R0Color, fillOpacity: 0.6}),\n                Plot.text(\n                    mean_age_dots,\n                    {x: \"R0\", y: \"R0_mean_age\", text: (d) => `R0 = ${d.R0}`, dx: -60, dy: 30, fontWeight: \"bold\", fill: R0Color}\n                )\n                ] :\n            null,\n            Plot.arrow(mean_age_dots, {x1: \"R0\", x2: \"R0\", y1: \"arrow_start\", y2: \"arrow_end\", strokeWidth: 4, headLength: 5, inset: 15}),\n        ]\n    });\n\n  return plot;\n}\n"},{"methodName":"interpretQuiet","source":"shinyInput('reset')"},{"methodName":"interpretQuiet","source":"shinyInput('R0')"},{"methodName":"interpretQuiet","source":"shinyInput('vacc')"},{"methodName":"interpretQuiet","source":"shinyInput('lifeexp')"}]}
</script><script type="module">
if (window.location.protocol === "file:") { alert("The OJS runtime does not work with file:// URLs. Please use a web server to view this document."); }
window._ojs.paths.runtimeToDoc = "../..";
window._ojs.paths.runtimeToRoot = "../..";
window._ojs.paths.docToRoot = "";
window._ojs.selfContained = false;
window._ojs.runtime.interpretFromScriptTags();
</script><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./L05_age-structure.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Age Structured Models</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./L06_parameter-estimation.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Parameter Estimation</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb90" data-shortcodes="false" data-code-line-numbers=""><pre class="sourceCode numberSource markdown number-lines code-with-copy"><code class="sourceCode markdown"><span id="cb90-1"><a href="#cb90-1"></a><span class="co">---</span></span>
<span id="cb90-2"><a href="#cb90-2"></a><span class="an">subtitle:</span><span class="co"> "Heterogeneity and Age Structure in SIR Models"</span></span>
<span id="cb90-3"><a href="#cb90-3"></a><span class="an">abstract-title:</span><span class="co"> ""</span></span>
<span id="cb90-4"><a href="#cb90-4"></a><span class="an">abstract:</span><span class="co"> |</span></span>
<span id="cb90-5"><a href="#cb90-5"></a><span class="co">    *Materials adapted from Helen Wearing and Aaron King [@kingAgeStructuredModels2011]*</span></span>
<span id="cb90-6"><a href="#cb90-6"></a><span class="an">execute:</span></span>
<span id="cb90-7"><a href="#cb90-7"></a><span class="co">    warning: false</span></span>
<span id="cb90-8"><a href="#cb90-8"></a><span class="an">metadata-files:</span><span class="co"> </span></span>
<span id="cb90-9"><a href="#cb90-9"></a><span class="co">    - metadata/matthewferrari.yml</span></span>
<span id="cb90-10"><a href="#cb90-10"></a><span class="co">    - metadata/mathjax-packages.yml</span></span>
<span id="cb90-11"><a href="#cb90-11"></a><span class="co">---</span></span>
<span id="cb90-12"><a href="#cb90-12"></a></span>
<span id="cb90-13"><a href="#cb90-13"></a><span class="fu"># R Session 02</span></span>
<span id="cb90-14"><a href="#cb90-14"></a><span class="fu">## Load Packages</span></span>
<span id="cb90-15"><a href="#cb90-15"></a></span>
<span id="cb90-18"><a href="#cb90-18"></a><span class="in">```{r}</span></span>
<span id="cb90-19"><a href="#cb90-19"></a><span class="fu">library</span>(diagram)</span>
<span id="cb90-20"><a href="#cb90-20"></a><span class="fu">library</span>(deSolve)</span>
<span id="cb90-21"><a href="#cb90-21"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb90-22"><a href="#cb90-22"></a><span class="fu">library</span>(gt)</span>
<span id="cb90-23"><a href="#cb90-23"></a><span class="fu">library</span>(rio)</span>
<span id="cb90-24"><a href="#cb90-24"></a><span class="in">```</span></span>
<span id="cb90-25"><a href="#cb90-25"></a></span>
<span id="cb90-28"><a href="#cb90-28"></a><span class="in">```{r}</span></span>
<span id="cb90-29"><a href="#cb90-29"></a><span class="fu">theme_set</span>(<span class="fu">theme_minimal</span>())</span>
<span id="cb90-30"><a href="#cb90-30"></a><span class="in">```</span></span>
<span id="cb90-31"><a href="#cb90-31"></a></span>
<span id="cb90-32"><a href="#cb90-32"></a><span class="fu">## A Model With 2 Classes</span></span>
<span id="cb90-33"><a href="#cb90-33"></a></span>
<span id="cb90-34"><a href="#cb90-34"></a>We'll start with the simplest mechanistic model of two classes we can think of, which has separate classes for two groups $a$ and $b$. These groups could represent different socioeconomic classes, for example.</span>
<span id="cb90-35"><a href="#cb90-35"></a></span>
<span id="cb90-38"><a href="#cb90-38"></a><span class="in">```{r}</span></span>
<span id="cb90-39"><a href="#cb90-39"></a><span class="co">#| echo: false</span></span>
<span id="cb90-40"><a href="#cb90-40"></a><span class="co">#| column: body</span></span>
<span id="cb90-41"><a href="#cb90-41"></a><span class="co">#| out-width: 100%</span></span>
<span id="cb90-42"><a href="#cb90-42"></a>elpos <span class="ot">&lt;-</span> <span class="fu">rbind</span>(</span>
<span id="cb90-43"><a href="#cb90-43"></a>    <span class="at">Sa =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">3</span>),</span>
<span id="cb90-44"><a href="#cb90-44"></a>    <span class="at">Sb =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">3</span>),</span>
<span id="cb90-45"><a href="#cb90-45"></a>    <span class="at">Ia =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>),</span>
<span id="cb90-46"><a href="#cb90-46"></a>    <span class="at">Ib =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">2</span>),</span>
<span id="cb90-47"><a href="#cb90-47"></a>    <span class="at">Ra =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>),</span>
<span id="cb90-48"><a href="#cb90-48"></a>    <span class="at">Rb =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">1</span>)</span>
<span id="cb90-49"><a href="#cb90-49"></a>)</span>
<span id="cb90-50"><a href="#cb90-50"></a></span>
<span id="cb90-51"><a href="#cb90-51"></a>elpos[, <span class="dv">1</span>] <span class="ot">&lt;-</span> (<span class="dv">2</span> <span class="sc">*</span> elpos[, <span class="dv">1</span>] <span class="sc">-</span> <span class="dv">1</span>) <span class="sc">/</span> <span class="dv">8</span></span>
<span id="cb90-52"><a href="#cb90-52"></a>elpos[, <span class="dv">2</span>] <span class="ot">&lt;-</span> (<span class="dv">2</span> <span class="sc">*</span> elpos[, <span class="dv">2</span>] <span class="sc">-</span> <span class="dv">1</span>) <span class="sc">/</span> <span class="dv">6</span></span>
<span id="cb90-53"><a href="#cb90-53"></a></span>
<span id="cb90-54"><a href="#cb90-54"></a>fromto <span class="ot">&lt;-</span> <span class="fu">rbind</span>(</span>
<span id="cb90-55"><a href="#cb90-55"></a>    <span class="at">SaIa =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>),</span>
<span id="cb90-56"><a href="#cb90-56"></a>    <span class="at">SbIb =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">4</span>),</span>
<span id="cb90-57"><a href="#cb90-57"></a>    <span class="at">IaRa =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">5</span>),</span>
<span id="cb90-58"><a href="#cb90-58"></a>    <span class="at">IbRb =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">6</span>)</span>
<span id="cb90-59"><a href="#cb90-59"></a>)</span>
<span id="cb90-60"><a href="#cb90-60"></a></span>
<span id="cb90-61"><a href="#cb90-61"></a>op <span class="ot">&lt;-</span> <span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>))</span>
<span id="cb90-62"><a href="#cb90-62"></a></span>
<span id="cb90-63"><a href="#cb90-63"></a>diagram<span class="sc">::</span><span class="fu">openplotmat</span>(<span class="at">asp =</span> <span class="fl">0.9</span>)</span>
<span id="cb90-64"><a href="#cb90-64"></a></span>
<span id="cb90-65"><a href="#cb90-65"></a>arrpos <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">nrow =</span> <span class="fu">nrow</span>(fromto))</span>
<span id="cb90-66"><a href="#cb90-66"></a></span>
<span id="cb90-67"><a href="#cb90-67"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq_len</span>(<span class="fu">nrow</span>(fromto))){</span>
<span id="cb90-68"><a href="#cb90-68"></a>    arrpos[i,] <span class="ot">&lt;-</span> diagram<span class="sc">::</span><span class="fu">straightarrow</span>(</span>
<span id="cb90-69"><a href="#cb90-69"></a>        <span class="at">to =</span> elpos[fromto[i, <span class="dv">2</span>], ],</span>
<span id="cb90-70"><a href="#cb90-70"></a>        <span class="at">from =</span> elpos[fromto[i, <span class="dv">1</span>], ],</span>
<span id="cb90-71"><a href="#cb90-71"></a>        <span class="at">lwd =</span> <span class="dv">2</span>,</span>
<span id="cb90-72"><a href="#cb90-72"></a>        <span class="at">arr.pos =</span> <span class="fl">0.65</span>,</span>
<span id="cb90-73"><a href="#cb90-73"></a>        <span class="at">arr.length =</span> <span class="fl">0.5</span></span>
<span id="cb90-74"><a href="#cb90-74"></a>    )</span>
<span id="cb90-75"><a href="#cb90-75"></a>}</span>
<span id="cb90-76"><a href="#cb90-76"></a></span>
<span id="cb90-77"><a href="#cb90-77"></a><span class="co"># Add labels to the boxes</span></span>
<span id="cb90-78"><a href="#cb90-78"></a>purrr<span class="sc">::</span><span class="fu">walk</span>(</span>
<span id="cb90-79"><a href="#cb90-79"></a>    <span class="fu">c</span>(<span class="st">"Sa"</span>, <span class="st">"Sb"</span>, <span class="st">"Ia"</span>, <span class="st">"Ib"</span>, <span class="st">"Ra"</span>, <span class="st">"Rb"</span>),</span>
<span id="cb90-80"><a href="#cb90-80"></a>    <span class="at">.f =</span> <span class="cf">function</span>(.x) {</span>
<span id="cb90-81"><a href="#cb90-81"></a>        diagram<span class="sc">::</span><span class="fu">textrect</span>(</span>
<span id="cb90-82"><a href="#cb90-82"></a>            elpos[.x, ],</span>
<span id="cb90-83"><a href="#cb90-83"></a>            <span class="fl">0.07</span>,</span>
<span id="cb90-84"><a href="#cb90-84"></a>            <span class="fl">0.07</span>,</span>
<span id="cb90-85"><a href="#cb90-85"></a>            <span class="at">lab =</span> .x,</span>
<span id="cb90-86"><a href="#cb90-86"></a>            <span class="at">box.col =</span> <span class="fu">gray</span>(<span class="fl">0.7</span>),</span>
<span id="cb90-87"><a href="#cb90-87"></a>            <span class="at">shadow.col =</span> <span class="fu">gray</span>(<span class="fl">0.4</span>),</span>
<span id="cb90-88"><a href="#cb90-88"></a>            <span class="at">shadow.size =</span> <span class="fl">0.01</span>,</span>
<span id="cb90-89"><a href="#cb90-89"></a>            <span class="at">cex =</span> <span class="dv">2</span></span>
<span id="cb90-90"><a href="#cb90-90"></a>        )</span>
<span id="cb90-91"><a href="#cb90-91"></a>    }</span>
<span id="cb90-92"><a href="#cb90-92"></a>)</span>
<span id="cb90-93"><a href="#cb90-93"></a></span>
<span id="cb90-94"><a href="#cb90-94"></a><span class="co"># Add labels to the arrows</span></span>
<span id="cb90-95"><a href="#cb90-95"></a>purrr<span class="sc">::</span><span class="fu">pwalk</span>(</span>
<span id="cb90-96"><a href="#cb90-96"></a>    <span class="at">.l =</span> <span class="fu">list</span>(</span>
<span id="cb90-97"><a href="#cb90-97"></a>        <span class="at">state =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>,</span>
<span id="cb90-98"><a href="#cb90-98"></a>        <span class="at">xadjust =</span> <span class="fu">rep</span>(<span class="fl">0.05</span>, <span class="dv">4</span>),</span>
<span id="cb90-99"><a href="#cb90-99"></a>        <span class="at">yadjust =</span> <span class="fu">rep</span>(<span class="sc">-</span><span class="fl">0.175</span>, <span class="dv">4</span>),</span>
<span id="cb90-100"><a href="#cb90-100"></a>        <span class="at">label =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"lambda"</span>, <span class="st">"gamma"</span>), <span class="dv">2</span>),</span>
<span id="cb90-101"><a href="#cb90-101"></a>        <span class="at">label_subscript =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"a"</span>, <span class="st">"b"</span>), <span class="dv">2</span>)</span>
<span id="cb90-102"><a href="#cb90-102"></a>    ),</span>
<span id="cb90-103"><a href="#cb90-103"></a>    <span class="at">.f =</span> <span class="cf">function</span>(state, xadjust, yadjust, label, label_subscript) {</span>
<span id="cb90-104"><a href="#cb90-104"></a>        <span class="fu">text</span>(</span>
<span id="cb90-105"><a href="#cb90-105"></a>            elpos[state, <span class="dv">1</span>] <span class="sc">+</span> xadjust,</span>
<span id="cb90-106"><a href="#cb90-106"></a>            elpos[state, <span class="dv">2</span>] <span class="sc">+</span> yadjust,</span>
<span id="cb90-107"><a href="#cb90-107"></a>            <span class="fu">bquote</span>(.(<span class="fu">ensym</span>(label))[.(label_subscript)]),</span>
<span id="cb90-108"><a href="#cb90-108"></a>            <span class="at">cex =</span> <span class="dv">2</span></span>
<span id="cb90-109"><a href="#cb90-109"></a>        )</span>
<span id="cb90-110"><a href="#cb90-110"></a>    }</span>
<span id="cb90-111"><a href="#cb90-111"></a>)</span>
<span id="cb90-112"><a href="#cb90-112"></a></span>
<span id="cb90-113"><a href="#cb90-113"></a><span class="fu">par</span>(op)</span>
<span id="cb90-114"><a href="#cb90-114"></a><span class="in">```</span></span>
<span id="cb90-115"><a href="#cb90-115"></a></span>
<span id="cb90-116"><a href="#cb90-116"></a>Which can be written in equations as,</span>
<span id="cb90-117"><a href="#cb90-117"></a>$$</span>
<span id="cb90-118"><a href="#cb90-118"></a>\begin{aligned}</span>
<span id="cb90-119"><a href="#cb90-119"></a>    \frac{\dd{S_a}}{\dd{t}} &amp;= -\lambda_a\,S_a \phantom{-\gamma\,I_b}<span class="sc">\\</span></span>
<span id="cb90-120"><a href="#cb90-120"></a>    \frac{\dd{S_b}}{\dd{t}} &amp;= -\lambda_b\,S_b \phantom{-\gamma\,I_b}<span class="sc">\\</span></span>
<span id="cb90-121"><a href="#cb90-121"></a>    \frac{\dd{I_a}}{\dd{t}} &amp;= \phantom{-}\lambda_a\,S_a -\gamma\,I_a<span class="sc">\\</span></span>
<span id="cb90-122"><a href="#cb90-122"></a>    \frac{\dd{I_b}}{\dd{t}} &amp;= \phantom{-}\lambda_b\,S_b-\gamma\,I_b<span class="sc">\\</span></span>
<span id="cb90-123"><a href="#cb90-123"></a>    \frac{\dd{R_a}}{\dd{t}} &amp;= \phantom{-\lambda_a\,S_b}+\gamma\,I_a<span class="sc">\\</span></span>
<span id="cb90-124"><a href="#cb90-124"></a>    \frac{\dd{R_b}}{\dd{t}} &amp;= \phantom{-\lambda_a\,S_b}+\gamma\,I_b<span class="sc">\\</span></span>
<span id="cb90-125"><a href="#cb90-125"></a>  \end{aligned}</span>
<span id="cb90-126"><a href="#cb90-126"></a>$$</span>
<span id="cb90-127"><a href="#cb90-127"></a></span>
<span id="cb90-128"><a href="#cb90-128"></a>The $\lambda$s denote the group-specific force of infections:</span>
<span id="cb90-129"><a href="#cb90-129"></a></span>
<span id="cb90-130"><a href="#cb90-130"></a>$$</span>
<span id="cb90-131"><a href="#cb90-131"></a>\begin{aligned}</span>
<span id="cb90-132"><a href="#cb90-132"></a>        \lambda_a &amp;= \beta_{aa}\,I_a+\beta_{ab}\,I_b<span class="sc">\\</span></span>
<span id="cb90-133"><a href="#cb90-133"></a>        \lambda_b &amp;= \beta_{ba}\,I_a+\beta_{bb}\,I_b</span>
<span id="cb90-134"><a href="#cb90-134"></a>\end{aligned}</span>
<span id="cb90-135"><a href="#cb90-135"></a>$$</span>
<span id="cb90-136"><a href="#cb90-136"></a></span>
<span id="cb90-137"><a href="#cb90-137"></a>In this model, each population can infect each other but the infection moves through the populations separately.</span>
<span id="cb90-138"><a href="#cb90-138"></a>Let's simulate such a model.</span>
<span id="cb90-139"><a href="#cb90-139"></a>To make things concrete, we'll assume that the transmission rates $\beta$ are greater within groups than between them.</span>
<span id="cb90-140"><a href="#cb90-140"></a></span>
<span id="cb90-143"><a href="#cb90-143"></a><span class="in">```{r}</span></span>
<span id="cb90-144"><a href="#cb90-144"></a><span class="co"># Create a named parameter vector that we can index by name in the model</span></span>
<span id="cb90-145"><a href="#cb90-145"></a>ab_params <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb90-146"><a href="#cb90-146"></a>    <span class="at">beta_within =</span> <span class="fl">0.025</span>,</span>
<span id="cb90-147"><a href="#cb90-147"></a>    <span class="at">beta_between =</span> <span class="fl">0.005</span>,</span>
<span id="cb90-148"><a href="#cb90-148"></a>    <span class="at">recovery =</span> <span class="dv">10</span></span>
<span id="cb90-149"><a href="#cb90-149"></a>)</span>
<span id="cb90-150"><a href="#cb90-150"></a><span class="in">```</span></span>
<span id="cb90-151"><a href="#cb90-151"></a></span>
<span id="cb90-154"><a href="#cb90-154"></a><span class="in">```{r}</span></span>
<span id="cb90-155"><a href="#cb90-155"></a><span class="co"># Here we set up the ODE model that matches the equations above</span></span>
<span id="cb90-156"><a href="#cb90-156"></a>ab_model <span class="ot">&lt;-</span> <span class="cf">function</span> (t, x, p, ...) {</span>
<span id="cb90-157"><a href="#cb90-157"></a>    <span class="co"># Unpack the state variables</span></span>
<span id="cb90-158"><a href="#cb90-158"></a>    Sa <span class="ot">&lt;-</span> x[<span class="st">"Sa"</span>]</span>
<span id="cb90-159"><a href="#cb90-159"></a>    Sb <span class="ot">&lt;-</span> x[<span class="st">"Sb"</span>]</span>
<span id="cb90-160"><a href="#cb90-160"></a>    Ia <span class="ot">&lt;-</span> x[<span class="st">"Ia"</span>]</span>
<span id="cb90-161"><a href="#cb90-161"></a>    Ib <span class="ot">&lt;-</span> x[<span class="st">"Ib"</span>]</span>
<span id="cb90-162"><a href="#cb90-162"></a>    </span>
<span id="cb90-163"><a href="#cb90-163"></a>    <span class="co"># Unpack the parameters</span></span>
<span id="cb90-164"><a href="#cb90-164"></a>    beta_within <span class="ot">&lt;-</span> p[<span class="st">"beta_within"</span>]</span>
<span id="cb90-165"><a href="#cb90-165"></a>    beta_between <span class="ot">&lt;-</span> p[<span class="st">"beta_between"</span>]</span>
<span id="cb90-166"><a href="#cb90-166"></a>    recovery <span class="ot">&lt;-</span> p[<span class="st">"recovery"</span>]</span>
<span id="cb90-167"><a href="#cb90-167"></a></span>
<span id="cb90-168"><a href="#cb90-168"></a>    <span class="co"># group A force of infection</span></span>
<span id="cb90-169"><a href="#cb90-169"></a>    lambda_a <span class="ot">&lt;-</span> beta_within <span class="sc">*</span> Ia <span class="sc">+</span> beta_between <span class="sc">*</span> Ib</span>
<span id="cb90-170"><a href="#cb90-170"></a></span>
<span id="cb90-171"><a href="#cb90-171"></a>    <span class="co"># group B force of infection</span></span>
<span id="cb90-172"><a href="#cb90-172"></a>    lambda_b <span class="ot">&lt;-</span> beta_within <span class="sc">*</span> Ib <span class="sc">+</span> beta_between <span class="sc">*</span> Ia</span>
<span id="cb90-173"><a href="#cb90-173"></a>    </span>
<span id="cb90-174"><a href="#cb90-174"></a>    <span class="co"># The ODEs</span></span>
<span id="cb90-175"><a href="#cb90-175"></a>    dSadt <span class="ot">&lt;-</span> <span class="sc">-</span> lambda_a <span class="sc">*</span> Sa</span>
<span id="cb90-176"><a href="#cb90-176"></a>    dSbdt <span class="ot">&lt;-</span> <span class="sc">-</span> lambda_b <span class="sc">*</span> Sb</span>
<span id="cb90-177"><a href="#cb90-177"></a>    dIadt <span class="ot">&lt;-</span> lambda_a <span class="sc">*</span> Sa <span class="sc">-</span> recovery <span class="sc">*</span> Ia</span>
<span id="cb90-178"><a href="#cb90-178"></a>    dIbdt <span class="ot">&lt;-</span> lambda_b <span class="sc">*</span> Sb <span class="sc">-</span> recovery <span class="sc">*</span> Ib</span>
<span id="cb90-179"><a href="#cb90-179"></a>    dRadt <span class="ot">&lt;-</span> recovery <span class="sc">*</span> Ia</span>
<span id="cb90-180"><a href="#cb90-180"></a>    dRbdt <span class="ot">&lt;-</span> recovery <span class="sc">*</span> Ib</span>
<span id="cb90-181"><a href="#cb90-181"></a></span>
<span id="cb90-182"><a href="#cb90-182"></a>    <span class="co"># Return the derivatives</span></span>
<span id="cb90-183"><a href="#cb90-183"></a>    <span class="fu">list</span>(<span class="fu">c</span>(</span>
<span id="cb90-184"><a href="#cb90-184"></a>        dSadt,</span>
<span id="cb90-185"><a href="#cb90-185"></a>        dSbdt,</span>
<span id="cb90-186"><a href="#cb90-186"></a>        dIadt,</span>
<span id="cb90-187"><a href="#cb90-187"></a>        dIbdt,</span>
<span id="cb90-188"><a href="#cb90-188"></a>        dRadt,</span>
<span id="cb90-189"><a href="#cb90-189"></a>        dRbdt</span>
<span id="cb90-190"><a href="#cb90-190"></a>    ))</span>
<span id="cb90-191"><a href="#cb90-191"></a>}</span>
<span id="cb90-192"><a href="#cb90-192"></a><span class="in">```</span></span>
<span id="cb90-193"><a href="#cb90-193"></a></span>
<span id="cb90-196"><a href="#cb90-196"></a><span class="in">```{r}</span></span>
<span id="cb90-197"><a href="#cb90-197"></a><span class="co"># initial conditions</span></span>
<span id="cb90-198"><a href="#cb90-198"></a>ab_yinit <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="at">Sa =</span> <span class="dv">1000</span>, <span class="at">Sb =</span> <span class="dv">2000</span>, <span class="at">Ia =</span> <span class="dv">1</span>, <span class="at">Ib =</span> <span class="dv">1</span>, <span class="at">Ra =</span> <span class="dv">0</span>, <span class="at">Rb =</span> <span class="dv">0</span>)</span>
<span id="cb90-199"><a href="#cb90-199"></a></span>
<span id="cb90-200"><a href="#cb90-200"></a><span class="co"># Run the ODE solver from the deSolve package</span></span>
<span id="cb90-201"><a href="#cb90-201"></a>ab_sol <span class="ot">&lt;-</span> deSolve<span class="sc">::</span><span class="fu">ode</span>(</span>
<span id="cb90-202"><a href="#cb90-202"></a>    <span class="at">y =</span> ab_yinit,</span>
<span id="cb90-203"><a href="#cb90-203"></a>    <span class="at">times =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">2</span>, <span class="at">by =</span> <span class="fl">0.001</span>),</span>
<span id="cb90-204"><a href="#cb90-204"></a>    <span class="at">func =</span> ab_model,</span>
<span id="cb90-205"><a href="#cb90-205"></a>    <span class="at">parms =</span> ab_params,</span>
<span id="cb90-206"><a href="#cb90-206"></a>)</span>
<span id="cb90-207"><a href="#cb90-207"></a><span class="in">```</span></span>
<span id="cb90-208"><a href="#cb90-208"></a></span>
<span id="cb90-211"><a href="#cb90-211"></a><span class="in">```{r}</span></span>
<span id="cb90-212"><a href="#cb90-212"></a>ab_df <span class="ot">&lt;-</span> ab_sol <span class="sc">%&gt;%</span></span>
<span id="cb90-213"><a href="#cb90-213"></a>    <span class="co"># Convert the solution to a tibble for manipulation</span></span>
<span id="cb90-214"><a href="#cb90-214"></a>    <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb90-215"><a href="#cb90-215"></a>    <span class="co"># Create and modify columns</span></span>
<span id="cb90-216"><a href="#cb90-216"></a>    <span class="fu">mutate</span>(</span>
<span id="cb90-217"><a href="#cb90-217"></a>        <span class="co"># Convert all columns into type numeric</span></span>
<span id="cb90-218"><a href="#cb90-218"></a>        <span class="fu">across</span>(<span class="fu">everything</span>(), as.numeric),</span>
<span id="cb90-219"><a href="#cb90-219"></a>        <span class="co"># Create new columns to track pop sizes in each group</span></span>
<span id="cb90-220"><a href="#cb90-220"></a>        <span class="at">Na =</span> Sa <span class="sc">+</span> Ia <span class="sc">+</span> Ra,</span>
<span id="cb90-221"><a href="#cb90-221"></a>        <span class="at">Nb =</span> Sb <span class="sc">+</span> Ib <span class="sc">+</span> Rb</span>
<span id="cb90-222"><a href="#cb90-222"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb90-223"><a href="#cb90-223"></a>    <span class="co"># Go from a wide to long dataframe for ggplot</span></span>
<span id="cb90-224"><a href="#cb90-224"></a>    <span class="fu">pivot_longer</span>(</span>
<span id="cb90-225"><a href="#cb90-225"></a>        <span class="at">cols =</span> <span class="sc">-</span>time,</span>
<span id="cb90-226"><a href="#cb90-226"></a>        <span class="at">names_to =</span> <span class="fu">c</span>(<span class="st">"state"</span>, <span class="st">"group"</span>),</span>
<span id="cb90-227"><a href="#cb90-227"></a>        <span class="at">names_sep =</span> <span class="dv">1</span>,</span>
<span id="cb90-228"><a href="#cb90-228"></a>        <span class="at">values_to =</span> <span class="st">"value"</span></span>
<span id="cb90-229"><a href="#cb90-229"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb90-230"><a href="#cb90-230"></a>    <span class="co"># Clean pivoted columns for ordered plots</span></span>
<span id="cb90-231"><a href="#cb90-231"></a>    <span class="fu">mutate</span>(</span>
<span id="cb90-232"><a href="#cb90-232"></a>        <span class="at">state =</span> <span class="fu">factor</span>(state, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"S"</span>, <span class="st">"I"</span>, <span class="st">"R"</span>, <span class="st">"N"</span>)),</span>
<span id="cb90-233"><a href="#cb90-233"></a>        <span class="at">group =</span> <span class="fu">paste</span>(<span class="st">"Group"</span>, <span class="fu">str_to_upper</span>(group))</span>
<span id="cb90-234"><a href="#cb90-234"></a>    )</span>
<span id="cb90-235"><a href="#cb90-235"></a><span class="in">```</span></span>
<span id="cb90-236"><a href="#cb90-236"></a></span>
<span id="cb90-239"><a href="#cb90-239"></a><span class="in">```{r}</span></span>
<span id="cb90-240"><a href="#cb90-240"></a><span class="co">#| column: body</span></span>
<span id="cb90-241"><a href="#cb90-241"></a><span class="co">#| out-width: 100%</span></span>
<span id="cb90-242"><a href="#cb90-242"></a><span class="co"># Create a vector of colors to be used throughout the ggplots</span></span>
<span id="cb90-243"><a href="#cb90-243"></a>SIRcolors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"#1f77b4"</span>, <span class="st">"#ff7f0e"</span>, <span class="st">"#FF3851"</span>, <span class="st">"#591099"</span>)</span>
<span id="cb90-244"><a href="#cb90-244"></a></span>
<span id="cb90-245"><a href="#cb90-245"></a><span class="fu">ggplot</span>(ab_df, <span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> value, <span class="at">color =</span> state)) <span class="sc">+</span></span>
<span id="cb90-246"><a href="#cb90-246"></a>    <span class="fu">geom_line</span>(<span class="at">linewidth =</span> <span class="fl">1.5</span>) <span class="sc">+</span></span>
<span id="cb90-247"><a href="#cb90-247"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span>group, <span class="at">scales =</span> <span class="st">"free_y"</span>) <span class="sc">+</span></span>
<span id="cb90-248"><a href="#cb90-248"></a>    <span class="fu">scale_color_manual</span>(</span>
<span id="cb90-249"><a href="#cb90-249"></a>        <span class="at">values =</span> SIRcolors,</span>
<span id="cb90-250"><a href="#cb90-250"></a>        <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"Susceptible"</span>, <span class="st">"Infected"</span>, <span class="st">"Recovered"</span>, <span class="st">"Total"</span>)</span>
<span id="cb90-251"><a href="#cb90-251"></a>    ) <span class="sc">+</span></span>
<span id="cb90-252"><a href="#cb90-252"></a>    <span class="fu">labs</span>(</span>
<span id="cb90-253"><a href="#cb90-253"></a>        <span class="at">x =</span> <span class="st">"Time"</span>,</span>
<span id="cb90-254"><a href="#cb90-254"></a>        <span class="at">y =</span> <span class="st">"Number of individuals"</span>,</span>
<span id="cb90-255"><a href="#cb90-255"></a>        <span class="at">color =</span> <span class="st">"State"</span></span>
<span id="cb90-256"><a href="#cb90-256"></a>    ) <span class="sc">+</span></span>
<span id="cb90-257"><a href="#cb90-257"></a>    <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span>
<span id="cb90-258"><a href="#cb90-258"></a><span class="in">```</span></span>
<span id="cb90-259"><a href="#cb90-259"></a></span>
<span id="cb90-260"><a href="#cb90-260"></a>::: {.callout-question}</span>
<span id="cb90-261"><a href="#cb90-261"></a>Despite using the same transmission rates, the epidemic in group B is much larger than in group A.</span>
<span id="cb90-262"><a href="#cb90-262"></a>Why do you think this is?</span>
<span id="cb90-263"><a href="#cb90-263"></a>:::</span>
<span id="cb90-264"><a href="#cb90-264"></a></span>
<span id="cb90-265"><a href="#cb90-265"></a>Now let's plot the proportion of individuals in each state for the two groups.</span>
<span id="cb90-266"><a href="#cb90-266"></a></span>
<span id="cb90-269"><a href="#cb90-269"></a><span class="in">```{r}</span></span>
<span id="cb90-270"><a href="#cb90-270"></a>ab_df_props <span class="ot">&lt;-</span> ab_df <span class="sc">%&gt;%</span></span>
<span id="cb90-271"><a href="#cb90-271"></a>    <span class="co"># Remove total pop count as we only want the group-specific values</span></span>
<span id="cb90-272"><a href="#cb90-272"></a>    <span class="fu">filter</span>(state <span class="sc">!=</span> <span class="st">"N"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb90-273"><a href="#cb90-273"></a>    <span class="fu">mutate</span>(</span>
<span id="cb90-274"><a href="#cb90-274"></a>        <span class="co"># Concatenate the state variable and the group letter for each row</span></span>
<span id="cb90-275"><a href="#cb90-275"></a>        <span class="at">state_group =</span> <span class="fu">paste0</span>(state, <span class="fu">str_extract_all</span>(group, <span class="st">"[^Group ]"</span>)),</span>
<span id="cb90-276"><a href="#cb90-276"></a>        <span class="co"># Factor new variable for nicer plotting</span></span>
<span id="cb90-277"><a href="#cb90-277"></a>        <span class="at">state_group =</span> <span class="fu">factor</span>(state_group, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"RA"</span>, <span class="st">"RB"</span>, <span class="st">"IA"</span>, <span class="st">"IB"</span>, <span class="st">"SA"</span>, <span class="st">"SB"</span>))</span>
<span id="cb90-278"><a href="#cb90-278"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb90-279"><a href="#cb90-279"></a>    <span class="co"># Group by time and state_group so we can calculate the relevant proportions over time</span></span>
<span id="cb90-280"><a href="#cb90-280"></a>    <span class="fu">group_by</span>(time, state_group) <span class="sc">%&gt;%</span></span>
<span id="cb90-281"><a href="#cb90-281"></a>    <span class="fu">mutate</span>(</span>
<span id="cb90-282"><a href="#cb90-282"></a>        <span class="at">prop =</span> value <span class="sc">/</span> <span class="fu">sum</span>(ab_yinit)</span>
<span id="cb90-283"><a href="#cb90-283"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb90-284"><a href="#cb90-284"></a>    <span class="fu">ungroup</span>()</span>
<span id="cb90-285"><a href="#cb90-285"></a><span class="in">```</span></span>
<span id="cb90-286"><a href="#cb90-286"></a></span>
<span id="cb90-289"><a href="#cb90-289"></a><span class="in">```{r}</span></span>
<span id="cb90-290"><a href="#cb90-290"></a><span class="co">#| column: body</span></span>
<span id="cb90-291"><a href="#cb90-291"></a><span class="co">#| out-width: 100%</span></span>
<span id="cb90-292"><a href="#cb90-292"></a><span class="co"># Create new vectors of colors as using 6: one of each for A and J groups</span></span>
<span id="cb90-293"><a href="#cb90-293"></a>Scolors <span class="ot">&lt;-</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">3</span>, <span class="st">"Blues"</span>)[<span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">3</span>)]</span>
<span id="cb90-294"><a href="#cb90-294"></a>Icolors <span class="ot">&lt;-</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">3</span>, <span class="st">"Oranges"</span>)[<span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">3</span>)]</span>
<span id="cb90-295"><a href="#cb90-295"></a>Rcolors <span class="ot">&lt;-</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">3</span>, <span class="st">"Greens"</span>)[<span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">3</span>)]</span>
<span id="cb90-296"><a href="#cb90-296"></a></span>
<span id="cb90-297"><a href="#cb90-297"></a><span class="fu">ggplot</span>(ab_df_props, <span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> prop, <span class="at">fill =</span> state_group)) <span class="sc">+</span></span>
<span id="cb90-298"><a href="#cb90-298"></a>    <span class="fu">geom_area</span>() <span class="sc">+</span></span>
<span id="cb90-299"><a href="#cb90-299"></a>    <span class="fu">scale_fill_manual</span>(</span>
<span id="cb90-300"><a href="#cb90-300"></a>        <span class="at">values =</span> <span class="fu">c</span>(Scolors, Icolors, Rcolors),</span>
<span id="cb90-301"><a href="#cb90-301"></a>        <span class="at">limits =</span> <span class="fu">c</span>(<span class="st">"SA"</span>, <span class="st">"SB"</span>, <span class="st">"IA"</span>, <span class="st">"IB"</span>, <span class="st">"RA"</span>, <span class="st">"RB"</span>),</span>
<span id="cb90-302"><a href="#cb90-302"></a>    ) <span class="sc">+</span></span>
<span id="cb90-303"><a href="#cb90-303"></a>    <span class="fu">labs</span>(</span>
<span id="cb90-304"><a href="#cb90-304"></a>        <span class="at">x =</span> <span class="st">"Time"</span>,</span>
<span id="cb90-305"><a href="#cb90-305"></a>        <span class="at">y =</span> <span class="st">"Proportion of individuals"</span>,</span>
<span id="cb90-306"><a href="#cb90-306"></a>        <span class="at">fill =</span> <span class="st">"State"</span></span>
<span id="cb90-307"><a href="#cb90-307"></a>    ) <span class="sc">+</span></span>
<span id="cb90-308"><a href="#cb90-308"></a>    <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span>
<span id="cb90-309"><a href="#cb90-309"></a><span class="in">```</span></span>
<span id="cb90-310"><a href="#cb90-310"></a></span>
<span id="cb90-311"><a href="#cb90-311"></a><span class="fu">## A Model With 2 Age Classes</span></span>
<span id="cb90-312"><a href="#cb90-312"></a></span>
<span id="cb90-313"><a href="#cb90-313"></a>Note that age is a special kind of heterogeneity in an epidemic model because individuals necessarily move from one class (younger) to another class (older) in a directional fashion that is independent of the infection and recovery process.</span>
<span id="cb90-314"><a href="#cb90-314"></a></span>
<span id="cb90-315"><a href="#cb90-315"></a><span class="co">&lt;!-- Today's first lecture showed how force of infection can vary with age. --&gt;</span></span>
<span id="cb90-316"><a href="#cb90-316"></a></span>
<span id="cb90-317"><a href="#cb90-317"></a><span class="co">&lt;!-- What sort of mechanisms might give rise to these effects? --&gt;</span></span>
<span id="cb90-318"><a href="#cb90-318"></a></span>
<span id="cb90-319"><a href="#cb90-319"></a><span class="co">&lt;!-- Here we'll see to what extent we can infer these mechanisms on the basis of age-specific incidence and seroprevalence data. --&gt;</span></span>
<span id="cb90-320"><a href="#cb90-320"></a></span>
<span id="cb90-321"><a href="#cb90-321"></a>We'll start by introducing age into the model above.</span>
<span id="cb90-322"><a href="#cb90-322"></a>So now $a$ becomes juveniles and $b$ becomes adults.</span>
<span id="cb90-323"><a href="#cb90-323"></a>And, independent of the disease process, juveniles (of any category) age into adults.</span>
<span id="cb90-324"><a href="#cb90-324"></a>Additionally, new juveniles are added through births (always first susceptible) and old individuals are lost to death.</span>
<span id="cb90-325"><a href="#cb90-325"></a></span>
<span id="cb90-328"><a href="#cb90-328"></a><span class="in">```{r}</span></span>
<span id="cb90-329"><a href="#cb90-329"></a><span class="co">#| echo: false</span></span>
<span id="cb90-330"><a href="#cb90-330"></a><span class="co">#| column: body</span></span>
<span id="cb90-331"><a href="#cb90-331"></a><span class="co">#| out-width: 100%</span></span>
<span id="cb90-332"><a href="#cb90-332"></a>elpos <span class="ot">&lt;-</span> <span class="fu">rbind</span>(</span>
<span id="cb90-333"><a href="#cb90-333"></a>    <span class="at">B =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>),</span>
<span id="cb90-334"><a href="#cb90-334"></a>    <span class="at">Sj =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">3</span>),</span>
<span id="cb90-335"><a href="#cb90-335"></a>    <span class="at">Sa =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">3</span>),</span>
<span id="cb90-336"><a href="#cb90-336"></a>    <span class="at">Ij =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>),</span>
<span id="cb90-337"><a href="#cb90-337"></a>    <span class="at">Ia =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">2</span>),</span>
<span id="cb90-338"><a href="#cb90-338"></a>    <span class="at">Rj =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>),</span>
<span id="cb90-339"><a href="#cb90-339"></a>    <span class="at">Ra =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">1</span>),</span>
<span id="cb90-340"><a href="#cb90-340"></a>    <span class="at">Ds =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">3</span>),</span>
<span id="cb90-341"><a href="#cb90-341"></a>    <span class="at">Di =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">2</span>),</span>
<span id="cb90-342"><a href="#cb90-342"></a>    <span class="at">Dr =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">1</span>)</span>
<span id="cb90-343"><a href="#cb90-343"></a>)</span>
<span id="cb90-344"><a href="#cb90-344"></a></span>
<span id="cb90-345"><a href="#cb90-345"></a>elpos[, <span class="dv">2</span>] <span class="ot">&lt;-</span> (<span class="dv">2</span> <span class="sc">*</span> elpos[, <span class="dv">2</span>] <span class="sc">-</span> <span class="dv">1</span>) <span class="sc">/</span> <span class="dv">6</span></span>
<span id="cb90-346"><a href="#cb90-346"></a>elpos[, <span class="dv">1</span>] <span class="ot">&lt;-</span> (<span class="dv">2</span> <span class="sc">*</span> elpos[, <span class="dv">1</span>] <span class="sc">-</span> <span class="dv">1</span>) <span class="sc">/</span> <span class="dv">8</span></span>
<span id="cb90-347"><a href="#cb90-347"></a></span>
<span id="cb90-348"><a href="#cb90-348"></a>fromto <span class="ot">&lt;-</span> <span class="fu">rbind</span>(</span>
<span id="cb90-349"><a href="#cb90-349"></a>    <span class="at">BSj =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>),</span>
<span id="cb90-350"><a href="#cb90-350"></a>    <span class="at">RaD =</span> <span class="fu">c</span>(<span class="dv">7</span>, <span class="dv">10</span>),</span>
<span id="cb90-351"><a href="#cb90-351"></a>    <span class="at">SjSa =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">3</span>),</span>
<span id="cb90-352"><a href="#cb90-352"></a>    <span class="at">IjIa =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">5</span>),</span>
<span id="cb90-353"><a href="#cb90-353"></a>    <span class="at">RjRa =</span> <span class="fu">c</span>(<span class="dv">6</span>, <span class="dv">7</span>),</span>
<span id="cb90-354"><a href="#cb90-354"></a>    <span class="at">SjIj =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">4</span>),</span>
<span id="cb90-355"><a href="#cb90-355"></a>    <span class="at">SaIa =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">5</span>),</span>
<span id="cb90-356"><a href="#cb90-356"></a>    <span class="at">IjRj =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">6</span>),</span>
<span id="cb90-357"><a href="#cb90-357"></a>    <span class="at">IaRa =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">7</span>),</span>
<span id="cb90-358"><a href="#cb90-358"></a>    <span class="at">SaD =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">8</span>),</span>
<span id="cb90-359"><a href="#cb90-359"></a>    <span class="at">IaD =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">9</span>)</span>
<span id="cb90-360"><a href="#cb90-360"></a>)</span>
<span id="cb90-361"><a href="#cb90-361"></a></span>
<span id="cb90-362"><a href="#cb90-362"></a>op <span class="ot">&lt;-</span> <span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>))</span>
<span id="cb90-363"><a href="#cb90-363"></a></span>
<span id="cb90-364"><a href="#cb90-364"></a>diagram<span class="sc">::</span><span class="fu">openplotmat</span>(<span class="at">asp =</span> <span class="fl">1.0</span>)</span>
<span id="cb90-365"><a href="#cb90-365"></a></span>
<span id="cb90-366"><a href="#cb90-366"></a>arrpos <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">nrow =</span> <span class="fu">nrow</span>(fromto))</span>
<span id="cb90-367"><a href="#cb90-367"></a></span>
<span id="cb90-368"><a href="#cb90-368"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq_len</span>(<span class="fu">nrow</span>(fromto)))</span>
<span id="cb90-369"><a href="#cb90-369"></a>    arrpos[i, ] <span class="ot">&lt;-</span> diagram<span class="sc">::</span><span class="fu">straightarrow</span>(</span>
<span id="cb90-370"><a href="#cb90-370"></a>        <span class="at">to =</span> elpos[fromto[i, <span class="dv">2</span>], ],</span>
<span id="cb90-371"><a href="#cb90-371"></a>        <span class="at">from =</span> elpos[fromto[i, <span class="dv">1</span>], ],</span>
<span id="cb90-372"><a href="#cb90-372"></a>        <span class="at">lwd =</span> <span class="dv">2</span>,</span>
<span id="cb90-373"><a href="#cb90-373"></a>        <span class="at">arr.pos =</span> <span class="fl">0.65</span>,</span>
<span id="cb90-374"><a href="#cb90-374"></a>        <span class="at">arr.length =</span> <span class="fl">0.5</span></span>
<span id="cb90-375"><a href="#cb90-375"></a>    )</span>
<span id="cb90-376"><a href="#cb90-376"></a></span>
<span id="cb90-377"><a href="#cb90-377"></a>purrr<span class="sc">::</span><span class="fu">walk</span>(</span>
<span id="cb90-378"><a href="#cb90-378"></a>    <span class="fu">c</span>(<span class="st">"B"</span>, <span class="st">"Sj"</span>, <span class="st">"Sa"</span>, <span class="st">"Ij"</span>, <span class="st">"Ia"</span>, <span class="st">"Rj"</span>, <span class="st">"Ra"</span>, <span class="st">"Ds"</span>, <span class="st">"Di"</span>, <span class="st">"Dr"</span>),</span>
<span id="cb90-379"><a href="#cb90-379"></a>    <span class="at">.f =</span> <span class="cf">function</span>(.x) {</span>
<span id="cb90-380"><a href="#cb90-380"></a>        diagram<span class="sc">::</span><span class="fu">textrect</span>(</span>
<span id="cb90-381"><a href="#cb90-381"></a>            elpos[.x, ],</span>
<span id="cb90-382"><a href="#cb90-382"></a>            <span class="fl">0.07</span>,</span>
<span id="cb90-383"><a href="#cb90-383"></a>            <span class="fl">0.07</span>,</span>
<span id="cb90-384"><a href="#cb90-384"></a>            <span class="at">lab =</span> .x,</span>
<span id="cb90-385"><a href="#cb90-385"></a>            <span class="at">box.col =</span> <span class="fu">gray</span>(<span class="fl">0.7</span>),</span>
<span id="cb90-386"><a href="#cb90-386"></a>            <span class="at">shadow.col =</span> <span class="fu">gray</span>(<span class="fl">0.4</span>),</span>
<span id="cb90-387"><a href="#cb90-387"></a>            <span class="at">shadow.size =</span> <span class="fl">0.01</span>,</span>
<span id="cb90-388"><a href="#cb90-388"></a>            <span class="at">cex =</span> <span class="dv">2</span></span>
<span id="cb90-389"><a href="#cb90-389"></a>        )</span>
<span id="cb90-390"><a href="#cb90-390"></a>    }</span>
<span id="cb90-391"><a href="#cb90-391"></a>)</span>
<span id="cb90-392"><a href="#cb90-392"></a></span>
<span id="cb90-393"><a href="#cb90-393"></a>purrr<span class="sc">::</span><span class="fu">pwalk</span>(</span>
<span id="cb90-394"><a href="#cb90-394"></a>    <span class="at">.l =</span> <span class="fu">list</span>(</span>
<span id="cb90-395"><a href="#cb90-395"></a>        <span class="at">state =</span> <span class="fu">c</span>(<span class="fu">seq</span>(<span class="dv">2</span>, <span class="dv">7</span>, <span class="at">by =</span> <span class="dv">1</span>), <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>),</span>
<span id="cb90-396"><a href="#cb90-396"></a>        <span class="at">xadjust =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="fl">0.125</span>, <span class="dv">6</span>), <span class="fu">rep</span>(<span class="fl">0.05</span>, <span class="dv">4</span>)),</span>
<span id="cb90-397"><a href="#cb90-397"></a>        <span class="at">yadjust =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="fl">0.05</span>, <span class="dv">6</span>), <span class="fu">rep</span>(<span class="sc">-</span><span class="fl">0.175</span>, <span class="dv">4</span>)),</span>
<span id="cb90-398"><a href="#cb90-398"></a>        <span class="at">label =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"alpha"</span>, <span class="st">"mu"</span>), <span class="dv">3</span>), <span class="fu">rep</span>(<span class="st">"lambda"</span>, <span class="dv">2</span>), <span class="fu">rep</span>(<span class="st">"gamma"</span>, <span class="dv">2</span>)),</span>
<span id="cb90-399"><a href="#cb90-399"></a>        <span class="at">label_subscript =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">""</span>, <span class="dv">6</span>), <span class="st">"J"</span>, <span class="st">"A"</span>, <span class="fu">rep</span>(<span class="st">""</span>, <span class="dv">2</span>))</span>
<span id="cb90-400"><a href="#cb90-400"></a>    ),</span>
<span id="cb90-401"><a href="#cb90-401"></a>    <span class="at">.f =</span> <span class="cf">function</span>(state, xadjust, yadjust, label, label_subscript) {</span>
<span id="cb90-402"><a href="#cb90-402"></a>        <span class="fu">text</span>(</span>
<span id="cb90-403"><a href="#cb90-403"></a>            elpos[state, <span class="dv">1</span>] <span class="sc">+</span> xadjust,</span>
<span id="cb90-404"><a href="#cb90-404"></a>            elpos[state, <span class="dv">2</span>] <span class="sc">+</span> yadjust,</span>
<span id="cb90-405"><a href="#cb90-405"></a>            <span class="fu">bquote</span>(.(<span class="fu">ensym</span>(label))[.(label_subscript)]),</span>
<span id="cb90-406"><a href="#cb90-406"></a>            <span class="at">cex =</span> <span class="dv">2</span></span>
<span id="cb90-407"><a href="#cb90-407"></a>        )</span>
<span id="cb90-408"><a href="#cb90-408"></a>    }</span>
<span id="cb90-409"><a href="#cb90-409"></a>)</span>
<span id="cb90-410"><a href="#cb90-410"></a></span>
<span id="cb90-411"><a href="#cb90-411"></a><span class="fu">par</span>(op)</span>
<span id="cb90-412"><a href="#cb90-412"></a><span class="in">```</span></span>
<span id="cb90-413"><a href="#cb90-413"></a></span>
<span id="cb90-414"><a href="#cb90-414"></a>We can do this very simply using the same ingredients that go into the basic SIR model.</span>
<span id="cb90-415"><a href="#cb90-415"></a>In that model, the waiting times in the S and I classes are exponential.</span>
<span id="cb90-416"><a href="#cb90-416"></a>Let's assume the same thing about the aging process.</span>
<span id="cb90-417"><a href="#cb90-417"></a>We'll also add in births into the juvenile susceptible class and deaths from the adult classes.</span>
<span id="cb90-418"><a href="#cb90-418"></a></span>
<span id="cb90-419"><a href="#cb90-419"></a>$$</span>
<span id="cb90-420"><a href="#cb90-420"></a>  \begin{aligned}</span>
<span id="cb90-421"><a href="#cb90-421"></a>    \frac{\dd{S_J}}{\dd{t}} &amp;= B -\lambda_J\,S_J \phantom{- \gamma\,I_A} -\alpha\,S_J \phantom{-\mu\,S_A}<span class="sc">\\</span></span>
<span id="cb90-422"><a href="#cb90-422"></a>    \frac{\dd{S_A}}{\dd{t}} &amp;= \phantom{B} - \lambda_A\,S_A \phantom{- \gamma\,I_A} +\alpha\,S_J -\mu\,S_A<span class="sc">\\</span></span>
<span id="cb90-423"><a href="#cb90-423"></a>    \frac{\dd{I_J}}{\dd{t}} &amp;= \phantom{B} +\lambda_J\,S_J - \gamma\,I_J -\alpha\,I_J \phantom{-\mu\,S_A}<span class="sc">\\</span></span>
<span id="cb90-424"><a href="#cb90-424"></a>    \frac{\dd{I_A}}{\dd{t}} &amp;= \phantom{B} +\lambda_A\,S_A - \gamma\,I_A + \alpha\,I_J - \mu\,I_A<span class="sc">\\</span></span>
<span id="cb90-425"><a href="#cb90-425"></a>    \frac{\dd{R_J}}{\dd{t}} &amp;= \phantom{B - \lambda_J\,S_A} + \gamma\,I_J - \alpha\,R_J \phantom{- \mu\,S_A}<span class="sc">\\</span></span>
<span id="cb90-426"><a href="#cb90-426"></a>    \frac{\dd{R_A}}{\dd{t}} &amp;= \phantom{B - \lambda_J\,S_A} + \gamma\,I_A + \alpha\,R_J -\mu\,R_A<span class="sc">\\</span></span>
<span id="cb90-427"><a href="#cb90-427"></a>  \end{aligned}</span>
<span id="cb90-428"><a href="#cb90-428"></a>$$</span>
<span id="cb90-429"><a href="#cb90-429"></a></span>
<span id="cb90-430"><a href="#cb90-430"></a>Now, let's simulate this model, under the same assumptions about transmission rates as above.</span>
<span id="cb90-431"><a href="#cb90-431"></a></span>
<span id="cb90-434"><a href="#cb90-434"></a><span class="in">```{r}</span></span>
<span id="cb90-435"><a href="#cb90-435"></a><span class="co"># define the parameters for the demographic model</span></span>
<span id="cb90-436"><a href="#cb90-436"></a>demog_params <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb90-437"><a href="#cb90-437"></a>    <span class="at">beta_within =</span> <span class="fl">0.004</span>,</span>
<span id="cb90-438"><a href="#cb90-438"></a>    <span class="at">beta_between =</span> <span class="fl">0.002</span>,</span>
<span id="cb90-439"><a href="#cb90-439"></a>    <span class="at">recovery =</span> <span class="dv">10</span>,</span>
<span id="cb90-440"><a href="#cb90-440"></a>    <span class="at">births =</span> <span class="dv">100</span>,</span>
<span id="cb90-441"><a href="#cb90-441"></a>    <span class="co"># Width of age bands in years</span></span>
<span id="cb90-442"><a href="#cb90-442"></a>    <span class="at">age_band_j =</span> <span class="dv">20</span>,</span>
<span id="cb90-443"><a href="#cb90-443"></a>    <span class="at">age_band_a =</span> <span class="dv">60</span></span>
<span id="cb90-444"><a href="#cb90-444"></a>)</span>
<span id="cb90-445"><a href="#cb90-445"></a><span class="in">```</span></span>
<span id="cb90-446"><a href="#cb90-446"></a></span>
<span id="cb90-449"><a href="#cb90-449"></a><span class="in">```{r}</span></span>
<span id="cb90-450"><a href="#cb90-450"></a>demog_model <span class="ot">&lt;-</span> <span class="cf">function</span> (t, x, p, ...) {</span>
<span id="cb90-451"><a href="#cb90-451"></a>    <span class="co"># Unpack states</span></span>
<span id="cb90-452"><a href="#cb90-452"></a>    Sj <span class="ot">&lt;-</span> x[<span class="st">"Sj"</span>]</span>
<span id="cb90-453"><a href="#cb90-453"></a>    Sa <span class="ot">&lt;-</span> x[<span class="st">"Sa"</span>]</span>
<span id="cb90-454"><a href="#cb90-454"></a>    Ij <span class="ot">&lt;-</span> x[<span class="st">"Ij"</span>]</span>
<span id="cb90-455"><a href="#cb90-455"></a>    Ia <span class="ot">&lt;-</span> x[<span class="st">"Ia"</span>]</span>
<span id="cb90-456"><a href="#cb90-456"></a>    Rj <span class="ot">&lt;-</span> x[<span class="st">"Rj"</span>]</span>
<span id="cb90-457"><a href="#cb90-457"></a>    Ra <span class="ot">&lt;-</span> x[<span class="st">"Ra"</span>]</span>
<span id="cb90-458"><a href="#cb90-458"></a></span>
<span id="cb90-459"><a href="#cb90-459"></a>    <span class="co"># Unpack parameters from vector</span></span>
<span id="cb90-460"><a href="#cb90-460"></a>    beta_within <span class="ot">&lt;-</span> p[<span class="st">"beta_within"</span>]</span>
<span id="cb90-461"><a href="#cb90-461"></a>    beta_between <span class="ot">&lt;-</span> p[<span class="st">"beta_between"</span>]</span>
<span id="cb90-462"><a href="#cb90-462"></a>    recovery <span class="ot">&lt;-</span> p[<span class="st">"recovery"</span>]</span>
<span id="cb90-463"><a href="#cb90-463"></a>    births <span class="ot">&lt;-</span> p[<span class="st">"births"</span>]</span>
<span id="cb90-464"><a href="#cb90-464"></a>    <span class="co"># Calculate rate of aging from each age group</span></span>
<span id="cb90-465"><a href="#cb90-465"></a>    aging_j <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> p[<span class="st">"age_band_j"</span>]</span>
<span id="cb90-466"><a href="#cb90-466"></a>    aging_a <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> p[<span class="st">"age_band_a"</span>]</span>
<span id="cb90-467"><a href="#cb90-467"></a></span>
<span id="cb90-468"><a href="#cb90-468"></a>    <span class="co"># juv. force of infection</span></span>
<span id="cb90-469"><a href="#cb90-469"></a>    lambda_j <span class="ot">&lt;-</span> beta_within <span class="sc">*</span> Ij <span class="sc">+</span> beta_between <span class="sc">*</span> Ia</span>
<span id="cb90-470"><a href="#cb90-470"></a></span>
<span id="cb90-471"><a href="#cb90-471"></a>    <span class="co"># adult. force of infection</span></span>
<span id="cb90-472"><a href="#cb90-472"></a>    lambda_a <span class="ot">&lt;-</span> beta_within <span class="sc">*</span> Ia <span class="sc">+</span> beta_between <span class="sc">*</span> Ij</span>
<span id="cb90-473"><a href="#cb90-473"></a></span>
<span id="cb90-474"><a href="#cb90-474"></a>    <span class="co"># Calculate the ODEs</span></span>
<span id="cb90-475"><a href="#cb90-475"></a>    dSjdt <span class="ot">&lt;-</span> births <span class="sc">-</span> (lambda_j <span class="sc">*</span> Sj) <span class="sc">-</span> (aging_j <span class="sc">*</span> Sj)</span>
<span id="cb90-476"><a href="#cb90-476"></a>    dSadt <span class="ot">&lt;-</span> <span class="sc">-</span>(lambda_a <span class="sc">*</span> Sa) <span class="sc">+</span> (aging_j <span class="sc">*</span> Sj) <span class="sc">-</span> (aging_a <span class="sc">*</span> Sa)</span>
<span id="cb90-477"><a href="#cb90-477"></a>    dIjdt <span class="ot">&lt;-</span> (lambda_j <span class="sc">*</span> Sj) <span class="sc">-</span> (recovery <span class="sc">*</span> Ij) <span class="sc">-</span> (aging_j <span class="sc">*</span> Ij)</span>
<span id="cb90-478"><a href="#cb90-478"></a>    dIadt <span class="ot">&lt;-</span> (lambda_a <span class="sc">*</span> Sa) <span class="sc">-</span> (recovery <span class="sc">*</span> Ia) <span class="sc">+</span> (aging_j <span class="sc">*</span> Ij) <span class="sc">-</span> (aging_a <span class="sc">*</span> Ia)</span>
<span id="cb90-479"><a href="#cb90-479"></a>    dRjdt <span class="ot">&lt;-</span> (recovery <span class="sc">*</span> Ij) <span class="sc">-</span> (aging_j <span class="sc">*</span> Rj)</span>
<span id="cb90-480"><a href="#cb90-480"></a>    dRadt <span class="ot">&lt;-</span> (recovery <span class="sc">*</span> Ia) <span class="sc">+</span> (aging_j <span class="sc">*</span> Rj) <span class="sc">-</span> (aging_a <span class="sc">*</span> Ra)</span>
<span id="cb90-481"><a href="#cb90-481"></a></span>
<span id="cb90-482"><a href="#cb90-482"></a>    <span class="co"># Return the ODEs</span></span>
<span id="cb90-483"><a href="#cb90-483"></a>    <span class="fu">list</span>(<span class="fu">c</span>(</span>
<span id="cb90-484"><a href="#cb90-484"></a>        dSjdt,</span>
<span id="cb90-485"><a href="#cb90-485"></a>        dSadt,</span>
<span id="cb90-486"><a href="#cb90-486"></a>        dIjdt,</span>
<span id="cb90-487"><a href="#cb90-487"></a>        dIadt,</span>
<span id="cb90-488"><a href="#cb90-488"></a>        dRjdt,</span>
<span id="cb90-489"><a href="#cb90-489"></a>        dRadt</span>
<span id="cb90-490"><a href="#cb90-490"></a>    ))</span>
<span id="cb90-491"><a href="#cb90-491"></a>}</span>
<span id="cb90-492"><a href="#cb90-492"></a><span class="in">```</span></span>
<span id="cb90-493"><a href="#cb90-493"></a></span>
<span id="cb90-494"><a href="#cb90-494"></a>Note that in this function, $\mu=$ <span class="in">`aging_a`</span> $=$ <span class="in">`1 / p["age_band_a"]`</span>, i.e., death, is just like another age class.</span>
<span id="cb90-495"><a href="#cb90-495"></a></span>
<span id="cb90-498"><a href="#cb90-498"></a><span class="in">```{r}</span></span>
<span id="cb90-499"><a href="#cb90-499"></a><span class="co"># initial conditions</span></span>
<span id="cb90-500"><a href="#cb90-500"></a>demog_yinit <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="at">Sj =</span> <span class="dv">2000</span>, <span class="at">Sa =</span> <span class="dv">3000</span>, <span class="at">Ij =</span> <span class="dv">0</span>, <span class="at">Ia =</span> <span class="dv">1</span>, <span class="at">Rj =</span> <span class="dv">0</span>, <span class="at">Ra =</span> <span class="dv">1000</span>)</span>
<span id="cb90-501"><a href="#cb90-501"></a></span>
<span id="cb90-502"><a href="#cb90-502"></a><span class="co"># Solve the demographic model</span></span>
<span id="cb90-503"><a href="#cb90-503"></a>demog_sol <span class="ot">&lt;-</span> deSolve<span class="sc">::</span><span class="fu">ode</span>(</span>
<span id="cb90-504"><a href="#cb90-504"></a>    <span class="at">y =</span> demog_yinit,</span>
<span id="cb90-505"><a href="#cb90-505"></a>    <span class="at">times =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">200</span>, <span class="at">by =</span> <span class="fl">0.1</span>),</span>
<span id="cb90-506"><a href="#cb90-506"></a>    <span class="at">func =</span> demog_model,</span>
<span id="cb90-507"><a href="#cb90-507"></a>    <span class="at">parms =</span> demog_params</span>
<span id="cb90-508"><a href="#cb90-508"></a>)</span>
<span id="cb90-509"><a href="#cb90-509"></a></span>
<span id="cb90-510"><a href="#cb90-510"></a>demog_df <span class="ot">&lt;-</span> demog_sol <span class="sc">%&gt;%</span></span>
<span id="cb90-511"><a href="#cb90-511"></a>    <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb90-512"><a href="#cb90-512"></a>    <span class="fu">mutate</span>(</span>
<span id="cb90-513"><a href="#cb90-513"></a>        <span class="fu">across</span>(<span class="fu">everything</span>(), as.numeric),</span>
<span id="cb90-514"><a href="#cb90-514"></a>        <span class="at">Nj =</span> Sj <span class="sc">+</span> Ij <span class="sc">+</span> Rj,</span>
<span id="cb90-515"><a href="#cb90-515"></a>        <span class="at">Na =</span> Sa <span class="sc">+</span> Ia <span class="sc">+</span> Ra,</span>
<span id="cb90-516"><a href="#cb90-516"></a>        <span class="co"># Calculate total population as need for proportional area plots</span></span>
<span id="cb90-517"><a href="#cb90-517"></a>        <span class="at">N =</span> Nj <span class="sc">+</span> Na</span>
<span id="cb90-518"><a href="#cb90-518"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb90-519"><a href="#cb90-519"></a>    <span class="fu">pivot_longer</span>(</span>
<span id="cb90-520"><a href="#cb90-520"></a>        <span class="at">cols =</span> <span class="sc">-</span><span class="fu">c</span>(time, N),</span>
<span id="cb90-521"><a href="#cb90-521"></a>        <span class="at">names_to =</span> <span class="fu">c</span>(<span class="st">"state"</span>, <span class="st">"group"</span>),</span>
<span id="cb90-522"><a href="#cb90-522"></a>        <span class="at">names_sep =</span> <span class="dv">1</span>,</span>
<span id="cb90-523"><a href="#cb90-523"></a>        <span class="at">values_to =</span> <span class="st">"value"</span></span>
<span id="cb90-524"><a href="#cb90-524"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb90-525"><a href="#cb90-525"></a>    <span class="fu">mutate</span>(</span>
<span id="cb90-526"><a href="#cb90-526"></a>        <span class="at">state =</span> <span class="fu">factor</span>(state, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"S"</span>, <span class="st">"I"</span>, <span class="st">"R"</span>, <span class="st">"N"</span>)),</span>
<span id="cb90-527"><a href="#cb90-527"></a>        <span class="at">group =</span> <span class="fu">paste</span>(<span class="st">"Group"</span>, <span class="fu">str_to_upper</span>(group))</span>
<span id="cb90-528"><a href="#cb90-528"></a>    )</span>
<span id="cb90-529"><a href="#cb90-529"></a><span class="in">```</span></span>
<span id="cb90-530"><a href="#cb90-530"></a></span>
<span id="cb90-531"><a href="#cb90-531"></a><span class="kw">&lt;div</span> <span class="er">class</span><span class="ot">=</span><span class="st">"exercise"</span><span class="kw">&gt;</span></span>
<span id="cb90-532"><a href="#cb90-532"></a></span>
<span id="cb90-533"><a href="#cb90-533"></a><span class="fu">### Exercise 1: Use this code to plot the number of susceptible, infected, and recovered individuals over time</span></span>
<span id="cb90-534"><a href="#cb90-534"></a></span>
<span id="cb90-535"><a href="#cb90-535"></a><span class="kw">&lt;/div&gt;</span></span>
<span id="cb90-536"><a href="#cb90-536"></a></span>
<span id="cb90-539"><a href="#cb90-539"></a><span class="in">```{r}</span></span>
<span id="cb90-540"><a href="#cb90-540"></a><span class="co">#| column: body</span></span>
<span id="cb90-541"><a href="#cb90-541"></a><span class="co">#| out-width: 100%</span></span>
<span id="cb90-542"><a href="#cb90-542"></a><span class="fu">ggplot</span>(demog_df, <span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> value, <span class="at">color =</span> state)) <span class="sc">+</span></span>
<span id="cb90-543"><a href="#cb90-543"></a>    <span class="fu">geom_line</span>(<span class="at">linewidth =</span> <span class="fl">1.5</span>) <span class="sc">+</span></span>
<span id="cb90-544"><a href="#cb90-544"></a>    <span class="fu">facet_wrap</span>(</span>
<span id="cb90-545"><a href="#cb90-545"></a>        <span class="sc">~</span>group, <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">scales =</span> <span class="st">"free_y"</span>,</span>
<span id="cb90-546"><a href="#cb90-546"></a>        <span class="at">labeller =</span> <span class="fu">as_labeller</span>(<span class="fu">c</span>(</span>
<span id="cb90-547"><a href="#cb90-547"></a>            <span class="st">`</span><span class="at">Group A</span><span class="st">`</span> <span class="ot">=</span> <span class="st">"Adults"</span>,</span>
<span id="cb90-548"><a href="#cb90-548"></a>            <span class="st">`</span><span class="at">Group J</span><span class="st">`</span> <span class="ot">=</span> <span class="st">"Juveniles"</span></span>
<span id="cb90-549"><a href="#cb90-549"></a>        ))</span>
<span id="cb90-550"><a href="#cb90-550"></a>    ) <span class="sc">+</span></span>
<span id="cb90-551"><a href="#cb90-551"></a>    <span class="fu">scale_color_manual</span>(</span>
<span id="cb90-552"><a href="#cb90-552"></a>        <span class="at">values =</span> SIRcolors,</span>
<span id="cb90-553"><a href="#cb90-553"></a>        <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"Susceptible"</span>, <span class="st">"Infected"</span>, <span class="st">"Recovered"</span>, <span class="st">"Total"</span>)</span>
<span id="cb90-554"><a href="#cb90-554"></a>    ) <span class="sc">+</span></span>
<span id="cb90-555"><a href="#cb90-555"></a>    <span class="fu">labs</span>(</span>
<span id="cb90-556"><a href="#cb90-556"></a>        <span class="at">x =</span> <span class="st">"Time"</span>,</span>
<span id="cb90-557"><a href="#cb90-557"></a>        <span class="at">y =</span> <span class="st">"Number of individuals"</span>,</span>
<span id="cb90-558"><a href="#cb90-558"></a>        <span class="at">color =</span> <span class="st">"State"</span></span>
<span id="cb90-559"><a href="#cb90-559"></a>    ) <span class="sc">+</span></span>
<span id="cb90-560"><a href="#cb90-560"></a>    <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span>
<span id="cb90-561"><a href="#cb90-561"></a><span class="in">```</span></span>
<span id="cb90-562"><a href="#cb90-562"></a></span>
<span id="cb90-563"><a href="#cb90-563"></a>Note that now that births are replenishing susceptibles, infection persists. The results of the above are plotted here:</span>
<span id="cb90-564"><a href="#cb90-564"></a></span>
<span id="cb90-565"><a href="#cb90-565"></a>Now let's plot the proportion of individuals in each state for the two groups.</span>
<span id="cb90-566"><a href="#cb90-566"></a></span>
<span id="cb90-569"><a href="#cb90-569"></a><span class="in">```{r}</span></span>
<span id="cb90-570"><a href="#cb90-570"></a><span class="co"># Calculate the proportions in each state and group at each time point</span></span>
<span id="cb90-571"><a href="#cb90-571"></a>demog_df_props <span class="ot">&lt;-</span> demog_df <span class="sc">%&gt;%</span></span>
<span id="cb90-572"><a href="#cb90-572"></a>    <span class="fu">filter</span>(state <span class="sc">!=</span> <span class="st">"N"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb90-573"><a href="#cb90-573"></a>    <span class="fu">mutate</span>(</span>
<span id="cb90-574"><a href="#cb90-574"></a>        <span class="at">state_group =</span> <span class="fu">paste0</span>(state, <span class="fu">str_extract_all</span>(group, <span class="st">"[^Group ]"</span>)),</span>
<span id="cb90-575"><a href="#cb90-575"></a>        <span class="at">state_group =</span> <span class="fu">factor</span>(</span>
<span id="cb90-576"><a href="#cb90-576"></a>          state_group,</span>
<span id="cb90-577"><a href="#cb90-577"></a>          <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"RJ"</span>, <span class="st">"RA"</span>, <span class="st">"IJ"</span>, <span class="st">"IA"</span>, <span class="st">"SJ"</span>, <span class="st">"SA"</span>)</span>
<span id="cb90-578"><a href="#cb90-578"></a>        )</span>
<span id="cb90-579"><a href="#cb90-579"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb90-580"><a href="#cb90-580"></a>    <span class="fu">group_by</span>(time, state_group) <span class="sc">%&gt;%</span></span>
<span id="cb90-581"><a href="#cb90-581"></a>    <span class="fu">mutate</span>(</span>
<span id="cb90-582"><a href="#cb90-582"></a>        <span class="co"># Calculate the proportion of the total population, not the group pop</span></span>
<span id="cb90-583"><a href="#cb90-583"></a>        <span class="at">prop =</span> value <span class="sc">/</span> N</span>
<span id="cb90-584"><a href="#cb90-584"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb90-585"><a href="#cb90-585"></a>    <span class="fu">ungroup</span>()</span>
<span id="cb90-586"><a href="#cb90-586"></a><span class="in">```</span></span>
<span id="cb90-587"><a href="#cb90-587"></a></span>
<span id="cb90-590"><a href="#cb90-590"></a><span class="in">```{r}</span></span>
<span id="cb90-591"><a href="#cb90-591"></a><span class="co">#| column: body</span></span>
<span id="cb90-592"><a href="#cb90-592"></a><span class="co">#| out-width: 100%</span></span>
<span id="cb90-593"><a href="#cb90-593"></a><span class="fu">ggplot</span>(demog_df_props, <span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> prop, <span class="at">fill =</span> state_group)) <span class="sc">+</span></span>
<span id="cb90-594"><a href="#cb90-594"></a>    <span class="fu">geom_area</span>() <span class="sc">+</span></span>
<span id="cb90-595"><a href="#cb90-595"></a>    <span class="fu">scale_fill_manual</span>(</span>
<span id="cb90-596"><a href="#cb90-596"></a>        <span class="at">values =</span> <span class="fu">c</span>(Scolors, Icolors, Rcolors),</span>
<span id="cb90-597"><a href="#cb90-597"></a>        <span class="at">limits =</span> <span class="fu">c</span>(<span class="st">"SJ"</span>, <span class="st">"SA"</span>, <span class="st">"IJ"</span>, <span class="st">"IA"</span>, <span class="st">"RJ"</span>, <span class="st">"RA"</span>)</span>
<span id="cb90-598"><a href="#cb90-598"></a>    ) <span class="sc">+</span></span>
<span id="cb90-599"><a href="#cb90-599"></a>    <span class="fu">labs</span>(</span>
<span id="cb90-600"><a href="#cb90-600"></a>        <span class="at">x =</span> <span class="st">"Time"</span>,</span>
<span id="cb90-601"><a href="#cb90-601"></a>        <span class="at">y =</span> <span class="st">"Proportion of individuals"</span>,</span>
<span id="cb90-602"><a href="#cb90-602"></a>        <span class="at">fill =</span> <span class="st">"State"</span></span>
<span id="cb90-603"><a href="#cb90-603"></a>    ) <span class="sc">+</span></span>
<span id="cb90-604"><a href="#cb90-604"></a>    <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span>
<span id="cb90-605"><a href="#cb90-605"></a><span class="in">```</span></span>
<span id="cb90-606"><a href="#cb90-606"></a></span>
<span id="cb90-607"><a href="#cb90-607"></a>Now let's plot the equilibrium seroprevalence for each age group.</span>
<span id="cb90-608"><a href="#cb90-608"></a></span>
<span id="cb90-611"><a href="#cb90-611"></a><span class="in">```{r}</span></span>
<span id="cb90-612"><a href="#cb90-612"></a><span class="co"># Select the last row (time point) of the data frame</span></span>
<span id="cb90-613"><a href="#cb90-613"></a>demog_equil_seroprev <span class="ot">&lt;-</span> <span class="fu">tail</span>(demog_df) <span class="sc">%&gt;%</span></span>
<span id="cb90-614"><a href="#cb90-614"></a>    <span class="fu">mutate</span>(</span>
<span id="cb90-615"><a href="#cb90-615"></a>        <span class="co"># Calculate the proportion of individuals in each state and age group</span></span>
<span id="cb90-616"><a href="#cb90-616"></a>        <span class="at">prop =</span> value <span class="sc">/</span> <span class="fu">sum</span>(value),</span>
<span id="cb90-617"><a href="#cb90-617"></a>        <span class="co"># Relabel groups for plots</span></span>
<span id="cb90-618"><a href="#cb90-618"></a>        <span class="at">group =</span> <span class="fu">case_when</span>(group <span class="sc">==</span> <span class="st">"Group J"</span> <span class="sc">~</span> <span class="st">"Juveniles"</span>, <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"Adults"</span>),</span>
<span id="cb90-619"><a href="#cb90-619"></a>        <span class="at">group =</span> <span class="fu">factor</span>(group, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Juveniles"</span>, <span class="st">"Adults"</span>)),</span>
<span id="cb90-620"><a href="#cb90-620"></a>        <span class="at">.by =</span> group</span>
<span id="cb90-621"><a href="#cb90-621"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb90-622"><a href="#cb90-622"></a>    <span class="fu">filter</span>(state <span class="sc">==</span> <span class="st">"R"</span>)</span>
<span id="cb90-623"><a href="#cb90-623"></a><span class="in">```</span></span>
<span id="cb90-624"><a href="#cb90-624"></a></span>
<span id="cb90-627"><a href="#cb90-627"></a><span class="in">```{r}</span></span>
<span id="cb90-628"><a href="#cb90-628"></a><span class="co">#| column: body</span></span>
<span id="cb90-629"><a href="#cb90-629"></a><span class="co">#| out-width: 100%</span></span>
<span id="cb90-630"><a href="#cb90-630"></a><span class="co"># Create vector of colors to distinguish between age groups</span></span>
<span id="cb90-631"><a href="#cb90-631"></a>age_group_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"#2980B9"</span>, <span class="st">"#154360"</span>)</span>
<span id="cb90-632"><a href="#cb90-632"></a></span>
<span id="cb90-633"><a href="#cb90-633"></a><span class="fu">ggplot</span>(demog_equil_seroprev, <span class="fu">aes</span>(<span class="at">x =</span> group, <span class="at">y =</span> prop, <span class="at">fill =</span> group)) <span class="sc">+</span></span>
<span id="cb90-634"><a href="#cb90-634"></a>    <span class="fu">geom_col</span>(<span class="at">position =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb90-635"><a href="#cb90-635"></a>    <span class="fu">scale_fill_manual</span>(</span>
<span id="cb90-636"><a href="#cb90-636"></a>        <span class="at">values =</span> age_group_colors</span>
<span id="cb90-637"><a href="#cb90-637"></a>    ) <span class="sc">+</span></span>
<span id="cb90-638"><a href="#cb90-638"></a>    <span class="fu">labs</span>(</span>
<span id="cb90-639"><a href="#cb90-639"></a>        <span class="at">x =</span> <span class="st">"Age group"</span>,</span>
<span id="cb90-640"><a href="#cb90-640"></a>        <span class="at">y =</span> <span class="st">"Equilibrium seroprevalence"</span>,</span>
<span id="cb90-641"><a href="#cb90-641"></a>        <span class="at">fill =</span> <span class="st">"Age group"</span></span>
<span id="cb90-642"><a href="#cb90-642"></a>    ) <span class="sc">+</span></span>
<span id="cb90-643"><a href="#cb90-643"></a>    <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb90-644"><a href="#cb90-644"></a><span class="in">```</span></span>
<span id="cb90-645"><a href="#cb90-645"></a></span>
<span id="cb90-646"><a href="#cb90-646"></a>One thing we are often interested in is the $R_0$ of a system.</span>
<span id="cb90-647"><a href="#cb90-647"></a>The details are beyond the scope of this workshop and are not required to complete the exercises in this worksheet, but we have outlined them in @sec-simple-ngm, particularly in @eq-simple-ngm, at the end of this page.</span>
<span id="cb90-648"><a href="#cb90-648"></a></span>
<span id="cb90-651"><a href="#cb90-651"></a><span class="in">```{r}</span></span>
<span id="cb90-652"><a href="#cb90-652"></a><span class="co">#| echo: false</span></span>
<span id="cb90-653"><a href="#cb90-653"></a><span class="co"># Calculate the rate of aging out of each age group</span></span>
<span id="cb90-654"><a href="#cb90-654"></a>alpha <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> demog_params[<span class="st">"age_band_j"</span>]</span>
<span id="cb90-655"><a href="#cb90-655"></a>mu <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> demog_params[<span class="st">"age_band_a"</span>]</span>
<span id="cb90-656"><a href="#cb90-656"></a></span>
<span id="cb90-657"><a href="#cb90-657"></a><span class="co"># Calculate the dfe pop sizes</span></span>
<span id="cb90-658"><a href="#cb90-658"></a>n <span class="ot">&lt;-</span> demog_params[<span class="st">"births"</span>] <span class="sc">/</span> <span class="fu">c</span>(alpha, mu)</span>
<span id="cb90-659"><a href="#cb90-659"></a></span>
<span id="cb90-660"><a href="#cb90-660"></a><span class="co"># Create the transmission matrix</span></span>
<span id="cb90-661"><a href="#cb90-661"></a>beta_demog <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(</span>
<span id="cb90-662"><a href="#cb90-662"></a>    demog_params[<span class="st">"beta_within"</span>],</span>
<span id="cb90-663"><a href="#cb90-663"></a>    demog_params[<span class="st">"beta_between"</span>],</span>
<span id="cb90-664"><a href="#cb90-664"></a>    demog_params[<span class="st">"beta_between"</span>],</span>
<span id="cb90-665"><a href="#cb90-665"></a>    demog_params[<span class="st">"beta_within"</span>]</span>
<span id="cb90-666"><a href="#cb90-666"></a>    ),</span>
<span id="cb90-667"><a href="#cb90-667"></a>    <span class="at">nrow =</span> <span class="dv">2</span>,</span>
<span id="cb90-668"><a href="#cb90-668"></a>    <span class="at">ncol =</span> <span class="dv">2</span></span>
<span id="cb90-669"><a href="#cb90-669"></a>)</span>
<span id="cb90-670"><a href="#cb90-670"></a></span>
<span id="cb90-671"><a href="#cb90-671"></a><span class="co"># Calculate the next generation matrix</span></span>
<span id="cb90-672"><a href="#cb90-672"></a>ba_ngm <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb90-673"><a href="#cb90-673"></a>    <span class="fu">c</span>(</span>
<span id="cb90-674"><a href="#cb90-674"></a>        n[<span class="dv">1</span>] <span class="sc">*</span> (beta_demog[<span class="dv">1</span>, <span class="dv">1</span>] <span class="sc">/</span> (demog_params[<span class="st">"recovery"</span>] <span class="sc">+</span> alpha)) <span class="sc">+</span></span>
<span id="cb90-675"><a href="#cb90-675"></a>            alpha <span class="sc">/</span> (demog_params[<span class="st">"recovery"</span>] <span class="sc">+</span> mu) <span class="sc">*</span></span>
<span id="cb90-676"><a href="#cb90-676"></a>            n[<span class="dv">1</span>] <span class="sc">*</span> beta_demog[<span class="dv">1</span>, <span class="dv">2</span>] <span class="sc">/</span> (demog_params[<span class="st">"recovery"</span>] <span class="sc">+</span> mu),</span>
<span id="cb90-677"><a href="#cb90-677"></a>        </span>
<span id="cb90-678"><a href="#cb90-678"></a>        n[<span class="dv">2</span>] <span class="sc">*</span> beta_demog[<span class="dv">2</span>, <span class="dv">1</span>] <span class="sc">/</span> (demog_params[<span class="st">"recovery"</span>] <span class="sc">+</span> alpha) <span class="sc">+</span></span>
<span id="cb90-679"><a href="#cb90-679"></a>            alpha <span class="sc">/</span> (demog_params[<span class="st">"recovery"</span>] <span class="sc">+</span> mu) <span class="sc">*</span></span>
<span id="cb90-680"><a href="#cb90-680"></a>            n[<span class="dv">2</span>] <span class="sc">*</span> (beta_demog[<span class="dv">2</span>, <span class="dv">2</span>] <span class="sc">/</span> (demog_params[<span class="st">"recovery"</span>] <span class="sc">+</span> mu)),</span>
<span id="cb90-681"><a href="#cb90-681"></a></span>
<span id="cb90-682"><a href="#cb90-682"></a>        n[<span class="dv">1</span>] <span class="sc">*</span> beta_demog[<span class="dv">1</span>, <span class="dv">2</span>] <span class="sc">/</span> (demog_params[<span class="st">"recovery"</span>] <span class="sc">+</span> mu),</span>
<span id="cb90-683"><a href="#cb90-683"></a></span>
<span id="cb90-684"><a href="#cb90-684"></a>        n[<span class="dv">2</span>] <span class="sc">*</span> beta_demog[<span class="dv">2</span>, <span class="dv">2</span>] <span class="sc">/</span> (demog_params[<span class="st">"recovery"</span>] <span class="sc">+</span> mu)</span>
<span id="cb90-685"><a href="#cb90-685"></a>    ),</span>
<span id="cb90-686"><a href="#cb90-686"></a>    <span class="at">nrow =</span> <span class="dv">2</span>,</span>
<span id="cb90-687"><a href="#cb90-687"></a>    <span class="at">ncol =</span> <span class="dv">2</span></span>
<span id="cb90-688"><a href="#cb90-688"></a>)</span>
<span id="cb90-689"><a href="#cb90-689"></a></span>
<span id="cb90-690"><a href="#cb90-690"></a><span class="co"># Calculate R0 by taking the **spectral trace** i.e., the largest non-negative Eigenvalue of the NGM</span></span>
<span id="cb90-691"><a href="#cb90-691"></a>ba_ngm <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">max</span>(<span class="fu">Re</span>(<span class="fu">eigen</span>(ba_ngm, <span class="at">only.values =</span> <span class="cn">TRUE</span>)<span class="sc">$</span>values)), <span class="at">digits =</span> <span class="dv">2</span>)</span>
<span id="cb90-692"><a href="#cb90-692"></a><span class="in">```</span></span>
<span id="cb90-693"><a href="#cb90-693"></a></span>
<span id="cb90-694"><a href="#cb90-694"></a>In our system, $R_0 =$ <span class="in">`r ba_ngm`</span>.</span>
<span id="cb90-695"><a href="#cb90-695"></a></span>
<span id="cb90-696"><a href="#cb90-696"></a><span class="fu">## Getting more realistic: adding more age classes</span></span>
<span id="cb90-697"><a href="#cb90-697"></a></span>
<span id="cb90-698"><a href="#cb90-698"></a>In the models above, the aging process follows an exponential distribution, which means that whether an individual is 1\~year old or 10 years old, the chance of them becoming an adult is the same!</span>
<span id="cb90-699"><a href="#cb90-699"></a>To improve on this, we can assume that the time a juvenile must wait before becoming an adult follows a gamma distribution.</span>
<span id="cb90-700"><a href="#cb90-700"></a>This is equivalent to saying that the waiting time is a sum of some number of exponential distributions.</span>
<span id="cb90-701"><a href="#cb90-701"></a>This suggests that we can achieve such a distribution by adding age classes to the model, so that becoming an adult means passing through some number of stages.</span>
<span id="cb90-702"><a href="#cb90-702"></a>We'll use 30 age classes, and since they don't have to be of equal duration, we'll assume that they're not.</span>
<span id="cb90-703"><a href="#cb90-703"></a>Specifically, we'll have 20 1-yr age classes to take us up to adulthood and break adults into 10 age classes of 5\~yr duration each. The last age class covers age 66-80.</span>
<span id="cb90-704"><a href="#cb90-704"></a></span>
<span id="cb90-705"><a href="#cb90-705"></a>Now, when we had just two age classes, we could write out each of the equations easily enough, but now that we're going to have 30, we'll need to be more systematic.</span>
<span id="cb90-706"><a href="#cb90-706"></a>In particular, we'll need to think of $\beta$ as a matrix of transmission rates.</span>
<span id="cb90-707"><a href="#cb90-707"></a>Let's see how to define such a matrix in <span class="in">`R`</span>.</span>
<span id="cb90-708"><a href="#cb90-708"></a>So that we don't change too many things all at once, let's keep the same contact structure as in the juvenile-adult model.</span>
<span id="cb90-709"><a href="#cb90-709"></a></span>
<span id="cb90-712"><a href="#cb90-712"></a><span class="in">```{r}</span></span>
<span id="cb90-713"><a href="#cb90-713"></a><span class="co"># Set up the parameters for model that incorporates a more realistic age matrix</span></span>
<span id="cb90-714"><a href="#cb90-714"></a>ages_params <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb90-715"><a href="#cb90-715"></a>    <span class="at">beta_j =</span> <span class="fl">0.02</span>,</span>
<span id="cb90-716"><a href="#cb90-716"></a>    <span class="at">beta_a =</span> <span class="fl">0.01</span>,</span>
<span id="cb90-717"><a href="#cb90-717"></a>    <span class="at">beta_aj =</span> <span class="fl">0.01</span> <span class="sc">/</span> <span class="dv">2</span>,</span>
<span id="cb90-718"><a href="#cb90-718"></a>    <span class="at">recovery =</span> <span class="dv">10</span>,</span>
<span id="cb90-719"><a href="#cb90-719"></a>    <span class="at">births =</span> <span class="dv">100</span></span>
<span id="cb90-720"><a href="#cb90-720"></a>)</span>
<span id="cb90-721"><a href="#cb90-721"></a></span>
<span id="cb90-722"><a href="#cb90-722"></a><span class="co"># Create a vector of ages</span></span>
<span id="cb90-723"><a href="#cb90-723"></a>ages <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">seq</span>(<span class="dv">1</span>, <span class="dv">20</span>, <span class="at">by =</span> <span class="dv">1</span>), <span class="fu">seq</span>(<span class="dv">25</span>, <span class="dv">65</span>, <span class="at">by =</span> <span class="dv">5</span>), <span class="dv">80</span>)</span>
<span id="cb90-724"><a href="#cb90-724"></a></span>
<span id="cb90-725"><a href="#cb90-725"></a><span class="co"># Calculate the widths of the age bands</span></span>
<span id="cb90-726"><a href="#cb90-726"></a>da_ages <span class="ot">&lt;-</span> <span class="fu">diff</span>(<span class="fu">c</span>(<span class="dv">0</span>, ages))                  <span class="co"># widths of age classes</span></span>
<span id="cb90-727"><a href="#cb90-727"></a></span>
<span id="cb90-728"><a href="#cb90-728"></a><span class="co"># set up a matrix of contact rates between classes -- more contact within juveniles and adults than between</span></span>
<span id="cb90-729"><a href="#cb90-729"></a>ages_beta_mat <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">nrow =</span> <span class="dv">30</span>, <span class="at">ncol =</span> <span class="dv">30</span>)</span>
<span id="cb90-730"><a href="#cb90-730"></a></span>
<span id="cb90-731"><a href="#cb90-731"></a><span class="co"># transmission rate for juveniles</span></span>
<span id="cb90-732"><a href="#cb90-732"></a>ages_beta_mat[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>] <span class="ot">&lt;-</span> ages_params[<span class="st">"beta_j"</span>]</span>
<span id="cb90-733"><a href="#cb90-733"></a></span>
<span id="cb90-734"><a href="#cb90-734"></a><span class="co"># transmission rate for adults</span></span>
<span id="cb90-735"><a href="#cb90-735"></a>ages_beta_mat[<span class="dv">21</span><span class="sc">:</span><span class="dv">30</span>, <span class="dv">21</span><span class="sc">:</span><span class="dv">30</span>] <span class="ot">&lt;-</span> ages_params[<span class="st">"beta_a"</span>]</span>
<span id="cb90-736"><a href="#cb90-736"></a></span>
<span id="cb90-737"><a href="#cb90-737"></a><span class="co"># lower transmission rate between juveniles and adults</span></span>
<span id="cb90-738"><a href="#cb90-738"></a>ages_beta_mat[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="dv">21</span><span class="sc">:</span><span class="dv">30</span>] <span class="ot">&lt;-</span> ages_params[<span class="st">"beta_aj"</span>]</span>
<span id="cb90-739"><a href="#cb90-739"></a></span>
<span id="cb90-740"><a href="#cb90-740"></a><span class="co"># lower transmission rate between juveniles and adults</span></span>
<span id="cb90-741"><a href="#cb90-741"></a>ages_beta_mat[<span class="dv">21</span><span class="sc">:</span><span class="dv">30</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>] <span class="ot">&lt;-</span> ages_params[<span class="st">"beta_aj"</span>]</span>
<span id="cb90-742"><a href="#cb90-742"></a><span class="in">```</span></span>
<span id="cb90-743"><a href="#cb90-743"></a></span>
<span id="cb90-744"><a href="#cb90-744"></a><span class="kw">&lt;a</span> <span class="er">id</span><span class="ot">=</span><span class="st">"fig-beta-mat"</span><span class="kw">&gt;&lt;/a&gt;</span></span>
<span id="cb90-745"><a href="#cb90-745"></a></span>
<span id="cb90-748"><a href="#cb90-748"></a><span class="in">```{r}</span></span>
<span id="cb90-749"><a href="#cb90-749"></a><span class="co">#| column: body</span></span>
<span id="cb90-750"><a href="#cb90-750"></a><span class="co">#| out-width: 100%</span></span>
<span id="cb90-751"><a href="#cb90-751"></a>ages_beta_mat <span class="sc">%&gt;%</span></span>
<span id="cb90-752"><a href="#cb90-752"></a>    <span class="co"># Turn into a data.frame so we can use ggplot()</span></span>
<span id="cb90-753"><a href="#cb90-753"></a>    <span class="fu">as.data.frame.table</span>() <span class="sc">%&gt;%</span></span>
<span id="cb90-754"><a href="#cb90-754"></a>    <span class="fu">mutate</span>(</span>
<span id="cb90-755"><a href="#cb90-755"></a>        <span class="at">age_contactor =</span> <span class="fu">rep</span>(ages, <span class="dv">30</span>),</span>
<span id="cb90-756"><a href="#cb90-756"></a>        <span class="co"># Repeat each age in ages vector 30 times, and place in a list, then unlist </span></span>
<span id="cb90-757"><a href="#cb90-757"></a>        <span class="co"># to be stacked into a single dataframe column</span></span>
<span id="cb90-758"><a href="#cb90-758"></a>        <span class="at">age_contactee =</span> <span class="fu">unlist</span>(purrr<span class="sc">::</span><span class="fu">map</span>(<span class="at">.x =</span> ages, <span class="sc">~</span><span class="fu">rep</span>(.x, <span class="dv">30</span>))),</span>
<span id="cb90-759"><a href="#cb90-759"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb90-760"><a href="#cb90-760"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> age_contactor, <span class="at">y =</span> age_contactee, <span class="at">z =</span> Freq)) <span class="sc">+</span></span>
<span id="cb90-761"><a href="#cb90-761"></a>    <span class="fu">geom_contour_filled</span>(<span class="at">bins =</span> <span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb90-762"><a href="#cb90-762"></a>    <span class="fu">scale_fill_brewer</span>(</span>
<span id="cb90-763"><a href="#cb90-763"></a>        <span class="at">palette =</span> <span class="st">"Reds"</span>,</span>
<span id="cb90-764"><a href="#cb90-764"></a>        <span class="co"># Don't drop unused bins as useful for ensuring extremities properly binned</span></span>
<span id="cb90-765"><a href="#cb90-765"></a>        <span class="at">drop =</span> <span class="cn">FALSE</span></span>
<span id="cb90-766"><a href="#cb90-766"></a>    ) <span class="sc">+</span></span>
<span id="cb90-767"><a href="#cb90-767"></a>    <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Age of Contactor"</span>, <span class="at">y =</span> <span class="st">"Age of Contactee"</span>)</span>
<span id="cb90-768"><a href="#cb90-768"></a><span class="in">```</span></span>
<span id="cb90-769"><a href="#cb90-769"></a></span>
<span id="cb90-770"><a href="#cb90-770"></a>::: {.callout-note collapse="false"}</span>
<span id="cb90-771"><a href="#cb90-771"></a>You could also use the <span class="in">`filled.contour()`</span> base-R function to plot the beta matrix without needing to do any dataframe modifications, but it doesn't look as nice ...</span>
<span id="cb90-772"><a href="#cb90-772"></a></span>
<span id="cb90-775"><a href="#cb90-775"></a><span class="in">```{r}</span></span>
<span id="cb90-776"><a href="#cb90-776"></a><span class="co">#| code-fold: true</span></span>
<span id="cb90-777"><a href="#cb90-777"></a><span class="fu">filled.contour</span>(</span>
<span id="cb90-778"><a href="#cb90-778"></a>    ages, ages, ages_beta_mat,</span>
<span id="cb90-779"><a href="#cb90-779"></a>    <span class="at">plot.title =</span> <span class="fu">title</span>(</span>
<span id="cb90-780"><a href="#cb90-780"></a>        <span class="at">xlab =</span> <span class="st">"Age of Contactor"</span>,</span>
<span id="cb90-781"><a href="#cb90-781"></a>        <span class="at">ylab =</span> <span class="st">"Age of Contactee"</span></span>
<span id="cb90-782"><a href="#cb90-782"></a>    )</span>
<span id="cb90-783"><a href="#cb90-783"></a>)</span>
<span id="cb90-784"><a href="#cb90-784"></a><span class="in">```</span></span>
<span id="cb90-785"><a href="#cb90-785"></a>:::</span>
<span id="cb90-786"><a href="#cb90-786"></a></span>
<span id="cb90-787"><a href="#cb90-787"></a>We'll assume that, at the time of introduction, all children are susceptible, as are adults over 45, but that individuals aged 20--45 have seen the pathogen before and are immune.</span>
<span id="cb90-788"><a href="#cb90-788"></a>The vector <span class="in">`yinit`</span> expresses these initial conditions.</span>
<span id="cb90-789"><a href="#cb90-789"></a></span>
<span id="cb90-792"><a href="#cb90-792"></a><span class="in">```{r}</span></span>
<span id="cb90-793"><a href="#cb90-793"></a><span class="co"># Create a long vector of initial states with only one</span></span>
<span id="cb90-794"><a href="#cb90-794"></a><span class="co"># initial infection in age 50</span></span>
<span id="cb90-795"><a href="#cb90-795"></a>demog_yinit_ages <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb90-796"><a href="#cb90-796"></a>    <span class="at">S =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">100</span>, <span class="dv">20</span>), <span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">5</span>), <span class="fu">rep</span>(<span class="dv">200</span>, <span class="dv">5</span>)),</span>
<span id="cb90-797"><a href="#cb90-797"></a>    <span class="at">I =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">25</span>), <span class="dv">1</span>, <span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">4</span>)),</span>
<span id="cb90-798"><a href="#cb90-798"></a>    <span class="at">R =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">20</span>), <span class="fu">rep</span>(<span class="dv">1000</span>, <span class="dv">5</span>), <span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">5</span>))</span>
<span id="cb90-799"><a href="#cb90-799"></a>)</span>
<span id="cb90-800"><a href="#cb90-800"></a><span class="in">```</span></span>
<span id="cb90-801"><a href="#cb90-801"></a></span>
<span id="cb90-802"><a href="#cb90-802"></a>Note that we're starting out with 1 infected individual in the 26th age class (age 50).</span>
<span id="cb90-803"><a href="#cb90-803"></a></span>
<span id="cb90-804"><a href="#cb90-804"></a>The codes that follow will be a bit easier to follow if we introduce some indexes that will allow us to pick out certain bits of the <span class="in">`yinit`</span> vector.</span>
<span id="cb90-805"><a href="#cb90-805"></a></span>
<span id="cb90-808"><a href="#cb90-808"></a><span class="in">```{r}</span></span>
<span id="cb90-809"><a href="#cb90-809"></a><span class="co"># Create vectors of indices relating to each state (ordered S1-80, I1-80, R1-80)</span></span>
<span id="cb90-810"><a href="#cb90-810"></a>sindex <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span></span>
<span id="cb90-811"><a href="#cb90-811"></a>iindex <span class="ot">&lt;-</span> <span class="dv">31</span><span class="sc">:</span><span class="dv">60</span></span>
<span id="cb90-812"><a href="#cb90-812"></a>rindex <span class="ot">&lt;-</span> <span class="dv">61</span><span class="sc">:</span><span class="dv">90</span></span>
<span id="cb90-813"><a href="#cb90-813"></a><span class="co"># Create vectors of indices relating to age group</span></span>
<span id="cb90-814"><a href="#cb90-814"></a>juvies <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span></span>
<span id="cb90-815"><a href="#cb90-815"></a>adults <span class="ot">&lt;-</span> <span class="dv">21</span><span class="sc">:</span><span class="dv">30</span></span>
<span id="cb90-816"><a href="#cb90-816"></a><span class="in">```</span></span>
<span id="cb90-817"><a href="#cb90-817"></a></span>
<span id="cb90-818"><a href="#cb90-818"></a>Now, to capture the aging process, it's convenient to define another matrix to hold the rates of movement between age classes.</span>
<span id="cb90-819"><a href="#cb90-819"></a>Generally, this matrix would look like this:</span>
<span id="cb90-820"><a href="#cb90-820"></a></span>
<span id="cb90-821"><a href="#cb90-821"></a>$$</span>
<span id="cb90-822"><a href="#cb90-822"></a>\begin{pmatrix}</span>
<span id="cb90-823"><a href="#cb90-823"></a>    -\alpha_1 &amp; 0 &amp; 0 &amp; \cdots &amp; 0<span class="sc">\\</span></span>
<span id="cb90-824"><a href="#cb90-824"></a>    \alpha_1 &amp; -\alpha_2 &amp; 0 &amp; \cdots &amp; 0<span class="sc">\\</span></span>
<span id="cb90-825"><a href="#cb90-825"></a>    0 &amp; \alpha_2 &amp; -\alpha_3 &amp; \cdots &amp; 0<span class="sc">\\</span></span>
<span id="cb90-826"><a href="#cb90-826"></a>    \vdots &amp;  &amp; \ddots &amp; \ddots &amp; \vdots <span class="sc">\\</span></span>
<span id="cb90-827"><a href="#cb90-827"></a>    0 &amp; \cdots &amp; &amp; \alpha_{29} &amp; -\alpha_{30}<span class="sc">\\</span></span>
<span id="cb90-828"><a href="#cb90-828"></a>\end{pmatrix}</span>
<span id="cb90-829"><a href="#cb90-829"></a>$${#eq-aging-mat}</span>
<span id="cb90-830"><a href="#cb90-830"></a></span>
<span id="cb90-831"><a href="#cb90-831"></a></span>
<span id="cb90-834"><a href="#cb90-834"></a><span class="in">```{r}</span></span>
<span id="cb90-835"><a href="#cb90-835"></a><span class="co"># Create a diagonal matrix that holds the rates of aging out of each age class</span></span>
<span id="cb90-836"><a href="#cb90-836"></a><span class="co"># The rows represent the age class you're in, the columns represent the age class</span></span>
<span id="cb90-837"><a href="#cb90-837"></a><span class="co"># you're moving to</span></span>
<span id="cb90-838"><a href="#cb90-838"></a>aging_mat <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="sc">-</span><span class="dv">1</span> <span class="sc">/</span> da_ages)</span>
<span id="cb90-839"><a href="#cb90-839"></a></span>
<span id="cb90-840"><a href="#cb90-840"></a><span class="co"># Fill in the rates of aging into each age class</span></span>
<span id="cb90-841"><a href="#cb90-841"></a>aging_mat[<span class="fu">row</span>(aging_mat) <span class="sc">-</span> <span class="fu">col</span>(aging_mat) <span class="sc">==</span> <span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> <span class="fu">head</span>(da_ages, <span class="sc">-</span><span class="dv">1</span>)</span>
<span id="cb90-842"><a href="#cb90-842"></a><span class="in">```</span></span>
<span id="cb90-843"><a href="#cb90-843"></a></span>
<span id="cb90-844"><a href="#cb90-844"></a>Have a look at the aging matrix, for example by doing:</span>
<span id="cb90-845"><a href="#cb90-845"></a></span>
<span id="cb90-848"><a href="#cb90-848"></a><span class="in">```{r}</span></span>
<span id="cb90-849"><a href="#cb90-849"></a><span class="co"># Move fast through the 1-year age classes - negatives are moves out, positives are moves in</span></span>
<span id="cb90-850"><a href="#cb90-850"></a><span class="co"># Cannot move between non-adjacent age classes</span></span>
<span id="cb90-851"><a href="#cb90-851"></a>aging_mat[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb90-852"><a href="#cb90-852"></a><span class="in">```</span></span>
<span id="cb90-853"><a href="#cb90-853"></a></span>
<span id="cb90-856"><a href="#cb90-856"></a><span class="in">```{r}</span></span>
<span id="cb90-857"><a href="#cb90-857"></a><span class="co"># Move slowly between the wider age classes</span></span>
<span id="cb90-858"><a href="#cb90-858"></a>aging_mat[<span class="dv">25</span><span class="sc">:</span><span class="dv">30</span>, <span class="dv">25</span><span class="sc">:</span><span class="dv">30</span>]</span>
<span id="cb90-859"><a href="#cb90-859"></a><span class="in">```</span></span>
<span id="cb90-860"><a href="#cb90-860"></a></span>
<span id="cb90-863"><a href="#cb90-863"></a><span class="in">```{r}</span></span>
<span id="cb90-864"><a href="#cb90-864"></a><span class="co">#| column: body</span></span>
<span id="cb90-865"><a href="#cb90-865"></a><span class="co">#| out-width: 100%</span></span>
<span id="cb90-866"><a href="#cb90-866"></a>aging_mat <span class="sc">%&gt;%</span></span>
<span id="cb90-867"><a href="#cb90-867"></a>    <span class="fu">as.data.frame.table</span>() <span class="sc">%&gt;%</span></span>
<span id="cb90-868"><a href="#cb90-868"></a>    <span class="fu">mutate</span>(</span>
<span id="cb90-869"><a href="#cb90-869"></a>        <span class="at">age_recipient =</span> <span class="fu">rep</span>(ages, <span class="dv">30</span>),</span>
<span id="cb90-870"><a href="#cb90-870"></a>        <span class="at">age_source =</span> <span class="fu">unlist</span>(purrr<span class="sc">::</span><span class="fu">map</span>(<span class="at">.x =</span> ages, <span class="sc">~</span><span class="fu">rep</span>(.x, <span class="dv">30</span>)))</span>
<span id="cb90-871"><a href="#cb90-871"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb90-872"><a href="#cb90-872"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> age_source, <span class="at">y =</span> age_recipient, <span class="at">z =</span> Freq)) <span class="sc">+</span></span>
<span id="cb90-873"><a href="#cb90-873"></a>    <span class="fu">geom_contour_filled</span>() <span class="sc">+</span></span>
<span id="cb90-874"><a href="#cb90-874"></a>    <span class="fu">scale_fill_brewer</span>(</span>
<span id="cb90-875"><a href="#cb90-875"></a>        <span class="at">palette =</span> <span class="st">"RdBu"</span>,</span>
<span id="cb90-876"><a href="#cb90-876"></a>        <span class="at">drop =</span> <span class="cn">FALSE</span></span>
<span id="cb90-877"><a href="#cb90-877"></a>    ) <span class="sc">+</span></span>
<span id="cb90-878"><a href="#cb90-878"></a>    <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Source Age Group"</span>, <span class="at">y =</span> <span class="st">"Recipient Age Group"</span>)</span>
<span id="cb90-879"><a href="#cb90-879"></a><span class="in">```</span></span>
<span id="cb90-880"><a href="#cb90-880"></a></span>
<span id="cb90-881"><a href="#cb90-881"></a><span class="kw">&lt;div</span> <span class="er">class</span><span class="ot">=</span><span class="st">"exercise"</span><span class="kw">&gt;</span></span>
<span id="cb90-882"><a href="#cb90-882"></a></span>
<span id="cb90-883"><a href="#cb90-883"></a><span class="fu">### Exercise 2: What can you say about its structure? How are the different age groups in contact with each other?</span></span>
<span id="cb90-884"><a href="#cb90-884"></a></span>
<span id="cb90-885"><a href="#cb90-885"></a><span class="kw">&lt;/div&gt;</span></span>
<span id="cb90-886"><a href="#cb90-886"></a></span>
<span id="cb90-887"><a href="#cb90-887"></a>Now we can put the pieces together to write a simulator for the age-structured SIR dynamics.</span>
<span id="cb90-888"><a href="#cb90-888"></a></span>
<span id="cb90-891"><a href="#cb90-891"></a><span class="in">```{r}</span></span>
<span id="cb90-892"><a href="#cb90-892"></a><span class="co"># Using a list instead of a vector to hold the parameters, as ages_beta_mat and aging </span></span>
<span id="cb90-893"><a href="#cb90-893"></a><span class="co"># are both matrices, so we want to keep them as matrices, rather than flattening</span></span>
<span id="cb90-894"><a href="#cb90-894"></a>multistage_params <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb90-895"><a href="#cb90-895"></a>    <span class="at">beta_mat =</span> ages_beta_mat,</span>
<span id="cb90-896"><a href="#cb90-896"></a>    <span class="at">recovery =</span> ages_params[<span class="st">"recovery"</span>],</span>
<span id="cb90-897"><a href="#cb90-897"></a>    <span class="at">births =</span> ages_params[<span class="st">"births"</span>],</span>
<span id="cb90-898"><a href="#cb90-898"></a>    <span class="at">aging_mat =</span> aging_mat</span>
<span id="cb90-899"><a href="#cb90-899"></a>)</span>
<span id="cb90-900"><a href="#cb90-900"></a></span>
<span id="cb90-901"><a href="#cb90-901"></a>multistage_model <span class="ot">&lt;-</span> <span class="cf">function</span> (t, x, p, ...) {</span>
<span id="cb90-902"><a href="#cb90-902"></a>    <span class="co"># Unpack all states from the vector using the relevant indices</span></span>
<span id="cb90-903"><a href="#cb90-903"></a>    s <span class="ot">&lt;-</span> x[sindex]</span>
<span id="cb90-904"><a href="#cb90-904"></a>    i <span class="ot">&lt;-</span> x[iindex]</span>
<span id="cb90-905"><a href="#cb90-905"></a>    r <span class="ot">&lt;-</span> x[rindex]</span>
<span id="cb90-906"><a href="#cb90-906"></a>    </span>
<span id="cb90-907"><a href="#cb90-907"></a>    <span class="co"># Unpack parameters</span></span>
<span id="cb90-908"><a href="#cb90-908"></a>    beta_mat <span class="ot">&lt;-</span> p[[<span class="st">"beta_mat"</span>]]</span>
<span id="cb90-909"><a href="#cb90-909"></a>    recovery <span class="ot">&lt;-</span> p[[<span class="st">"recovery"</span>]]</span>
<span id="cb90-910"><a href="#cb90-910"></a>    births <span class="ot">&lt;-</span> p[[<span class="st">"births"</span>]]</span>
<span id="cb90-911"><a href="#cb90-911"></a>    aging_mat <span class="ot">&lt;-</span> p[[<span class="st">"aging_mat"</span>]]</span>
<span id="cb90-912"><a href="#cb90-912"></a></span>
<span id="cb90-913"><a href="#cb90-913"></a>    <span class="co"># Calculate force of infection using matrix multiplication</span></span>
<span id="cb90-914"><a href="#cb90-914"></a>    lambda <span class="ot">&lt;-</span> beta_mat <span class="sc">%*%</span> i</span>
<span id="cb90-915"><a href="#cb90-915"></a>    </span>
<span id="cb90-916"><a href="#cb90-916"></a>    <span class="co"># Calculate the ODEs at every time step</span></span>
<span id="cb90-917"><a href="#cb90-917"></a>    <span class="co"># Note that R add element-wise for vectors i.e. lambda * s results</span></span>
<span id="cb90-918"><a href="#cb90-918"></a>    <span class="co"># in a vector length 30 (30 age groups), as does aging_mat %*% s,</span></span>
<span id="cb90-919"><a href="#cb90-919"></a>    <span class="co"># so v1[i] + v2[i] for i in 1:30</span></span>
<span id="cb90-920"><a href="#cb90-920"></a>    dsdt <span class="ot">&lt;-</span> <span class="sc">-</span>lambda <span class="sc">*</span> s <span class="sc">+</span> aging_mat <span class="sc">%*%</span> s</span>
<span id="cb90-921"><a href="#cb90-921"></a>    didt <span class="ot">&lt;-</span> lambda <span class="sc">*</span> s <span class="sc">+</span> aging_mat <span class="sc">%*%</span> i <span class="sc">-</span> recovery <span class="sc">*</span> i</span>
<span id="cb90-922"><a href="#cb90-922"></a>    drdt <span class="ot">&lt;-</span> aging_mat <span class="sc">%*%</span> r <span class="sc">+</span> recovery <span class="sc">*</span> i </span>
<span id="cb90-923"><a href="#cb90-923"></a>    <span class="co"># Add the birth rate to the first age group</span></span>
<span id="cb90-924"><a href="#cb90-924"></a>    dsdt[<span class="dv">1</span>] <span class="ot">&lt;-</span> dsdt[<span class="dv">1</span>] <span class="sc">+</span> births</span>
<span id="cb90-925"><a href="#cb90-925"></a>    </span>
<span id="cb90-926"><a href="#cb90-926"></a>    <span class="co"># Return the ODEs in a list</span></span>
<span id="cb90-927"><a href="#cb90-927"></a>    <span class="fu">list</span>(<span class="fu">c</span>(dsdt, didt, drdt))</span>
<span id="cb90-928"><a href="#cb90-928"></a></span>
<span id="cb90-929"><a href="#cb90-929"></a>}</span>
<span id="cb90-930"><a href="#cb90-930"></a><span class="in">```</span></span>
<span id="cb90-931"><a href="#cb90-931"></a></span>
<span id="cb90-932"><a href="#cb90-932"></a>We can plug this into <span class="in">`ode`</span> just as we did the simpler models to simulate an epidemic.</span>
<span id="cb90-933"><a href="#cb90-933"></a>We'll then plot the epidemic curve.</span>
<span id="cb90-934"><a href="#cb90-934"></a></span>
<span id="cb90-937"><a href="#cb90-937"></a><span class="in">```{r}</span></span>
<span id="cb90-938"><a href="#cb90-938"></a><span class="co"># Solve the model with a realistic age matrix</span></span>
<span id="cb90-939"><a href="#cb90-939"></a>multistage_sol <span class="ot">&lt;-</span> deSolve<span class="sc">::</span><span class="fu">ode</span>(</span>
<span id="cb90-940"><a href="#cb90-940"></a>    <span class="at">y =</span> demog_yinit_ages,</span>
<span id="cb90-941"><a href="#cb90-941"></a>    <span class="at">times =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">100</span>, <span class="at">by =</span> <span class="fl">0.1</span>),</span>
<span id="cb90-942"><a href="#cb90-942"></a>    <span class="at">func =</span> multistage_model,</span>
<span id="cb90-943"><a href="#cb90-943"></a>    <span class="at">parms =</span> multistage_params</span>
<span id="cb90-944"><a href="#cb90-944"></a>)</span>
<span id="cb90-945"><a href="#cb90-945"></a></span>
<span id="cb90-946"><a href="#cb90-946"></a><span class="co"># Extract all infected age groups at all time points into a new vector</span></span>
<span id="cb90-947"><a href="#cb90-947"></a>multistage_infecteds <span class="ot">&lt;-</span> multistage_sol[, <span class="dv">1</span> <span class="sc">+</span> iindex]</span>
<span id="cb90-948"><a href="#cb90-948"></a><span class="in">```</span></span>
<span id="cb90-949"><a href="#cb90-949"></a></span>
<span id="cb90-952"><a href="#cb90-952"></a><span class="in">```{r}</span></span>
<span id="cb90-953"><a href="#cb90-953"></a><span class="co"># Create a dataframe of the sum of infectious individuals in Juv/Adult age groups</span></span>
<span id="cb90-954"><a href="#cb90-954"></a><span class="co"># at each time point</span></span>
<span id="cb90-955"><a href="#cb90-955"></a>multistage_df <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb90-956"><a href="#cb90-956"></a>        <span class="co"># Get all times from model run</span></span>
<span id="cb90-957"><a href="#cb90-957"></a>        <span class="at">time =</span> multistage_sol[, <span class="dv">1</span>],</span>
<span id="cb90-958"><a href="#cb90-958"></a>        <span class="co"># At each timepoint, apply the sum function to all juvenile infected individuals</span></span>
<span id="cb90-959"><a href="#cb90-959"></a>        <span class="at">Juveniles =</span> <span class="fu">apply</span>(multistage_infecteds[, juvies], <span class="dv">1</span>, sum),</span>
<span id="cb90-960"><a href="#cb90-960"></a>        <span class="co"># At each timepoint, apply the sum function to all adult infected individuals</span></span>
<span id="cb90-961"><a href="#cb90-961"></a>        <span class="at">Adults =</span> <span class="fu">apply</span>(multistage_infecteds[, adults], <span class="dv">1</span>, sum)</span>
<span id="cb90-962"><a href="#cb90-962"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb90-963"><a href="#cb90-963"></a>    <span class="co"># Pivot to create a long dataframe that works with ggplot</span></span>
<span id="cb90-964"><a href="#cb90-964"></a>    <span class="fu">pivot_longer</span>(</span>
<span id="cb90-965"><a href="#cb90-965"></a>        <span class="at">cols =</span> <span class="fu">c</span>(Juveniles, Adults),</span>
<span id="cb90-966"><a href="#cb90-966"></a>        <span class="at">names_to =</span> <span class="st">"age_group"</span>,</span>
<span id="cb90-967"><a href="#cb90-967"></a>        <span class="at">values_to =</span> <span class="st">"infections"</span></span>
<span id="cb90-968"><a href="#cb90-968"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb90-969"><a href="#cb90-969"></a>    <span class="co"># Turn new pivoted variable into a factor to plot nicely</span></span>
<span id="cb90-970"><a href="#cb90-970"></a>    <span class="fu">mutate</span>(</span>
<span id="cb90-971"><a href="#cb90-971"></a>        <span class="at">age_group =</span> <span class="fu">factor</span>(age_group, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Juveniles"</span>, <span class="st">"Adults"</span>))</span>
<span id="cb90-972"><a href="#cb90-972"></a>    )</span>
<span id="cb90-973"><a href="#cb90-973"></a><span class="in">```</span></span>
<span id="cb90-974"><a href="#cb90-974"></a></span>
<span id="cb90-977"><a href="#cb90-977"></a><span class="in">```{r}</span></span>
<span id="cb90-978"><a href="#cb90-978"></a><span class="co">#| column: body</span></span>
<span id="cb90-979"><a href="#cb90-979"></a><span class="co">#| out-width: 100%</span></span>
<span id="cb90-980"><a href="#cb90-980"></a><span class="fu">ggplot</span>(multistage_df, <span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> infections, <span class="at">color =</span> age_group)) <span class="sc">+</span></span>
<span id="cb90-981"><a href="#cb90-981"></a>    <span class="fu">geom_line</span>(<span class="at">linewidth =</span> <span class="fl">1.5</span>) <span class="sc">+</span></span>
<span id="cb90-982"><a href="#cb90-982"></a>    <span class="fu">scale_color_manual</span>(</span>
<span id="cb90-983"><a href="#cb90-983"></a>        <span class="at">values =</span> age_group_colors</span>
<span id="cb90-984"><a href="#cb90-984"></a>    ) <span class="sc">+</span></span>
<span id="cb90-985"><a href="#cb90-985"></a>    <span class="fu">labs</span>(</span>
<span id="cb90-986"><a href="#cb90-986"></a>        <span class="at">x =</span> <span class="st">"Time"</span>,</span>
<span id="cb90-987"><a href="#cb90-987"></a>        <span class="at">y =</span> <span class="st">"Number of infections"</span>,</span>
<span id="cb90-988"><a href="#cb90-988"></a>        <span class="at">color =</span> <span class="st">"Age group"</span></span>
<span id="cb90-989"><a href="#cb90-989"></a>    )</span>
<span id="cb90-990"><a href="#cb90-990"></a><span class="in">```</span></span>
<span id="cb90-991"><a href="#cb90-991"></a></span>
<span id="cb90-992"><a href="#cb90-992"></a>Let's mimic a situation where we have cross-sectional seroprevalence data (e.g. measures of antibodies that tell you someone is in the R class).</span>
<span id="cb90-993"><a href="#cb90-993"></a>In using such data, we'd typically assume that the system was at equilibrium.</span>
<span id="cb90-994"><a href="#cb90-994"></a></span>
<span id="cb90-995"><a href="#cb90-995"></a><span class="kw">&lt;div</span> <span class="er">class</span><span class="ot">=</span><span class="st">"exercise"</span><span class="kw">&gt;</span></span>
<span id="cb90-996"><a href="#cb90-996"></a></span>
<span id="cb90-997"><a href="#cb90-997"></a><span class="fu">### Exercise 3: What does the equilibrium age-specific seroprevalence look like in this example? {#sec-ex-3}</span></span>
<span id="cb90-998"><a href="#cb90-998"></a></span>
<span id="cb90-999"><a href="#cb90-999"></a><span class="kw">&lt;/div&gt;</span></span>
<span id="cb90-1000"><a href="#cb90-1000"></a></span>
<span id="cb90-1001"><a href="#cb90-1001"></a>Use the code below to display the age-specific seroprevalence (i.e., the seroprevalence for each age group at equilibrium)</span>
<span id="cb90-1002"><a href="#cb90-1002"></a></span>
<span id="cb90-1005"><a href="#cb90-1005"></a><span class="in">```{r}</span></span>
<span id="cb90-1006"><a href="#cb90-1006"></a><span class="co"># Get the last values for all individuals. drop() removes the column name, [-1] removes the time value</span></span>
<span id="cb90-1007"><a href="#cb90-1007"></a>multistage_equil <span class="ot">&lt;-</span> <span class="fu">drop</span>(<span class="fu">tail</span>(multistage_sol, <span class="dv">1</span>))[<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb90-1008"><a href="#cb90-1008"></a><span class="co"># Calculate the equilibrium pop sizes of each age group</span></span>
<span id="cb90-1009"><a href="#cb90-1009"></a>multistage_equil_n <span class="ot">&lt;-</span> multistage_equil[sindex] <span class="sc">+</span></span>
<span id="cb90-1010"><a href="#cb90-1010"></a>    multistage_equil[iindex] <span class="sc">+</span></span>
<span id="cb90-1011"><a href="#cb90-1011"></a>    multistage_equil[rindex]</span>
<span id="cb90-1012"><a href="#cb90-1012"></a></span>
<span id="cb90-1013"><a href="#cb90-1013"></a><span class="co"># Calculate equilibrium seroprevalence for each age group</span></span>
<span id="cb90-1014"><a href="#cb90-1014"></a>multistage_equil_seroprev <span class="ot">&lt;-</span> multistage_equil[rindex] <span class="sc">/</span> multistage_equil_n</span>
<span id="cb90-1015"><a href="#cb90-1015"></a></span>
<span id="cb90-1016"><a href="#cb90-1016"></a><span class="co"># Create a dataframe to store equilibrium seroprev for plotting</span></span>
<span id="cb90-1017"><a href="#cb90-1017"></a>multistage_equil_seroprev_df <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb90-1018"><a href="#cb90-1018"></a>    <span class="at">age =</span> ages,</span>
<span id="cb90-1019"><a href="#cb90-1019"></a>    <span class="at">seroprev =</span> multistage_equil_seroprev,</span>
<span id="cb90-1020"><a href="#cb90-1020"></a>    <span class="at">width =</span> da_ages</span>
<span id="cb90-1021"><a href="#cb90-1021"></a>)</span>
<span id="cb90-1022"><a href="#cb90-1022"></a><span class="in">```</span></span>
<span id="cb90-1023"><a href="#cb90-1023"></a></span>
<span id="cb90-1026"><a href="#cb90-1026"></a><span class="in">```{r}</span></span>
<span id="cb90-1027"><a href="#cb90-1027"></a><span class="co">#| column: body</span></span>
<span id="cb90-1028"><a href="#cb90-1028"></a><span class="co">#| out-width: 100%</span></span>
<span id="cb90-1029"><a href="#cb90-1029"></a><span class="fu">ggplot</span>(multistage_equil_seroprev_df, <span class="fu">aes</span>(<span class="at">x =</span> age, <span class="at">y =</span> seroprev, <span class="at">fill =</span> age)) <span class="sc">+</span></span>
<span id="cb90-1030"><a href="#cb90-1030"></a>    <span class="co"># Set column width to width of age bands, and justify to start at lower bound</span></span>
<span id="cb90-1031"><a href="#cb90-1031"></a>    <span class="fu">geom_col</span>(<span class="at">width =</span> multistage_equil_seroprev_df<span class="sc">$</span>width, <span class="at">just =</span> <span class="fl">1.0</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb90-1032"><a href="#cb90-1032"></a>    <span class="fu">labs</span>(</span>
<span id="cb90-1033"><a href="#cb90-1033"></a>        <span class="at">x =</span> <span class="st">"Age"</span>,</span>
<span id="cb90-1034"><a href="#cb90-1034"></a>        <span class="at">y =</span> <span class="st">"Seroprevalence"</span></span>
<span id="cb90-1035"><a href="#cb90-1035"></a>    ) <span class="sc">+</span></span>
<span id="cb90-1036"><a href="#cb90-1036"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">80</span>, <span class="dv">10</span>)) <span class="sc">+</span></span>
<span id="cb90-1037"><a href="#cb90-1037"></a>    <span class="fu">scale_fill_continuous</span>(<span class="at">low =</span> age_group_colors[<span class="dv">1</span>], <span class="at">high =</span> age_group_colors[<span class="dv">2</span>])</span>
<span id="cb90-1038"><a href="#cb90-1038"></a><span class="in">```</span></span>
<span id="cb90-1039"><a href="#cb90-1039"></a></span>
<span id="cb90-1040"><a href="#cb90-1040"></a>::: {.callout-note title="QUESTION"}</span>
<span id="cb90-1041"><a href="#cb90-1041"></a>At what age does the seroprevalence reach 75%?</span>
<span id="cb90-1042"><a href="#cb90-1042"></a>:::</span>
<span id="cb90-1043"><a href="#cb90-1043"></a></span>
<span id="cb90-1044"><a href="#cb90-1044"></a>Let's also compute $R_0$.</span>
<span id="cb90-1045"><a href="#cb90-1045"></a>And because we've added a lot of age structure, with transitions between the age groups, we can't just copy and paste the previous Next Generation Matrix code (from @sec-simple-ngm).</span>
<span id="cb90-1046"><a href="#cb90-1046"></a>As before, the details of this computation are out of the workshop's scope, but they are outlined in @sec-age-structure-ngm.</span>
<span id="cb90-1047"><a href="#cb90-1047"></a>We have created a function to calculate R0 for an age-structured SIR and have added some comments, but read @sec-age-structure-ngm for the full details and reasoning.</span>
<span id="cb90-1048"><a href="#cb90-1048"></a></span>
<span id="cb90-1051"><a href="#cb90-1051"></a><span class="in">```{r}</span></span>
<span id="cb90-1052"><a href="#cb90-1052"></a><span class="co"># Calculate the stable disease-free age distribution. Could also simulate without any infections.</span></span>
<span id="cb90-1053"><a href="#cb90-1053"></a>multistage_stable_n <span class="ot">&lt;-</span> <span class="fu">solve</span>(</span>
<span id="cb90-1054"><a href="#cb90-1054"></a>    aging_mat,</span>
<span id="cb90-1055"><a href="#cb90-1055"></a>    <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span> <span class="sc">*</span> multistage_params[[<span class="st">"births"</span>]], <span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">29</span>))</span>
<span id="cb90-1056"><a href="#cb90-1056"></a>)</span>
<span id="cb90-1057"><a href="#cb90-1057"></a><span class="in">```</span></span>
<span id="cb90-1058"><a href="#cb90-1058"></a></span>
<span id="cb90-1059"><a href="#cb90-1059"></a><span class="kw">&lt;a</span> <span class="er">id</span><span class="ot">=</span><span class="st">"ngm-function"</span><span class="kw">&gt;&lt;/a&gt;</span></span>
<span id="cb90-1060"><a href="#cb90-1060"></a></span>
<span id="cb90-1063"><a href="#cb90-1063"></a><span class="in">```{r}</span></span>
<span id="cb90-1064"><a href="#cb90-1064"></a><span class="co">#' Calculate R0</span></span>
<span id="cb90-1065"><a href="#cb90-1065"></a><span class="co">#'</span></span>
<span id="cb90-1066"><a href="#cb90-1066"></a><span class="co">#' Calculate the R0 of an SIR model using the next generation matrix approach described in @heffernanPerspectivesBasicReproductive2005</span></span>
<span id="cb90-1067"><a href="#cb90-1067"></a><span class="co">#'</span></span>
<span id="cb90-1068"><a href="#cb90-1068"></a><span class="co">#' @param beta_mat A matrix of beta parameter values</span></span>
<span id="cb90-1069"><a href="#cb90-1069"></a><span class="co">#' @param stable_n_mat A matrix of the stable age distributions</span></span>
<span id="cb90-1070"><a href="#cb90-1070"></a><span class="co">#' @param aging_mat A matrix of the aging rates between age compartments</span></span>
<span id="cb90-1071"><a href="#cb90-1071"></a><span class="co">#' @param recovery_rate The recover rate parameter value (of type double)</span></span>
<span id="cb90-1072"><a href="#cb90-1072"></a><span class="co">#'</span></span>
<span id="cb90-1073"><a href="#cb90-1073"></a><span class="co">#' @return The R0 value as type double</span></span>
<span id="cb90-1074"><a href="#cb90-1074"></a><span class="co">#' @examples</span></span>
<span id="cb90-1075"><a href="#cb90-1075"></a><span class="co">#' calculate_R0(</span></span>
<span id="cb90-1076"><a href="#cb90-1076"></a><span class="co">#'    beta_mat = multistage_params[["beta_mat"]],</span></span>
<span id="cb90-1077"><a href="#cb90-1077"></a><span class="co">#'    stable_n_mat = multistage_stable_n,</span></span>
<span id="cb90-1078"><a href="#cb90-1078"></a><span class="co">#'    aging_mat = multistage_params[["aging_mat"]],</span></span>
<span id="cb90-1079"><a href="#cb90-1079"></a><span class="co">#'    recovery_rate = multistage_params[["recovery"]]</span></span>
<span id="cb90-1080"><a href="#cb90-1080"></a><span class="co">#')</span></span>
<span id="cb90-1081"><a href="#cb90-1081"></a>calculate_R0 <span class="ot">&lt;-</span> <span class="cf">function</span>(beta_mat, stable_n_mat, aging_mat, recovery_rate) {</span>
<span id="cb90-1082"><a href="#cb90-1082"></a>    <span class="co"># evaluate new inf jac pde at dfe</span></span>
<span id="cb90-1083"><a href="#cb90-1083"></a>    f_mat <span class="ot">&lt;-</span> beta_mat <span class="sc">*</span> stable_n_mat</span>
<span id="cb90-1084"><a href="#cb90-1084"></a></span>
<span id="cb90-1085"><a href="#cb90-1085"></a>    <span class="co"># set off-diag of non-inf transition jac pde to neg aging of prev age group</span></span>
<span id="cb90-1086"><a href="#cb90-1086"></a>    <span class="co"># (use aging matrix as already calculated in correct places)</span></span>
<span id="cb90-1087"><a href="#cb90-1087"></a>    v_mat <span class="ot">&lt;-</span> <span class="sc">-</span>aging_mat</span>
<span id="cb90-1088"><a href="#cb90-1088"></a>    <span class="co"># Update the diagonal of non-inf transition jac to add recovery rate</span></span>
<span id="cb90-1089"><a href="#cb90-1089"></a>    <span class="fu">diag</span>(v_mat) <span class="ot">&lt;-</span> <span class="fu">diag</span>(v_mat) <span class="sc">+</span> recovery_rate</span>
<span id="cb90-1090"><a href="#cb90-1090"></a></span>
<span id="cb90-1091"><a href="#cb90-1091"></a>    <span class="do">## Alternative method of calculating using age bands directly</span></span>
<span id="cb90-1092"><a href="#cb90-1092"></a>    <span class="co"># v_mat &lt;- diag(recovery_rate + 1 / da_ages)</span></span>
<span id="cb90-1093"><a href="#cb90-1093"></a>    <span class="co"># v_mat[row(v_mat) - col(v_mat) == 1] &lt;- - 1 / head(da_ages, -1)</span></span>
<span id="cb90-1094"><a href="#cb90-1094"></a>    </span>
<span id="cb90-1095"><a href="#cb90-1095"></a>    <span class="co"># spectral trace</span></span>
<span id="cb90-1096"><a href="#cb90-1096"></a>    R0 <span class="ot">&lt;-</span> <span class="fu">max</span>(<span class="fu">Re</span>(<span class="fu">eigen</span>(<span class="fu">solve</span>(v_mat, f_mat), <span class="at">only.values =</span> <span class="cn">TRUE</span>)<span class="sc">$</span>values))   </span>
<span id="cb90-1097"><a href="#cb90-1097"></a>    </span>
<span id="cb90-1098"><a href="#cb90-1098"></a>    <span class="fu">return</span>(R0)</span>
<span id="cb90-1099"><a href="#cb90-1099"></a>}</span>
<span id="cb90-1100"><a href="#cb90-1100"></a><span class="in">```</span></span>
<span id="cb90-1101"><a href="#cb90-1101"></a></span>
<span id="cb90-1104"><a href="#cb90-1104"></a><span class="in">```{r}</span></span>
<span id="cb90-1105"><a href="#cb90-1105"></a><span class="fu">calculate_R0</span>(</span>
<span id="cb90-1106"><a href="#cb90-1106"></a>    <span class="at">beta_mat =</span> multistage_params[[<span class="st">"beta_mat"</span>]],</span>
<span id="cb90-1107"><a href="#cb90-1107"></a>    <span class="at">stable_n_mat =</span> multistage_stable_n,</span>
<span id="cb90-1108"><a href="#cb90-1108"></a>    <span class="at">aging_mat =</span> multistage_params[[<span class="st">"aging_mat"</span>]],</span>
<span id="cb90-1109"><a href="#cb90-1109"></a>    <span class="at">recovery_rate =</span> multistage_params[[<span class="st">"recovery"</span>]]</span>
<span id="cb90-1110"><a href="#cb90-1110"></a>)</span>
<span id="cb90-1111"><a href="#cb90-1111"></a><span class="in">```</span></span>
<span id="cb90-1112"><a href="#cb90-1112"></a></span>
<span id="cb90-1113"><a href="#cb90-1113"></a><span class="kw">&lt;div</span> <span class="er">class</span><span class="ot">=</span><span class="st">"exercise"</span><span class="kw">&gt;</span></span>
<span id="cb90-1114"><a href="#cb90-1114"></a></span>
<span id="cb90-1115"><a href="#cb90-1115"></a><span class="fu">### Exercise 4: Updating the contact matrix</span></span>
<span id="cb90-1116"><a href="#cb90-1116"></a></span>
<span id="cb90-1117"><a href="#cb90-1117"></a><span class="kw">&lt;/div&gt;</span></span>
<span id="cb90-1118"><a href="#cb90-1118"></a></span>
<span id="cb90-1119"><a href="#cb90-1119"></a>::: {.callout-important}</span>
<span id="cb90-1120"><a href="#cb90-1120"></a>You will need to read and edit the following code carefully so that it runs with your updated parameters.</span>
<span id="cb90-1121"><a href="#cb90-1121"></a>We have highlighted the relevant lines in the code chunks, so hopefully you won't miss them, though make sure you do copy all the code!</span>
<span id="cb90-1122"><a href="#cb90-1122"></a>:::</span>
<span id="cb90-1123"><a href="#cb90-1123"></a></span>
<span id="cb90-1124"><a href="#cb90-1124"></a><span class="fu">#### Change the juvenile-juvenile contact rate to be 0.025.</span></span>
<span id="cb90-1125"><a href="#cb90-1125"></a></span>
<span id="cb90-1126"><a href="#cb90-1126"></a></span>
<span id="cb90-1129"><a href="#cb90-1129"></a><span class="in">```{r}</span></span>
<span id="cb90-1130"><a href="#cb90-1130"></a><span class="co">#| echo: false</span></span>
<span id="cb90-1131"><a href="#cb90-1131"></a>update_age_beta_mat <span class="ot">&lt;-</span> ages_beta_mat</span>
<span id="cb90-1132"><a href="#cb90-1132"></a>update_age_beta_mat[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>] <span class="ot">&lt;-</span> <span class="fl">0.025</span></span>
<span id="cb90-1133"><a href="#cb90-1133"></a></span>
<span id="cb90-1134"><a href="#cb90-1134"></a>update_age_params <span class="ot">&lt;-</span> multistage_params</span>
<span id="cb90-1135"><a href="#cb90-1135"></a>update_age_params[[<span class="st">"beta_mat"</span>]] <span class="ot">&lt;-</span> update_age_beta_mat</span>
<span id="cb90-1136"><a href="#cb90-1136"></a><span class="in">```</span></span>
<span id="cb90-1137"><a href="#cb90-1137"></a></span>
<span id="cb90-1140"><a href="#cb90-1140"></a><span class="in">```{r}</span></span>
<span id="cb90-1141"><a href="#cb90-1141"></a><span class="co">#| eval: false</span></span>
<span id="cb90-1142"><a href="#cb90-1142"></a><span class="co">#| source-line-numbers: "2"</span></span>
<span id="cb90-1143"><a href="#cb90-1143"></a>update_age_beta_mat <span class="ot">&lt;-</span> ages_beta_mat</span>
<span id="cb90-1144"><a href="#cb90-1144"></a>update_age_beta_mat[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>] <span class="ot">&lt;-</span> ?</span>
<span id="cb90-1145"><a href="#cb90-1145"></a></span>
<span id="cb90-1146"><a href="#cb90-1146"></a>update_age_params <span class="ot">&lt;-</span> multistage_params</span>
<span id="cb90-1147"><a href="#cb90-1147"></a>update_age_params[[<span class="st">"beta_mat"</span>]] <span class="ot">&lt;-</span> update_age_beta_mat</span>
<span id="cb90-1148"><a href="#cb90-1148"></a><span class="in">```</span></span>
<span id="cb90-1149"><a href="#cb90-1149"></a></span>
<span id="cb90-1150"><a href="#cb90-1150"></a><span class="fu">#### Simulate and plot the age-structured SIR dynamics under your assumptions and record how the age-specific seroprevalence has changed.</span></span>
<span id="cb90-1151"><a href="#cb90-1151"></a></span>
<span id="cb90-1154"><a href="#cb90-1154"></a><span class="in">```{r}</span></span>
<span id="cb90-1155"><a href="#cb90-1155"></a><span class="co">#| echo: false</span></span>
<span id="cb90-1156"><a href="#cb90-1156"></a>update_age_sol <span class="ot">&lt;-</span> deSolve<span class="sc">::</span><span class="fu">ode</span>(</span>
<span id="cb90-1157"><a href="#cb90-1157"></a>    <span class="at">y =</span> demog_yinit_ages,</span>
<span id="cb90-1158"><a href="#cb90-1158"></a>    <span class="at">times =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">400</span>, <span class="at">by =</span> <span class="fl">0.1</span>),</span>
<span id="cb90-1159"><a href="#cb90-1159"></a>    <span class="at">func =</span> multistage_model,</span>
<span id="cb90-1160"><a href="#cb90-1160"></a>    <span class="at">parms =</span> update_age_params</span>
<span id="cb90-1161"><a href="#cb90-1161"></a>)</span>
<span id="cb90-1162"><a href="#cb90-1162"></a></span>
<span id="cb90-1163"><a href="#cb90-1163"></a><span class="co"># Get the time series for each infectious age group</span></span>
<span id="cb90-1164"><a href="#cb90-1164"></a>update_age_infecteds <span class="ot">&lt;-</span> update_age_sol[, <span class="dv">1</span> <span class="sc">+</span> iindex]</span>
<span id="cb90-1165"><a href="#cb90-1165"></a></span>
<span id="cb90-1166"><a href="#cb90-1166"></a>update_age_equil <span class="ot">&lt;-</span> <span class="fu">drop</span>(<span class="fu">tail</span>(update_age_sol, <span class="dv">1</span>))[<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb90-1167"><a href="#cb90-1167"></a></span>
<span id="cb90-1168"><a href="#cb90-1168"></a>update_age_equil_n <span class="ot">&lt;-</span> update_age_equil[sindex] <span class="sc">+</span></span>
<span id="cb90-1169"><a href="#cb90-1169"></a>    update_age_equil[iindex] <span class="sc">+</span></span>
<span id="cb90-1170"><a href="#cb90-1170"></a>    update_age_equil[rindex]</span>
<span id="cb90-1171"><a href="#cb90-1171"></a></span>
<span id="cb90-1172"><a href="#cb90-1172"></a>update_age_equil_seroprev <span class="ot">&lt;-</span> update_age_equil[rindex] <span class="sc">/</span> update_age_equil_n</span>
<span id="cb90-1173"><a href="#cb90-1173"></a></span>
<span id="cb90-1174"><a href="#cb90-1174"></a>update_age_equil_seroprev_df <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb90-1175"><a href="#cb90-1175"></a>    <span class="at">age =</span> ages,</span>
<span id="cb90-1176"><a href="#cb90-1176"></a>    <span class="at">seroprev =</span> update_age_equil_seroprev,</span>
<span id="cb90-1177"><a href="#cb90-1177"></a>    <span class="at">width =</span> da_ages</span>
<span id="cb90-1178"><a href="#cb90-1178"></a>)</span>
<span id="cb90-1179"><a href="#cb90-1179"></a><span class="in">```</span></span>
<span id="cb90-1180"><a href="#cb90-1180"></a></span>
<span id="cb90-1181"><a href="#cb90-1181"></a></span>
<span id="cb90-1184"><a href="#cb90-1184"></a><span class="in">```{r}</span></span>
<span id="cb90-1185"><a href="#cb90-1185"></a><span class="co">#| eval: false</span></span>
<span id="cb90-1186"><a href="#cb90-1186"></a><span class="co">#| source-line-numbers: "5,15-17,21"</span></span>
<span id="cb90-1187"><a href="#cb90-1187"></a>update_age_sol <span class="ot">&lt;-</span> deSolve<span class="sc">::</span><span class="fu">ode</span>(</span>
<span id="cb90-1188"><a href="#cb90-1188"></a>    <span class="at">y =</span> demog_yinit_ages,</span>
<span id="cb90-1189"><a href="#cb90-1189"></a>    <span class="at">times =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">400</span>, <span class="at">by =</span> <span class="fl">0.1</span>),</span>
<span id="cb90-1190"><a href="#cb90-1190"></a>    <span class="at">func =</span> multistage_model,</span>
<span id="cb90-1191"><a href="#cb90-1191"></a>    <span class="at">parms =</span> ?</span>
<span id="cb90-1192"><a href="#cb90-1192"></a>)</span>
<span id="cb90-1193"><a href="#cb90-1193"></a></span>
<span id="cb90-1194"><a href="#cb90-1194"></a><span class="co"># Get the time series for each infectious age group</span></span>
<span id="cb90-1195"><a href="#cb90-1195"></a>update_age_infecteds <span class="ot">&lt;-</span> update_age_sol[, <span class="dv">1</span> <span class="sc">+</span> iindex]</span>
<span id="cb90-1196"><a href="#cb90-1196"></a></span>
<span id="cb90-1197"><a href="#cb90-1197"></a><span class="co"># Get the last values in the time series</span></span>
<span id="cb90-1198"><a href="#cb90-1198"></a>update_age_equil <span class="ot">&lt;-</span> <span class="fu">drop</span>(<span class="fu">tail</span>(update_age_sol, <span class="dv">1</span>))[<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb90-1199"><a href="#cb90-1199"></a></span>
<span id="cb90-1200"><a href="#cb90-1200"></a><span class="co"># Calculate the number of individuals in each age group at the final timepoint</span></span>
<span id="cb90-1201"><a href="#cb90-1201"></a>update_age_equil_n <span class="ot">&lt;-</span> update_age_equil[ ? ] <span class="sc">+</span></span>
<span id="cb90-1202"><a href="#cb90-1202"></a>    update_age_equil[ ? ] <span class="sc">+</span></span>
<span id="cb90-1203"><a href="#cb90-1203"></a>    update_age_equil[ ? ]</span>
<span id="cb90-1204"><a href="#cb90-1204"></a></span>
<span id="cb90-1205"><a href="#cb90-1205"></a><span class="co"># Calculate final seroprevalence</span></span>
<span id="cb90-1206"><a href="#cb90-1206"></a><span class="co"># Hint: You need PREVIOUSLY infected individuals</span></span>
<span id="cb90-1207"><a href="#cb90-1207"></a>update_age_equil_seroprev <span class="ot">&lt;-</span> update_age_equil[ ? ] <span class="sc">/</span> update_age_equil_n</span>
<span id="cb90-1208"><a href="#cb90-1208"></a></span>
<span id="cb90-1209"><a href="#cb90-1209"></a>update_age_equil_seroprev_df <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb90-1210"><a href="#cb90-1210"></a>    <span class="at">age =</span> ages,</span>
<span id="cb90-1211"><a href="#cb90-1211"></a>    <span class="at">seroprev =</span> update_age_equil_seroprev,</span>
<span id="cb90-1212"><a href="#cb90-1212"></a>    <span class="at">width =</span> da_ages</span>
<span id="cb90-1213"><a href="#cb90-1213"></a>)</span>
<span id="cb90-1214"><a href="#cb90-1214"></a></span>
<span id="cb90-1215"><a href="#cb90-1215"></a><span class="fu">ggplot</span>(update_age_equil_seroprev_df, <span class="fu">aes</span>(<span class="at">x =</span> age, <span class="at">y =</span> seroprev, <span class="at">fill =</span> age)) <span class="sc">+</span></span>
<span id="cb90-1216"><a href="#cb90-1216"></a>    <span class="co"># Set column width to width of age bands, and justify to start at lower bound</span></span>
<span id="cb90-1217"><a href="#cb90-1217"></a>    <span class="fu">geom_col</span>(<span class="at">width =</span> update_age_equil_seroprev_df<span class="sc">$</span>width, <span class="at">just =</span> <span class="fl">1.0</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb90-1218"><a href="#cb90-1218"></a>    <span class="fu">labs</span>(</span>
<span id="cb90-1219"><a href="#cb90-1219"></a>        <span class="at">x =</span> <span class="st">"Age"</span>,</span>
<span id="cb90-1220"><a href="#cb90-1220"></a>        <span class="at">y =</span> <span class="st">"Seroprevalence"</span></span>
<span id="cb90-1221"><a href="#cb90-1221"></a>    ) <span class="sc">+</span></span>
<span id="cb90-1222"><a href="#cb90-1222"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">80</span>, <span class="dv">10</span>)) <span class="sc">+</span></span>
<span id="cb90-1223"><a href="#cb90-1223"></a>    <span class="fu">scale_fill_continuous</span>(<span class="at">low =</span> age_group_colors[<span class="dv">1</span>], <span class="at">high =</span> age_group_colors[<span class="dv">2</span>])</span>
<span id="cb90-1224"><a href="#cb90-1224"></a><span class="in">```</span></span>
<span id="cb90-1225"><a href="#cb90-1225"></a></span>
<span id="cb90-1226"><a href="#cb90-1226"></a>::: {.callout-note title="QUESTION"}</span>
<span id="cb90-1227"><a href="#cb90-1227"></a>At what age does the seroprevalence reach 75%?</span>
<span id="cb90-1228"><a href="#cb90-1228"></a>How does this compare to the answer in @sec-ex-3?</span>
<span id="cb90-1229"><a href="#cb90-1229"></a>:::</span>
<span id="cb90-1230"><a href="#cb90-1230"></a></span>
<span id="cb90-1231"><a href="#cb90-1231"></a><span class="fu">#### Compute $R_0$ for your assumptions.</span></span>
<span id="cb90-1232"><a href="#cb90-1232"></a></span>
<span id="cb90-1233"><a href="#cb90-1233"></a>As described previously, the calculation for $R_0$ is difficult due to all the age categories and transitions.</span>
<span id="cb90-1234"><a href="#cb90-1234"></a>Use the <span class="in">`calculate_R0()`</span> function we <span class="co">[</span><span class="ot">defined earlier</span><span class="co">](#ngm-function)</span> to calculate $R_0$ for our updated system.</span>
<span id="cb90-1235"><a href="#cb90-1235"></a></span>
<span id="cb90-1238"><a href="#cb90-1238"></a><span class="in">```{r}</span></span>
<span id="cb90-1239"><a href="#cb90-1239"></a><span class="co">#| eval: false</span></span>
<span id="cb90-1240"><a href="#cb90-1240"></a><span class="co">#| source-line-numbers: "2-5"</span></span>
<span id="cb90-1241"><a href="#cb90-1241"></a><span class="fu">calculate_R0</span>(</span>
<span id="cb90-1242"><a href="#cb90-1242"></a>    <span class="at">beta_mat =</span> ?,</span>
<span id="cb90-1243"><a href="#cb90-1243"></a>    <span class="at">stable_n_mat =</span> ?,</span>
<span id="cb90-1244"><a href="#cb90-1244"></a>    <span class="at">aging_mat =</span> ?,</span>
<span id="cb90-1245"><a href="#cb90-1245"></a>    <span class="at">recovery_rate =</span> ?</span>
<span id="cb90-1246"><a href="#cb90-1246"></a>)</span>
<span id="cb90-1247"><a href="#cb90-1247"></a><span class="in">```</span></span>
<span id="cb90-1248"><a href="#cb90-1248"></a></span>
<span id="cb90-1251"><a href="#cb90-1251"></a><span class="in">```{r}</span></span>
<span id="cb90-1252"><a href="#cb90-1252"></a><span class="co">#| echo: false</span></span>
<span id="cb90-1253"><a href="#cb90-1253"></a>update_age_R0 <span class="ot">&lt;-</span> <span class="fu">round</span>(</span>
<span id="cb90-1254"><a href="#cb90-1254"></a>    <span class="fu">calculate_R0</span>(</span>
<span id="cb90-1255"><a href="#cb90-1255"></a>        <span class="at">beta_mat =</span> update_age_params[[<span class="st">"beta_mat"</span>]],</span>
<span id="cb90-1256"><a href="#cb90-1256"></a>        <span class="at">stable_n_mat =</span> multistage_stable_n,</span>
<span id="cb90-1257"><a href="#cb90-1257"></a>        <span class="at">aging_mat =</span> update_age_params[[<span class="st">"aging_mat"</span>]],</span>
<span id="cb90-1258"><a href="#cb90-1258"></a>        <span class="at">recovery_rate =</span> update_age_params[[<span class="st">"recovery"</span>]]</span>
<span id="cb90-1259"><a href="#cb90-1259"></a>    ),</span>
<span id="cb90-1260"><a href="#cb90-1260"></a>    <span class="at">digits =</span> <span class="dv">2</span></span>
<span id="cb90-1261"><a href="#cb90-1261"></a>)</span>
<span id="cb90-1262"><a href="#cb90-1262"></a><span class="in">```</span></span>
<span id="cb90-1263"><a href="#cb90-1263"></a></span>
<span id="cb90-1264"><a href="#cb90-1264"></a>If you've done everything correctly, you should get $R_0 =$ <span class="in">`r update_age_R0`</span>.</span>
<span id="cb90-1265"><a href="#cb90-1265"></a>This is higher than previously.</span>
<span id="cb90-1266"><a href="#cb90-1266"></a>Does this match your intuition, given our changes to the beta matrix?</span>
<span id="cb90-1267"><a href="#cb90-1267"></a></span>
<span id="cb90-1268"><a href="#cb90-1268"></a>:::{.callout-warning}</span>
<span id="cb90-1269"><a href="#cb90-1269"></a>The following is Aaron's original code which provides slightly different results.</span>
<span id="cb90-1270"><a href="#cb90-1270"></a>:::</span>
<span id="cb90-1271"><a href="#cb90-1271"></a></span>
<span id="cb90-1274"><a href="#cb90-1274"></a><span class="in">```{r}</span></span>
<span id="cb90-1275"><a href="#cb90-1275"></a>F_mat <span class="ot">&lt;-</span> <span class="fu">diag</span>(update_age_equil_n) <span class="sc">%*%</span> update_age_beta_mat <span class="sc">+</span></span>
<span id="cb90-1276"><a href="#cb90-1276"></a>    update_age_params[[<span class="st">"aging_mat"</span>]] <span class="sc">-</span></span>
<span id="cb90-1277"><a href="#cb90-1277"></a>    <span class="fu">diag</span>(<span class="fu">diag</span>(update_age_params[[<span class="st">"aging_mat"</span>]]))</span>
<span id="cb90-1278"><a href="#cb90-1278"></a></span>
<span id="cb90-1279"><a href="#cb90-1279"></a>V_mat <span class="ot">&lt;-</span> <span class="fu">diag</span>(update_age_params[[<span class="st">"recovery"</span>]] <span class="sc">-</span> <span class="fu">diag</span>(update_age_params[[<span class="st">"aging_mat"</span>]]))</span>
<span id="cb90-1280"><a href="#cb90-1280"></a></span>
<span id="cb90-1281"><a href="#cb90-1281"></a><span class="co"># R0 approximately 6.53</span></span>
<span id="cb90-1282"><a href="#cb90-1282"></a><span class="fu">max</span>(<span class="fu">Re</span>(<span class="fu">eigen</span>(<span class="fu">solve</span>(V_mat, F_mat), <span class="at">only.values =</span> <span class="cn">TRUE</span>)<span class="sc">$</span>values))</span>
<span id="cb90-1283"><a href="#cb90-1283"></a><span class="in">```</span></span>
<span id="cb90-1284"><a href="#cb90-1284"></a></span>
<span id="cb90-1285"><a href="#cb90-1285"></a><span class="fu">## R0 and the Mean Age of Infection</span></span>
<span id="cb90-1286"><a href="#cb90-1286"></a></span>
<span id="cb90-1287"><a href="#cb90-1287"></a>To develop some intuition about the relationship between $R_0$ and the mean age of infection, let's play with an interactive plot.</span>
<span id="cb90-1288"><a href="#cb90-1288"></a>We will assume that the population is completely susceptible and that the force of infection is constant.</span>
<span id="cb90-1289"><a href="#cb90-1289"></a>We'll also assume that there is heterogenous mixing i.e. no age structure.</span>
<span id="cb90-1290"><a href="#cb90-1290"></a></span>
<span id="cb90-1291"><a href="#cb90-1291"></a>As we've seen in <span class="co">[</span><span class="ot">Matt's lecture on age structure</span><span class="co">](L05_age-structure.qmd)</span>, we can calculate the mean age of infection using the equation below:</span>
<span id="cb90-1292"><a href="#cb90-1292"></a></span>
<span id="cb90-1293"><a href="#cb90-1293"></a>$$</span>
<span id="cb90-1294"><a href="#cb90-1294"></a>A \approx \frac{L}{R_E - 1}</span>
<span id="cb90-1295"><a href="#cb90-1295"></a>$${#eq-mean-age}</span>
<span id="cb90-1296"><a href="#cb90-1296"></a></span>
<span id="cb90-1297"><a href="#cb90-1297"></a>where $L$ is the life expectancy $\left(L = \frac{1}{\mu}\right)$ and $R_E$ is the effective reproductive number ($R_E = R_0 * (1 - p)$ where $p$ is the fraction of individuals vaccinated).</span>
<span id="cb90-1298"><a href="#cb90-1298"></a></span>
<span id="cb90-1299"><a href="#cb90-1299"></a>::: {.callout-note collapse="false"}</span>
<span id="cb90-1300"><a href="#cb90-1300"></a>See @sec-mean-age-r-code for code you can run in <span class="in">`R`</span> to investigate the relationship between $R_0$, vaccination coverage, life expectancy, and the mean age of infection.</span>
<span id="cb90-1301"><a href="#cb90-1301"></a>:::</span>
<span id="cb90-1302"><a href="#cb90-1302"></a></span>
<span id="cb90-1303"><a href="#cb90-1303"></a><span class="kw">&lt;div</span> <span class="er">class</span><span class="ot">=</span><span class="st">"exercise"</span><span class="kw">&gt;</span></span>
<span id="cb90-1304"><a href="#cb90-1304"></a></span>
<span id="cb90-1305"><a href="#cb90-1305"></a><span class="fu">### Exercise 5: mean age of infection interactions</span></span>
<span id="cb90-1306"><a href="#cb90-1306"></a></span>
<span id="cb90-1307"><a href="#cb90-1307"></a><span class="kw">&lt;/div&gt;</span></span>
<span id="cb90-1308"><a href="#cb90-1308"></a></span>
<span id="cb90-1309"><a href="#cb90-1309"></a><span class="fu">#### When you increase $R_0$ from 2.0 to 4.0, what happens to the mean age of infection?</span></span>
<span id="cb90-1310"><a href="#cb90-1310"></a><span class="fu">##### Is there a linear change? If not, why not?</span></span>
<span id="cb90-1311"><a href="#cb90-1311"></a><span class="fu">#### With $R_0$ to 4.0, approximately what level of vaccination coverage is required for a mean age of infection of 40 years?</span></span>
<span id="cb90-1312"><a href="#cb90-1312"></a><span class="fu">#### Leaving $R_0$ and vaccination coverage the same, decrease the life expectancy to 50 years. What happens to the mean age of infection?</span></span>
<span id="cb90-1313"><a href="#cb90-1313"></a><span class="fu">##### If it changed, why do you think it did?</span></span>
<span id="cb90-1314"><a href="#cb90-1314"></a></span>
<span id="cb90-1315"><a href="#cb90-1315"></a><span class="kw">&lt;br&gt;</span></span>
<span id="cb90-1316"><a href="#cb90-1316"></a></span>
<span id="cb90-1319"><a href="#cb90-1319"></a><span class="in">```{ojs}</span></span>
<span id="cb90-1320"><a href="#cb90-1320"></a><span class="co">//| echo: false</span></span>
<span id="cb90-1321"><a href="#cb90-1321"></a>init_R0 <span class="op">=</span> <span class="fl">2.0</span></span>
<span id="cb90-1322"><a href="#cb90-1322"></a>init_vacc <span class="op">=</span> <span class="fl">0.0</span></span>
<span id="cb90-1323"><a href="#cb90-1323"></a>init_lifeexp <span class="op">=</span> <span class="dv">75</span></span>
<span id="cb90-1324"><a href="#cb90-1324"></a><span class="in">```</span></span>
<span id="cb90-1325"><a href="#cb90-1325"></a></span>
<span id="cb90-1328"><a href="#cb90-1328"></a><span class="in">```{ojs}</span></span>
<span id="cb90-1329"><a href="#cb90-1329"></a><span class="co">//| echo: false</span></span>
<span id="cb90-1330"><a href="#cb90-1330"></a><span class="kw">function</span> <span class="fu">set</span>(input<span class="op">,</span> value) {</span>
<span id="cb90-1331"><a href="#cb90-1331"></a>  input<span class="op">.</span><span class="at">value</span> <span class="op">=</span> value<span class="op">;</span></span>
<span id="cb90-1332"><a href="#cb90-1332"></a>  input<span class="op">.</span><span class="fu">dispatchEvent</span>(<span class="kw">new</span> <span class="bu">Event</span>(<span class="st">"input"</span><span class="op">,</span> {<span class="dt">bubbles</span><span class="op">:</span> <span class="kw">true</span>}))<span class="op">;</span></span>
<span id="cb90-1333"><a href="#cb90-1333"></a>}</span>
<span id="cb90-1334"><a href="#cb90-1334"></a><span class="in">```</span></span>
<span id="cb90-1335"><a href="#cb90-1335"></a></span>
<span id="cb90-1338"><a href="#cb90-1338"></a><span class="in">```{ojs}</span></span>
<span id="cb90-1339"><a href="#cb90-1339"></a><span class="co">//| echo: false</span></span>
<span id="cb90-1340"><a href="#cb90-1340"></a><span class="co">//| panel: sidebar</span></span>
<span id="cb90-1341"><a href="#cb90-1341"></a>viewof reset <span class="op">=</span> Inputs<span class="op">.</span><span class="fu">button</span>([</span>
<span id="cb90-1342"><a href="#cb90-1342"></a>  [<span class="st">"Reset all sliders"</span><span class="op">,</span> () <span class="kw">=&gt;</span> {</span>
<span id="cb90-1343"><a href="#cb90-1343"></a>    <span class="kw">set</span>(viewof R0<span class="op">,</span> init_R0)</span>
<span id="cb90-1344"><a href="#cb90-1344"></a>    <span class="kw">set</span>(viewof vacc<span class="op">,</span> init_vacc)</span>
<span id="cb90-1345"><a href="#cb90-1345"></a>    <span class="kw">set</span>(viewof lifeexp<span class="op">,</span> init_lifeexp)</span>
<span id="cb90-1346"><a href="#cb90-1346"></a>  }]</span>
<span id="cb90-1347"><a href="#cb90-1347"></a>])</span>
<span id="cb90-1348"><a href="#cb90-1348"></a>viewof R0 <span class="op">=</span> Inputs<span class="op">.</span><span class="fu">range</span>(</span>
<span id="cb90-1349"><a href="#cb90-1349"></a>    [<span class="fl">1.0</span><span class="op">,</span> <span class="fl">10.0</span>]<span class="op">,</span></span>
<span id="cb90-1350"><a href="#cb90-1350"></a>    {<span class="dt">value</span><span class="op">:</span> <span class="fl">2.0</span><span class="op">,</span> <span class="dt">step</span><span class="op">:</span> <span class="fl">0.01</span><span class="op">,</span> <span class="dt">label</span><span class="op">:</span> <span class="fu">md</span><span class="vs">`</span><span class="sc">${</span><span class="fu">tex</span><span class="vs">`R_0`</span><span class="sc">}</span><span class="vs">`</span>}</span>
<span id="cb90-1351"><a href="#cb90-1351"></a>)</span>
<span id="cb90-1352"><a href="#cb90-1352"></a></span>
<span id="cb90-1353"><a href="#cb90-1353"></a>viewof vacc <span class="op">=</span> Inputs<span class="op">.</span><span class="fu">range</span>(</span>
<span id="cb90-1354"><a href="#cb90-1354"></a>    [<span class="fl">0.0</span><span class="op">,</span> <span class="fl">1.0</span>]<span class="op">,</span></span>
<span id="cb90-1355"><a href="#cb90-1355"></a>    {<span class="dt">value</span><span class="op">:</span> <span class="fl">0.0</span><span class="op">,</span> <span class="dt">step</span><span class="op">:</span> <span class="fl">0.01</span><span class="op">,</span> <span class="dt">label</span><span class="op">:</span> <span class="st">"Vaccination coverage"</span>}</span>
<span id="cb90-1356"><a href="#cb90-1356"></a>)</span>
<span id="cb90-1357"><a href="#cb90-1357"></a></span>
<span id="cb90-1358"><a href="#cb90-1358"></a>viewof lifeexp <span class="op">=</span> Inputs<span class="op">.</span><span class="fu">range</span>(</span>
<span id="cb90-1359"><a href="#cb90-1359"></a>    [<span class="dv">50</span><span class="op">,</span> <span class="dv">100</span>]<span class="op">,</span></span>
<span id="cb90-1360"><a href="#cb90-1360"></a>    {<span class="dt">value</span><span class="op">:</span> <span class="dv">75</span><span class="op">,</span> <span class="dt">step</span><span class="op">:</span> <span class="dv">1</span><span class="op">,</span> <span class="dt">label</span><span class="op">:</span> <span class="st">"Life expectancy"</span>}</span>
<span id="cb90-1361"><a href="#cb90-1361"></a>)</span>
<span id="cb90-1362"><a href="#cb90-1362"></a></span>
<span id="cb90-1363"><a href="#cb90-1363"></a><span class="fu">md</span><span class="vs">`</span><span class="sc">${</span><span class="fu">tex</span><span class="vs">`R_E = </span><span class="sc">${</span>Re_str<span class="sc">}</span><span class="vs">`</span><span class="sc">}</span><span class="vs">`</span></span>
<span id="cb90-1364"><a href="#cb90-1364"></a><span class="fu">md</span><span class="vs">`</span><span class="sc">${</span><span class="fu">tex</span><span class="vs">`</span><span class="sc">\t</span><span class="vs">ext{Mean age of infection} = </span><span class="sc">${</span>Re_mean_age_str<span class="sc">}</span><span class="vs">`</span><span class="sc">}</span><span class="vs">`</span></span>
<span id="cb90-1365"><a href="#cb90-1365"></a><span class="in">```</span></span>
<span id="cb90-1366"><a href="#cb90-1366"></a></span>
<span id="cb90-1369"><a href="#cb90-1369"></a><span class="in">```{ojs}</span></span>
<span id="cb90-1370"><a href="#cb90-1370"></a><span class="co">//| echo: false</span></span>
<span id="cb90-1371"><a href="#cb90-1371"></a>Re <span class="op">=</span> R0 <span class="op">*</span> (<span class="dv">1</span> <span class="op">-</span> vacc)</span>
<span id="cb90-1372"><a href="#cb90-1372"></a>Re_str <span class="op">=</span> Re<span class="op">.</span><span class="fu">toPrecision</span>(<span class="dv">4</span>)<span class="op">.</span><span class="fu">toLocaleString</span>()</span>
<span id="cb90-1373"><a href="#cb90-1373"></a><span class="in">```</span></span>
<span id="cb90-1374"><a href="#cb90-1374"></a></span>
<span id="cb90-1377"><a href="#cb90-1377"></a><span class="in">```{ojs}</span></span>
<span id="cb90-1378"><a href="#cb90-1378"></a><span class="co">//| echo: false</span></span>
<span id="cb90-1379"><a href="#cb90-1379"></a><span class="kw">function</span> <span class="fu">calc_mean_age</span>(Re<span class="op">,</span> lifeexp) {</span>
<span id="cb90-1380"><a href="#cb90-1380"></a>    <span class="cf">if</span>(Re <span class="op">&gt;=</span> <span class="dv">1</span>) {</span>
<span id="cb90-1381"><a href="#cb90-1381"></a>        <span class="kw">var</span> mean_age <span class="op">=</span> (lifeexp <span class="op">/</span> (Re <span class="op">-</span> <span class="dv">1</span>))</span>
<span id="cb90-1382"><a href="#cb90-1382"></a>    } <span class="cf">else</span> {</span>
<span id="cb90-1383"><a href="#cb90-1383"></a>        <span class="kw">var</span> mean_age <span class="op">=</span> <span class="kw">Infinity</span></span>
<span id="cb90-1384"><a href="#cb90-1384"></a>    }</span>
<span id="cb90-1385"><a href="#cb90-1385"></a>    <span class="cf">return</span> mean_age</span>
<span id="cb90-1386"><a href="#cb90-1386"></a>}</span>
<span id="cb90-1387"><a href="#cb90-1387"></a><span class="in">```</span></span>
<span id="cb90-1388"><a href="#cb90-1388"></a></span>
<span id="cb90-1389"><a href="#cb90-1389"></a></span>
<span id="cb90-1392"><a href="#cb90-1392"></a><span class="in">```{ojs}</span></span>
<span id="cb90-1393"><a href="#cb90-1393"></a><span class="co">//| echo: false</span></span>
<span id="cb90-1394"><a href="#cb90-1394"></a>R0_mean_age <span class="op">=</span> <span class="fu">calc_mean_age</span>(R0<span class="op">,</span> lifeexp)</span>
<span id="cb90-1395"><a href="#cb90-1395"></a>Re_mean_age <span class="op">=</span> <span class="fu">calc_mean_age</span>(Re<span class="op">,</span> lifeexp)</span>
<span id="cb90-1396"><a href="#cb90-1396"></a>Re_mean_age_str <span class="op">=</span> Re_mean_age<span class="op">.</span><span class="fu">toPrecision</span>(<span class="dv">4</span>)<span class="op">.</span><span class="fu">toLocaleString</span>()</span>
<span id="cb90-1397"><a href="#cb90-1397"></a><span class="in">```</span></span>
<span id="cb90-1398"><a href="#cb90-1398"></a></span>
<span id="cb90-1399"><a href="#cb90-1399"></a></span>
<span id="cb90-1402"><a href="#cb90-1402"></a><span class="in">```{ojs}</span></span>
<span id="cb90-1403"><a href="#cb90-1403"></a><span class="co">//| echo: false</span></span>
<span id="cb90-1404"><a href="#cb90-1404"></a><span class="im">import</span> { aq<span class="op">,</span> op } <span class="im">from</span> <span class="st">'@uwdata/arquero'</span></span>
<span id="cb90-1405"><a href="#cb90-1405"></a><span class="in">```</span></span>
<span id="cb90-1406"><a href="#cb90-1406"></a></span>
<span id="cb90-1409"><a href="#cb90-1409"></a><span class="in">```{ojs}</span></span>
<span id="cb90-1410"><a href="#cb90-1410"></a><span class="co">//| echo: false</span></span>
<span id="cb90-1411"><a href="#cb90-1411"></a><span class="kw">function</span> <span class="fu">calc_mean_age_arr</span>(vacc<span class="op">,</span> lifeexp<span class="op">,</span> R0_min<span class="op">,</span> R0_max<span class="op">,</span> dR0) {</span>
<span id="cb90-1412"><a href="#cb90-1412"></a>    <span class="kw">var</span> R0_sim <span class="op">=</span> R0_min</span>
<span id="cb90-1413"><a href="#cb90-1413"></a></span>
<span id="cb90-1414"><a href="#cb90-1414"></a>    <span class="kw">var</span> R0 <span class="op">=</span> []</span>
<span id="cb90-1415"><a href="#cb90-1415"></a>    <span class="kw">var</span> Re <span class="op">=</span> []</span>
<span id="cb90-1416"><a href="#cb90-1416"></a>    <span class="kw">var</span> R0_mean_age <span class="op">=</span> []</span>
<span id="cb90-1417"><a href="#cb90-1417"></a>    <span class="kw">var</span> Re_mean_age <span class="op">=</span> []</span>
<span id="cb90-1418"><a href="#cb90-1418"></a></span>
<span id="cb90-1419"><a href="#cb90-1419"></a>    <span class="cf">for</span> (R0_sim <span class="op">=</span> R0_min<span class="op">;</span> R0_sim <span class="op">&lt;=</span> R0_max<span class="op">;</span> R0_sim <span class="op">+=</span> dR0) {</span>
<span id="cb90-1420"><a href="#cb90-1420"></a>        <span class="kw">var</span> Re_sim <span class="op">=</span> R0_sim <span class="op">*</span> (<span class="dv">1</span> <span class="op">-</span> vacc)</span>
<span id="cb90-1421"><a href="#cb90-1421"></a>        <span class="kw">var</span> R0_mean_age_sim <span class="op">=</span> <span class="fu">calc_mean_age</span>(R0_sim<span class="op">,</span> lifeexp)</span>
<span id="cb90-1422"><a href="#cb90-1422"></a>        <span class="kw">var</span> Re_mean_age_sim <span class="op">=</span> <span class="fu">calc_mean_age</span>(Re_sim<span class="op">,</span> lifeexp)</span>
<span id="cb90-1423"><a href="#cb90-1423"></a></span>
<span id="cb90-1424"><a href="#cb90-1424"></a>        R0<span class="op">.</span><span class="fu">push</span>(R0_sim)</span>
<span id="cb90-1425"><a href="#cb90-1425"></a>        Re<span class="op">.</span><span class="fu">push</span>(Re_sim)</span>
<span id="cb90-1426"><a href="#cb90-1426"></a>        R0_mean_age<span class="op">.</span><span class="fu">push</span>(R0_mean_age_sim)</span>
<span id="cb90-1427"><a href="#cb90-1427"></a>        Re_mean_age<span class="op">.</span><span class="fu">push</span>(Re_mean_age_sim)</span>
<span id="cb90-1428"><a href="#cb90-1428"></a>    }</span>
<span id="cb90-1429"><a href="#cb90-1429"></a></span>
<span id="cb90-1430"><a href="#cb90-1430"></a>    <span class="cf">return</span> {</span>
<span id="cb90-1431"><a href="#cb90-1431"></a>        <span class="dt">Re</span><span class="op">:</span> aq<span class="op">.</span><span class="fu">table</span>({</span>
<span id="cb90-1432"><a href="#cb90-1432"></a>                <span class="dt">R0</span><span class="op">:</span> R0<span class="op">,</span></span>
<span id="cb90-1433"><a href="#cb90-1433"></a>                <span class="dt">mean_age</span><span class="op">:</span> Re_mean_age</span>
<span id="cb90-1434"><a href="#cb90-1434"></a>            })<span class="op">.</span><span class="fu">filter</span>((d) <span class="kw">=&gt;</span> d<span class="op">.</span><span class="at">mean_age</span> <span class="op">&lt;=</span> <span class="dv">100</span>)<span class="op">,</span></span>
<span id="cb90-1435"><a href="#cb90-1435"></a>        <span class="dt">R0</span><span class="op">:</span> aq<span class="op">.</span><span class="fu">table</span>({</span>
<span id="cb90-1436"><a href="#cb90-1436"></a>                <span class="dt">R0</span><span class="op">:</span> R0<span class="op">,</span></span>
<span id="cb90-1437"><a href="#cb90-1437"></a>                <span class="dt">mean_age</span><span class="op">:</span> R0_mean_age</span>
<span id="cb90-1438"><a href="#cb90-1438"></a>            })<span class="op">.</span><span class="fu">filter</span>((d) <span class="kw">=&gt;</span> d<span class="op">.</span><span class="at">mean_age</span> <span class="op">&lt;=</span> <span class="dv">100</span>)</span>
<span id="cb90-1439"><a href="#cb90-1439"></a>    }</span>
<span id="cb90-1440"><a href="#cb90-1440"></a>}</span>
<span id="cb90-1441"><a href="#cb90-1441"></a><span class="in">```</span></span>
<span id="cb90-1442"><a href="#cb90-1442"></a></span>
<span id="cb90-1445"><a href="#cb90-1445"></a><span class="in">```{ojs}</span></span>
<span id="cb90-1446"><a href="#cb90-1446"></a><span class="co">//| echo: false</span></span>
<span id="cb90-1447"><a href="#cb90-1447"></a>mean_age_arrs <span class="op">=</span> <span class="fu">calc_mean_age_arr</span>(vacc<span class="op">,</span> lifeexp<span class="op">,</span> <span class="fl">1.0</span><span class="op">,</span> <span class="fl">10.0</span><span class="op">,</span> <span class="fl">0.01</span>)</span>
<span id="cb90-1448"><a href="#cb90-1448"></a><span class="in">```</span></span>
<span id="cb90-1449"><a href="#cb90-1449"></a></span>
<span id="cb90-1452"><a href="#cb90-1452"></a><span class="in">```{ojs}</span></span>
<span id="cb90-1453"><a href="#cb90-1453"></a><span class="co">//| echo: false</span></span>
<span id="cb90-1454"><a href="#cb90-1454"></a>mean_age_dots <span class="op">=</span> [({</span>
<span id="cb90-1455"><a href="#cb90-1455"></a>    <span class="dt">arrow_start</span><span class="op">:</span> R0_mean_age <span class="op">&lt;=</span> <span class="dv">100</span> <span class="op">?</span> R0_mean_age <span class="op">:</span> <span class="dv">100</span><span class="op">,</span></span>
<span id="cb90-1456"><a href="#cb90-1456"></a>    <span class="dt">arrow_end</span><span class="op">:</span> Re_mean_age <span class="op">&lt;=</span> <span class="dv">100</span> <span class="op">?</span> Re_mean_age <span class="op">:</span> <span class="dv">100</span><span class="op">,</span></span>
<span id="cb90-1457"><a href="#cb90-1457"></a>    <span class="dt">R0</span><span class="op">:</span> R0<span class="op">.</span><span class="fu">toPrecision</span>(<span class="dv">3</span>)<span class="op">,</span></span>
<span id="cb90-1458"><a href="#cb90-1458"></a>    <span class="dt">Re</span><span class="op">:</span> Re<span class="op">.</span><span class="fu">toPrecision</span>(<span class="dv">3</span>)<span class="op">,</span></span>
<span id="cb90-1459"><a href="#cb90-1459"></a>    R0_mean_age<span class="op">,</span></span>
<span id="cb90-1460"><a href="#cb90-1460"></a>    Re_mean_age</span>
<span id="cb90-1461"><a href="#cb90-1461"></a>})]</span>
<span id="cb90-1462"><a href="#cb90-1462"></a><span class="in">```</span></span>
<span id="cb90-1463"><a href="#cb90-1463"></a></span>
<span id="cb90-1466"><a href="#cb90-1466"></a><span class="in">```{ojs}</span></span>
<span id="cb90-1467"><a href="#cb90-1467"></a><span class="co">//| echo: false</span></span>
<span id="cb90-1468"><a href="#cb90-1468"></a><span class="co">//| panel: fill</span></span>
<span id="cb90-1469"><a href="#cb90-1469"></a>{</span>
<span id="cb90-1470"><a href="#cb90-1470"></a>    <span class="kw">let</span> R0Color <span class="op">=</span> <span class="st">"#1f77b4"</span></span>
<span id="cb90-1471"><a href="#cb90-1471"></a>    <span class="kw">let</span> ReColor <span class="op">=</span> <span class="st">"#ff7f0e"</span></span>
<span id="cb90-1472"><a href="#cb90-1472"></a></span>
<span id="cb90-1473"><a href="#cb90-1473"></a>    <span class="kw">let</span> plot <span class="op">=</span> Plot<span class="op">.</span><span class="fu">plot</span>({</span>
<span id="cb90-1474"><a href="#cb90-1474"></a>        <span class="dt">color</span><span class="op">:</span> {</span>
<span id="cb90-1475"><a href="#cb90-1475"></a>            <span class="dt">legend</span><span class="op">:</span> <span class="kw">true</span><span class="op">,</span></span>
<span id="cb90-1476"><a href="#cb90-1476"></a>            <span class="dt">domain</span><span class="op">:</span> [<span class="st">"Unvaccinated"</span><span class="op">,</span> <span class="st">"Vaccinated"</span>]<span class="op">,</span></span>
<span id="cb90-1477"><a href="#cb90-1477"></a>            <span class="dt">range</span><span class="op">:</span> [<span class="st">"#1f77b4"</span><span class="op">,</span> <span class="st">"#ff7f0e"</span>]</span>
<span id="cb90-1478"><a href="#cb90-1478"></a>        }<span class="op">,</span></span>
<span id="cb90-1479"><a href="#cb90-1479"></a>        <span class="dt">style</span><span class="op">:</span> {<span class="dt">fontSize</span><span class="op">:</span> <span class="st">"20px"</span>}<span class="op">,</span></span>
<span id="cb90-1480"><a href="#cb90-1480"></a>        <span class="dt">marginLeft</span><span class="op">:</span> <span class="dv">65</span><span class="op">,</span></span>
<span id="cb90-1481"><a href="#cb90-1481"></a>        <span class="dt">marginTop</span><span class="op">:</span> <span class="dv">40</span><span class="op">,</span></span>
<span id="cb90-1482"><a href="#cb90-1482"></a>        <span class="dt">marginBottom</span><span class="op">:</span> <span class="dv">55</span><span class="op">,</span></span>
<span id="cb90-1483"><a href="#cb90-1483"></a>        <span class="dt">grid</span><span class="op">:</span> <span class="kw">true</span><span class="op">,</span></span>
<span id="cb90-1484"><a href="#cb90-1484"></a>        <span class="dt">width</span><span class="op">:</span> <span class="dv">800</span><span class="op">,</span></span>
<span id="cb90-1485"><a href="#cb90-1485"></a>        <span class="dt">height</span><span class="op">:</span> <span class="dv">670</span><span class="op">,</span></span>
<span id="cb90-1486"><a href="#cb90-1486"></a>        <span class="dt">x</span><span class="op">:</span> {<span class="dt">label</span><span class="op">:</span> <span class="st">"R0"</span><span class="op">,</span> <span class="dt">domain</span><span class="op">:</span> [<span class="dv">0</span><span class="op">,</span> <span class="dv">10</span>]}<span class="op">,</span></span>
<span id="cb90-1487"><a href="#cb90-1487"></a>        <span class="dt">y</span><span class="op">:</span> {<span class="dt">label</span><span class="op">:</span> <span class="st">"Mean Age of Infection"</span><span class="op">,</span> <span class="dt">domain</span><span class="op">:</span> [<span class="dv">0</span><span class="op">,</span> <span class="dv">100</span>]}<span class="op">,</span></span>
<span id="cb90-1488"><a href="#cb90-1488"></a>        <span class="dt">marks</span><span class="op">:</span> [</span>
<span id="cb90-1489"><a href="#cb90-1489"></a>            Plot<span class="op">.</span><span class="fu">line</span>(mean_age_arrs<span class="op">.</span><span class="at">Re</span><span class="op">,</span> {<span class="dt">x</span><span class="op">:</span> <span class="st">"R0"</span><span class="op">,</span> <span class="dt">y</span><span class="op">:</span> <span class="st">"mean_age"</span><span class="op">,</span> <span class="dt">stroke</span><span class="op">:</span> ReColor<span class="op">,</span> <span class="dt">strokeWidth</span><span class="op">:</span> <span class="dv">6</span>})<span class="op">,</span></span>
<span id="cb90-1490"><a href="#cb90-1490"></a>            Plot<span class="op">.</span><span class="fu">line</span>(mean_age_arrs<span class="op">.</span><span class="at">R0</span><span class="op">,</span> {<span class="dt">x</span><span class="op">:</span> <span class="st">"R0"</span><span class="op">,</span> <span class="dt">y</span><span class="op">:</span> <span class="st">"mean_age"</span><span class="op">,</span> <span class="dt">stroke</span><span class="op">:</span> R0Color<span class="op">,</span> <span class="dt">strokeWidth</span><span class="op">:</span> <span class="dv">6</span>})<span class="op">,</span></span>
<span id="cb90-1491"><a href="#cb90-1491"></a>            Re_mean_age <span class="op">&lt;=</span> <span class="dv">100</span> <span class="op">?</span></span>
<span id="cb90-1492"><a href="#cb90-1492"></a>                [</span>
<span id="cb90-1493"><a href="#cb90-1493"></a>                    vacc <span class="op">&gt;</span> <span class="fl">0.00</span> <span class="op">?</span> Plot<span class="op">.</span><span class="fu">dot</span>(mean_age_dots<span class="op">,</span> {<span class="dt">x</span><span class="op">:</span> <span class="st">"R0"</span><span class="op">,</span> <span class="dt">y</span><span class="op">:</span> <span class="st">"Re_mean_age"</span><span class="op">,</span> <span class="dt">r</span><span class="op">:</span> <span class="dv">12</span><span class="op">,</span> <span class="dt">stroke</span><span class="op">:</span> ReColor<span class="op">,</span> <span class="dt">fill</span><span class="op">:</span>     ReColor<span class="op">,</span>  <span class="dt">fillOpacity</span><span class="op">:</span> <span class="fl">0.6</span>}) <span class="op">:</span> <span class="kw">null</span><span class="op">,</span></span>
<span id="cb90-1494"><a href="#cb90-1494"></a>                    Plot<span class="op">.</span><span class="fu">text</span>(</span>
<span id="cb90-1495"><a href="#cb90-1495"></a>                        mean_age_dots<span class="op">,</span></span>
<span id="cb90-1496"><a href="#cb90-1496"></a>                        {<span class="dt">x</span><span class="op">:</span> <span class="st">"R0"</span><span class="op">,</span> <span class="dt">y</span><span class="op">:</span> <span class="st">"Re_mean_age"</span><span class="op">,</span> <span class="dt">text</span><span class="op">:</span> (d) <span class="kw">=&gt;</span> <span class="vs">`Re = </span><span class="sc">${</span>d<span class="op">.</span><span class="at">Re</span><span class="sc">}</span><span class="vs">`</span><span class="op">,</span> <span class="dt">dx</span><span class="op">:</span> <span class="dv">55</span><span class="op">,</span> <span class="dt">dy</span><span class="op">:</span> <span class="op">-</span><span class="dv">25</span><span class="op">,</span> <span class="dt">fontWeight</span><span class="op">:</span> <span class="st">"bold"</span><span class="op">,</span> <span class="dt">fill</span><span class="op">:</span> ReColor}</span>
<span id="cb90-1497"><a href="#cb90-1497"></a>                    )</span>
<span id="cb90-1498"><a href="#cb90-1498"></a>                ] <span class="op">:</span></span>
<span id="cb90-1499"><a href="#cb90-1499"></a>            <span class="kw">null</span><span class="op">,</span></span>
<span id="cb90-1500"><a href="#cb90-1500"></a>            R0_mean_age <span class="op">&lt;=</span> <span class="dv">100</span> <span class="op">?</span></span>
<span id="cb90-1501"><a href="#cb90-1501"></a>                [</span>
<span id="cb90-1502"><a href="#cb90-1502"></a>                Plot<span class="op">.</span><span class="fu">dot</span>(mean_age_dots<span class="op">,</span> {<span class="dt">x</span><span class="op">:</span> <span class="st">"R0"</span><span class="op">,</span> <span class="dt">y</span><span class="op">:</span> <span class="st">"R0_mean_age"</span><span class="op">,</span> <span class="dt">r</span><span class="op">:</span> <span class="dv">12</span><span class="op">,</span> <span class="dt">stroke</span><span class="op">:</span> R0Color<span class="op">,</span> <span class="dt">fill</span><span class="op">:</span> R0Color<span class="op">,</span> <span class="dt">fillOpacity</span><span class="op">:</span> <span class="fl">0.6</span>})<span class="op">,</span></span>
<span id="cb90-1503"><a href="#cb90-1503"></a>                Plot<span class="op">.</span><span class="fu">text</span>(</span>
<span id="cb90-1504"><a href="#cb90-1504"></a>                    mean_age_dots<span class="op">,</span></span>
<span id="cb90-1505"><a href="#cb90-1505"></a>                    {<span class="dt">x</span><span class="op">:</span> <span class="st">"R0"</span><span class="op">,</span> <span class="dt">y</span><span class="op">:</span> <span class="st">"R0_mean_age"</span><span class="op">,</span> <span class="dt">text</span><span class="op">:</span> (d) <span class="kw">=&gt;</span> <span class="vs">`R0 = </span><span class="sc">${</span>d<span class="op">.</span><span class="at">R0</span><span class="sc">}</span><span class="vs">`</span><span class="op">,</span> <span class="dt">dx</span><span class="op">:</span> <span class="op">-</span><span class="dv">60</span><span class="op">,</span> <span class="dt">dy</span><span class="op">:</span> <span class="dv">30</span><span class="op">,</span> <span class="dt">fontWeight</span><span class="op">:</span> <span class="st">"bold"</span><span class="op">,</span> <span class="dt">fill</span><span class="op">:</span> R0Color}</span>
<span id="cb90-1506"><a href="#cb90-1506"></a>                )</span>
<span id="cb90-1507"><a href="#cb90-1507"></a>                ] <span class="op">:</span></span>
<span id="cb90-1508"><a href="#cb90-1508"></a>            <span class="kw">null</span><span class="op">,</span></span>
<span id="cb90-1509"><a href="#cb90-1509"></a>            Plot<span class="op">.</span><span class="fu">arrow</span>(mean_age_dots<span class="op">,</span> {<span class="dt">x1</span><span class="op">:</span> <span class="st">"R0"</span><span class="op">,</span> <span class="dt">x2</span><span class="op">:</span> <span class="st">"R0"</span><span class="op">,</span> <span class="dt">y1</span><span class="op">:</span> <span class="st">"arrow_start"</span><span class="op">,</span> <span class="dt">y2</span><span class="op">:</span> <span class="st">"arrow_end"</span><span class="op">,</span> <span class="dt">strokeWidth</span><span class="op">:</span> <span class="dv">4</span><span class="op">,</span> <span class="dt">headLength</span><span class="op">:</span> <span class="dv">5</span><span class="op">,</span> <span class="dt">inset</span><span class="op">:</span> <span class="dv">15</span>})<span class="op">,</span></span>
<span id="cb90-1510"><a href="#cb90-1510"></a>        ]</span>
<span id="cb90-1511"><a href="#cb90-1511"></a>    })<span class="op">;</span></span>
<span id="cb90-1512"><a href="#cb90-1512"></a></span>
<span id="cb90-1513"><a href="#cb90-1513"></a>  <span class="cf">return</span> plot<span class="op">;</span></span>
<span id="cb90-1514"><a href="#cb90-1514"></a>}</span>
<span id="cb90-1515"><a href="#cb90-1515"></a><span class="in">```</span></span>
<span id="cb90-1516"><a href="#cb90-1516"></a></span>
<span id="cb90-1517"><a href="#cb90-1517"></a><span class="fu">## Bonus Materials</span></span>
<span id="cb90-1518"><a href="#cb90-1518"></a><span class="fu">### Calculating $R_0$ with the Next Generation Matrix</span></span>
<span id="cb90-1519"><a href="#cb90-1519"></a><span class="fu">#### Simple model structure {#sec-simple-ngm}</span></span>
<span id="cb90-1520"><a href="#cb90-1520"></a></span>
<span id="cb90-1521"><a href="#cb90-1521"></a>To compute $R_0$, we need to know the stable age distribution (the relative proportion in the juvenile and adult age classes) of the population, which we can find by solving for the disease-free equilibrium: $S_J^*=B/\alpha$ and $S_A^*=B/\mu$.</span>
<span id="cb90-1522"><a href="#cb90-1522"></a>With the stable age distribution, we can calculate $R_0$ by constructing the next generation matrix.</span>
<span id="cb90-1523"><a href="#cb90-1523"></a>The code below outlines how the next generation matrix is constructed using the $\alpha$ (aging from juvenile to adult), $\mu$ (death), $n$ (total births), $\gamma$ (recovery), $da$ (width of age groups in years), and $\beta$ (transmission) parameters. </span>
<span id="cb90-1524"><a href="#cb90-1524"></a></span>
<span id="cb90-1525"><a href="#cb90-1525"></a></span>
<span id="cb90-1526"><a href="#cb90-1526"></a>The next generation matrix is a matrix that specifies how many new age-specific infections are generated by a typical infected individual of each age class (in a fully susceptible population).</span>
<span id="cb90-1527"><a href="#cb90-1527"></a>For example, let's consider an infected adult and ask how many new juvenile infections it generates: this is the product of the number of susceptible juveniles (from the stable age distribution), the per capita transmission rate from adults to juveniles and the average duration of infection, i.e. $S_J^* \times \beta_{JA} \times 1/ (\gamma+\mu)$.</span>
<span id="cb90-1528"><a href="#cb90-1528"></a>This forms one element of our next generation matrix.</span>
<span id="cb90-1529"><a href="#cb90-1529"></a>The other elements look very similar, except there are extra terms when we consider an infected juvenile because there is a (very small) chance they may age during the infectious period and therefore cause new infections as an adult:</span>
<span id="cb90-1530"><a href="#cb90-1530"></a></span>
<span id="cb90-1531"><a href="#cb90-1531"></a>$$</span>
<span id="cb90-1532"><a href="#cb90-1532"></a>\mathrm{NGM} = \begin{pmatrix}</span>
<span id="cb90-1533"><a href="#cb90-1533"></a>        \frac{S_J^* \beta_{JJ}}{(\gamma + \alpha)} +</span>
<span id="cb90-1534"><a href="#cb90-1534"></a>        \frac{\alpha}{(\gamma+\mu)} \frac{S_J^* \beta_{JA}}{(\gamma + \mu)} &amp;</span>
<span id="cb90-1535"><a href="#cb90-1535"></a>        \frac{S_J^* \beta_{JA}}{(\gamma + \mu)} <span class="sc">\\</span></span>
<span id="cb90-1536"><a href="#cb90-1536"></a>        \frac{S_A^* \beta_{AJ}}{(\gamma + \alpha)} + </span>
<span id="cb90-1537"><a href="#cb90-1537"></a>            \frac{\alpha}{(\gamma + \mu)} \frac{S_A^*\beta_{AA}}{(\gamma+\mu)} &amp;</span>
<span id="cb90-1538"><a href="#cb90-1538"></a>            \frac{S_A^* \beta_{AA}}{(\gamma + \mu)} </span>
<span id="cb90-1539"><a href="#cb90-1539"></a>    \end{pmatrix}</span>
<span id="cb90-1540"><a href="#cb90-1540"></a>$$ {#eq-simple-ngm}</span>
<span id="cb90-1541"><a href="#cb90-1541"></a></span>
<span id="cb90-1542"><a href="#cb90-1542"></a>$R_0$ can then be computed as the dominant eigenvalue (i.e., the one with the largest real part) of this matrix. Let's take an example from a model with 2 age classes, from above. First, let's define the components of the next generation matrix:</span>
<span id="cb90-1543"><a href="#cb90-1543"></a></span>
<span id="cb90-1546"><a href="#cb90-1546"></a><span class="in">```{r}</span></span>
<span id="cb90-1547"><a href="#cb90-1547"></a>ngm_params <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb90-1548"><a href="#cb90-1548"></a>    <span class="at">beta_within =</span> <span class="fl">0.011</span>,</span>
<span id="cb90-1549"><a href="#cb90-1549"></a>    <span class="at">beta_between =</span> <span class="fl">0.005</span>,</span>
<span id="cb90-1550"><a href="#cb90-1550"></a>    <span class="at">age_band_j =</span> <span class="dv">20</span>,</span>
<span id="cb90-1551"><a href="#cb90-1551"></a>    <span class="at">age_band_a =</span> <span class="dv">60</span>,</span>
<span id="cb90-1552"><a href="#cb90-1552"></a>    <span class="at">recovery =</span> <span class="dv">10</span></span>
<span id="cb90-1553"><a href="#cb90-1553"></a>)</span>
<span id="cb90-1554"><a href="#cb90-1554"></a></span>
<span id="cb90-1555"><a href="#cb90-1555"></a>alpha_ngm <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> ngm_params[<span class="st">"age_band_j"</span>]</span>
<span id="cb90-1556"><a href="#cb90-1556"></a>mu_ngm <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> ngm_params[<span class="st">"age_band_a"</span>]</span>
<span id="cb90-1557"><a href="#cb90-1557"></a>n_ngm <span class="ot">&lt;-</span> demog_params[<span class="st">"births"</span>] <span class="sc">/</span> <span class="fu">c</span>(alpha_ngm, mu_ngm)</span>
<span id="cb90-1558"><a href="#cb90-1558"></a></span>
<span id="cb90-1559"><a href="#cb90-1559"></a>beta_ngm <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(</span>
<span id="cb90-1560"><a href="#cb90-1560"></a>    ngm_params[<span class="st">"beta_within"</span>],</span>
<span id="cb90-1561"><a href="#cb90-1561"></a>    ngm_params[<span class="st">"beta_between"</span>],</span>
<span id="cb90-1562"><a href="#cb90-1562"></a>    ngm_params[<span class="st">"beta_between"</span>],</span>
<span id="cb90-1563"><a href="#cb90-1563"></a>    ngm_params[<span class="st">"beta_within"</span>]</span>
<span id="cb90-1564"><a href="#cb90-1564"></a>    ),</span>
<span id="cb90-1565"><a href="#cb90-1565"></a>    <span class="at">nrow =</span> <span class="dv">2</span>,</span>
<span id="cb90-1566"><a href="#cb90-1566"></a>    <span class="at">ncol =</span> <span class="dv">2</span></span>
<span id="cb90-1567"><a href="#cb90-1567"></a>)</span>
<span id="cb90-1568"><a href="#cb90-1568"></a><span class="in">```</span></span>
<span id="cb90-1569"><a href="#cb90-1569"></a></span>
<span id="cb90-1570"><a href="#cb90-1570"></a>The Next Generation Matrix can be calculated in <span class="in">`R`</span> as:</span>
<span id="cb90-1571"><a href="#cb90-1571"></a></span>
<span id="cb90-1574"><a href="#cb90-1574"></a><span class="in">```{r}</span></span>
<span id="cb90-1575"><a href="#cb90-1575"></a>ngm <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb90-1576"><a href="#cb90-1576"></a>    <span class="fu">c</span>(</span>
<span id="cb90-1577"><a href="#cb90-1577"></a>        n_ngm[<span class="dv">1</span>] <span class="sc">*</span> (beta_ngm[<span class="dv">1</span>, <span class="dv">1</span>] <span class="sc">/</span> (ngm_params[<span class="st">"recovery"</span>] <span class="sc">+</span> alpha_ngm)) <span class="sc">+</span></span>
<span id="cb90-1578"><a href="#cb90-1578"></a>            alpha_ngm <span class="sc">/</span> (ngm_params[<span class="st">"recovery"</span>] <span class="sc">+</span> mu_ngm) <span class="sc">*</span></span>
<span id="cb90-1579"><a href="#cb90-1579"></a>            n_ngm[<span class="dv">1</span>] <span class="sc">*</span> beta_ngm[<span class="dv">1</span>, <span class="dv">2</span>] <span class="sc">/</span> (ngm_params[<span class="st">"recovery"</span>] <span class="sc">+</span> mu_ngm),</span>
<span id="cb90-1580"><a href="#cb90-1580"></a>        </span>
<span id="cb90-1581"><a href="#cb90-1581"></a>        n_ngm[<span class="dv">2</span>] <span class="sc">*</span> beta_ngm[<span class="dv">2</span>, <span class="dv">1</span>] <span class="sc">/</span> (ngm_params[<span class="st">"recovery"</span>] <span class="sc">+</span> alpha_ngm) <span class="sc">+</span></span>
<span id="cb90-1582"><a href="#cb90-1582"></a>            alpha_ngm <span class="sc">/</span> (ngm_params[<span class="st">"recovery"</span>] <span class="sc">+</span> mu_ngm) <span class="sc">*</span></span>
<span id="cb90-1583"><a href="#cb90-1583"></a>            n_ngm[<span class="dv">2</span>] <span class="sc">*</span> (beta_ngm[<span class="dv">2</span>, <span class="dv">2</span>] <span class="sc">/</span> (ngm_params[<span class="st">"recovery"</span>] <span class="sc">+</span> mu_ngm)),</span>
<span id="cb90-1584"><a href="#cb90-1584"></a></span>
<span id="cb90-1585"><a href="#cb90-1585"></a>        n_ngm[<span class="dv">1</span>] <span class="sc">*</span> beta_ngm[<span class="dv">1</span>, <span class="dv">2</span>] <span class="sc">/</span> (ngm_params[<span class="st">"recovery"</span>] <span class="sc">+</span> mu_ngm),</span>
<span id="cb90-1586"><a href="#cb90-1586"></a></span>
<span id="cb90-1587"><a href="#cb90-1587"></a>        n_ngm[<span class="dv">2</span>] <span class="sc">*</span> beta_ngm[<span class="dv">2</span>, <span class="dv">2</span>] <span class="sc">/</span> (ngm_params[<span class="st">"recovery"</span>] <span class="sc">+</span> mu_ngm)</span>
<span id="cb90-1588"><a href="#cb90-1588"></a>    ),</span>
<span id="cb90-1589"><a href="#cb90-1589"></a>    <span class="at">nrow =</span> <span class="dv">2</span>,</span>
<span id="cb90-1590"><a href="#cb90-1590"></a>    <span class="at">ncol =</span> <span class="dv">2</span></span>
<span id="cb90-1591"><a href="#cb90-1591"></a>)</span>
<span id="cb90-1592"><a href="#cb90-1592"></a><span class="in">```</span></span>
<span id="cb90-1593"><a href="#cb90-1593"></a></span>
<span id="cb90-1594"><a href="#cb90-1594"></a>We can then calculate the eigenvalues and eigenvectors of this matrix:</span>
<span id="cb90-1595"><a href="#cb90-1595"></a></span>
<span id="cb90-1598"><a href="#cb90-1598"></a><span class="in">```{r}</span></span>
<span id="cb90-1599"><a href="#cb90-1599"></a><span class="fu">eigen</span>(ngm)</span>
<span id="cb90-1600"><a href="#cb90-1600"></a><span class="in">```</span></span>
<span id="cb90-1601"><a href="#cb90-1601"></a></span>
<span id="cb90-1602"><a href="#cb90-1602"></a>We can also choose to just output the eigenvalues:</span>
<span id="cb90-1603"><a href="#cb90-1603"></a></span>
<span id="cb90-1606"><a href="#cb90-1606"></a><span class="in">```{r}</span></span>
<span id="cb90-1607"><a href="#cb90-1607"></a><span class="fu">eigen</span>(ngm, <span class="at">only.values =</span> <span class="cn">TRUE</span>)</span>
<span id="cb90-1608"><a href="#cb90-1608"></a><span class="in">```</span></span>
<span id="cb90-1609"><a href="#cb90-1609"></a></span>
<span id="cb90-1610"><a href="#cb90-1610"></a>Finally, let's print $R_0$:</span>
<span id="cb90-1611"><a href="#cb90-1611"></a></span>
<span id="cb90-1614"><a href="#cb90-1614"></a><span class="in">```{r}</span></span>
<span id="cb90-1615"><a href="#cb90-1615"></a><span class="fu">max</span>(</span>
<span id="cb90-1616"><a href="#cb90-1616"></a>    <span class="fu">Re</span>(</span>
<span id="cb90-1617"><a href="#cb90-1617"></a>        <span class="fu">eigen</span>(ngm, <span class="at">only.values =</span> <span class="cn">TRUE</span>)<span class="sc">$</span>values</span>
<span id="cb90-1618"><a href="#cb90-1618"></a>    )</span>
<span id="cb90-1619"><a href="#cb90-1619"></a>)</span>
<span id="cb90-1620"><a href="#cb90-1620"></a><span class="in">```</span></span>
<span id="cb90-1621"><a href="#cb90-1621"></a></span>
<span id="cb90-1622"><a href="#cb90-1622"></a><span class="fu">#### Age-structured NGM {#sec-age-structure-ngm}</span></span>
<span id="cb90-1623"><a href="#cb90-1623"></a></span>
<span id="cb90-1624"><a href="#cb90-1624"></a>Many times it would be impractical to write out the NGM: there are often too many compartments in an age-structured model.</span>
<span id="cb90-1625"><a href="#cb90-1625"></a>In this instance, we want to use a slightly different approach, but the underlying principles are the same: each element of the NGM balances the number of new infections expected to be produced with the rates of individuals coming in and out of that compartment.</span>
<span id="cb90-1626"><a href="#cb90-1626"></a></span>
<span id="cb90-1627"><a href="#cb90-1627"></a><span class="fu">##### Stable Age Distribution</span></span>
<span id="cb90-1628"><a href="#cb90-1628"></a></span>
<span id="cb90-1629"><a href="#cb90-1629"></a>The first thing we need, as before, is the stable age distribution i.e., the disease-free equilibrium.</span>
<span id="cb90-1630"><a href="#cb90-1630"></a>There are two ways we can do this:</span>
<span id="cb90-1631"><a href="#cb90-1631"></a></span>
<span id="cb90-1632"><a href="#cb90-1632"></a><span class="ss">1. </span>Simulate the model without any infections for a sufficiently long time (simple, but less accurate)</span>
<span id="cb90-1633"><a href="#cb90-1633"></a><span class="ss">2. </span>Do the math.</span>
<span id="cb90-1634"><a href="#cb90-1634"></a></span>
<span id="cb90-1635"><a href="#cb90-1635"></a><span class="fu">###### Disease-Free Simulation</span></span>
<span id="cb90-1636"><a href="#cb90-1636"></a></span>
<span id="cb90-1639"><a href="#cb90-1639"></a><span class="in">```{r}</span></span>
<span id="cb90-1640"><a href="#cb90-1640"></a><span class="co"># Set up initial conditions without any infections</span></span>
<span id="cb90-1641"><a href="#cb90-1641"></a>multistage_sonly_yinit <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb90-1642"><a href="#cb90-1642"></a>    <span class="at">S =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">250</span>, <span class="dv">30</span>)),</span>
<span id="cb90-1643"><a href="#cb90-1643"></a>    <span class="at">I =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">30</span>)),</span>
<span id="cb90-1644"><a href="#cb90-1644"></a>    <span class="at">R =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">30</span>))</span>
<span id="cb90-1645"><a href="#cb90-1645"></a>)</span>
<span id="cb90-1646"><a href="#cb90-1646"></a></span>
<span id="cb90-1647"><a href="#cb90-1647"></a><span class="co"># Solve disease free sim to get dfe</span></span>
<span id="cb90-1648"><a href="#cb90-1648"></a>multistage_sonly_sol <span class="ot">&lt;-</span> deSolve<span class="sc">::</span><span class="fu">ode</span>(</span>
<span id="cb90-1649"><a href="#cb90-1649"></a>    <span class="at">y =</span> multistage_sonly_yinit,</span>
<span id="cb90-1650"><a href="#cb90-1650"></a>    <span class="at">times =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">300</span>, <span class="at">by =</span> <span class="dv">1</span>),</span>
<span id="cb90-1651"><a href="#cb90-1651"></a>    <span class="at">func =</span> multistage_model,</span>
<span id="cb90-1652"><a href="#cb90-1652"></a>    <span class="at">parms =</span> multistage_params</span>
<span id="cb90-1653"><a href="#cb90-1653"></a>)</span>
<span id="cb90-1654"><a href="#cb90-1654"></a></span>
<span id="cb90-1655"><a href="#cb90-1655"></a><span class="co"># Calculate population size at each time point and save to dataframe</span></span>
<span id="cb90-1656"><a href="#cb90-1656"></a>multistage_sonly_pop <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb90-1657"><a href="#cb90-1657"></a>    <span class="at">time =</span> multistage_sonly_sol[, <span class="dv">1</span>],</span>
<span id="cb90-1658"><a href="#cb90-1658"></a>    <span class="at">pop =</span> <span class="fu">apply</span>(multistage_sonly_sol[, <span class="sc">-</span><span class="dv">1</span>], <span class="dv">1</span> , sum)</span>
<span id="cb90-1659"><a href="#cb90-1659"></a>)</span>
<span id="cb90-1660"><a href="#cb90-1660"></a><span class="in">```</span></span>
<span id="cb90-1661"><a href="#cb90-1661"></a></span>
<span id="cb90-1664"><a href="#cb90-1664"></a><span class="in">```{r}</span></span>
<span id="cb90-1665"><a href="#cb90-1665"></a><span class="co">#| column: body</span></span>
<span id="cb90-1666"><a href="#cb90-1666"></a><span class="co">#| out-width: 100%</span></span>
<span id="cb90-1667"><a href="#cb90-1667"></a><span class="fu">ggplot</span>(multistage_sonly_pop, <span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> pop)) <span class="sc">+</span></span>
<span id="cb90-1668"><a href="#cb90-1668"></a>    <span class="fu">geom_area</span>(<span class="at">fill =</span> SIRcolors[<span class="dv">4</span>], <span class="at">alpha =</span> <span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb90-1669"><a href="#cb90-1669"></a>    <span class="fu">labs</span>(</span>
<span id="cb90-1670"><a href="#cb90-1670"></a>        <span class="at">x =</span> <span class="st">"Time"</span>,</span>
<span id="cb90-1671"><a href="#cb90-1671"></a>        <span class="at">y =</span> <span class="st">"Population size"</span></span>
<span id="cb90-1672"><a href="#cb90-1672"></a>    )</span>
<span id="cb90-1673"><a href="#cb90-1673"></a><span class="in">```</span></span>
<span id="cb90-1674"><a href="#cb90-1674"></a></span>
<span id="cb90-1675"><a href="#cb90-1675"></a><span class="fu">###### Doing the Math</span></span>
<span id="cb90-1676"><a href="#cb90-1676"></a></span>
<span id="cb90-1677"><a href="#cb90-1677"></a>Alternatively, we can get the stable age distribution by finding the population structure that balances the birth, aging, and death processes.</span>
<span id="cb90-1678"><a href="#cb90-1678"></a>We have already seen the aging matrix in @eq-aging-mat, and at equilibrium, we have the matrix equation</span>
<span id="cb90-1679"><a href="#cb90-1679"></a></span>
<span id="cb90-1680"><a href="#cb90-1680"></a>$$</span>
<span id="cb90-1681"><a href="#cb90-1681"></a>\begin{pmatrix}</span>
<span id="cb90-1682"><a href="#cb90-1682"></a>    -\alpha_1 &amp; 0 &amp; 0 &amp; \cdots &amp; 0<span class="sc">\\</span></span>
<span id="cb90-1683"><a href="#cb90-1683"></a>    \alpha_1 &amp; -\alpha_2 &amp; 0 &amp; \cdots &amp; 0<span class="sc">\\</span></span>
<span id="cb90-1684"><a href="#cb90-1684"></a>    0 &amp; \alpha_2 &amp; -\alpha_3 &amp; \cdots &amp; 0<span class="sc">\\</span></span>
<span id="cb90-1685"><a href="#cb90-1685"></a>    \vdots &amp;  &amp; \ddots &amp; \ddots &amp; \vdots <span class="sc">\\</span></span>
<span id="cb90-1686"><a href="#cb90-1686"></a>    0 &amp; \cdots &amp; &amp; \alpha_{29} &amp; -\alpha_{30}<span class="sc">\\</span></span>
<span id="cb90-1687"><a href="#cb90-1687"></a>\end{pmatrix} .</span>
<span id="cb90-1688"><a href="#cb90-1688"></a>\begin{pmatrix}</span>
<span id="cb90-1689"><a href="#cb90-1689"></a>    n_1 <span class="sc">\\</span> n_2 <span class="sc">\\</span> n_3 <span class="sc">\\</span> \vdots <span class="sc">\\</span> n_{30}</span>
<span id="cb90-1690"><a href="#cb90-1690"></a>\end{pmatrix} +</span>
<span id="cb90-1691"><a href="#cb90-1691"></a>\begin{pmatrix}</span>
<span id="cb90-1692"><a href="#cb90-1692"></a>    B <span class="sc">\\</span> 0 <span class="sc">\\</span> 0 <span class="sc">\\</span> \vdots <span class="sc">\\</span> 0</span>
<span id="cb90-1693"><a href="#cb90-1693"></a>\end{pmatrix} =</span>
<span id="cb90-1694"><a href="#cb90-1694"></a>\begin{pmatrix}</span>
<span id="cb90-1695"><a href="#cb90-1695"></a>    0 <span class="sc">\\</span> 0 <span class="sc">\\</span> 0 <span class="sc">\\</span> \vdots <span class="sc">\\</span> 0</span>
<span id="cb90-1696"><a href="#cb90-1696"></a>\end{pmatrix}</span>
<span id="cb90-1697"><a href="#cb90-1697"></a>$$</span>
<span id="cb90-1698"><a href="#cb90-1698"></a></span>
<span id="cb90-1699"><a href="#cb90-1699"></a>To solve this equation in <span class="in">`R`</span>, we can do</span>
<span id="cb90-1700"><a href="#cb90-1700"></a></span>
<span id="cb90-1703"><a href="#cb90-1703"></a><span class="in">```{r}</span></span>
<span id="cb90-1704"><a href="#cb90-1704"></a><span class="co"># solve(a, b) solves the equation a %*% x = b for x, so rearrange equation above so b is on the RHS of the equation</span></span>
<span id="cb90-1705"><a href="#cb90-1705"></a>multistage_stable_n <span class="ot">&lt;-</span> <span class="fu">solve</span>(</span>
<span id="cb90-1706"><a href="#cb90-1706"></a>    aging_mat,</span>
<span id="cb90-1707"><a href="#cb90-1707"></a>    <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span> <span class="sc">*</span> multistage_params[[<span class="st">"births"</span>]], <span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">29</span>))</span>
<span id="cb90-1708"><a href="#cb90-1708"></a>)</span>
<span id="cb90-1709"><a href="#cb90-1709"></a></span>
<span id="cb90-1710"><a href="#cb90-1710"></a>multistage_stable_n</span>
<span id="cb90-1711"><a href="#cb90-1711"></a><span class="in">```</span></span>
<span id="cb90-1712"><a href="#cb90-1712"></a></span>
<span id="cb90-1715"><a href="#cb90-1715"></a><span class="in">```{r}</span></span>
<span id="cb90-1716"><a href="#cb90-1716"></a><span class="co"># Check the final pop value of the S-only sim is equal to the sum of the stable age distribution calculated above</span></span>
<span id="cb90-1717"><a href="#cb90-1717"></a><span class="fu">round</span>(<span class="fu">tail</span>(multistage_sonly_pop<span class="sc">$</span>pop, <span class="dv">1</span>)) <span class="sc">==</span> <span class="fu">sum</span>(multistage_stable_n)</span>
<span id="cb90-1718"><a href="#cb90-1718"></a><span class="in">```</span></span>
<span id="cb90-1719"><a href="#cb90-1719"></a></span>
<span id="cb90-1720"><a href="#cb90-1720"></a>The following lines then compute $R_0$ using the next generation matrix method.</span>
<span id="cb90-1721"><a href="#cb90-1721"></a>This calculation comes from a recipe described in detail previously <span class="co">[</span><span class="ot">@diekmann2000mathematical; @heesterbeekBriefHistoryR02002; @bjornstadAdvancedNextGenerationMatrix2018; @heffernanPerspectivesBasicReproductive2005; @hurfordNextgenerationToolsEvolutionary2009</span><span class="co">]</span> (we would recommend starting with <span class="co">[</span><span class="ot">@bjornstadAdvancedNextGenerationMatrix2018; and @heffernanPerspectivesBasicReproductive2005</span><span class="co">]</span>).</span>
<span id="cb90-1722"><a href="#cb90-1722"></a></span>
<span id="cb90-1723"><a href="#cb90-1723"></a>::: {.callout-warning}</span>
<span id="cb90-1724"><a href="#cb90-1724"></a>This is Aaron's original code and gives slightly different results than the calculations below.</span>
<span id="cb90-1725"><a href="#cb90-1725"></a>:::</span>
<span id="cb90-1726"><a href="#cb90-1726"></a></span>
<span id="cb90-1729"><a href="#cb90-1729"></a><span class="in">```{r}</span></span>
<span id="cb90-1730"><a href="#cb90-1730"></a>F_mat <span class="ot">&lt;-</span> <span class="fu">diag</span>(multistage_stable_n) <span class="sc">%*%</span> multistage_params[[<span class="st">"beta_mat"</span>]] <span class="sc">+</span></span>
<span id="cb90-1731"><a href="#cb90-1731"></a>    multistage_params[[<span class="st">"aging_mat"</span>]] <span class="sc">-</span></span>
<span id="cb90-1732"><a href="#cb90-1732"></a>    <span class="fu">diag</span>(<span class="fu">diag</span>(multistage_params[[<span class="st">"aging_mat"</span>]]))</span>
<span id="cb90-1733"><a href="#cb90-1733"></a></span>
<span id="cb90-1734"><a href="#cb90-1734"></a>V_mat <span class="ot">&lt;-</span> <span class="fu">diag</span>(multistage_params[[<span class="st">"recovery"</span>]] <span class="sc">-</span> <span class="fu">diag</span>(multistage_params[[<span class="st">"aging_mat"</span>]]))</span>
<span id="cb90-1735"><a href="#cb90-1735"></a></span>
<span id="cb90-1736"><a href="#cb90-1736"></a><span class="fu">max</span>(</span>
<span id="cb90-1737"><a href="#cb90-1737"></a>    <span class="fu">Re</span>(</span>
<span id="cb90-1738"><a href="#cb90-1738"></a>        <span class="fu">eigen</span>(<span class="fu">solve</span>(V_mat, F_mat), <span class="at">only.values =</span> <span class="cn">TRUE</span>)<span class="sc">$</span>values</span>
<span id="cb90-1739"><a href="#cb90-1739"></a>    )</span>
<span id="cb90-1740"><a href="#cb90-1740"></a>)</span>
<span id="cb90-1741"><a href="#cb90-1741"></a><span class="in">```</span></span>
<span id="cb90-1742"><a href="#cb90-1742"></a></span>
<span id="cb90-1743"><a href="#cb90-1743"></a>The steps below are copied from <span class="co">[</span><span class="ot">@bjornstadAdvancedNextGenerationMatrix2018</span><span class="co">]</span></span>
<span id="cb90-1744"><a href="#cb90-1744"></a></span>
<span id="cb90-1745"><a href="#cb90-1745"></a><span class="ss">1. </span>Identify all n infected compartments</span>
<span id="cb90-1746"><a href="#cb90-1746"></a><span class="ss">2. </span>Construct a n × 1 matrix, $\mathbf{F}$, that contains expressions for all completely new infections entering each infected compartment</span>
<span id="cb90-1747"><a href="#cb90-1747"></a><span class="ss">3. </span>Construct a n × 1 matrix, $\mathbf{V^−}$, that contains expressions for all losses out of each infected compartment</span>
<span id="cb90-1748"><a href="#cb90-1748"></a><span class="ss">4. </span>Construct a n × 1 matrix, $\mathbf{V^+}$, that contains expressions for all gains into each infected compartment that does not represent new infections but transfers among infectious classes</span>
<span id="cb90-1749"><a href="#cb90-1749"></a><span class="ss">5. </span>Construct a n × 1 matrix, $\mathbf{V} = \mathbf{V^−} − \mathbf{V^+}$</span>
<span id="cb90-1750"><a href="#cb90-1750"></a><span class="ss">6. </span>Generate two n × n Jacobian matrices $f$ and $v$ that are the partial derivatives of $\mathbf{F}$ and $\mathbf{V}$ with respect to the $n$ infectious state variables</span>
<span id="cb90-1751"><a href="#cb90-1751"></a><span class="ss">7. </span>Evaluate the matrices at the disease free equilibrium (dfe), and finally</span>
<span id="cb90-1752"><a href="#cb90-1752"></a><span class="ss">8. </span>$R_0$ is the spectral trace (greatest non-negative real eigenvalue) of $\mathbf{fv}^{−1}|_{\text{dfe}}$.</span>
<span id="cb90-1753"><a href="#cb90-1753"></a></span>
<span id="cb90-1754"><a href="#cb90-1754"></a>Working through these steps looks like this:</span>
<span id="cb90-1755"><a href="#cb90-1755"></a></span>
<span id="cb90-1756"><a href="#cb90-1756"></a><span class="ss">1. </span>Our only infected compartments are the $I_i$ states, for each age group ($i \in <span class="co">[</span><span class="ot">1, 30</span><span class="co">]</span>$)</span>
<span id="cb90-1757"><a href="#cb90-1757"></a>To start, let's write out our differential equation:</span>
<span id="cb90-1758"><a href="#cb90-1758"></a></span>
<span id="cb90-1759"><a href="#cb90-1759"></a>\begin{equation}</span>
<span id="cb90-1760"><a href="#cb90-1760"></a>    \frac{\dd{I_i}}{\dd{t}} = \lambda_i S_i - \gamma I_i + \alpha_{i-1} I_{i-1} - \alpha_i I_i</span>
<span id="cb90-1761"><a href="#cb90-1761"></a>\end{equation}</span>
<span id="cb90-1762"><a href="#cb90-1762"></a></span>
<span id="cb90-1763"><a href="#cb90-1763"></a></span>
<span id="cb90-1764"><a href="#cb90-1764"></a></span>
<span id="cb90-1765"><a href="#cb90-1765"></a><span class="ss">2. </span>We'll now calculate $\mathbf{F}$ and $\mathbf{f}$</span>
<span id="cb90-1766"><a href="#cb90-1766"></a></span>
<span id="cb90-1767"><a href="#cb90-1767"></a></span>
<span id="cb90-1768"><a href="#cb90-1768"></a>\begin{align*}</span>
<span id="cb90-1769"><a href="#cb90-1769"></a>    \mathbf{F} &amp;= \begin{pmatrix}</span>
<span id="cb90-1770"><a href="#cb90-1770"></a>        \lambda_1 S_1 + \cancelto{0}{\alpha_0 I_0} <span class="sc">\\</span></span>
<span id="cb90-1771"><a href="#cb90-1771"></a>        \vdots <span class="sc">\\</span></span>
<span id="cb90-1772"><a href="#cb90-1772"></a>        \lambda_{30} S_{30} + \alpha_{29} I_{29}</span>
<span id="cb90-1773"><a href="#cb90-1773"></a>    \end{pmatrix} <span class="sc">\\</span> <span class="sc">\\</span></span>
<span id="cb90-1774"><a href="#cb90-1774"></a>    \mathbf{F} &amp;= \begin{pmatrix}</span>
<span id="cb90-1775"><a href="#cb90-1775"></a>        \left(\beta_{1, 1} I_1 + \cdots + \beta_{1, 30} I_{30} \right)S_1 <span class="sc">\\</span></span>
<span id="cb90-1776"><a href="#cb90-1776"></a>        \vdots <span class="sc">\\</span></span>
<span id="cb90-1777"><a href="#cb90-1777"></a>        \left(\beta_{30, 1} I_1 + \cdots + \beta_{30, 30} I_{30} \right) S_{30} + \alpha_{29} I_{29}</span>
<span id="cb90-1778"><a href="#cb90-1778"></a>    \end{pmatrix} <span class="sc">\\</span></span>
<span id="cb90-1779"><a href="#cb90-1779"></a>    \mathbf{f} &amp;= \begin{pmatrix}</span>
<span id="cb90-1780"><a href="#cb90-1780"></a>        \frac{\partial F_1}{\partial I_1} &amp; \cdots &amp; \frac{\partial F_1}{\partial I_{30}} <span class="sc">\\</span></span>
<span id="cb90-1781"><a href="#cb90-1781"></a>        \vdots &amp; \ddots &amp; \vdots <span class="sc">\\</span></span>
<span id="cb90-1782"><a href="#cb90-1782"></a>        \frac{\partial F_{30}}{\partial I_1} &amp; \cdots &amp; \frac{\partial F_{30}}{\partial I_{30}}</span>
<span id="cb90-1783"><a href="#cb90-1783"></a>    \end{pmatrix} &amp; \frac{\partial F_1}{\partial I_1} &amp;= \frac{\partial}{\partial I_1} \left( \beta_{1, 1} I_1 + \cancelto{0}{\beta_{1, 2} I_2 + \cdots + \beta_{1, 30} I_{30}}\right) S_1 <span class="sc">\\</span></span>
<span id="cb90-1784"><a href="#cb90-1784"></a>    &amp; &amp; \frac{\partial F_1}{\partial I_1} &amp;= \beta_{1, 1} S_1 <span class="sc">\\</span></span>
<span id="cb90-1785"><a href="#cb90-1785"></a>    \mathbf{f} &amp;= \begin{pmatrix}</span>
<span id="cb90-1786"><a href="#cb90-1786"></a>        \beta_{1, 1} S_1 &amp; \cdots &amp; \beta_{1, 30} S_1 <span class="sc">\\</span></span>
<span id="cb90-1787"><a href="#cb90-1787"></a>        \vdots &amp; \ddots &amp; \vdots <span class="sc">\\</span></span>
<span id="cb90-1788"><a href="#cb90-1788"></a>        \beta_{30, 1} S_{30} &amp; \cdots &amp; \beta_{30, 30} S_{30}</span>
<span id="cb90-1789"><a href="#cb90-1789"></a>    \end{pmatrix}</span>
<span id="cb90-1790"><a href="#cb90-1790"></a>\end{align*}</span>
<span id="cb90-1791"><a href="#cb90-1791"></a></span>
<span id="cb90-1792"><a href="#cb90-1792"></a><span class="ss">3. </span>Now let's calculate $\mathbf{V^-}$, $\mathbf{V^+}$, $\mathbf{V}$, and $\mathbf{v}$</span>
<span id="cb90-1793"><a href="#cb90-1793"></a></span>
<span id="cb90-1794"><a href="#cb90-1794"></a>\begin{align*}</span>
<span id="cb90-1795"><a href="#cb90-1795"></a>    \mathbf{V^-} &amp;= \begin{pmatrix}</span>
<span id="cb90-1796"><a href="#cb90-1796"></a>        \gamma I_1 + \alpha_1 I_1 <span class="sc">\\</span></span>
<span id="cb90-1797"><a href="#cb90-1797"></a>        \gamma I_2 + \alpha_2 I_2 <span class="sc">\\</span></span>
<span id="cb90-1798"><a href="#cb90-1798"></a>        \vdots <span class="sc">\\</span></span>
<span id="cb90-1799"><a href="#cb90-1799"></a>        \gamma I_{30} + \alpha_{30} I_{30}</span>
<span id="cb90-1800"><a href="#cb90-1800"></a>    \end{pmatrix} &amp; \mathbf{V^+} &amp;= \begin{pmatrix}</span>
<span id="cb90-1801"><a href="#cb90-1801"></a>        \cancelto{0}{\alpha_0 I_0} <span class="sc">\\</span></span>
<span id="cb90-1802"><a href="#cb90-1802"></a>        \alpha_1 I_1 <span class="sc">\\</span></span>
<span id="cb90-1803"><a href="#cb90-1803"></a>        \vdots <span class="sc">\\</span></span>
<span id="cb90-1804"><a href="#cb90-1804"></a>        \alpha_{29} I_{29}</span>
<span id="cb90-1805"><a href="#cb90-1805"></a>    \end{pmatrix} <span class="sc">\\</span> <span class="sc">\\</span></span>
<span id="cb90-1806"><a href="#cb90-1806"></a>    \mathbf{V} &amp;= \mathbf{V^-} - \mathbf{V^+} = \begin{pmatrix}</span>
<span id="cb90-1807"><a href="#cb90-1807"></a>        \gamma I_1 + \alpha_1 I_1 <span class="sc">\\</span></span>
<span id="cb90-1808"><a href="#cb90-1808"></a>        \gamma I_2 + \alpha_2 I_2 - \alpha_1 I_1<span class="sc">\\</span></span>
<span id="cb90-1809"><a href="#cb90-1809"></a>        \vdots <span class="sc">\\</span></span>
<span id="cb90-1810"><a href="#cb90-1810"></a>        \gamma I_{30} + \alpha_{30} I_{30} - \alpha_{29} I_{29}</span>
<span id="cb90-1811"><a href="#cb90-1811"></a>    \end{pmatrix} <span class="sc">\\</span> <span class="sc">\\</span></span>
<span id="cb90-1812"><a href="#cb90-1812"></a>    \mathbf{v} &amp;= \begin{pmatrix}</span>
<span id="cb90-1813"><a href="#cb90-1813"></a>        \pdv{V_1}{I_1} &amp; \cdots &amp; \pdv{V_1}{I_{30}} <span class="sc">\\</span></span>
<span id="cb90-1814"><a href="#cb90-1814"></a>        \vdots &amp; \ddots &amp; \vdots <span class="sc">\\</span></span>
<span id="cb90-1815"><a href="#cb90-1815"></a>        \pdv{V_{30}}{I_1} &amp; \cdots &amp; \pdv{V_{30}}{I_{30}}</span>
<span id="cb90-1816"><a href="#cb90-1816"></a>    \end{pmatrix} &amp; \pdv{V_1}{I_1} &amp;= \pdv{I_1} I_1 \left( \gamma + \alpha_1  \right)<span class="sc">\\</span></span>
<span id="cb90-1817"><a href="#cb90-1817"></a>    &amp; &amp; \pdv{V_1}{I_1} &amp;= \gamma + \alpha_1 <span class="sc">\\</span> <span class="sc">\\</span></span>
<span id="cb90-1818"><a href="#cb90-1818"></a>    &amp; &amp; \pdv{V_2}{I_1} &amp;= \pdv{I_1} \left( \cancelto{0}{I_2 \left( \gamma + \alpha_2  \right)} - \alpha_1 I_1 \right)<span class="sc">\\</span></span>
<span id="cb90-1819"><a href="#cb90-1819"></a>    &amp; &amp; \pdv{V_2}{I_1} &amp;= - \alpha_1 <span class="sc">\\</span> <span class="sc">\\</span></span>
<span id="cb90-1820"><a href="#cb90-1820"></a>    &amp; &amp; \pdv{V_1}{I_2} &amp;= \pdv{I_2} \cancelto{0}{I_1 \left( \gamma + \alpha_1  \right)} <span class="sc">\\</span></span>
<span id="cb90-1821"><a href="#cb90-1821"></a>    &amp; &amp; \pdv{V_1}{I_2} &amp;= 0 <span class="sc">\\</span> <span class="sc">\\</span></span>
<span id="cb90-1822"><a href="#cb90-1822"></a>    \mathbf{v} &amp;= \begin{pmatrix}</span>
<span id="cb90-1823"><a href="#cb90-1823"></a>        \gamma + \alpha_1 &amp; 0 &amp;  \cdots  &amp; 0<span class="sc">\\</span></span>
<span id="cb90-1824"><a href="#cb90-1824"></a><span class="ss">        - </span>\alpha_1 &amp; \gamma + \alpha_2 &amp; \cdots &amp; 0 <span class="sc">\\</span></span>
<span id="cb90-1825"><a href="#cb90-1825"></a>        \vdots &amp; \ddots &amp; \ddots &amp; \vdots <span class="sc">\\</span></span>
<span id="cb90-1826"><a href="#cb90-1826"></a>        0 &amp; \cdots &amp; - \alpha_{29} &amp; \gamma + \alpha_{30}</span>
<span id="cb90-1827"><a href="#cb90-1827"></a>    \end{pmatrix} </span>
<span id="cb90-1828"><a href="#cb90-1828"></a>\end{align*}</span>
<span id="cb90-1829"><a href="#cb90-1829"></a></span>
<span id="cb90-1830"><a href="#cb90-1830"></a><span class="ss">4. </span>To evaluate $\mathbf{f}$ and $\mathbf{v}$ at the disease-free equilibrium, we can use the results from our previous calculations.</span>
<span id="cb90-1831"><a href="#cb90-1831"></a>$\mathbf{v}$ doesn't have any state terms in the equation, so it is already evaluated at $\text{dfe}$.</span>
<span id="cb90-1832"><a href="#cb90-1832"></a>$\mathbf{f}|_{\text{dfe}}$ involves subsituting $S_i$ for the equilibrium population distribution that balances the births and aging processes.</span>
<span id="cb90-1833"><a href="#cb90-1833"></a></span>
<span id="cb90-1834"><a href="#cb90-1834"></a>This translates to the function we <span class="co">[</span><span class="ot">defined earlier</span><span class="co">](#ngm-function)</span></span>
<span id="cb90-1835"><a href="#cb90-1835"></a></span>
<span id="cb90-1836"><a href="#cb90-1836"></a><span class="fu">### Mean age of infection `R` code {#sec-mean-age-r-code}</span></span>
<span id="cb90-1837"><a href="#cb90-1837"></a></span>
<span id="cb90-1838"><a href="#cb90-1838"></a>Now let's look at how we can investigate our the relationships between the mean age of infection and $R_0$ and the vaccination coverage using <span class="in">`R`</span>.</span>
<span id="cb90-1839"><a href="#cb90-1839"></a>Unlike the interactive plot that simply uses @eq-mean-age to calculate the mean age of infection, we will use a more realistic age-structured model.</span>
<span id="cb90-1840"><a href="#cb90-1840"></a></span>
<span id="cb90-1841"><a href="#cb90-1841"></a>Let's return to the earlier models with an age-class mixing matrix.</span>
<span id="cb90-1842"><a href="#cb90-1842"></a>But this time, we'll calculate $R_0$, the mean age of infection, and the number of cases that occur in individuals between 15-35 years as we increase the contact rate.</span>
<span id="cb90-1843"><a href="#cb90-1843"></a></span>
<span id="cb90-1844"><a href="#cb90-1844"></a>Recall from the rubella and congenital rubella syndrome (CRS) example that the risk of severe disease outcomes depends on the risk of infection in reproductive age women (here we'll use individuals between 15 and 35 years as a proxy; in reality we would want to account for the differential rate of reproduction at different ages, including those above 35 years).</span>
<span id="cb90-1845"><a href="#cb90-1845"></a>Recall also that increasing vaccination reduces $R_E$ -- for simplicity here, so we don't have to add vaccination into the code, we'll simply change $R_0$ because we already know that will give us outcomes that are dynamically equivalent to increasing the proportion of children born who are vaccinated.</span>
<span id="cb90-1846"><a href="#cb90-1846"></a>We'll then calculate how the mean age of infection changes, and specifically how the absolute number of cases among individuals between the ages of 15-35 (as a proxy for reproductive age women) changes. </span>
<span id="cb90-1847"><a href="#cb90-1847"></a>To do so, we'll make a loop and evaluate the code for each of 10 decreaing levels of mixing (which will reduce $R_0$ and we can interpret as analogous to the reduction in $R_E$ that would result from increasing vaccination).</span>
<span id="cb90-1848"><a href="#cb90-1848"></a></span>
<span id="cb90-1849"><a href="#cb90-1849"></a>::: {.callout-note title="Note about <span class="in">`map()`</span>" collapse="false"}</span>
<span id="cb90-1850"><a href="#cb90-1850"></a>As you may have noticed previously, we often use the <span class="in">`map_*()`</span> series of functions.</span>
<span id="cb90-1851"><a href="#cb90-1851"></a>We'll use that again here (<span class="in">`map_dfr()`</span>).</span>
<span id="cb90-1852"><a href="#cb90-1852"></a>The full reasons are too complicated to get into here, but broadly speaking, the <span class="in">`map_*()`</span> functions provide us guarantees over the output of our loops.</span>
<span id="cb90-1853"><a href="#cb90-1853"></a>If it runs, we know that something didn't get silently skipped, and that out output vector/list/dataframe is the same length as the inputs.</span>
<span id="cb90-1854"><a href="#cb90-1854"></a>The same can not be said for <span class="in">`for()`</span> loops, and the base <span class="in">`apply`</span> functions are more awkward to work with as they don't have a consistent syntax across the family of functions.</span>
<span id="cb90-1855"><a href="#cb90-1855"></a></span>
<span id="cb90-1856"><a href="#cb90-1856"></a>To learn more, read <span class="co">[</span><span class="ot">this section</span><span class="co">]()</span> of our <span class="in">`R`</span> primer.</span>
<span id="cb90-1857"><a href="#cb90-1857"></a>:::</span>
<span id="cb90-1858"><a href="#cb90-1858"></a></span>
<span id="cb90-1861"><a href="#cb90-1861"></a><span class="in">```{r}</span></span>
<span id="cb90-1862"><a href="#cb90-1862"></a><span class="co"># Create vector of scalings to reduce R0</span></span>
<span id="cb90-1863"><a href="#cb90-1863"></a>scale_contact <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">1</span>, .<span class="dv">2</span>, <span class="at">length =</span> <span class="dv">10</span>)</span>
<span id="cb90-1864"><a href="#cb90-1864"></a></span>
<span id="cb90-1865"><a href="#cb90-1865"></a><span class="co"># Create a new transmission matrix</span></span>
<span id="cb90-1866"><a href="#cb90-1866"></a>beta_low <span class="ot">&lt;-</span> <span class="fl">0.007</span></span>
<span id="cb90-1867"><a href="#cb90-1867"></a>beta_medium <span class="ot">&lt;-</span> <span class="fl">0.02</span></span>
<span id="cb90-1868"><a href="#cb90-1868"></a>beta_high <span class="ot">&lt;-</span> <span class="fl">0.03</span></span>
<span id="cb90-1869"><a href="#cb90-1869"></a></span>
<span id="cb90-1870"><a href="#cb90-1870"></a>beta_mat <span class="ot">&lt;-</span> <span class="fu">matrix</span>(beta_low, <span class="at">nrow =</span> <span class="dv">30</span>, <span class="at">ncol =</span> <span class="dv">30</span>)</span>
<span id="cb90-1871"><a href="#cb90-1871"></a>beta_mat[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>] <span class="ot">&lt;-</span> beta_medium</span>
<span id="cb90-1872"><a href="#cb90-1872"></a>beta_mat[<span class="dv">6</span><span class="sc">:</span><span class="dv">16</span>, <span class="dv">6</span><span class="sc">:</span><span class="dv">16</span>] <span class="ot">&lt;-</span> beta_high</span>
<span id="cb90-1873"><a href="#cb90-1873"></a></span>
<span id="cb90-1874"><a href="#cb90-1874"></a>scaled_params <span class="ot">&lt;-</span> multistage_params</span>
<span id="cb90-1875"><a href="#cb90-1875"></a><span class="in">```</span></span>
<span id="cb90-1876"><a href="#cb90-1876"></a></span>
<span id="cb90-1879"><a href="#cb90-1879"></a><span class="in">```{r}</span></span>
<span id="cb90-1880"><a href="#cb90-1880"></a><span class="co"># Create a dataframe where each row relates to a different R0 value</span></span>
<span id="cb90-1881"><a href="#cb90-1881"></a>R0_mean_age_contacts_df <span class="ot">&lt;-</span> <span class="fu">map_dfr</span>(</span>
<span id="cb90-1882"><a href="#cb90-1882"></a>    <span class="co"># Apply the function to each item in the vector of R0 scaling factors</span></span>
<span id="cb90-1883"><a href="#cb90-1883"></a>    <span class="at">.x =</span> scale_contact,</span>
<span id="cb90-1884"><a href="#cb90-1884"></a>    <span class="at">.f =</span> <span class="cf">function</span>(.x) {</span>
<span id="cb90-1885"><a href="#cb90-1885"></a>        <span class="co"># Scale contacts</span></span>
<span id="cb90-1886"><a href="#cb90-1886"></a>        scaled_beta_mat <span class="ot">&lt;-</span> beta_mat <span class="sc">*</span> .x</span>
<span id="cb90-1887"><a href="#cb90-1887"></a></span>
<span id="cb90-1888"><a href="#cb90-1888"></a>        <span class="co"># Set up parameters</span></span>
<span id="cb90-1889"><a href="#cb90-1889"></a>        scaled_params[[<span class="st">"beta_mat"</span>]] <span class="ot">&lt;-</span> scaled_beta_mat</span>
<span id="cb90-1890"><a href="#cb90-1890"></a></span>
<span id="cb90-1891"><a href="#cb90-1891"></a>        <span class="co"># Solve the model</span></span>
<span id="cb90-1892"><a href="#cb90-1892"></a>        sol <span class="ot">&lt;-</span> deSolve<span class="sc">::</span><span class="fu">ode</span>(</span>
<span id="cb90-1893"><a href="#cb90-1893"></a>            <span class="at">y =</span> demog_yinit_ages,</span>
<span id="cb90-1894"><a href="#cb90-1894"></a>            <span class="at">times =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">400</span>, <span class="at">by =</span> <span class="fl">0.1</span>),</span>
<span id="cb90-1895"><a href="#cb90-1895"></a>            <span class="at">func =</span> multistage_model,</span>
<span id="cb90-1896"><a href="#cb90-1896"></a>            <span class="at">parms =</span> scaled_params</span>
<span id="cb90-1897"><a href="#cb90-1897"></a>        )</span>
<span id="cb90-1898"><a href="#cb90-1898"></a></span>
<span id="cb90-1899"><a href="#cb90-1899"></a>        <span class="co"># Get stable age distribution</span></span>
<span id="cb90-1900"><a href="#cb90-1900"></a>        stable_n <span class="ot">&lt;-</span> <span class="fu">solve</span>(</span>
<span id="cb90-1901"><a href="#cb90-1901"></a>            scaled_params[[<span class="st">"aging_mat"</span>]],</span>
<span id="cb90-1902"><a href="#cb90-1902"></a>            <span class="sc">-</span><span class="fu">c</span>(scaled_params[[<span class="st">"births"</span>]], <span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">29</span>))</span>
<span id="cb90-1903"><a href="#cb90-1903"></a>        )</span>
<span id="cb90-1904"><a href="#cb90-1904"></a></span>
<span id="cb90-1905"><a href="#cb90-1905"></a>        R0 <span class="ot">&lt;-</span> <span class="fu">calculate_R0</span>(</span>
<span id="cb90-1906"><a href="#cb90-1906"></a>            <span class="at">beta_mat =</span> scaled_params[[<span class="st">"beta_mat"</span>]],</span>
<span id="cb90-1907"><a href="#cb90-1907"></a>            <span class="at">stable_n_mat =</span> stable_n,</span>
<span id="cb90-1908"><a href="#cb90-1908"></a>            <span class="at">aging_mat =</span> scaled_params[[<span class="st">"aging_mat"</span>]],</span>
<span id="cb90-1909"><a href="#cb90-1909"></a>            <span class="at">recovery =</span> scaled_params[[<span class="st">"recovery"</span>]]</span>
<span id="cb90-1910"><a href="#cb90-1910"></a>        )</span>
<span id="cb90-1911"><a href="#cb90-1911"></a></span>
<span id="cb90-1912"><a href="#cb90-1912"></a>        <span class="co"># Get final number of infected individuals for each I class</span></span>
<span id="cb90-1913"><a href="#cb90-1913"></a>        infecteds <span class="ot">&lt;-</span> sol[<span class="fu">dim</span>(sol)[<span class="dv">1</span>], <span class="dv">1</span> <span class="sc">+</span> iindex]</span>
<span id="cb90-1914"><a href="#cb90-1914"></a></span>
<span id="cb90-1915"><a href="#cb90-1915"></a>        <span class="co"># Calculate mean age of infection</span></span>
<span id="cb90-1916"><a href="#cb90-1916"></a>        mean_age <span class="ot">&lt;-</span> <span class="fu">sum</span>(ages <span class="sc">*</span> infecteds <span class="sc">/</span> <span class="fu">sum</span>(infecteds))</span>
<span id="cb90-1917"><a href="#cb90-1917"></a></span>
<span id="cb90-1918"><a href="#cb90-1918"></a>        <span class="co"># Calculate sum of cases between 15-35 years </span></span>
<span id="cb90-1919"><a href="#cb90-1919"></a>        <span class="co"># recall, from the figures above, that that this is the equilibrium prevalence</span></span>
<span id="cb90-1920"><a href="#cb90-1920"></a>        <span class="co"># of infection in these age classes, or the average numbef of individuals that are infected </span></span>
<span id="cb90-1921"><a href="#cb90-1921"></a>        <span class="co"># at any given time in these age classes. Note that we're not differentiating between </span></span>
<span id="cb90-1922"><a href="#cb90-1922"></a>        <span class="co"># individuals who can and cannot get pregnant here. So we're making an implicit assumption that</span></span>
<span id="cb90-1923"><a href="#cb90-1923"></a>        <span class="co"># there no difference in the risk of rubella infection in these groups so that if prevalence goes</span></span>
<span id="cb90-1924"><a href="#cb90-1924"></a>        <span class="co"># up in one group, it goes up in the other.</span></span>
<span id="cb90-1925"><a href="#cb90-1925"></a>        sum_cases <span class="ot">&lt;-</span> <span class="fu">sum</span>(infecteds[<span class="dv">15</span><span class="sc">:</span><span class="dv">23</span>])</span>
<span id="cb90-1926"><a href="#cb90-1926"></a></span>
<span id="cb90-1927"><a href="#cb90-1927"></a>        <span class="co"># Return a dataframe with the values</span></span>
<span id="cb90-1928"><a href="#cb90-1928"></a>        <span class="fu">return</span>(<span class="fu">tibble</span>(R0, mean_age, sum_cases))</span>
<span id="cb90-1929"><a href="#cb90-1929"></a>    }</span>
<span id="cb90-1930"><a href="#cb90-1930"></a>)</span>
<span id="cb90-1931"><a href="#cb90-1931"></a><span class="in">```</span></span>
<span id="cb90-1932"><a href="#cb90-1932"></a></span>
<span id="cb90-1933"><a href="#cb90-1933"></a>Now we can make a table of the results and plot mean age and the sum of cases between 15-35 years of age as a function of $R_0$.</span>
<span id="cb90-1934"><a href="#cb90-1934"></a></span>
<span id="cb90-1937"><a href="#cb90-1937"></a><span class="in">```{r}</span></span>
<span id="cb90-1938"><a href="#cb90-1938"></a><span class="co">#| column: body</span></span>
<span id="cb90-1939"><a href="#cb90-1939"></a><span class="co">#| out-width: 100%</span></span>
<span id="cb90-1940"><a href="#cb90-1940"></a><span class="co"># Create a table from the dataframe</span></span>
<span id="cb90-1941"><a href="#cb90-1941"></a><span class="fu">gt</span>(R0_mean_age_contacts_df) <span class="sc">%&gt;%</span></span>
<span id="cb90-1942"><a href="#cb90-1942"></a>    <span class="fu">fmt_number</span>(</span>
<span id="cb90-1943"><a href="#cb90-1943"></a>        <span class="at">columns =</span> <span class="fu">everything</span>(),</span>
<span id="cb90-1944"><a href="#cb90-1944"></a>        <span class="at">decimals =</span> <span class="dv">2</span></span>
<span id="cb90-1945"><a href="#cb90-1945"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb90-1946"><a href="#cb90-1946"></a>    <span class="co"># Relabel the column headers</span></span>
<span id="cb90-1947"><a href="#cb90-1947"></a>    <span class="fu">cols_label</span>(</span>
<span id="cb90-1948"><a href="#cb90-1948"></a>        <span class="at">R0 =</span> <span class="fu">md</span>(<span class="st">"**R0**"</span>),</span>
<span id="cb90-1949"><a href="#cb90-1949"></a>        <span class="at">mean_age =</span> <span class="fu">md</span>(<span class="st">"**Mean age of&lt;br&gt;infection**"</span>),</span>
<span id="cb90-1950"><a href="#cb90-1950"></a>        <span class="at">sum_cases =</span> <span class="fu">md</span>(<span class="st">"**Total cases between&lt;br&gt;15-35 years**"</span>)</span>
<span id="cb90-1951"><a href="#cb90-1951"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb90-1952"><a href="#cb90-1952"></a>    <span class="co"># Apply style to the table with gray alternating rows</span></span>
<span id="cb90-1953"><a href="#cb90-1953"></a>    <span class="fu">opt_stylize</span>(<span class="at">style =</span> <span class="dv">1</span>, <span class="at">color =</span> <span class="st">'gray'</span>) <span class="sc">%&gt;%</span></span>
<span id="cb90-1954"><a href="#cb90-1954"></a>    <span class="co"># Increate space between columns</span></span>
<span id="cb90-1955"><a href="#cb90-1955"></a>    <span class="fu">opt_horizontal_padding</span>(<span class="at">scale =</span> <span class="dv">3</span>) <span class="sc">%&gt;%</span></span>
<span id="cb90-1956"><a href="#cb90-1956"></a>    <span class="fu">cols_align</span>(<span class="st">"center"</span>)</span>
<span id="cb90-1957"><a href="#cb90-1957"></a><span class="in">```</span></span>
<span id="cb90-1958"><a href="#cb90-1958"></a></span>
<span id="cb90-1961"><a href="#cb90-1961"></a><span class="in">```{r}</span></span>
<span id="cb90-1962"><a href="#cb90-1962"></a><span class="co">#| column: body</span></span>
<span id="cb90-1963"><a href="#cb90-1963"></a><span class="co">#| out-width: 100%</span></span>
<span id="cb90-1964"><a href="#cb90-1964"></a>R0_mean_age_contacts_df <span class="sc">%&gt;%</span></span>
<span id="cb90-1965"><a href="#cb90-1965"></a>    <span class="co"># Convert to long data frame for facet plotting</span></span>
<span id="cb90-1966"><a href="#cb90-1966"></a>    <span class="fu">pivot_longer</span>(<span class="sc">-</span>R0, <span class="at">names_to =</span> <span class="st">"metric"</span>, <span class="at">values_to =</span> <span class="st">"value"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb90-1967"><a href="#cb90-1967"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> R0, <span class="at">y =</span> value)) <span class="sc">+</span></span>
<span id="cb90-1968"><a href="#cb90-1968"></a>    <span class="fu">geom_line</span>(<span class="at">color =</span> <span class="st">"slategray4"</span>) <span class="sc">+</span></span>
<span id="cb90-1969"><a href="#cb90-1969"></a>    <span class="fu">geom_point</span>(<span class="at">shape =</span> <span class="dv">21</span>, <span class="at">size =</span> <span class="dv">5</span>, <span class="at">fill =</span> <span class="st">"slategray4"</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>) <span class="sc">+</span></span>
<span id="cb90-1970"><a href="#cb90-1970"></a>    <span class="fu">facet_wrap</span>(</span>
<span id="cb90-1971"><a href="#cb90-1971"></a>        <span class="sc">~</span>metric,</span>
<span id="cb90-1972"><a href="#cb90-1972"></a>        <span class="at">scales =</span> <span class="st">"free_y"</span>,</span>
<span id="cb90-1973"><a href="#cb90-1973"></a>        <span class="at">labeller =</span> <span class="fu">as_labeller</span>(<span class="fu">c</span>(</span>
<span id="cb90-1974"><a href="#cb90-1974"></a>            <span class="at">mean_age =</span> <span class="st">"Mean Age of Infection"</span>,</span>
<span id="cb90-1975"><a href="#cb90-1975"></a>            <span class="at">sum_cases =</span> <span class="st">"Total cases between 15-35 years"</span></span>
<span id="cb90-1976"><a href="#cb90-1976"></a>        ))</span>
<span id="cb90-1977"><a href="#cb90-1977"></a>    ) <span class="sc">+</span></span>
<span id="cb90-1978"><a href="#cb90-1978"></a>    <span class="fu">labs</span>(</span>
<span id="cb90-1979"><a href="#cb90-1979"></a>        <span class="at">x =</span> <span class="st">"R0"</span>,</span>
<span id="cb90-1980"><a href="#cb90-1980"></a>        <span class="at">y =</span> <span class="st">"Value"</span></span>
<span id="cb90-1981"><a href="#cb90-1981"></a>    )</span>
<span id="cb90-1982"><a href="#cb90-1982"></a><span class="in">```</span></span>
<span id="cb90-1983"><a href="#cb90-1983"></a></span>
<span id="cb90-1984"><a href="#cb90-1984"></a><span class="fu">### What do real contact networks look like?</span></span>
<span id="cb90-1985"><a href="#cb90-1985"></a></span>
<span id="cb90-1986"><a href="#cb90-1986"></a>The POLYMOD study <span class="co">[</span><span class="ot">@mossongSocialContactsMixing2008a</span><span class="co">]</span> was a journal-based look into the contact network in contemporary European society.</span>
<span id="cb90-1987"><a href="#cb90-1987"></a>Let's have a look what these data tell us about the contact structure.</span>
<span id="cb90-1988"><a href="#cb90-1988"></a></span>
<span id="cb90-1991"><a href="#cb90-1991"></a><span class="in">```{r}</span></span>
<span id="cb90-1992"><a href="#cb90-1992"></a>mossong_cont_net <span class="ot">&lt;-</span> rio<span class="sc">::</span><span class="fu">import</span>(<span class="st">"https://raw.githubusercontent.com/arnold-c/SISMID-Module-02_2023/main/data/mossong-matrix.csv"</span>)</span>
<span id="cb90-1993"><a href="#cb90-1993"></a><span class="co"># mossong_cont_net &lt;- rio::import(here::here("data", "mossong-matrix.csv"))</span></span>
<span id="cb90-1994"><a href="#cb90-1994"></a></span>
<span id="cb90-1995"><a href="#cb90-1995"></a>mossong_ages <span class="ot">&lt;-</span> <span class="fu">unique</span>(mossong_cont_net<span class="sc">$</span>contactor)</span>
<span id="cb90-1996"><a href="#cb90-1996"></a>mossong_cont_net<span class="sc">$</span>contactor <span class="ot">&lt;-</span> <span class="fu">ordered</span>(mossong_cont_net<span class="sc">$</span>contactor, <span class="at">levels =</span> mossong_ages)</span>
<span id="cb90-1997"><a href="#cb90-1997"></a>mossong_cont_net<span class="sc">$</span>contactee <span class="ot">&lt;-</span> <span class="fu">ordered</span>(mossong_cont_net<span class="sc">$</span>contactee, <span class="at">levels =</span> mossong_ages)</span>
<span id="cb90-1998"><a href="#cb90-1998"></a><span class="in">```</span></span>
<span id="cb90-1999"><a href="#cb90-1999"></a></span>
<span id="cb90-2000"><a href="#cb90-2000"></a>Since contacts are symmetric, we'll need to estimate the symmetric contact matrix.</span>
<span id="cb90-2001"><a href="#cb90-2001"></a></span>
<span id="cb90-2004"><a href="#cb90-2004"></a><span class="in">```{r}</span></span>
<span id="cb90-2005"><a href="#cb90-2005"></a>mossong_mat <span class="ot">&lt;-</span> mossong_cont_net <span class="sc">%&gt;%</span></span>
<span id="cb90-2006"><a href="#cb90-2006"></a>    <span class="fu">pivot_wider</span>(</span>
<span id="cb90-2007"><a href="#cb90-2007"></a>        <span class="at">names_from =</span> contactor,</span>
<span id="cb90-2008"><a href="#cb90-2008"></a>        <span class="at">values_from =</span> contact.rate</span>
<span id="cb90-2009"><a href="#cb90-2009"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb90-2010"><a href="#cb90-2010"></a>    <span class="fu">select</span>(<span class="sc">-</span>contactee) <span class="sc">%&gt;%</span></span>
<span id="cb90-2011"><a href="#cb90-2011"></a>    <span class="fu">as.matrix</span>()</span>
<span id="cb90-2012"><a href="#cb90-2012"></a></span>
<span id="cb90-2013"><a href="#cb90-2013"></a><span class="fu">rownames</span>(mossong_mat) <span class="ot">&lt;-</span> mossong_ages</span>
<span id="cb90-2014"><a href="#cb90-2014"></a></span>
<span id="cb90-2015"><a href="#cb90-2015"></a><span class="co"># Create a symmetrical contact matrix</span></span>
<span id="cb90-2016"><a href="#cb90-2016"></a>mossong_mat_sym <span class="ot">&lt;-</span> (mossong_mat <span class="sc">+</span> <span class="fu">t</span>(mossong_mat)) <span class="sc">/</span> <span class="dv">2</span></span>
<span id="cb90-2017"><a href="#cb90-2017"></a><span class="in">```</span></span>
<span id="cb90-2018"><a href="#cb90-2018"></a></span>
<span id="cb90-2019"><a href="#cb90-2019"></a>Here we'll use the <span class="in">`filled.contour`</span> function to visualize the contact matrix, to show you an alternative way of visualizing contact matrices.</span>
<span id="cb90-2020"><a href="#cb90-2020"></a>Notices that we are using the raw matrix object, not a long dataframe, as previously.</span>
<span id="cb90-2021"><a href="#cb90-2021"></a></span>
<span id="cb90-2024"><a href="#cb90-2024"></a><span class="in">```{r}</span></span>
<span id="cb90-2025"><a href="#cb90-2025"></a><span class="co">#| column: body</span></span>
<span id="cb90-2026"><a href="#cb90-2026"></a><span class="co">#| out-width: 100%</span></span>
<span id="cb90-2027"><a href="#cb90-2027"></a><span class="fu">filled.contour</span>(</span>
<span id="cb90-2028"><a href="#cb90-2028"></a>    ages, ages, <span class="fu">log10</span>(mossong_mat),</span>
<span id="cb90-2029"><a href="#cb90-2029"></a>    <span class="at">plot.title =</span> <span class="fu">title</span>(</span>
<span id="cb90-2030"><a href="#cb90-2030"></a>        <span class="at">main =</span> <span class="st">"Log10 of Raw Contact Rate"</span>,</span>
<span id="cb90-2031"><a href="#cb90-2031"></a>        <span class="at">xlab =</span> <span class="st">"Age of Contactor"</span>,</span>
<span id="cb90-2032"><a href="#cb90-2032"></a>        <span class="at">ylab =</span> <span class="st">"Age of Contactee"</span></span>
<span id="cb90-2033"><a href="#cb90-2033"></a>    )</span>
<span id="cb90-2034"><a href="#cb90-2034"></a>)</span>
<span id="cb90-2035"><a href="#cb90-2035"></a><span class="in">```</span></span>
<span id="cb90-2036"><a href="#cb90-2036"></a></span>
<span id="cb90-2039"><a href="#cb90-2039"></a><span class="in">```{r}</span></span>
<span id="cb90-2040"><a href="#cb90-2040"></a><span class="co">#| column: body</span></span>
<span id="cb90-2041"><a href="#cb90-2041"></a><span class="co">#| out-width: 100%</span></span>
<span id="cb90-2042"><a href="#cb90-2042"></a><span class="fu">filled.contour</span>(</span>
<span id="cb90-2043"><a href="#cb90-2043"></a>    ages, ages, <span class="fu">log10</span>(mossong_mat_sym),</span>
<span id="cb90-2044"><a href="#cb90-2044"></a>    <span class="at">plot.title =</span> <span class="fu">title</span>(</span>
<span id="cb90-2045"><a href="#cb90-2045"></a>        <span class="at">main =</span> <span class="st">"Log10 of Symmetrical Contact Rate"</span>,</span>
<span id="cb90-2046"><a href="#cb90-2046"></a>        <span class="at">xlab =</span> <span class="st">"Age of Contactor"</span>,</span>
<span id="cb90-2047"><a href="#cb90-2047"></a>        <span class="at">ylab =</span> <span class="st">"Age of Contactee"</span></span>
<span id="cb90-2048"><a href="#cb90-2048"></a>    )</span>
<span id="cb90-2049"><a href="#cb90-2049"></a>)</span>
<span id="cb90-2050"><a href="#cb90-2050"></a><span class="in">```</span></span>
<span id="cb90-2051"><a href="#cb90-2051"></a></span>
<span id="cb90-2054"><a href="#cb90-2054"></a><span class="in">```{r}</span></span>
<span id="cb90-2055"><a href="#cb90-2055"></a>mossong_cont_sums <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb90-2056"><a href="#cb90-2056"></a>        <span class="at">age =</span> <span class="fu">factor</span>(mossong_ages, <span class="at">levels =</span> mossong_ages),</span>
<span id="cb90-2057"><a href="#cb90-2057"></a>        <span class="at">contactees =</span> <span class="fu">rowSums</span>(mossong_mat),</span>
<span id="cb90-2058"><a href="#cb90-2058"></a>        <span class="at">contactors =</span> <span class="fu">colSums</span>(mossong_mat)</span>
<span id="cb90-2059"><a href="#cb90-2059"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb90-2060"><a href="#cb90-2060"></a>    <span class="fu">pivot_longer</span>(<span class="sc">-</span>age, <span class="at">names_to =</span> <span class="st">"type"</span>, <span class="at">values_to =</span> <span class="st">"total_contacts"</span>) </span>
<span id="cb90-2061"><a href="#cb90-2061"></a><span class="in">```</span></span>
<span id="cb90-2062"><a href="#cb90-2062"></a></span>
<span id="cb90-2065"><a href="#cb90-2065"></a><span class="in">```{r}</span></span>
<span id="cb90-2066"><a href="#cb90-2066"></a><span class="co">#| column: body</span></span>
<span id="cb90-2067"><a href="#cb90-2067"></a><span class="co">#| out-width: 100%</span></span>
<span id="cb90-2068"><a href="#cb90-2068"></a><span class="fu">ggplot</span>(</span>
<span id="cb90-2069"><a href="#cb90-2069"></a>        mossong_cont_sums,</span>
<span id="cb90-2070"><a href="#cb90-2070"></a>        <span class="fu">aes</span>(<span class="at">x =</span> age, <span class="at">y =</span> total_contacts, <span class="at">color =</span> type, <span class="at">fill =</span> type, <span class="at">group =</span> type)</span>
<span id="cb90-2071"><a href="#cb90-2071"></a>    ) <span class="sc">+</span></span>
<span id="cb90-2072"><a href="#cb90-2072"></a>    <span class="fu">geom_path</span>(<span class="at">linewidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb90-2073"><a href="#cb90-2073"></a>    <span class="fu">geom_point</span>(</span>
<span id="cb90-2074"><a href="#cb90-2074"></a>        <span class="at">position =</span> <span class="st">"identity"</span>,</span>
<span id="cb90-2075"><a href="#cb90-2075"></a>        <span class="at">alpha =</span> <span class="fl">0.8</span>,</span>
<span id="cb90-2076"><a href="#cb90-2076"></a>        <span class="at">shape =</span> <span class="dv">21</span>,</span>
<span id="cb90-2077"><a href="#cb90-2077"></a>        <span class="at">size =</span> <span class="dv">4</span></span>
<span id="cb90-2078"><a href="#cb90-2078"></a>    ) <span class="sc">+</span></span>
<span id="cb90-2079"><a href="#cb90-2079"></a>    <span class="fu">scale_color_manual</span>(</span>
<span id="cb90-2080"><a href="#cb90-2080"></a>        <span class="at">values =</span> <span class="fu">c</span>(<span class="st">"slategray4"</span>, <span class="st">"navy"</span>),</span>
<span id="cb90-2081"><a href="#cb90-2081"></a>        <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"Contactees"</span>, <span class="st">"Contactors"</span>),</span>
<span id="cb90-2082"><a href="#cb90-2082"></a>        <span class="at">aesthetics =</span> <span class="fu">c</span>(<span class="st">"color"</span>, <span class="st">"fill"</span>)</span>
<span id="cb90-2083"><a href="#cb90-2083"></a>    ) <span class="sc">+</span></span>
<span id="cb90-2084"><a href="#cb90-2084"></a>    <span class="fu">guides</span>(<span class="at">color =</span> <span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb90-2085"><a href="#cb90-2085"></a>    <span class="fu">labs</span>(</span>
<span id="cb90-2086"><a href="#cb90-2086"></a>        <span class="at">x =</span> <span class="st">"Age"</span>,</span>
<span id="cb90-2087"><a href="#cb90-2087"></a>        <span class="at">y =</span> <span class="st">"Total contacts"</span>,</span>
<span id="cb90-2088"><a href="#cb90-2088"></a>        <span class="at">fill =</span> <span class="st">"Type of contact"</span></span>
<span id="cb90-2089"><a href="#cb90-2089"></a>    ) <span class="sc">+</span></span>
<span id="cb90-2090"><a href="#cb90-2090"></a>    <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span>
<span id="cb90-2091"><a href="#cb90-2091"></a><span class="in">```</span></span>
<span id="cb90-2092"><a href="#cb90-2092"></a></span>
<span id="cb90-2093"><a href="#cb90-2093"></a>While this matrix tells us how many contacts are made per year by an individual of each age, it doesn't tell us anything about the probability that a contact results in communication of infection.</span>
<span id="cb90-2094"><a href="#cb90-2094"></a>Let's assume that each contact has a constant probability $q$ of resulting in a transmission event.</span>
<span id="cb90-2095"><a href="#cb90-2095"></a></span>
<span id="cb90-2098"><a href="#cb90-2098"></a><span class="in">```{r}</span></span>
<span id="cb90-2099"><a href="#cb90-2099"></a>q <span class="ot">&lt;-</span> <span class="fl">3e-5</span></span>
<span id="cb90-2100"><a href="#cb90-2100"></a>mossong_beta_mat <span class="ot">&lt;-</span> q <span class="sc">*</span> mossong_mat_sym</span>
<span id="cb90-2101"><a href="#cb90-2101"></a><span class="in">```</span></span>
<span id="cb90-2102"><a href="#cb90-2102"></a></span>
<span id="cb90-2105"><a href="#cb90-2105"></a><span class="in">```{r}</span></span>
<span id="cb90-2106"><a href="#cb90-2106"></a><span class="co">#| column: body</span></span>
<span id="cb90-2107"><a href="#cb90-2107"></a><span class="co">#| out-width: 100%</span></span>
<span id="cb90-2108"><a href="#cb90-2108"></a><span class="fu">filled.contour</span>(</span>
<span id="cb90-2109"><a href="#cb90-2109"></a>    ages, ages, <span class="fu">log10</span>(mossong_beta_mat),</span>
<span id="cb90-2110"><a href="#cb90-2110"></a>    <span class="at">plot.title =</span> <span class="fu">title</span>(</span>
<span id="cb90-2111"><a href="#cb90-2111"></a>        <span class="at">main =</span> <span class="st">"WAIFW matrix based on POLYMOD data"</span>,</span>
<span id="cb90-2112"><a href="#cb90-2112"></a>        <span class="at">xlab =</span> <span class="st">"Age"</span>,</span>
<span id="cb90-2113"><a href="#cb90-2113"></a>        <span class="at">ylab =</span> <span class="st">"Age"</span></span>
<span id="cb90-2114"><a href="#cb90-2114"></a>    )</span>
<span id="cb90-2115"><a href="#cb90-2115"></a>)</span>
<span id="cb90-2116"><a href="#cb90-2116"></a><span class="in">```</span></span>
<span id="cb90-2117"><a href="#cb90-2117"></a></span>
<span id="cb90-2118"><a href="#cb90-2118"></a>Now let's simulate the introduction of such a pathogen into a population characterized by this contact structure.</span>
<span id="cb90-2119"><a href="#cb90-2119"></a></span>
<span id="cb90-2122"><a href="#cb90-2122"></a><span class="in">```{r}</span></span>
<span id="cb90-2123"><a href="#cb90-2123"></a><span class="co"># Update the parameters with the POLYMOD-based beta matrix</span></span>
<span id="cb90-2124"><a href="#cb90-2124"></a>mossong_params <span class="ot">&lt;-</span> multistage_params</span>
<span id="cb90-2125"><a href="#cb90-2125"></a>mossong_params[[<span class="st">"beta_mat"</span>]] <span class="ot">&lt;-</span> mossong_beta_mat</span>
<span id="cb90-2126"><a href="#cb90-2126"></a></span>
<span id="cb90-2127"><a href="#cb90-2127"></a><span class="co"># Solve the model with the updated parameters</span></span>
<span id="cb90-2128"><a href="#cb90-2128"></a>mossong_sol <span class="ot">&lt;-</span> deSolve<span class="sc">::</span><span class="fu">ode</span>(</span>
<span id="cb90-2129"><a href="#cb90-2129"></a>    <span class="at">y =</span> demog_yinit_ages,</span>
<span id="cb90-2130"><a href="#cb90-2130"></a>    <span class="at">times =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">200</span>, <span class="at">by =</span> <span class="fl">0.5</span>),</span>
<span id="cb90-2131"><a href="#cb90-2131"></a>    <span class="at">func =</span> multistage_model,</span>
<span id="cb90-2132"><a href="#cb90-2132"></a>    <span class="at">parms =</span> mossong_params</span>
<span id="cb90-2133"><a href="#cb90-2133"></a>)</span>
<span id="cb90-2134"><a href="#cb90-2134"></a></span>
<span id="cb90-2135"><a href="#cb90-2135"></a><span class="co"># Extract the timeseries of infectious individuals</span></span>
<span id="cb90-2136"><a href="#cb90-2136"></a>mossong_infecteds <span class="ot">&lt;-</span> mossong_sol[, <span class="dv">1</span> <span class="sc">+</span> iindex]</span>
<span id="cb90-2137"><a href="#cb90-2137"></a></span>
<span id="cb90-2138"><a href="#cb90-2138"></a><span class="co"># Convert infectious individual timeseries to dataframe for plotting</span></span>
<span id="cb90-2139"><a href="#cb90-2139"></a>mossong_infecteds_df <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb90-2140"><a href="#cb90-2140"></a>        <span class="at">time =</span> mossong_sol[, <span class="dv">1</span>],</span>
<span id="cb90-2141"><a href="#cb90-2141"></a>        <span class="at">Juveniles =</span> <span class="fu">apply</span>(mossong_infecteds[, juvies], <span class="dv">1</span>, sum),</span>
<span id="cb90-2142"><a href="#cb90-2142"></a>        <span class="at">Adults =</span> <span class="fu">apply</span>(mossong_infecteds[, adults], <span class="dv">1</span>, sum)</span>
<span id="cb90-2143"><a href="#cb90-2143"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb90-2144"><a href="#cb90-2144"></a>    <span class="fu">pivot_longer</span>(</span>
<span id="cb90-2145"><a href="#cb90-2145"></a>        <span class="at">cols =</span> <span class="fu">c</span>(Juveniles, Adults),</span>
<span id="cb90-2146"><a href="#cb90-2146"></a>        <span class="at">names_to =</span> <span class="st">"age_group"</span>,</span>
<span id="cb90-2147"><a href="#cb90-2147"></a>        <span class="at">values_to =</span> <span class="st">"infections"</span></span>
<span id="cb90-2148"><a href="#cb90-2148"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb90-2149"><a href="#cb90-2149"></a>    <span class="fu">mutate</span>(</span>
<span id="cb90-2150"><a href="#cb90-2150"></a>        <span class="at">age_group =</span> <span class="fu">factor</span>(age_group, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Juveniles"</span>, <span class="st">"Adults"</span>))</span>
<span id="cb90-2151"><a href="#cb90-2151"></a>    )</span>
<span id="cb90-2152"><a href="#cb90-2152"></a><span class="in">```</span></span>
<span id="cb90-2153"><a href="#cb90-2153"></a></span>
<span id="cb90-2156"><a href="#cb90-2156"></a><span class="in">```{r}</span></span>
<span id="cb90-2157"><a href="#cb90-2157"></a><span class="co">#| column: body</span></span>
<span id="cb90-2158"><a href="#cb90-2158"></a><span class="co">#| out-width: 100%</span></span>
<span id="cb90-2159"><a href="#cb90-2159"></a><span class="fu">ggplot</span>(mossong_infecteds_df, <span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> infections, <span class="at">color =</span> age_group)) <span class="sc">+</span></span>
<span id="cb90-2160"><a href="#cb90-2160"></a>    <span class="fu">geom_line</span>(<span class="at">linewidth =</span> <span class="fl">1.5</span>) <span class="sc">+</span></span>
<span id="cb90-2161"><a href="#cb90-2161"></a>    <span class="fu">scale_color_manual</span>(</span>
<span id="cb90-2162"><a href="#cb90-2162"></a>        <span class="at">values =</span> age_group_colors</span>
<span id="cb90-2163"><a href="#cb90-2163"></a>    ) <span class="sc">+</span></span>
<span id="cb90-2164"><a href="#cb90-2164"></a>    <span class="fu">labs</span>(</span>
<span id="cb90-2165"><a href="#cb90-2165"></a>        <span class="at">x =</span> <span class="st">"Time"</span>,</span>
<span id="cb90-2166"><a href="#cb90-2166"></a>        <span class="at">y =</span> <span class="st">"Number of infections"</span>,</span>
<span id="cb90-2167"><a href="#cb90-2167"></a>        <span class="at">color =</span> <span class="st">"Age group"</span></span>
<span id="cb90-2168"><a href="#cb90-2168"></a>    )</span>
<span id="cb90-2169"><a href="#cb90-2169"></a><span class="in">```</span></span>
<span id="cb90-2170"><a href="#cb90-2170"></a></span>
<span id="cb90-2171"><a href="#cb90-2171"></a>As before, we can also look at the equilibrium seroprevalence</span>
<span id="cb90-2172"><a href="#cb90-2172"></a></span>
<span id="cb90-2175"><a href="#cb90-2175"></a><span class="in">```{r}</span></span>
<span id="cb90-2176"><a href="#cb90-2176"></a><span class="co"># Get last time point</span></span>
<span id="cb90-2177"><a href="#cb90-2177"></a>mossong_equil <span class="ot">&lt;-</span> <span class="fu">drop</span>(<span class="fu">tail</span>(mossong_sol, <span class="dv">1</span>))[<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb90-2178"><a href="#cb90-2178"></a></span>
<span id="cb90-2179"><a href="#cb90-2179"></a><span class="co"># Calculate number of individuals in each age group at end of simulation</span></span>
<span id="cb90-2180"><a href="#cb90-2180"></a>mossong_equil_n <span class="ot">&lt;-</span> mossong_equil[sindex] <span class="sc">+</span></span>
<span id="cb90-2181"><a href="#cb90-2181"></a>    mossong_equil[iindex] <span class="sc">+</span></span>
<span id="cb90-2182"><a href="#cb90-2182"></a>    mossong_equil[rindex]</span>
<span id="cb90-2183"><a href="#cb90-2183"></a></span>
<span id="cb90-2184"><a href="#cb90-2184"></a><span class="co"># Calculate equilibrium seroprevalence</span></span>
<span id="cb90-2185"><a href="#cb90-2185"></a>mossong_equil_seroprev <span class="ot">&lt;-</span> mossong_equil[rindex] <span class="sc">/</span> mossong_equil_n</span>
<span id="cb90-2186"><a href="#cb90-2186"></a></span>
<span id="cb90-2187"><a href="#cb90-2187"></a><span class="co"># Convert to dataframe for plotting</span></span>
<span id="cb90-2188"><a href="#cb90-2188"></a>mossong_equil_seroprev_df <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb90-2189"><a href="#cb90-2189"></a>    <span class="co"># We can reuse the ages vectors from before as they are the same</span></span>
<span id="cb90-2190"><a href="#cb90-2190"></a>    <span class="co"># as the POLYMOD data</span></span>
<span id="cb90-2191"><a href="#cb90-2191"></a>    <span class="at">age =</span> ages,</span>
<span id="cb90-2192"><a href="#cb90-2192"></a>    <span class="at">seroprev =</span> mossong_equil_seroprev,</span>
<span id="cb90-2193"><a href="#cb90-2193"></a>    <span class="at">width =</span> da_ages</span>
<span id="cb90-2194"><a href="#cb90-2194"></a>)</span>
<span id="cb90-2195"><a href="#cb90-2195"></a><span class="in">```</span></span>
<span id="cb90-2196"><a href="#cb90-2196"></a></span>
<span id="cb90-2199"><a href="#cb90-2199"></a><span class="in">```{r}</span></span>
<span id="cb90-2200"><a href="#cb90-2200"></a><span class="co">#| column: body</span></span>
<span id="cb90-2201"><a href="#cb90-2201"></a><span class="co">#| out-width: 100%</span></span>
<span id="cb90-2202"><a href="#cb90-2202"></a><span class="fu">ggplot</span>(mossong_equil_seroprev_df, <span class="fu">aes</span>(<span class="at">x =</span> age, <span class="at">y =</span> seroprev, <span class="at">fill =</span> age)) <span class="sc">+</span></span>
<span id="cb90-2203"><a href="#cb90-2203"></a>    <span class="fu">geom_col</span>(<span class="at">width =</span> mossong_equil_seroprev_df<span class="sc">$</span>width, <span class="at">just =</span> <span class="fl">1.0</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb90-2204"><a href="#cb90-2204"></a>    <span class="fu">labs</span>(</span>
<span id="cb90-2205"><a href="#cb90-2205"></a>        <span class="at">x =</span> <span class="st">"Age"</span>,</span>
<span id="cb90-2206"><a href="#cb90-2206"></a>        <span class="at">y =</span> <span class="st">"Seroprevalence"</span></span>
<span id="cb90-2207"><a href="#cb90-2207"></a>    ) <span class="sc">+</span></span>
<span id="cb90-2208"><a href="#cb90-2208"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">80</span>, <span class="dv">10</span>)) <span class="sc">+</span></span>
<span id="cb90-2209"><a href="#cb90-2209"></a>    <span class="fu">scale_fill_continuous</span>(<span class="at">low =</span> age_group_colors[<span class="dv">1</span>], <span class="at">high =</span> age_group_colors[<span class="dv">2</span>])</span>
<span id="cb90-2210"><a href="#cb90-2210"></a><span class="in">```</span></span>
<span id="cb90-2211"><a href="#cb90-2211"></a></span>
<span id="cb90-2212"><a href="#cb90-2212"></a>and compute the $R_0$ for this infection.</span>
<span id="cb90-2213"><a href="#cb90-2213"></a></span>
<span id="cb90-2216"><a href="#cb90-2216"></a><span class="in">```{r}</span></span>
<span id="cb90-2217"><a href="#cb90-2217"></a>mossong_stable_n <span class="ot">&lt;-</span> <span class="fu">solve</span>(</span>
<span id="cb90-2218"><a href="#cb90-2218"></a>    mossong_params[[<span class="st">"aging_mat"</span>]],</span>
<span id="cb90-2219"><a href="#cb90-2219"></a>    <span class="sc">-</span><span class="fu">c</span>(mossong_params[[<span class="st">"births"</span>]], <span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">29</span>))</span>
<span id="cb90-2220"><a href="#cb90-2220"></a>)</span>
<span id="cb90-2221"><a href="#cb90-2221"></a></span>
<span id="cb90-2222"><a href="#cb90-2222"></a><span class="fu">calculate_R0</span>(</span>
<span id="cb90-2223"><a href="#cb90-2223"></a>    <span class="at">beta_mat =</span> mossong_params[[<span class="st">"beta_mat"</span>]],</span>
<span id="cb90-2224"><a href="#cb90-2224"></a>    <span class="at">stable_n_mat =</span> mossong_stable_n,</span>
<span id="cb90-2225"><a href="#cb90-2225"></a>    <span class="at">aging_mat =</span> mossong_params[[<span class="st">"aging_mat"</span>]],</span>
<span id="cb90-2226"><a href="#cb90-2226"></a>    <span class="at">recovery =</span> mossong_params[[<span class="st">"recovery"</span>]]</span>
<span id="cb90-2227"><a href="#cb90-2227"></a>)</span>
<span id="cb90-2228"><a href="#cb90-2228"></a><span class="in">```</span></span>
<span id="cb90-2229"><a href="#cb90-2229"></a></span>
<span id="cb90-2230"><a href="#cb90-2230"></a>::: {.callout-note title="QUESTION"}</span>
<span id="cb90-2231"><a href="#cb90-2231"></a>How does this R0 value compare to the R0 value obtained from @sec-ex-3?</span>
<span id="cb90-2232"><a href="#cb90-2232"></a>:::</span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



<script src="site_libs/quarto-contrib/line-highlight-1.0.0/line-highlight.js" defer="true"></script>
</body></html>